// byted-tea-sdk@5.3.10/src/plugin/stay/alive.ts
import { isSupVisChange, beforePageUnload, isObject } from '../../util/tool'
import { getXpath } from '../track/path'

interface Option {
  aliveName?: string,
  params?: any
  mode?: string
  aliveTime?: number
  container?: any
  reportMode?: string
}
export default class Alive {
  collect: any
  config: any
  pageStartTime: number
  maxDuration: number = 12 * 60 * 60 * 1000
  sessionStartTime: number
  timerHandler: any
  url_path: string
  url: string
  title: string
  set_path: string
  set_url: string
  set_title: string
  aliveDTime: number = 60 * 1000
  aliveName: string
  disableCallback: any
  options: Option = { aliveName: 'predefine_page_alive', params: {}, container: null }
  focusState: boolean = true
  hard: boolean
  removeMode: any
  reportMode: string
  constructor(collect: any, config: any) {
    this.collect = collect
    this.config = config
    this.pageStartTime = Date.now()
    this.sessionStartTime = this.pageStartTime
    this.timerHandler = null
    if (isObject(config.enable_stay_duration)) {
      this.options = Object.assign(this.options, config.enable_stay_duration)
    }
    this.hard = this.options.mode && this.options.mode === 'hard'
    this.reportMode = this.options.reportMode || 'general'
    this.focusState = document.hasFocus()
    this.aliveDTime = this.options.aliveTime || 60 * 1000
  }
  setParams(url_path: string, title: string, url: string) {
    this.set_path = url_path
    this.set_url = url
    this.set_title = title
  }
  enable(url_path: string, title: string, url: string) {
    this.url_path = url_path
    this.url = url
    this.title = title
    this.disableCallback = this.enablePageAlive()
    if (this.hard) {
      this.removeMode = this.hardMode()
    }
  }

  disable() {
    this.disableCallback()
    this.pageStartTime = Date.now()
    if (this.hard) {
      this.removeMode && this.removeMode()
    }
  }

  sendEvent(leave: boolean, limited = false, blur?: boolean) {
    const duration = limited ? this.aliveDTime : Date.now() - this.sessionStartTime
    if (duration < 0 || duration > this.aliveDTime || (Date.now() - this.pageStartTime > this.maxDuration)) {
      return
    }
    // 未知情况，在document.visibilityState是hidden的时候也会定时上报
    if (limited && document.visibilityState === 'hidden') return;
    if (this.hard && !this.focusState) return
    let aliveData = {
      url_path: this.getParams('url_path'),
      title: this.getParams('title'),
      url: this.getParams('url'),
      duration: duration,
      is_support_visibility_change: isSupVisChange(),
      startTime: this.sessionStartTime,
      hidden: document.visibilityState,
      leave,
      mode: this.hard ? 'hard' : 'normal',
      focusState: this.focusState,
      blur,
      aliveType: this.options.container ? 'dom' : 'page',
      ...this.options.params
    }
    if (this.options.container) {
      const { element_path } = getXpath(this.options.container)
      aliveData = Object.assign(aliveData, { element_path, aliveType: this.options.container ? 'dom' : 'page' })
    }
    this.collect.beconEvent(this.options.aliveName, aliveData)
    // this.resetParams(location.pathname, location.href, document.title)
    this.sessionStartTime = Date.now()
  }

  getParams(type: string) {
    switch (type) {
      case 'url_path':
        return this.set_path || this.url_path || location.pathname
      case 'title':
        return this.set_title || this.title || document.title || location.pathname
      case 'url':
        return this.set_url || this.url || location.href
    }
  }
  setUpTimer() {
    if (this.timerHandler) clearInterval(this.timerHandler)
    return setInterval(() => {
      if (Date.now() - this.sessionStartTime > this.aliveDTime) {
        this.sendEvent(false, true)
      }
    }, 1000)
  }

  visibilitychange() {
    if (this.getDisplayStatus() === 'hidden') {
      if (this.timerHandler) {
        clearInterval(this.timerHandler)
        this.sendEvent(false, false)
      }
    } else if (this.getDisplayStatus() === 'visible') {
      // 重置时间
      this.sessionStartTime = Date.now()
      this.timerHandler = this.setUpTimer()
    }
  }
  getDisplayStatus() {
    if (this.options.container) {
      try {
        const style = window.getComputedStyle(this.options.container);
        if (style.display !== 'none' && style.visibility !== 'hidden' && style.opacity !== '0') {
          return 'visible'
        } else {
          return 'hidden'
        }
      } catch (e) {
        return 'visible'
      }
    } else {
      return document.visibilityState
    }
  }
  beforeunload() {
    if (!document.hidden) {
      this.sendEvent(true, false)
    }
  }
  muition() {
    const observer = new MutationObserver((mutations) => {
      mutations.forEach((mutation) => {
        // var newStyle = mutation.target.style;
        // console.log(mutation.target);
        // 检查元素是否可见
        this.visibilitychange()

      });
    });
    observer.observe(this.options.container, {
      attributes: true,
      attributeFilter: ['style']
    });
  }
  enablePageAlive() {
    this.timerHandler = this.setUpTimer()
    const change = this.visibilitychange.bind(this)
    const before = this.beforeunload.bind(this)
    if (this.options.container) {
      this.muition()
      // this.options.container.addEventListener('visibilitychange', change)
    } else {
      document.addEventListener('visibilitychange', change)
    }
    // beforePageUnload(before)
    // window.addEventListener('beforeunload', before('beforeunload'))
    window.addEventListener('pagehide', before)
    return () => {
      this.beforeunload()
      document.removeEventListener('visibilitychange', change)
      // window.removeEventListener('beforeunload', before)
      window.removeEventListener('pagehide', before)
    }
  }
  hardMode() {
    // 严格模式
    const blur = () => {
      if (this.timerHandler) {
        clearInterval(this.timerHandler)
        this.sendEvent(false, false, true)
      }
      this.focusState = false
    }
    const focus = () => {
      this.focusState = true
      this.sessionStartTime = Date.now()
      this.timerHandler = this.setUpTimer()
    }
    window.addEventListener("blur", blur);
    window.addEventListener("focus", focus);
    return () => {
      window.removeEventListener('blur', blur)
      window.removeEventListener('focus', focus)
    }
  }
}
