// byted-tea-sdk@5.3.10/src/collect/collect.ts
import { IInitParam, IConfigParam, EventParams } from '../../types/types'
import Hook, { THook, THookInfo } from '../plugin/hook';
import { isObject, isNumber, isString, getIndex, hashCode } from '../util/tool';
import ConfigManager from './config'
import Logger from '../util/log'
import Spider from '../plugin/spider/spider';
import Types, { DebuggerMesssge } from './hooktype';
import AppBridge from '../plugin/native'
import Event from './event'
import Token from './token'
import Session, { sessionId } from './session';
import { SDK_VERSION } from './constant'
import Storage from '../util/storage'
import RequestManager from '../plugin/request/request'
import Debugger from '../plugin/debug/debug';
import { PluginList } from './plugin'

type TEvent = any;

export type ProfileParams = {
  [key: string]: string | number | boolean | Array<any>;
};
export interface IPlugin {
  new();
  apply: (collect: Collector, options: IInitParam) => void
}
type Plugin = {
  name?: string;
  plugin: IPlugin;
  options?: any
}

export default class Collector {
  name: string
  hook: any
  inited: boolean
  logger: any
  spider: any
  Types: any
  configManager: any
  appBridge: any
  eventManager: any
  tokenManager: any
  sessionManager: any
  disableAutoPageView: boolean = false
  bridgeReport: boolean = false
  staging: boolean = false
  initConfig: any
  static plugins: Array<Plugin> = []
  // plugins: Array<Plugin> = []
  pluginInstances: Array<IPlugin> = []
  sended: boolean = false
  started: boolean = false
  remotePlugin: any
  eventFilter: any
  destroy: boolean = false
  sdkReady: boolean = false
  eventCheck: any
  adapters: Record<string, any> = {}
  debugger: any
  loadType: string = 'base'
  sdkStop: boolean = false
  requestManager: any
  pageviewParams: any = {}
  constructor(name: string) {
    this.name = name
    this.hook = new Hook()
    this.remotePlugin = new Map()
    this.Types = Types
    // this.adapters['fetch'] = fetch
    this.adapters['storage'] = Storage
    /**@@SDK_FULL
    this.loadType = 'full'
    @@SDK_FULL*/
  }
  static usePlugin(plugin: IPlugin, name?: string, options?: any) {
    // 本地加载
    if (name) {
      let check = false
      for (let i = 0, len = Collector.plugins.length; i < len; i++) {
        const p = Collector.plugins[i]
        if (p.name === name) {
          Collector.plugins[i].plugin = plugin;
          Collector.plugins[i].options = options || {};
          check = true
          break
        }
      }
      if (!check) {
        Collector.plugins.push({ name, plugin, options })
      }
    } else {
      Collector.plugins.push({ plugin })
    }
  }
  usePlugin(name: string, plugin?: any, instance?: string) {
    if (!name) return;
    if (this.loadType === 'full' && PluginList.includes(name)) {
      console.info(`your sdk version has ${name} plugin already ~`)
      return;
    }
    if (!plugin) {
      // 自带插件
      if (!this.remotePlugin.get(name)) {
        this.remotePlugin.set(name, 'sdk')
      }
    } else {
      if (typeof plugin === 'string') {
        // 远程插件
        if (!this.remotePlugin.get(name)) {
          this.remotePlugin.set(name, { src: plugin, call: instance })
        }
      } else {
        // 本地插件
        if (!this.remotePlugin.get(name)) {
          this.remotePlugin.set(name, { instance: plugin })
        }
      }
    }
  }
  init(initConfig: IInitParam) {
    this.logger = new Logger(this.name, initConfig.log)
    if (this.inited) {
      this.logger.warn(`[instance: ${this.name}], every instance's api: init,  can be call only one time!`)
      return;
    }
    if (!initConfig || !isObject(initConfig)) {
      this.logger.warn('init params error,please check!')
      return;
    }
    if (!initConfig.app_id || !isNumber(initConfig.app_id)) {
      this.logger.warn('app_id param is error, must be number, please check!')
      return;
    }
    if (initConfig.app_key && !isString(initConfig.app_key)) {
      this.logger.warn('app_key param is error, must be string, please check!')
      return;
    }
    if (!initConfig.channel_domain && ['cn', 'sg', 'va', 'my'].indexOf(initConfig.channel) === -1) {
      this.logger.warn('channel must be `cn`, `sg`,`va`, `my` !')
      initConfig.channel = 'cn';
    }
    this.spider = new Spider();
    if (this.spider.checkSpider(initConfig)) {
      this.logger.warn('your env may be a spider, can not report!');
      return;
    }
    this.configManager = new ConfigManager(this, initConfig)
    this.requestManager = new RequestManager(this, initConfig)
    this.appBridge = new AppBridge(initConfig, this.configManager)
    this.bridgeReport = this.appBridge.bridgeInject()
    this.debugger = new Debugger(this, initConfig)
    this.initConfig = initConfig
    this.emit(Types.Init)
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Init', data: initConfig, level: 'info', time: Date.now(), infoType: 'cus' })
    this.destroy = false;
    if (initConfig.disable_auto_pv) {
      this.disableAutoPageView = true;
    }
    if (!this.bridgeReport) {
      this.configManager.set({ app_id: initConfig.app_id })
      this.eventManager = new Event()
      this.tokenManager = new Token()
      this.sessionManager = new Session()
      Promise.all([
        new Promise<boolean>((resolve) => {
          this.once(Types.TokenComplete, () => {
            resolve(true)
          });
        }),
        new Promise<boolean>((resolve) => {
          this.once(Types.Start, () => {
            resolve(true)
          });
        }),
        new Promise<boolean>((resolve) => {
          this.once(Types.LogSettingReady, () => {
            resolve(true)
          });
        }),
      ]).then(() => {
        try {
          Collector.plugins.reduce((result, plugin) => {
            const { plugin: P, options } = plugin
            // const cloneConfig = JSON.parse(JSON.stringify(this.initConfig))
            const cloneOptions = Object.assign(this.initConfig, options)
            const instance = new P()
            instance.apply(this, cloneOptions)
            result.push(instance)
            this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: `SDK加载${plugin.name}插件`, level: 'info', time: Date.now(), infoType: 'sdk' });
            return result
          }, this.pluginInstances)
        } catch (e) {
          console.warn(`load plugin error, ${e.message}`)
          this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK加载插件发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
        }
        this.sdkReady = true;
        this.emit(Types.Ready);
        this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 用户信息初始化完成', time: Date.now(), secType: 'USER', level: 'info', data: this.configManager.get('user'), infoType: 'sdk' })
        this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 初始化完成', time: Date.now(), level: 'info', data: this.configManager.get('user'), infoType: 'sdk' })
        this.logger.info(`appid: ${initConfig.app_id}, userInfo:${JSON.stringify(this.configManager.get('user'))}`);
        this.logger.info(`appid: ${initConfig.app_id}, sdk is ready, version type is ${this.loadType}, version is ${SDK_VERSION}, you can report now !!!`);
        if (initConfig.disable_auto_pv) {
          this.disableAutoPageView = true;
        }
        try {
          if (this.loadType === 'full' && (initConfig.enable_ab_test || initConfig.autotrack)) {
            (window.opener || window.parent).postMessage('[tea-sdk]ready', '*');
          }
        } catch (e) {
          this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
        }
        this.pageView();
        this.on(Types.TokenChange, (tokenType: string) => {
          if (tokenType === 'webid') {
            this.pageView();
          }
          this.logger.info(`appid: ${initConfig.app_id} token change, new userInfo:${JSON.stringify(this.configManager.get('user'))}`)
          this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 设置了用户信息', time: Date.now(), secType: 'USER', level: 'info', data: this.configManager.get('user'), infoType: 'sdk' })
        })
        this.on(Types.TokenReset, () => {
          this.logger.info(`appid: ${initConfig.app_id} token reset, new userInfo:${JSON.stringify(this.configManager.get('user'))}`)
          this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 重置了用户信息', time: Date.now(), secType: 'USER', level: 'info', data: this.configManager.get('user'), infoType: 'sdk' })
        })

        this.on(Types.RouteChange, (info) => {
          if (info.init) return;
          if (initConfig.disable_route_report) return;
          this.pageView();
        })
      })
      this.tokenManager.apply(this, initConfig)
      this.eventManager.apply(this, initConfig)
      this.sessionManager.apply(this, initConfig)
    }
    this.inited = true
  }
  config(configs: IConfigParam) {
    if (!this.inited) {
      console.warn('config must be use after function init')
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK执行Config发生了异常', level: 'error', time: Date.now(), data: 'config must be use after function init', infoType: 'sdk' });
      return;
    }
    if (!configs || !isObject(configs)) {
      console.warn('config params is error, please check')
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK执行Config发生了异常', level: 'error', time: Date.now(), data: 'config params is error, please check', infoType: 'sdk' });
      return;
    }
    if (this.bridgeReport) {
      this.appBridge.setConfig(configs)
    } else {
      if (configs._staging_flag && configs._staging_flag === 1) {
        this.staging = true
      }
      if (configs.disable_auto_pv) {
        this.disableAutoPageView = true;
        delete configs['disable_auto_pv']
      }
      let newConfig = { ...configs }
      const persist = this.initConfig.configPersist || this.initConfig.config_persist || false;
      if (this.initConfig && persist) {
        let cacheConfig = this.configManager.getStore()
        if (cacheConfig) {
          newConfig = Object.assign(cacheConfig, configs)
        }
        this.configManager.setStore(configs)
      }
      const {
        web_id: webId,
        user_unique_id: userUniqueId,
        ...otherConfigs
      } = newConfig
      this.configManager.set(otherConfigs)
      if (newConfig.hasOwnProperty('web_id')) {
        this.emit(Types.ConfigWebId, newConfig['web_id'])
      }
      if (newConfig.hasOwnProperty('user_unique_id')) {
        this.emit(Types.ConfigUuid, newConfig['user_unique_id'])
      }
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Config', level: 'info', time: Date.now(), data: newConfig, infoType: 'cus' })
    }
  }
  setUserUniqueID(user_unique_id: any) {
    this.config({
      user_unique_id
    })
  }
  setHeaderInfo(key: string, value: any) {
    const headerInfo = {};
    headerInfo[key] = value;
    this.config(headerInfo);
  }
  removeHeaderInfo(key: string) {
    const headerInfo = {};
    headerInfo[key] = undefined;
    this.config(headerInfo);
  }
  setDomain(domain: string) {
    this.configManager && this.configManager.setDomain(domain)
    this.emit(Types.ConfigDomain)
  }
  getConfig(key?: string) {
    return this.configManager?.get(key)
  }

  send() {
    this.start();
  }

  start() {
    if (!this.inited || this.sended) return
    this.sended = true
    this.started = true
    this.emit(Types.Start)
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Start', level: 'info', time: Date.now(), infoType: 'cus' })
    if (this.bridgeReport) {
      this.pageView()
      this.emit(Types.Ready)
    }
  }

  event(event: string | TEvent[], params?: any) {
    try {
      const events = [];
      if (Array.isArray(event)) {
        event.forEach((each) => {
          const process = this.processEvent(each[0], each[1] || {})
          if (!process) return
          events.push(process)
        })
      } else {
        const process = this.processEvent(event, params)
        if (!process) return
        events.push(process)
      }
      if (this.bridgeReport) {
        events.forEach(item => {
          const { event, params } = item
          this.appBridge.onEventV3(event, JSON.stringify(params))
        })
      } else {
        if (events.length) {
          this.emit(Types.Event, events)
          this.emit(Types.SessionResetTime)
        }
      }
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Event', level: 'info', time: Date.now(), data: event, infoType: 'cus' });
    } catch (e) {
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Event发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
    }
  }
  beconEvent(event: string, params?: any) {
    if (Array.isArray(event)) {
      console.warn('beconEvent not support batch report, please check');
      return;
    }
    const events = [];
    const process = this.processEvent(event, params || {})
    if (!process) return
    events.push(process)
    if (events.length) {
      this.emit(Types.BeconEvent, events)
      this.emit(Types.SessionResetTime)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行Beacon Event', level: 'info', time: Date.now(), data: event, infoType: 'sdk' });
  }
  beaconEvent(event: string, params?: any) {
    this.beconEvent(event, params);
  }
  processEvent(_event: string, params: any = {}) {
    try {
      if (!_event) {
        console.warn('eventName is null， please check')
        return null
      }
      const EVT_REG = /^event\./
      let event = _event
      if (EVT_REG.test(_event)) {
        event = _event.slice(6)
      }
      let localParams = params
      if (typeof localParams !== 'object') {
        localParams = {}
      }
      if (localParams.profile) {
        delete localParams['profile']
      } else {
        localParams.event_index = getIndex()
      }
      let local_ms
      if (localParams.local_ms) {
        local_ms = localParams.local_ms
        delete localParams['local_ms']
      } else {
        local_ms = +new Date()
      }
      const evtData = {
        event,
        params: localParams,
        local_time_ms: local_ms,
        is_bav: this.initConfig && this.initConfig.autotrack ? 1 : 0
      }
      return evtData
    } catch (e) {
      this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
      return { event: _event, params }
    }
  }

  filterEvent(filter: any) {
    this.eventFilter = filter
  }

  on(type: string, hook: THook) {
    this.hook.on(type, hook);
  }

  once(type: string, hook: THook) {
    this.hook.once(type, hook);
  }

  off(type: string, hook?: THook) {
    this.hook.off(type, hook);
  }

  emit(type: string, info?: THookInfo, wait?: string) {
    this.hook.emit(type, info, wait);
  }
  set(type: string) {
    this.hook.set(type);
  }
  stop() {
    this.sdkStop = true;
  }
  reStart() {
    this.sdkStop = false;
  }
  pageView() {
    if (this.disableAutoPageView) return;
    this.predefinePageView();
  }
  setPageViewParams(params: any = {}) {
    this.pageviewParams = params;
  }
  predefinePageView(params: any = {}) {
    if (!this.inited) {
      console.warn('predefinePageView must be use after function init')
      return;
    }
    const defaultPvParams = {
      title: document.title || location.pathname,
      url: location.href,
      url_path: this.handlePath(),
      time: Date.now(),
      referrer: window.document.referrer,
      $is_first_time: `${this.configManager && this.configManager.is_first_time || false}`
    }
    const mergedParams = {
      ...defaultPvParams,
      ...params,
      ...this.pageviewParams
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 上报PV', level: 'info', time: Date.now(), infoType: 'sdk' });

    this.event('predefine_pageview', mergedParams)
  }
  handlePath() {
    let url_path = location.pathname
    try {
      if (this.initConfig.allow_hash) {
        const hash = new URL(location.href).hash.match(/#\/([^?#]+)/i);
        return url_path = hash && hash.length ? hash[1] : location.pathname;
      }
    } catch (e) {
      url_path = location.pathname;
    }
    return url_path
  }

  enableTransPort() {
    this.configManager.setEnableClientTransport(true);
  }

  disableTransPort() {
    this.configManager.setEnableClientTransport(false);
  }

  clearEventCache() {
    this.emit(Types.CleanEvents)
  }
  setWebIDviaUnionID(unionId: string) {
    if (!unionId) {
      return;
    }
    const webId = hashCode(unionId);
    this.config({
      web_id: `${webId}`,
      wechat_unionid: unionId,
    });
    this.emit(Types.CustomWebId)
  }
  setWebId(webid: string | number) {
    this.config({
      web_id: `${webid}`,
    });
  }
  setWebIDviaOpenID(openId: string) {
    if (!openId) {
      return;
    }
    const webId = hashCode(openId);
    this.config({
      web_id: `${webId}`,
      wechat_openid: openId,
    });
    this.emit(Types.CustomWebId)
  }

  setNativeAppId(appId: number) {
    if (!this.bridgeReport) return
    this.appBridge.setNativeAppId(appId)
  }

  setRequestHeaders(headerInfo: EventParams) {
    this.emit(Types.CustomHeader, headerInfo)
  }

  getSessionId() {
    return sessionId()
  }
  setSessionId(sessionId: string) {
    this.emit(Types.SessionReset, sessionId)
  }

  /** stay相关api */
  resetStayDuration(url_path?: string, title?: string, url?: string) {
    this.emit(Types.ResetStay, { url_path, title, url }, Types.Stay)

  }
  resetStayParams(url_path: string = '', title: string = '', url: string = '') {
    this.emit(Types.SetStay, { url_path, title, url }, Types.Stay)
  }
  getToken(callback: any, timeout?: number) {
    if (!this.inited) {
      console.warn('getToken must be use after function init')
      return;
    }
    let tokenReturn = false
    const callbackUser = (tobid?: string) => {
      if (tokenReturn) return
      tokenReturn = true
      const { user } = this.configManager.get()
      if (tobid) {
        user['tobid'] = tobid
        user['diss'.split('').reverse().join('')] = tobid
      }
      return callback({ ...user })
    }
    const getId = () => {
      this.tokenManager.getTobId().then(res => {
        callbackUser(res)
      })
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行GetToken', level: 'info', time: Date.now(), infoType: 'cus' });
    if (this.sdkReady) {
      getId()
      return
    }
    if (timeout) {
      setTimeout(() => {
        callbackUser()
      }, timeout)
    }
    this.on(Types.Ready, () => {
      getId()
    })
  }

  /** profile相关api */
  profileSet(profile: ProfileParams) {
    if (this.bridgeReport) {
      this.appBridge.profileSet(JSON.stringify(profile))
    } else {
      this.emit(Types.ProfileSet, profile, Types.Profile)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行profileSet', level: 'info', time: Date.now(), data: profile, infoType: 'cus' })
  }

  profileSetOnce(profile: ProfileParams) {
    if (this.bridgeReport) {
      this.appBridge.profileSetOnce(JSON.stringify(profile))
    } else {
      this.emit(Types.ProfileSetOnce, profile, Types.Profile)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行profileSetOnce', level: 'info', time: Date.now(), data: profile, infoType: 'cus' })
  }

  profileIncrement(profile: ProfileParams) {
    if (this.bridgeReport) {
      this.appBridge.profileIncrement(JSON.stringify(profile))
    } else {
      this.emit(Types.ProfileIncrement, profile, Types.Profile)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行profileIncrement', level: 'info', time: Date.now(), data: profile, infoType: 'cus' })
  }

  profileUnset(key: string) {
    if (this.bridgeReport) {
      this.appBridge.profileUnset(key)
    } else {
      this.emit(Types.ProfileUnset, key, Types.Profile)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行profileUnset', level: 'info', time: Date.now(), data: key, infoType: 'cus' })
  }

  profileAppend(profile: ProfileParams) {
    if (this.bridgeReport) {
      this.appBridge.profileAppend(JSON.stringify(profile))
    } else {
      this.emit(Types.ProfileAppend, profile, Types.Profile)
    }
    this.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行profileAppend', level: 'info', time: Date.now(), data: profile, infoType: 'cus' })
  }

  /**
   * track相关api
  */
  startTrackEvent(eventName: string) {
    if (!eventName) return;
    this.emit(Types.TrackDurationStart, eventName, Types.TrackDuration)
  }
  endTrackEvent(eventName: string, params: any = {}) {
    if (!eventName) return;
    this.emit(Types.TrackDurationEnd, {
      eventName,
      params
    }, Types.TrackDuration)
  }
  pauseTrackEvent(eventName: string) {
    if (!eventName) return;
    this.emit(Types.TrackDurationPause, eventName, Types.TrackDuration)
  }
  resumeTrackEvent(eventName: string) {
    if (!eventName) return;
    this.emit(Types.TrackDurationResume, eventName, Types.TrackDuration)
  }
  

  /** ab相关api */
  setExternalAbVersion(vids: string) {
    this.emit(
      Types.AbExternalVersion,
      typeof vids === 'string' && vids ? `${vids}`.trim() : null,
      Types.Ab
    )
  }

  getVar(name, defaultValue, callback) {
    this.emit(Types.AbVar, { name, defaultValue, callback }, Types.Ab)
  }

  getABconfig(params: any, callback: any) {
    this.emit(Types.AbConfig, { params, callback }, Types.Ab)
  }
  getABConfig(params: any, callback: any) {
    this.emit(Types.AbConfig, { params, callback }, Types.Ab)
  }
  getAbSdkVersion() {
    return this.configManager.getAbVersion()
  }
  onAbSdkVersionChange(linster: any) {
    this.emit(Types.AbVersionChangeOn, linster, Types.Ab)
    return () => {
      this.emit(Types.AbVersionChangeOff, linster, Types.Ab)
    }
  }
  offAbSdkVersionChange(linster?: any) {
    this.emit(Types.AbVersionChangeOff, linster, Types.Ab)
  }
  openOverlayer() {
    this.emit(Types.AbOpenLayer, '', Types.Ab)
  }

  closeOverlayer() {
    this.emit(Types.AbCloseLayer, '', Types.Ab)
  }

  getAllVars(callback) {
    this.emit(Types.AbAllVars, callback, Types.Ab)
  }

  destoryInstace() {
    if (this.destroy) return
    this.destroy = true
    this.off(Types.TokenComplete)
    this.emit(Types.DestoryInstance)
  }
  destroyInstance() {
    if (this.destroy) return
    this.destroy = true
    this.off(Types.TokenComplete)
    this.emit(Types.DestoryInstance)
  }
}
