// byted-tea-sdk@5.3.10/src/collect/event.ts
import Types from './hooktype';
import Storage from '../util/storage'
import { beforePageUnload, encodeBase64, decodeBase64, isIOS } from '../util/tool'
import { DebuggerMesssge } from './hooktype';
import EventCheck from '../plugin/check/check'
import { sessionId } from './session'
import LogSetting from '../plugin/logsetting/logsetting';


type TEvent = any;
export default class Event {
  collect: any
  config: any
  configManager: any
  eventKey: string
  beconKey: string
  abKey: string
  cacheStorgae: any
  localStorage: any
  reportTimeout: any
  maxReport: number
  reportTime: number
  timeout: number
  eventLimit: number = 50
  reportUrl: string
  enable_ttwebid: boolean = false
  eventCache: TEvent[] = []
  beconEventCache: TEvent[] = []
  eventCheck: any
  pageId: string
  refer_key: string
  logSetting: LogSetting
  whiteList: string[]
  blockList: string[]
  apply(collect: any, config: any) {
    this.collect = collect
    this.config = config
    this.configManager = collect.configManager
    this.eventCheck = new EventCheck(collect, config)
    this.cacheStorgae = new Storage(true)
    this.localStorage = new Storage(false)
    this.logSetting = new LogSetting()
    this.maxReport = config.max_report || 10
    this.reportTime = config.reportTime || config.report_time || 30
    this.timeout = config.timeout || 100000
    this.enable_ttwebid = config.enable_ttwebid
    this.reportUrl = config.report_url || this.configManager.getUrl('event')
    this.eventKey = `__tea_cache_events_${this.configManager.get('app_id')}`
    this.beconKey = `__tea_cache_events_becon_${this.configManager.get('app_id')}`
    this.abKey = `__tea_sdk_ab_version_${this.configManager.get('app_id')}`
    this.refer_key = `__tea_cache_refer_${this.configManager.get('app_id')}`
    this.pageId = sessionId()
    this.logSetting.apply(collect, config)
    this.collect.on(Types.Ready, () => {
      this.reportAll(false)
    })
    this.collect.on(Types.ConfigDomain, () => {
      this.reportUrl = this.configManager.getUrl('event')
    })

    this.collect.on(Types.Event, (events: any) => {
      this.event(events)
    });

    this.collect.on(Types.BeconEvent, (events: any) => {
      this.beconEvent(events)
    })
    this.collect.on(Types.CleanEvents, () => {
      // 清除当前的事件
      this.reportAll(false)
    })
    const remove = this.linster()
    this.collect.on(Types.DestoryInstance, () => {
      remove && remove()
    })
  }

  linster() {
    const leaveReport = () => {
      this.reportAll(true)
    }
    window.addEventListener('unload', leaveReport, false)
    beforePageUnload(leaveReport)
    const hiddenReport = () => {
      if (document.visibilityState === 'hidden') {
        this.reportAll(true)
      }
    }
    document.addEventListener('visibilitychange', hiddenReport, false)
    return () => {
      window.removeEventListener('unload', leaveReport)
      window.removeEventListener("pagehide", leaveReport, false)
      window.removeEventListener("beforeunload", leaveReport, false)
      document.removeEventListener('visibilitychange', hiddenReport, false)
    }
  }
  reportAll(becon?: boolean) {
    this.report(becon)
    this.reportBecon()
  }
  event(events: any) {
    try {
      if (this.collect.destroy) return
      if (this.collect.sdkStop) return
      if (this.config.enable_ios_sendbeacon && isIOS()) {
        this.beconEvent(events)
        return
      }
      const cache = this.cacheStorgae.getItem(this.eventKey) || []
      let newbatch = events
      if (this.config.batch_event) {
        newbatch = this.batchEvent(events)
      }
      const newCache = [...newbatch, ...cache]
      this.cacheStorgae.setItem(this.eventKey, newCache)
      if (this.reportTimeout) {
        clearTimeout(this.reportTimeout)
      }
      if (newCache.length >= this.maxReport) {
        this.report(false)
      } else {
        const _time = this.reportTime
        this.reportTimeout = setTimeout(() => {
          this.report(false)
          this.reportTimeout = null
        }, _time)
      }
    } catch (e) {
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
    }
  }
  beconEvent(events: any) {
    if (this.collect.destroy) return
    if (this.collect.sdkStop) return
    const cache = this.cacheStorgae.getItem(this.beconKey) || []
    const newCache = [...events, ...cache]
    this.cacheStorgae.setItem(this.beconKey, newCache)
    if (!this.collect.tokenManager.getReady()) return
    if (!this.collect.sdkReady) return
    this.cacheStorgae.removeItem(this.beconKey)
    this.send(this.split(this.merge(newCache)), true)
  }
  reportBecon() {
    if (this.collect.destroy) return
    if (this.collect.sdkStop) return
    if (!this.collect.tokenManager.getReady()) return
    if (!this.collect.sdkReady) return
    const cache = this.cacheStorgae.getItem(this.beconKey) || []
    if (!cache || !cache.length) return
    this.cacheStorgae.removeItem(this.beconKey)
    this.send(this.split(this.merge(cache)), true)
  }
  report(becon: boolean) {
    if (this.collect.destroy) return
    if (this.collect.sdkStop) return
    if (!this.collect.tokenManager.getReady()) return
    if (!this.collect.sdkReady) return
    const eventData = this.cacheStorgae.getItem(this.eventKey) || []
    if (!eventData.length) return
    this.cacheStorgae.removeItem(this.eventKey)
    if (!this.config.batch_event) {
      this.sliceEvent(eventData, becon);
    } else {
      this.send([eventData], becon)
    }
  }
  batchEvent(events: any) {
    let batch = []
    if (events.length > this.eventLimit) {
      for (let i = 0; i < events.length; i += this.eventLimit) {
        let result = []
        result = events.slice(i, i + this.eventLimit);
        const mergeData = this.merge(result);
        batch.push(mergeData);
      }
    } else {
      const mergeData = this.merge(events);
      batch = mergeData;
    }
    return batch
  }
  sliceEvent(events: any, becon: boolean) {
    if (events.length > this.eventLimit) {
      for (let i = 0; i < events.length; i += this.eventLimit) {
        let result = []
        result = events.slice(i, i + this.eventLimit);
        const mergeData = this.split(this.merge(result));
        this.send(mergeData, becon);
      }
    } else {
      const mergeData = this.split(this.merge(events));
      this.send(mergeData, becon);
    }
  }
  handleRefer() {
    let refer = ''
    try {
      if (this.config.spa || this.config.autotrack) {
        const cache_local = this.localStorage.getItem(this.refer_key) || {}
        if (cache_local.routeChange) {
          // 已经发生路由变化
          refer = cache_local.refer_key;
        } else {
          // 首页，用浏览器的refer
          refer = this.configManager.get('referrer');
        }
      } else {
        refer = this.configManager.get('referrer');
      }
    } catch (e) {
      refer = document.referrer;
    }
    return refer
  }
  merge(events: any, ignoreEvtParams?: boolean) {
    const { header, user } = this.configManager.get()
    if (this.config.enable_pageid) {
      header.custom.page_id = this.pageId;
    }
    // header.custom = this.customAllowFilter(header.custom)
    header.custom = JSON.stringify(header.custom)
    const evtParams = this.configManager.get('evtParams')
    const type = this.configManager.get('user_unique_id_type')
    let mergeData = []
    try {
      const mergeEvents = events.filter(item => {
        try {
          if (Object.keys(evtParams).length && !ignoreEvtParams) {
            item.params = { ...evtParams, ...item.params }
          }
          if (type) {
            item.params['$user_unique_id_type'] = type
          }
          const abCache = this.configManager.getAbCache();
          const abUserId = user[this.config.ab_user_mode] || user.user_unique_id;
          if (abCache && abCache.uuid && abCache.uuid === abUserId) {
            if (this.configManager.getAbVersion()) {
              item.ab_sdk_version = this.configManager.getAbVersion()
            }
          }
          item.session_id = this.collect.sessionManager.getSessionId()
          // item.params = this.paramsAllowFilter(item)
          item.params = JSON.stringify(item.params)
          return item;
        } catch (e) {
          console.warn(`filter merge event: ${item && item.event} error, pls check`, e && e.message)
          this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 处理Event数据异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
          return false;
        }
      })
      const resultEvent = JSON.parse(
        JSON.stringify({
          events: mergeEvents,
          user,
          header,
        }),
      );
      resultEvent.local_time = Math.floor(Date.now() / 1000);
      resultEvent.user_unique_type = this.config.enable_ttwebid ? this.config.user_unique_type : undefined
      resultEvent.verbose = 1
      mergeData.push(resultEvent)
    } catch (e) {
      console.warn('event merge error, pls check', e && e.message)
    }

    return mergeData
  }
  split(eventData: any) {
    eventData = eventData.map(item => {
      const _item = []
      _item.push(item)
      return _item
    })
    return eventData
  }
  blockEvent(eventName: string) {
    try {
      let block = this.blockList;
      if (this.config.enable_logsetting) {
        const resBlock = this.blockList;
        const settingConfig = this.logSetting.getSettingData('blocklist_events');
        const settingBlock = settingConfig.block_event;
        block = resBlock.concat(settingBlock.filter(item => resBlock.indexOf(item) === -1));
      }
      if (block && block.length) {
        if (block.includes(eventName)) {
          return false
        }
      }
      return true
    } catch (e) {
      return true
    }
  }
  whiteEvent(eventName: string) {
    try {
      if (this.whiteList?.length) {
        if (this.whiteList.includes(eventName)) {
          return true
        }
        return false
      }
      return true
    } catch (e) {
      return true
    }
  }
  customAllowFilter(custom: Record<string, any>, verify: boolean): string {
    try {
      if (verify) {
        return JSON.stringify(custom)
      }
      if (!this.config.enable_logsetting || !this.config.enable_logsetting_header_custom) {
        return JSON.stringify(custom)
      }
      const { header_custom_allow } = this.logSetting.getSettingData()
      const newCustom = {}
      if (header_custom_allow && header_custom_allow.length && Object.keys(custom).length) {
        const allowCustom = new Set(header_custom_allow)
        for (let key in custom) {
          if (allowCustom.has(key)) {
            newCustom[key] = custom[key]
          }
        }
      }
      return JSON.stringify(newCustom)
    } catch (e) {
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行customAllowFilter发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
      return JSON.stringify(custom)
    }
  }

  paramsAllowFilter(item: any, verify: boolean): string {
    try {
      if (verify) {
        return JSON.stringify(item.params)
      }
      if (!this.config.enable_logsetting || !this.config.enable_logsetting_params) {
        return JSON.stringify(item.params)
      }
      const { params_block } = this.logSetting.getSettingData()
      const { event, params } = item
      const newParams = {}
      if (params_block && params_block[event] && params_block[event].length) {
        const paramsList = new Set(params_block[event])
        for (let key in params) {
          if (!this.logSetting.getPredefineParamsList().has(key) && paramsList.has(key)) {
            continue
          } else {
            newParams[key] = params[key]
          }
        }
        return JSON.stringify(newParams)
      } else {
        return JSON.stringify(item.params)
      }
    } catch (e) {
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 执行paramsAllowFilter发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
      return JSON.stringify(item.params)
    }
  }

  send(events: any, becon: boolean) {
    if (!events.length) return;
    if (this.config.disable_track_event) return;
    events.forEach(originItem => {
      try {
        let filterItem = JSON.parse(JSON.stringify(originItem))
        if (this.config.filter) {
          filterItem = this.config.filter(filterItem)
          if (!filterItem) {
            console.warn('filter must return data !!')
          }
        }
        if (this.collect.eventFilter && filterItem) {
          filterItem = this.collect.eventFilter(filterItem)
          if (!filterItem) {
            console.warn('filterEvent api must return data !!')
          }
        }
        const reportItem = filterItem || originItem;
        const checkItem = JSON.parse(JSON.stringify(reportItem));
        // this.eventCheck.checkVerify(checkItem);
        if (!reportItem.length) return;
        let canReport = true;
        reportItem.forEach(item => {
          if (!item.events.length) {
            canReport = false
          }
        })
        if (!canReport) return
        this.collect.emit(Types.SubmitBefore, reportItem);
        // const encodeItem = this.config.disable_encryption ? reportItem : `data=${encodeBase64(JSON.stringify(reportItem))}`;
        this.collect.requestManager.useRequest({
          url: this.reportUrl,
          data: reportItem,
          success: (res, data) => {
            if (res && res.e !== 0) {
              this.collect.emit(Types.SubmitError, { type: 'f_data', eventData: data, errorCode: res.e, response: res });
              this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_EVENT, info: '埋点上报失败', time: Date.now(), data: checkItem, code: res.e, msg: res.m || '', failType: '数据异常失败', status: 'fail' })
            } else {
              this.collect.emit(Types.SubmitScuess, { eventData: data, res });
              this.collect.emit(Types.SubmitSuccess, { eventData: data, res });
              this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_EVENT, info: '埋点上报成功', time: Date.now(), data: checkItem, code: 200, status: 'success' })
            }
          },
          fail: (eventData, errorCode) => {
            this.configManager.get('reportErrorCallback')(eventData, errorCode)
            this.collect.emit(Types.SubmitError, { type: 'f_net', eventData, errorCode })
            this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_EVENT, info: '埋点上报网络异常', time: Date.now(), data: checkItem, code: errorCode, failType: '网络异常失败', status: 'fail' })
          },
          timeout: this.timeout,
          useBeacon: becon,
          withCredentials: this.enable_ttwebid,
          zip: true
        })
        this.collect.emit(Types.SubmitAfter, reportItem);
      } catch (e) {
        console.warn(`something error, ${JSON.stringify(e.stack)}`)
        this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK发送埋点发生了异常', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
      }
    })
  }
}
