// byted-tea-sdk@5.3.10/src/plugin/heatmap/index.ts
// 热力图加载组件
import { receiveMsg, dispatchMsg, init, addAllowdOrigin, processMsg } from '../../util/postMessage'
import { decodeUrl, loadScript } from '../../util/tool'
import { IInitParam } from '../../../types/types'
import { HOT_PIC_URL, SDK_VERSION, allowOrigins } from '../../collect/constant'
export default class HeatMap {
  private collector: any
  private config: IInitParam
  private dynamicEditorUrl: string
  private privateMode: boolean = false
  private hasLoad: boolean = false
  apply(collector: any, config: IInitParam) {
    this.collector = collector;
    this.config = config;
    this.dynamicEditorUrl = '';
    this.privateMode = config.channel_domain ? true : false;
    this.hasLoad = false;
    window.TEAVisualEditor = window.TEAVisualEditor || {};
    window.TEAVisualEditor.appId = config.app_id;
    this.dynamicUrl();
    this.init();
  }
  init() {
    addAllowdOrigin(allowOrigins);
    this.addMessage();
    this.reviceMessage();
  }
  reviceMessage() {
    try {
      receiveMsg('tea:openHeatMapCore', (event) => {
        const { app_id } = event.data;
        if (this.hasLoad) return;
        if (app_id && app_id !== this.config.app_id) return;
        let hotUrl = HOT_PIC_URL;
        /**@@PRIVITY
        if (this.privateMode) {
          hotUrl = this.dynamicEditorUrl ? `${this.dynamicEditorUrl}/heatmap-core.2.2.0` : `${hotUrl}.2.2.0`;
        } else {
          hotUrl = `${hotUrl}.2.2.0`;
        }
        @@PRIVITY*/
        this.loadScriptUrl({ event, url: `${hotUrl}.js?query=${Date.now()}` });
        this.hasLoad = true;
      })
    } catch (e) {
      console.log('openHeatMapCore error');
      this.collector.emit('sdk-inner-error', { type: 'sdk_error_load_heatmap', errorCode: 1000, msg: e.message || e.stack || 'sdk_error_load_heatmap' });
    }
  }
  addMessage() {
    window.addEventListener('message', processMsg, false);
  }
  loadScriptUrl({ event, url }) {
    loadScript(url, () => {
      dispatchMsg(event, `${url} ScriptloadSuccess`);
    },
      () => {
        if (event) {
          dispatchMsg(event, `${url} ScriptloadFail`);
        }
      });
  }
  dynamicUrl() {
    // 私有化场景下才动态识别热图地址
    if (this.privateMode) {
      let domain: any = '';
      let scriptSrc: any = '';
      try {
        const resourceList = window.performance.getEntriesByType('resource');
        if (resourceList && resourceList.length) {
          let remoteProtocol = location.protocol;
          remoteProtocol = remoteProtocol || 'https:';
          resourceList.forEach(item => {
            if (item['initiatorType'] === 'script') {
              if (item.name && item.name.indexOf('collect') !== -1) {
                scriptSrc = item.name;
              }
            }
          })
          if (!scriptSrc) {
            if (document.currentScript) {
              scriptSrc = document.currentScript['src'];
            }
          }
          if (scriptSrc) {
            domain = scriptSrc.split('/');
            if (domain && domain.length) {
              this.dynamicEditorUrl = `${remoteProtocol}/`
              for (let i = 2; i < domain.length; i++) {
                if (i === domain.length - 1) break;
                this.dynamicEditorUrl = this.dynamicEditorUrl + `/${domain[i]}`;
              }
              if (this.dynamicEditorUrl && this.dynamicEditorUrl.indexOf('/5.0')) {
                const editorAry = this.dynamicEditorUrl.split('/5.0');
                this.dynamicEditorUrl = editorAry[0] || this.dynamicEditorUrl;
              }
            }
          }
        }
      } catch (e) {
        this.collector.emit('sdk-inner-error', { type: 'sdk_error_parsing_remote_url', errorCode: 1000, msg: e.message || e.stack || 'sdk_error_parsing_remote_url' });
      }
    }
  }
}
