// @ies/starling_client_core@1.0.5/src/error.ts
export class StarlingError extends Error {
  public name = 'StarlingError'
  constructor(public code: number, public message: string, public originError?: Error) {
    super(message)
    this.code = code
    this.message = message
    this.originError = originError
    if ((Error as any).captureStackTrace) (Error as any).captureStackTrace(this, this.constructor)
  }
}

export const MAP_ERROR = {
  CHECK_VERSION_ERROR: {
    ERROR_CODE: 100,
    ERROR_MESSAGE: 'Request for checkVersion failed!'
  },
  LOAD_TEXT_ERROR: {
    ERROR_CODE: 101,
    ERROR_MESSAGE: 'Request for loadText failed!'
  },
  TEST_ERROR: {
    ERROR_CODE: 102,
    ERROR_MESSAGE: 'Request for loadTestText failed!'
  },
  API_PARAMS_ERROR: {
    ERROR_CODE: 103,
    ERROR_MESSAGE: 'Request failed, confirm your apiKey and namespace is correct'
  },
  PARAMS_ERROR: {
    ERROR_CODE: 110,
    ERROR_MESSAGE: 'Check your params, parameter apiKey, namespace and zoneHost is required!'
  },
  PARAMS_ERROR_2: {
    ERROR_CODE: 111,
    ERROR_MESSAGE: 'Check your params, parameter request and store is required!'
  },
  CACHE_ERROR: {
    ERROR_CODE: 112,
    ERROR_MESSAGE: 'Cache error!'
  },
  VERSION_NOT_FOUND: {
    ERROR_CODE: 120,
    ERROR_MESSAGE: 'Version is not found, please check your param [version]!'
  },
  OTHER_ERROR: {
    ERROR_CODE: 200,
    ERROR_MESSAGE: 'Something Wrong!'
  }
}
