// @ies/starling_client_core@1.0.5/src/index.ts
import { createOptions, Options, trackEvent, Texts, IAfterLoadPlugin } from './options'
import { Request } from './Request'
import { LocalCache } from './LocalCache'
import { TextManager, Result } from './TextManager'
import { StarlingError } from './error'

import './polyfill'

const MAP_EVENT_NAME = {
  EVENT_EXCEPTION: 'client_exception', // 异常事件
  EVENT_LOAD: 'client_load', // load 方法调用，开始，成功，失败
  EVENT_CACHE: 'client_cache', // 缓存命中情况
  EVENT_LOAD_TIME: 'client_load_time' // 文案获取时间
}

enum PluginType {
  AfterLoad = 'afterLoad'
}

type Config = Partial<Options>

class Starling {
  public options: Options
  private _textManager: TextManager
  private _tracker: trackEvent
  private afterLoadPlugins: IAfterLoadPlugin[] = []

  constructor(options: Options) {
    this.setConfig(options)
  }

  private _init(options: Options) {
    this.options = createOptions(options)
    const {
      apiKey,
      locale,
      zoneHost,
      request,
      timeout,
      fallbackLang,
      store,
      lazyUpdate,
      test,
      mode,
      format,
      cache,
      cdn
    } = this.options

    // 需要上报的公共参数- mode zone locale apiKey
    const commonParams = {
      mode: lazyUpdate ? 'lazy' : test ? 'test' : 'prod',
      zone: zoneHost,
      cdn,
      locale,
      apiKey
    }

    this._tracker = (name, attrs) => {
      try {
        // 打点失败了不应该影响正常业务
        this.options.tracker(name, { ...commonParams, ...attrs })
      } catch (error) {
        console.log('[starling-client]', error)
      }
    }

    const getReleaseVersion = this._getReleaseVersionByNamespace.bind(this)
    const requester = new Request({
      request,
      timeout,
      apiKey,
      locale,
      zoneHost,
      fallbackLang,
      mode,
      getReleaseVersion,
      cdn
    })
    const localCache = new LocalCache({
      apiKey,
      locale,
      store,
      tracker: this._tracker,
      getReleaseVersion,
      cache,
      cdn
    })

    this._textManager = new TextManager(requester, localCache, { locale, format })

    // init plugin
    this.initPlugin()
  }

  private initPlugin() {
    const plugins = this.options.plugins || []
    if (Array.isArray(plugins) && plugins.every(p => typeof p === 'object'))
      plugins.forEach(plugin => {
        if (plugin.type === PluginType.AfterLoad) {
          this.afterLoadPlugins.push(plugin)
        }
      })
  }

  private _getReleaseVersionByNamespace(ns: string) {
    const namespaces = this.options.namespace as string[]
    const releaseVersion = this.options.version

    let version = null
    if (Array.isArray(releaseVersion)) {
      const index = namespaces.findIndex(v => v === ns)
      version = releaseVersion[index]
    } else if (releaseVersion && typeof releaseVersion === 'object') {
      version = releaseVersion[ns]
    } else if (typeof releaseVersion === 'string') {
      version = releaseVersion
    }

    return version
  }

  /**
   * 用于修改 starling 配置项
   * @param config
   */
  setConfig(config: Config) {
    const options = Object.assign({}, this.options, config)
    this._init(options)
  }

  /**
   * 用于 lazy 模式下判断本地文案更新是否完成。
   *
   * @return {Promise<texts>} 返回本地更新的文案
   */
  async waitForRefresh(): Promise<Texts> {
    const res = await this._textManager.waitForRefresh()
    return res ? res.texts : null
  }

  private async _load() {
    const { namespace, lazyUpdate } = this.options
    const namespaces = namespace as Array<string>
    let result: Result

    if (lazyUpdate) {
      result = await this._textManager.lazyLoad(namespaces)
    } else {
      result = await this._textManager.load(namespaces)
    }
    return result
  }

  /**
   * 从远端按需拉取需要更新的文案，并返回文案。
   *
   * @param {(param: texts) => void} onSuccess 调用成功的回调函数
   * @param {(error: StarlingError)} onError 调用失败的回调函数
   * @return {Promise<texts>} 远端拉取的文案
   */
  async load(
    onSuccess?: (param: Record<string, string>) => void,
    onError?: (error: StarlingError) => void
  ): Promise<Texts> {
    const start = Date.now()
    const namespaces = this.options.namespace as string[]

    try {
      this._tracker(MAP_EVENT_NAME.EVENT_LOAD, { state: 'start', version: '__VERSION__' })

      const result = await this._load()

      const { diffNs, equalNs } = result
      let texts = result.texts

      const afterLoads = this.afterLoadPlugins
      if (Array.isArray(afterLoads) && afterLoads.length) {
        const processedText = await afterLoads.reduce((pre, cur) => {
          return pre.then(text => {
            if (typeof cur.process === 'function') {
              return cur.process(text || texts, this.options)
            }
            /* istanbul ignore next */
            return text
          })
        }, Promise.resolve(texts))
        texts = processedText || texts
      }

      onSuccess && onSuccess(texts)

      if (this.options.debug) {
        console.log('[starling-client] debug mode enabled, final texts:', texts)
      }

      this._tracker(MAP_EVENT_NAME.EVENT_LOAD_TIME, { time: Date.now() - start })
      this._tracker(MAP_EVENT_NAME.EVENT_LOAD, { state: 'success', version: '__VERSION__' })
      // 上报缓存命中率
      const diffLen = diffNs.length
      const equalLen = equalNs.length
      const namespacesLen = namespaces.length
      const ratio = (equalLen / namespacesLen).toFixed(2)
      this._tracker(MAP_EVENT_NAME.EVENT_CACHE, {
        diffLen,
        diff: diffNs.join(','),
        equalLen,
        equal: equalNs.join(','),
        namespaces: equalNs.join(','),
        namespacesLen,
        ratio
      })

      return texts
    } catch (err) {
      console.log('[starling-client]', err)
      this._tracker(MAP_EVENT_NAME.EVENT_LOAD, { state: 'fail' })
      this._tracker(MAP_EVENT_NAME.EVENT_EXCEPTION, {
        code: err.code,
        message: err.message,
        stack: err.stack
      })
      let cacheText: Texts | null = null
      try {
        const { texts } = await this._textManager._getCacheText(namespaces)
        if (texts && typeof texts === 'object') {
          cacheText = texts
          err.response = { data: texts }
        } else {
          /* istanbul ignore next */
          err.response = { data: {} }
        }
      } catch (error) {
        // 获取本地文案也出错了，什么都不做
        console.log('[starling-client] 获取本地文案也出错了，什么都不做')
      }
      // 如果开启了 fallbackToCache，并且本地有文案，就返回本地文案
      if (this.options.fallbackToCache && cacheText) {
        if (this.options.debug) {
          console.log(
            '[starling-client] fallbackToCache enabled, returning cached texts, final cacheText:',
            cacheText
          )
        }
        onSuccess && onSuccess(cacheText)
        return cacheText
      }
      // 如果用户传入了 onError，不应该再抛出一个异常
      // 那样用户还需要再处理失败的 Promise，duck 不必
      if (onError) {
        onError(err)
      } else {
        return Promise.reject(err)
      }
    }
  }
}

export default Starling
