// @ies/starling_client_core@1.0.5/src/LocalCache.ts
/**
 * @author fengweiyao
 * 模块负责与本地缓存进行交互，对外提供和 Request 相同的 api。
 */
import { Options, Texts, trackEvent } from './options'
import { MAP_ERROR } from './error'

export type CacheOptions = Required<
  Pick<Options, 'apiKey' | 'locale' | 'store' | 'cache' | 'cdn'>
> & {
  tracker?: trackEvent
  getReleaseVersion: (ns: string) => string
}
export class LocalCache {
  constructor(public options: CacheOptions) {
    this.options = options
  }

  private _setCache(key: string, val: any, errCb?: () => void): any {
    // if (this.options.cdn) {
    //   return
    // }
    try {
      return this.options.store.set(key, val)
    } catch (err) {
      const { ERROR_CODE } = MAP_ERROR.CACHE_ERROR
      this.options.tracker('cache_error', {
        code: ERROR_CODE,
        message: err.message,
        stack: err.stack
      })
      if (errCb) {
        errCb()
      }
      return
    }
  }

  private _getCache(key: string): any {
    // console.log('_getCache,key=',key)
    const { store, cache } = this.options
    try {
      // if (cdn) {
      //   // 使用 cdn 方案后保留缓存
      //   return
      // }
      return cache && store.get(key)
    } catch (err) {
      const { ERROR_CODE } = MAP_ERROR.CACHE_ERROR
      this.options.tracker('cache_error', {
        code: ERROR_CODE,
        message: err.message,
        stack: err.stack
      })
      return
    }
  }

  getVersion(namespace: string) {
    const { apiKey, locale, getReleaseVersion, cdn } = this.options

    let key = `version.${apiKey}.${namespace}.${locale}`
    const releaseVersion = getReleaseVersion(namespace)
    if (releaseVersion) {
      key += `.${releaseVersion}`
    }
    const text = this.getText(namespace)

    // 如果本地缓存被不小心清掉了，版本也不单独返回
    if (!text || typeof text !== 'object') {
      // cdn 场景会清除缓存，同时需要清除版本
      if (cdn && this.options.cdn) {
        this._getCache(key)
      }
      return null
    }
    // 使用localforge情况下， text会是promise，也需要返回版本号
    if (typeof text === 'object' && text.then) {
      return this._getCache(key)
    }

    // text 是空对象时也不返回版本
    if (typeof text === 'object' && Object.keys(text || {}).length === 0) {
      return null
    }

    return this._getCache(key)
  }

  getVersions(namespaces: Array<string>): Promise<Array<string>> {
    return Promise.all(namespaces.map(ns => this.getVersion(ns)))
  }

  getText(namespace: string) {
    const { apiKey, locale, getReleaseVersion } = this.options

    let key = `text.${apiKey}.${namespace}.${locale}`
    const releaseVersion = getReleaseVersion(namespace)
    if (releaseVersion) {
      key += `.${releaseVersion}`
    }
    return this._getCache(key)
  }

  // 允许 store 返回值为 promise，支持异步与缓存的交互
  getTexts(namespaces: Array<string>): Promise<Array<Texts>> {
    return Promise.all(namespaces.map(ns => this.getText(ns)).filter(Boolean))
  }

  // 更新指定的空间缓存
  async refresh(namespaces: string[], textRes: Array<Texts>, newVers: string[], langArr: string[]) {
    const { apiKey, getReleaseVersion, cache } = this.options
    const promises = []

    textRes.forEach((text, i) => {
      let textKey = `text.${apiKey}.${namespaces[i]}.${langArr[i]}`
      let versionKey = `version.${apiKey}.${namespaces[i]}.${langArr[i]}`

      const releaseVersion = getReleaseVersion(namespaces[i])
      if (releaseVersion) {
        textKey += `.${releaseVersion}`
        versionKey += `.${releaseVersion}`
      }

      if (cache) {
        promises.push(
          this._setCache(textKey, textRes[i], () => {
            this.options.store.set(textKey, {})
          })
        )
        promises.push(this._setCache(versionKey, newVers[i]))
      }
    })

    // 等待缓存更新完成
    await Promise.all(promises)

    const text = textRes.reduce((acc, cur) => {
      return Object.assign(acc, cur)
    }, {})

    return text
  }
}
