// @ies/starling_client_core@1.0.5/src/options.ts
import { StarlingError, MAP_ERROR } from './error'

export interface StoreAPI {
  get(key: string, optionalDefaultValue?: any): any
  set(key: string, value: any): any
  del(key: string): any
}

type RequireAtLeastOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyof T, Keys>> &
  {
    [K in Keys]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<Keys, K>>>
  }[Keys]

type Attrs = {
  [key: string]: any
}

export type trackEvent = (name: string, attrs: Attrs) => void

export type Mode = 'normal' | 'test' | 'gray'
export type FormatType = 'merged' | 'splitted'
export type RequestFn = (url: string, options: { timeout: number }) => Promise<Response>

export interface IAfterLoadPlugin {
  type: string
  process(text: Texts, options: Options): Promise<Texts>
}

interface CommonOptions {
  namespace: string | Array<string>
  // namespaces?: Array<string>; // TODO 处理参数
  api_key: string
  apiKey: string
  // 当支持 cdn 时，store 为可选，不需要本地缓存
  store?: StoreAPI
  request: RequestFn
  /**
   * 指定文案拉取的环境
   * normal 线上环境，test 测试环境，gray 灰度环境
   */
  mode?: Mode
  /**
   * 是否启用测试环境
   */
  test?: boolean
  /**
   * 最终返回文案的格式
   * merged 所有空间的文案全部合并到一个 map
   * splitted 每个空间的文案单独返回
   */
  format?: FormatType
  locale?: string
  timeout?: number
  lazyUpdate?: boolean
  fallbackLang?: Array<string>
  tracker?: trackEvent
  /**
   * 自定义发布版本
   * 有则根据对应版本拉取文案
   * 无则拉取最新版本的文案
   * 重：需要和namespace一一对应
   *
   * 支持格式
   * 1.字符串 1.0.0，namespace也为字符串
   * 2.数组 ['1.0.0']，namespace也为数组并一一对应
   * 3.对象 { name: '1.0.0' }，key为空间名称，设计这个格式的目的：担心用户有的空间走固定版本有的默认最新版本，这样设置就不需要管namespace顺序问题
   */
  version?: string | string[] | { [key: string]: string }
  plugins?: Array<IAfterLoadPlugin>
  cache?: boolean
  /** 是否开启debug模式，如果开启会在控制台打印所有的文案 */
  debug?: boolean
  /** 接口请求失败时是否返回缓存文案而非抛错，默认 false */
  fallbackToCache?: boolean
}

// 当 cdn 存在时 zoneHost 为可选
interface BaseOptionsWithCDN extends CommonOptions {
  cdn: string
  zoneHost?: string // 可选
}

// 当 cdn 不存在时 zoneHost 为必填
interface BaseOptionsWithoutCDN extends CommonOptions {
  cdn?: undefined // 确保 cdn 不存在
  /**
   * 指定域名，地区对应域名可查看:
   * https://bytedance.feishu.cn/wiki/wikcnrILBFaGcjbYQkfAFPljqsg#9CAlE8
   * 当支持 cdn 时，zoneHost 为可选，同时存在时优先走cdn 获取逻辑
   */
  zoneHost: string // 必填
}

// 合并以上两种情况
type BaseOptions = BaseOptionsWithCDN | BaseOptionsWithoutCDN

export type Options = RequireAtLeastOne<BaseOptions, 'api_key' | 'apiKey'>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Response = { data: { message: any; status: number } }

// 这几个别名的声明不知道放在哪里合适，暂时先放这里。
export type Version = string
export type Namspae = string
export type Texts = {
  [k: string]: string
}

const defaultOptions = {
  zone: 'CN',
  locale: 'zh',
  timeout: 15000,
  test: false,
  mode: 'normal',
  format: 'merged',
  lazyUpdate: false,
  fallbackLang: ['zh'],
  tracker: () => {},
  cache: true,
  cdn: '',
  debug: false,
  fallbackToCache: false
}

export function createOptions(options: Options) {
  options = Object.assign({}, defaultOptions, options)
  let { namespace, apiKey } = options
  const { request, store, zoneHost, test, cdn } = options

  if (test) options.mode = 'test'

  options.namespace = namespace = namespace
    ? Array.isArray(namespace)
      ? namespace
      : [namespace]
    : []
  options.apiKey = apiKey = apiKey || options.api_key

  if (!request || (!cdn && !store)) {
    const { ERROR_CODE, ERROR_MESSAGE } = MAP_ERROR.PARAMS_ERROR_2
    throw new StarlingError(ERROR_CODE, ERROR_MESSAGE, null)
  }

  if (!apiKey || !namespace.length || (!cdn && !zoneHost)) {
    const { ERROR_CODE, ERROR_MESSAGE } = MAP_ERROR.PARAMS_ERROR
    throw new StarlingError(ERROR_CODE, ERROR_MESSAGE, null)
  }

  return options
}
