// @ies/starling_client_core@1.0.5/src/Request.ts
/**
 * @author fengweiyao
 * 模块负责与远端进行交互，与 localCache 提供相同的对外 api。
 */
import { Options, Texts } from './options'
import { StarlingError, MAP_ERROR } from './error'

export type RequestOptions = Required<
  Pick<
    Options,
    'request' | 'timeout' | 'apiKey' | 'locale' | 'fallbackLang' | 'zoneHost' | 'cdn' | 'mode'
  >
> & { getReleaseVersion: (ns: string) => string }

interface TextResponse {
  updated: boolean
  data?: Texts
  version?: string
  lang?: string
}

export class Request {
  constructor(public options: RequestOptions) {
    this.options = options
  }

  async getText(namespace: string, localVersion: string): Promise<TextResponse> {
    const res = await this._request(namespace, localVersion)
    return res as TextResponse
  }

  getTexts(namespaces: Array<[string, string]>): Promise<Array<TextResponse>> {
    return Promise.all(namespaces.map(([ns, version]) => this.getText(ns, version)))
  }

  private async _request(namespace: string, localVersion: string): Promise<TextResponse> {
    /**
     * 根据版本获取文案不存在发布后文案还更新的情况
     * 所以只要本地存在对应版本文案就不再远程拉取
     * 本地构建一个假的成功请求欺骗一下上层逻辑
     */
    const { getReleaseVersion, cdn, fallbackLang, locale } = this.options
    const releaseVersion = getReleaseVersion(namespace)
    if (namespace && releaseVersion && localVersion) {
      return {} as TextResponse
    }

    const url = this.getUrl(namespace, localVersion)
    const { request, timeout } = this.options

    let res
    if (cdn) {
      // cdn 兜底逻辑由服务端转前端
      const localesToTry = [locale, ...fallbackLang]
      const errors = [] // 用于记录每次请求的错误
      const { ERROR_CODE, ERROR_MESSAGE } = MAP_ERROR.LOAD_TEXT_ERROR
      for (const loc of localesToTry) {
        try {
          const res = await request(this.getUrl(namespace, localVersion, loc), { timeout })
          const {
            data: { status, message }
          } = res
          if (+status === 200) return this._mapResponse(message) as TextResponse
          throw new StarlingError(ERROR_CODE, ERROR_MESSAGE)
        } catch (error) {
          errors.push(error)
          // 如果全部失败，则向上抛错
          if (errors.length === localesToTry.length) {
            return this._mapResponse({ data: {}, updated: false }) as TextResponse
          }
        }
      }
    } else {
      res = await request(url, { timeout })
    }

    const {
      data: { status, message }
    } = res
    // 200 表示成功，1000 表示参数异常
    if (+status === 200) return this._mapResponse(message) as TextResponse

    // 1001 releaseVersion 不存在
    if (+status === 1001) {
      const { ERROR_CODE, ERROR_MESSAGE } = MAP_ERROR.VERSION_NOT_FOUND
      throw new StarlingError(ERROR_CODE, ERROR_MESSAGE)
    }

    const { ERROR_CODE, ERROR_MESSAGE } = MAP_ERROR.API_PARAMS_ERROR
    throw new StarlingError(ERROR_CODE, ERROR_MESSAGE)
  }

  private getUrl(namespace: string, localVersion: string, fallbackLocale?: string): string {
    const { apiKey, locale, fallbackLang, zoneHost, mode, getReleaseVersion, cdn } = this.options
    const langs = [locale].concat(fallbackLang)

    let url = `${zoneHost}/check_and_get_text/${apiKey}/${mode}/${namespace}?lang=${langs}`

    const releaseVersion = getReleaseVersion(namespace)
    if (releaseVersion) {
      url = `${zoneHost}/text_by_release_version/${apiKey}/${mode}/${namespace}/${releaseVersion}?lang=${langs}`
    }

    if (
      typeof localVersion === 'string' &&
      (localVersion.includes('%22') || localVersion.includes('"'))
    ) {
      // 解码本地版本号，避免中文乱码，因为发现本地版本号可能会携带%22xxxx%22这种格式，所以需要先解码。
      let paramsLocalVersion = decodeURIComponent(localVersion)

      // 提取字符串中完全由数字组成的子串。
      paramsLocalVersion = paramsLocalVersion.match(/\d+/)?.[0] || ''

      if (paramsLocalVersion) url += `&version=${paramsLocalVersion}`
    } else {
      if (typeof localVersion === 'string') url += `&version=${localVersion}`
    }

    // cdn 方式不支持version 获取，统一改用 cdn 拉取
    if (cdn) {
      url = `${cdn}/${apiKey}/${mode}/${namespace}/${fallbackLocale || locale}.json`
    }

    return url
  }

  private _mapResponse(res: any) {
    if (!res) return {}
    return Object.keys(res).reduce((acc, key) => {
      const newKey = key.toLowerCase()
      acc[newKey] = res[key]
      return acc
    }, {})
  }
}
