// @ies/starling_client_core@1.0.5/src/TextManager.ts
/**
 * @author fengweiyao
 * @核心模块，负责计算需要更新的空间，并返回最新的文案，更新本地缓存。
 * 核心的步骤：
 * 1. 获取本地所有空间缓存的 version ，请求接口获取更新的文案
 * 2. 针对没有更新的空间，从本地获取文案
 * 3. 按照 format 对应格式，返回文案。
 *
 * 之所以单独抽离这个模块，是为了方便单元测试。
 */
import { Request } from './Request'
import { LocalCache } from './LocalCache'
import { Texts, Namspae, Options, FormatType } from './options'

function mergeTexts(arr1: Texts[], arr2?: Texts): Texts {
  const arr = arr2 ? arr1.concat(arr2) : arr1
  return arr.reduce((acc, cur) => {
    return Object.assign(acc, cur)
  }, {})
}

export type Result = {
  texts: Texts
  diffNs: Array<Namspae>
  equalNs: Array<Namspae>
}

export interface Namespace {
  namespace: string
  prior: number
  version?: string
  texts?: Texts
  lang?: string
}

export class NamespaceGroup {
  private _list: Array<Namespace> = []
  constructor(list: Array<Namespace> = []) {
    this._list = list
  }

  get length() {
    return this._list.length
  }
  getList() {
    return this._list
  }
  getNamespaces() {
    return this._list.map(item => item.namespace)
  }
  addNamespace(ns: Namespace) {
    this._list.push(ns)
  }
  insertTexts(textsArr: Array<Texts>) {
    this._list.forEach((ns, i) => {
      ns.texts = textsArr[i]
    })
  }
  sortByPrior() {
    this._list.sort((a, b) => a.prior - b.prior)
    return this._list
  }
  concat(nsg: NamespaceGroup) {
    return new NamespaceGroup(this._list.concat(nsg.getList()))
  }
}

type LoadResult = {
  texts: Texts
  diffNs: Array<string>
  equalNs: Array<string>
}

type TextManagerOptions = Required<Pick<Options, 'locale' | 'format'>>

export class TextManager {
  public remote: Request
  public local: LocalCache
  public options: TextManagerOptions
  private _promiseRefresh: Promise<LoadResult> = Promise.resolve(null)
  constructor(request: Request, localCache: LocalCache, options: TextManagerOptions) {
    this.remote = request
    this.local = localCache
    this.options = options
  }

  private async _checkVersions(
    namespaces: string[]
  ): Promise<{
    equalNs: NamespaceGroup
    diffNs: NamespaceGroup
  }> {
    const { locale } = this.options
    const localVersions = await this.local.getVersions(namespaces)
    const namespaceWithVersion = namespaces.map((ns, i) => [ns, localVersions[i]]) as Array<
      [string, string]
    >
    const resArr = await this.remote.getTexts(namespaceWithVersion)
    const equalNs = new NamespaceGroup()
    const diffNs = new NamespaceGroup()
    namespaces.forEach((ns, i) => {
      const { updated, version, data, lang } = resArr[i]
      if (!updated) {
        equalNs.addNamespace({ namespace: ns, prior: i, version: localVersions[i], lang: locale })
      } else {
        diffNs.addNamespace({ namespace: ns, prior: i, version, texts: data, lang })
      }
    })
    return { equalNs, diffNs }
  }

  waitForRefresh() {
    return this._promiseRefresh
  }

  _refresh(diffNs: NamespaceGroup) {
    const diffList = diffNs.getList()
    const nsArr = []
    const versionArr = []
    const txtArr = []
    const langArr = []
    diffList.forEach(i => {
      nsArr.push(i.namespace)
      versionArr.push(i.version)
      txtArr.push(i.texts)
      langArr.push(i.lang)
    })
    return this.local.refresh(nsArr, txtArr, versionArr, langArr)
  }

  async _getLocalTexts(nsg: NamespaceGroup) {
    const texts = await this.local.getTexts(nsg.getNamespaces())
    nsg.insertTexts(texts)
    return texts
  }

  async load(namespaces: string[]) {
    const { format } = this.options
    const { equalNs, diffNs } = await this._checkVersions(namespaces)
    equalNs.length ? await this._getLocalTexts(equalNs) : []

    const texts = this._formatTexts(format, diffNs.concat(equalNs))

    diffNs.length && (await this._refresh(diffNs))

    return {
      texts,
      diffNs: diffNs.getNamespaces(),
      equalNs: equalNs.getNamespaces()
    }
  }

  /**
   * 获取本地缓存文案
   * @param namespaces 空间列表
   */
  async _getCacheText(namespaces: string[]): Promise<{ texts: Texts; cacheAllLoaded: boolean }> {
    const { locale, format } = this.options
    const txtArr = await this.local.getTexts(namespaces)
    const versionArr = await this.local.getVersions(namespaces)
    const nsgList = txtArr
      .map((txt, i) => {
        if (!txt) return
        return {
          namespace: namespaces[i],
          version: versionArr[i],
          lang: locale,
          texts: txt
        } as Namespace
      })
      .filter(i => !!i)
    const nsg = new NamespaceGroup(nsgList)
    return {
      texts: this._formatTexts(format, nsg),
      cacheAllLoaded: nsgList.length === txtArr.length && txtArr.length === namespaces.length
    }
  }

  async lazyLoad(namespaces: string[]) {
    // 先获取本地文案
    // eslint-disable-next-line prefer-const
    let { texts, cacheAllLoaded } = await this._getCacheText(namespaces)

    const isLocalEmpty = Object.keys(texts).length === 0 || !cacheAllLoaded

    this._promiseRefresh = this.load(namespaces)
    let res: LoadResult
    // 本地文案为空
    if (isLocalEmpty) {
      // 等待获取远程文案
      res = await this._promiseRefresh
      texts = res.texts
    }

    return {
      texts,
      diffNs: isLocalEmpty ? res && res.diffNs : [],
      equalNs: isLocalEmpty ? res && res.equalNs : namespaces
    }
  }

  private _formatTexts(type: FormatType, nsg: NamespaceGroup) {
    if (type === 'splitted') {
      const list = nsg.getList()
      return list.reduce((acc, cur) => {
        const { namespace, lang, texts, version } = cur
        acc[namespace] = { lang, texts, version }
        return acc
      }, {})
    } else {
      // 保证文案合并的顺序，后面的空间会覆盖前面空间的文案。
      const textsArr = nsg.sortByPrior().map(i => i.texts)
      return mergeTexts(textsArr)
    }
  }
}
