// @rdservices/inject-runtime@0.0.17/dist/adapters/react/react-adapter.js
import { getFiberLabel } from "../../utils/get-fiber-label.js";
import { isStyledElement } from "../../utils/is-styled.js";
import { findDebugSource } from "./find-debug-source.js";
import { findFiberByHtmlElement } from "./find-fiber-by-element.js";
import { getFiberOwnBoundingBox } from "./get-fiber-own-bounding-box.js";
import { getComponentBox } from "./get-component-box.js";
function getElementInfo(found) {
    const fiber = findFiberByHtmlElement(found, false);
    if (fiber) {
        const { componentBox } = getComponentBox(fiber);
        const thisLabel = getFiberLabel(fiber, findDebugSource(fiber)?.source);
        if ([
            'Not found',
            'Anonymous'
        ].includes(thisLabel.label)) return null;
        if (isStyledElement(fiber)) thisLabel.label = `${thisLabel.label}(styled)`;
        thisLabel.label = `${thisLabel.label} - ${found.tagName.toLocaleLowerCase()}`;
        return {
            thisElement: {
                box: getFiberOwnBoundingBox(fiber) || found.getBoundingClientRect(),
                ...thisLabel
            },
            htmlElement: found,
            componentBox
        };
    }
    return null;
}
function fiberToPathItem(fiber) {
    const label = getFiberLabel(fiber, findDebugSource(fiber)?.source);
    return {
        title: label.label,
        link: label.link
    };
}
function getParentsPaths(element) {
    const fiber = findFiberByHtmlElement(element, false);
    if (fiber) {
        const pathItems = [];
        let currentFiber = fiber;
        pathItems.push(fiberToPathItem(currentFiber));
        while(currentFiber._debugOwner){
            currentFiber = currentFiber._debugOwner;
            pathItems.push(fiberToPathItem(currentFiber));
        }
        return pathItems;
    }
    return [];
}
const reactAdapter = {
    getElementInfo,
    getParentsPaths
};
export { getElementInfo, reactAdapter };
