// byted-tea-sdk@5.3.10/src/plugin/error/error.ts
import { SDK_TYPE, NOOP } from '../../collect/constant'
export default class Monitor {
  sdkReady: boolean
  config: any
  collect: any
  url: string
  apply(collect: any, config: any) {
    if (config.channel_domain || config.disable_track_event || config.disable_sdk_monitor) return;
    this.collect = collect
    this.config = config
    this.url = collect.configManager.getUrl('event')
    this.collect.on(collect.Types.Ready, () => {
      this.sdkOnload()
    })
    this.collect.on(collect.Types.SubmitError, ({ type, eventData, errorCode, response }) => {
      if (type !== 'f_data') return;
      this.sdkError(eventData, errorCode, response)
    })
    this.collect.on(collect.Types.SDKInnerError, ({ type, eventData, errorCode, msg }) => {
      this.sdkInnerError(type, eventData, errorCode, msg)
    })
  }
  sdkOnload() {
    try {
      const { header, user } = this.collect.configManager.get()
      const { app_id, app_name, sdk_version } = header
      const { web_id } = user
      const event = {
        event: 'onload',
        params: JSON.stringify({
          app_id,
          app_name: app_name || '',
          sdk_version,
          sdk_type: SDK_TYPE,
          sdk_config: this.config,
          sdk_desc: 'TOC',
          url: location.href
        }),
        local_time_ms: Date.now(),
      }
      const loadData = {
        events: [event],
        user: {
          user_unique_id: web_id
        },
        header: {},
      }
      setTimeout(() => {
        this.collect.requestManager.useXhr({
          url: this.url,
          data: [loadData],
          timeout: 30000,
          app_key: '566f58151b0ed37e',
        })
      }, 16)
    } catch (e) {
    }
  }
  sdkError(data, code, response) {
    try {
      const { user, header } = data[0]
      const flatEvents = []
      data.forEach((item) => {
        item.events.forEach((event) => {
          flatEvents.push(event)
        })
      })
      const errEvents = flatEvents.map(event => ({
        event: 'on_error',
        params: JSON.stringify({
          type: 'sdk_event_error',
          error_code: code,
          app_id: header.app_id,
          app_name: header.app_name || '',
          error_event: event.event,
          sdk_version: header.sdk_version,
          local_time_ms: event.local_time_ms,
          tea_event_index: Date.now(),
          params: event.params,
          header: header,
          user: user,
          err_msg: response.m || ''
        }),
        local_time_ms: Date.now(),
      }))
      const errData = {
        events: errEvents,
        user: {
          user_unique_id: user.user_unique_id,
        },
        header: {
        },
      }
      setTimeout(() => {
        this.collect.requestManager.useXhr({
          url: this.url,
          data: [errData],
          timeout: 30000,
          app_key: '566f58151b0ed37e',
        })
      }, 16)
    } catch (e) {
    }
  }
  sdkInnerError(type, data, code, msg) {
    try {
      const { user, header } = data[0]
      const flatEvents = []
      data.forEach((item) => {
        item.events.forEach((event) => {
          flatEvents.push(event)
        })
      })
      const errEvents = flatEvents.map(event => ({
        event: 'on_error',
        params: JSON.stringify({
          type,
          error_code: code,
          app_id: header.app_id,
          app_name: header.app_name || '',
          error_event: event.event,
          sdk_version: header.sdk_version,
          local_time_ms: event.local_time_ms,
          tea_event_index: Date.now(),
          params: event.params,
          header: header,
          user: user,
          ua: window.navigator.userAgent,
          err_msg: msg
        }),
        local_time_ms: Date.now(),
      }))
      const errData = {
        events: errEvents,
        user,
        header: {
        },
      }
      setTimeout(() => {
        this.collect.requestManager.useXhr({
          url: this.url,
          data: [errData],
          timeout: 30000,
          app_key: '566f58151b0ed37e',
        })
      }, 16)
    } catch (e) {
    }
  }
}
