// @rdservices/inject-runtime@0.0.17/dist/adapters/vue/vue-adapter.js
import { getVueComponentBoundingBox } from "./get-vnode-bounding-box.js";
function getElementInfo(found) {
    const parentComponent = found.__vueParentComponent;
    if (parentComponent) {
        if (!parentComponent.type) return null;
        const componentBBox = getVueComponentBoundingBox(parentComponent);
        const file = parentComponent.type.file || parentComponent.type.__file;
        const name = parentComponent.type.name || parentComponent.type.__name;
        if (file && name) return {
            thisElement: {
                box: found.getBoundingClientRect(),
                label: `${name} - ${found.tagName.toLocaleLowerCase()}`,
                link: {
                    column: 0,
                    line: 0,
                    filePath: file
                }
            },
            htmlElement: found,
            componentBox: componentBBox || found.getBoundingClientRect()
        };
    }
    return null;
}
function getParentsPaths(element) {
    const path = [];
    let currentElement = element;
    let previousComponentKey = null;
    do {
        if (currentElement) {
            const info = getElementInfo(currentElement);
            const currentComponentKey = JSON.stringify({
                label: info?.thisElement.label,
                ...info?.thisElement.link
            });
            if (info && currentComponentKey !== previousComponentKey) {
                previousComponentKey = currentComponentKey;
                const { link } = info.thisElement;
                const { label } = info.thisElement;
                if (link) path.push({
                    title: label,
                    link
                });
            }
        }
        currentElement = currentElement.parentElement;
    }while (currentElement);
    return path;
}
const vueAdapter = {
    getElementInfo,
    getParentsPaths
};
const vue_adapter = vueAdapter;
export { vue_adapter as default, getElementInfo };
