// @ies/starling_client_core@1.0.5/rollup.config.js
import typescript from 'rollup-plugin-typescript2'
import resolve from '@rollup/plugin-node-resolve'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import path from 'path'
import fs from 'fs'
import pkg from './package.json'
import replace from '@rollup/plugin-replace'
import { version } from './package.json'

function deleteall(path) {
  let files = []
  if (fs.existsSync(path)) {
    files = fs.readdirSync(path)
    files.forEach(function (file) {
      const curPath = path + '/' + file
      if (fs.statSync(curPath).isDirectory()) {
        deleteall(curPath)
      } else {
        fs.unlinkSync(curPath)
      }
    })
    fs.rmdirSync(path)
  }
}

deleteall(path.resolve(__dirname, './lib'))

const input = './src/index.ts'

export default [
  {
    input,
    external: ['axios', 'store'],
    output: [
      {
        file: pkg.module,
        format: 'es'
      },
      {
        file: pkg.main,
        format: 'cjs'
      }
    ],
    plugins: [
      replace({
        __VERSION__: version
      }),
      json(),
      resolve({
        browser: true
      }),
      typescript(),
      commonjs()
    ]
  }
]
