// byted-tea-sdk@5.3.10/src/util/storage.ts
import Cookies from './cookie';

const getCookie = (name?: string, domain?: string) => {
  try {
    let _matches
    if (name) {
      _matches = Cookies.get(name)
    } else {
      _matches = Cookies.get()
    }
    return _matches
  } catch (e) {
    return ''
  }
}

const setCookie = (name: string, value: any, expiresTime?: number, domain?: string) => {
  try {
    const _domain = domain || document.domain
    const timestamp = +new Date();
    const expires = expiresTime || 7 * 1000 * 60 * 60 * 24
    const furureTimestamp = timestamp + expires;
    Cookies.set(name, value, {
      expires: new Date(furureTimestamp),
      path: '/',
      domain: _domain
    })
  } catch (e) {
  }
}
class Memory {
  cache: any
  constructor() {
    this.cache = {}
  }
  setItem(cacheKey: string, data: any) {
    this.cache[cacheKey] = data
  }
  getItem(cacheKey: string) {
    return this.cache[cacheKey]
  }
  removeItem(cacheKey: string) {
    this.cache[cacheKey] = undefined
  }
  getCookie(name?: string, domain?: string) {
    return getCookie(name, domain)
  }
  setCookie(name: string, value: any, expiresTime?: number, domain?: string) {
    setCookie(name, value, expiresTime, domain)
  }
}

function isSupportLS() {
  try {
    localStorage.setItem('_ranger-test-key', 'hi')
    localStorage.getItem('_ranger-test-key')
    localStorage.removeItem('_ranger-test-key')
    return true
  } catch (e) {
    return false
  }
}
function isSupportSession() {
  try {
    sessionStorage.setItem('_ranger-test-key', 'hi')
    sessionStorage.getItem('_ranger-test-key')
    sessionStorage.removeItem('_ranger-test-key')
    return true
  } catch (e) {
    return false
  }
}

const local = {
  getItem(key: string) {
    try {
      var value = localStorage.getItem(key)
      let ret = value
      try {
        if (value && typeof value === 'string') {
          ret = JSON.parse(value)
        }
      } catch (e) { }
      return ret || {}
    } catch (e) { }
    return {}
  },
  setItem(key: string, value: any) {
    try {
      var stringValue = typeof value === 'string' ? value : JSON.stringify(value)
      localStorage.setItem(key, stringValue)
    } catch (e) { }
  },
  removeItem(key: string) {
    try {
      localStorage.removeItem(key)
    } catch (e) { }
  },
  getCookie(name?: string, domain?: string) {
    return getCookie(name, domain)
  },
  setCookie(name: string, value: any, expiresTime?: number, domain?: string) {
    setCookie(name, value, expiresTime, domain)
  },
  isSupportLS: isSupportLS(),
}
const session = {
  getItem(key: string) {
    try {
      var value = sessionStorage.getItem(key)
      let ret = value
      try {
        if (value && typeof value === 'string') {
          ret = JSON.parse(value)
        }
      } catch (e) { }

      return ret || {}
    } catch (e) { }
    return {}
  },
  setItem(key: string, value: any) {
    try {
      var stringValue = typeof value === 'string' ? value : JSON.stringify(value)
      sessionStorage.setItem(key, stringValue)
    } catch (e) { }
  },
  removeItem(key: string) {
    try {
      sessionStorage.removeItem(key)
    } catch (e) { }
  },
  getCookie(name?: string, domain?: string) {
    return getCookie(name, domain)
  },
  setCookie(name: string, value: any, expiresTime?: number, domain?: string) {
    setCookie(name, value, expiresTime, domain)
  },
  isSupportSession: isSupportSession(),
}

export default class Storage {
  _storage: any
  constructor(isCache, type?: string) {
    if (type && type === 'session') {
      this._storage = session
    } else {
      this._storage = !isCache && local.isSupportLS ? local : new Memory()
    }
  }
  getItem(key: string) {
    return this._storage.getItem(key)
  }
  setItem(key: string, value: any) {
    this._storage.setItem(key, value)
  }
  getCookie(name?: string, domain?: string) {
    return this._storage.getCookie(name, domain)
  }
  setCookie(key: string, value: any, expiresTime?: number, domain?: string) {
    this._storage.setCookie(key, value, expiresTime, domain)
  }
  removeItem(key: string) {
    this._storage.removeItem(key)
  }
}
