// byted-tea-sdk@5.3.10/src/plugin/visual/catch.ts
// 拦截页面上的xhr或者fetch请求
import { parseUrlQuery, decodeUrl } from '../../util/tool';

const filterUrl = ['1fz22z22z1nz21z4mz4bz4bz1kz1az21z4az28z1gz1hz1gz1cz18z1nz1gz4az1az1mz1k',
  '1fz22z22z1nz21z4mz4bz4bz1kz1az21z4az24z1mz1jz1az1cz18z1nz1nz1jz1mz1ez4az1az1mz1k',
  '1fz22z22z1nz21z4mz4bz4bz1kz1az21z4az1az22z1mz19z21z1lz21z21z1bz1iz4az1az1mz1k']
export default class InterceptorRequest {
  url: string
  _startTime: number
  catchFetch(cb) {
    const { fetch: originalFetch } = window;
    let isFilter = false
    window.fetch = async (...args) => {
      let [resource, config] = args;
      let resourceUrl
      try {
        if (typeof resource === 'string') {
          resourceUrl = resource
        } else if (resource instanceof Request) {
          resourceUrl = resource.url
        }
        filterUrl.forEach(url => {
          if (resourceUrl.includes(decodeUrl(url))) {
            isFilter = true
          }
        })
      } catch (e) {
        console.error(e);
      }
      const response = await originalFetch(resource, config);
      try {
        const json = () =>
          response
            .clone()
            .json()
            .then((data) => {
              try {
                if (!isFilter) {
                  const path = new URL(resourceUrl, window.location.origin)
                  cb({
                    url_path: path && path.pathname || resourceUrl,
                    params: config && config.method === 'POST' ? config.body && JSON.parse(config.body.toString()) : {},
                    response: data,
                    method: config && config.method || 'GET',
                    query: config && config.method === 'GET' ? parseUrlQuery(resourceUrl) : {}
                  })
                }
              } finally {
                return data
              }
            });
        response.json = json;
        return response;
      } finally {
        return response;
      }
    };
  }
  catchAjax(cb: any) {
    const xhr = XMLHttpRequest

    if (!xhr) return


    let protocol = window.location.protocol

    if (protocol === 'file:') return

    let _open = xhr.prototype.open
    let _send = xhr.prototype.send

    xhr.prototype.open = function (method, url) {
      this.url = url
      this._startTime = +new Date()
      return _open.apply(this, arguments)
    }

    xhr.prototype.send = function () {

      const ADD_EVENT_LISTENER = 'addEventListener'
      const ON_READY_STATE_CHANGE = 'onreadystatechange'
      const _dispatchEvent = (event) => {
        if (event) {
          const duration = +new Date() - this._startTime
          event.duration = duration
          let isFilter = false
          //  如果是sdk的接口本身的xhr不再进行拦截
          filterUrl.forEach(url => {
            const _url = decodeUrl(url)
            if (this.url && typeof this.url === 'string' && this.url.includes(_url)) {
              isFilter = true
            }
          })
          if (!isFilter) {
            let response;
            try {
              if (this.getAllResponseHeaders('content-type').indexOf('application/json') !== -1) {
                const responseType = event.currentTarget.responseType;
                response = event.currentTarget.response
                if (response) {
                  if (responseType !== 'json') {
                    response = JSON.parse(response)
                  }
                }
              }
              if (event.currentTarget.status === 200) {
                this.success && this.success(response)
                let params = event.currentTarget._data
                if (params instanceof FormData) return
                // 区分请求 Content_type： application/x-www-form-urlencoded or multipart/form-data
                const contentType = event.currentTarget._reqHeaders && event.currentTarget._reqHeaders['Content-Type'];
                params = contentType === 'application/x-www-form-urlencoded'
                  ? params
                  : (params && JSON.parse(params))
                const method = event.currentTarget._method
                const path = new URL(this.url, window.location.origin)
                cb({
                  url_path: path.pathname || path,
                  params,
                  response,
                  method: method || 'GET',
                  query: method && method === 'GET' ? parseUrlQuery(this.url) : {}
                })
              } else {
                this.fail && this.fail(response)
              }
            } catch (e) {
              console.log('catch xhr error')
            }
          }
        }
      }

      if (ADD_EVENT_LISTENER in this) {
        // xhr Level 2
        this[ADD_EVENT_LISTENER]('load', _dispatchEvent)
        this[ADD_EVENT_LISTENER]('error', _dispatchEvent)
        this[ADD_EVENT_LISTENER]('abort', _dispatchEvent)
      } else {
        const _originStateChange = this[ON_READY_STATE_CHANGE]
        this[ON_READY_STATE_CHANGE] = function (event) {
          if (this.readyState === 4) {
            _dispatchEvent(event)
          }
          if (_originStateChange) {
            _originStateChange.apply(this, arguments)
          }
        }
      }

      return _send.apply(this, arguments)
    }
  }
  getQuery() {

  }
}
