// byted-tea-sdk@5.3.10/src/plugin/visual/exposure.ts
import { getElementsFromXPath, getElementsFromXPathNext } from './selector'
import { getXpath } from '../track/path'
import localId from '../../util/local'
import { isArray } from '../../util/tool'
import Visual from './visual'

class Exposure {
  observer: any
  intersection: any
  elementList: any
  elememtMap: any
  observerInstance: any
  mutationInstance: any
  configList: any
  visual: Visual;
  constructor(visualInstance: any, configList: any) {
    try {
      this.visual = visualInstance
      this.configList = configList
      this.elementList = []
      this.elememtMap = new Map()
      this.configList.filter(item => item.trigger_type === 'view').forEach(item => item.uid = localId())
      if (IntersectionObserver) {
        this.observerInstance = new IntersectionObserver(this.observerCallback.bind(this), {
          threshold: [0, 0.01, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1],
        });
      } else {
        console.warn('your browser can not support exposure')
      }
      if (MutationObserver) {
        this.mutationInstance = new MutationObserver(this.initObserver.bind(this));
        this.mutationInstance.observe(document.body, {
          childList: true, attributes: false, subtree: true, attributeOldValue: false,
        });
      }
      this.initObserver()
    } catch (error) {
      console.error(error);
    }
  }
  initObserver() {
    try {
      const url = new URL(this.visual.repalceUrl(window.location.href));
      const { pathname, search, hash } = url;
      const patternStr = pathname + search + hash;
      if (this.observerInstance && this.configList) {
        this.configList.forEach(item => {
          if (item.trigger_type !== 'view') return
          if (!new RegExp(item.cond.url).test(patternStr)) return;
          const fieldSelector = item.element_config.field.includes('selector') ? 'selector' : 'path';
          const elementKey = fieldSelector === 'selector' ? (item.element_config.field === 'custom_selector' ? item.element_config.selector : (item.cond.similar_pattern ? `${item.element_config.selector}_${item.cond.similar_segment_selector}` : item.element_config.selector)) : `${item.element_config.path}_${item.element_config.position.join()}`
          const { elementInstanceList, exposureSimilar } = this.getElementsByElementConfig(item);
          this.elementList = this.elementList.filter(item => document.body.contains(item.elementInstance));
          if (isArray(elementInstanceList)) {
            elementInstanceList.forEach(element => {
              this.handelElementInstance(element, elementKey, item, exposureSimilar)
            })
          } else {
            this.handelElementInstance(elementInstanceList, elementKey, item, exposureSimilar)
          }
        })
      }
    } catch (error) {
      console.error(error);
    }
  }

  getElementsByElementConfig(item) {
    let elementInstanceList = [];
    let exposureSimilar = false
    try {
      const fieldSelector = item.element_config.field.includes('selector') ? 'selector' : 'path';
      if (fieldSelector === 'selector') {
        if (item.element_config.field === 'selector') {
          if (item.cond.similar_pattern) {
            exposureSimilar = true
            const querySelect = document.querySelector(item.element_config.selector)
            elementInstanceList = querySelect ? Array.from(querySelect.querySelectorAll(item.cond.similar_segment_selector)) : []
          } else {
            elementInstanceList = document.querySelector(item.element_config.selector)
          }
        } else {
          exposureSimilar = true
          elementInstanceList = Array.from(document.querySelectorAll(item.element_config.selector))
        }
      } else {
        if (item.cond.similar_pattern) {
          exposureSimilar = true
          elementInstanceList = getElementsFromXPathNext(item.element_config.path, item.element_config.position)
        } else {
          elementInstanceList = getElementsFromXPath(item.element_config.path, item.element_config.position)
        }
      }
    } catch (error) {
      console.error(error);
    } finally {
      return {
        elementInstanceList,
        exposureSimilar,
      }
    }
  }


  handelElementInstance(elementInstance: any, elementKey: string, item: any, exposureSimilar: boolean) {
    try {
      if (!elementInstance) return
      if (!elementInstance.entryKey) {
        // 没有entryKey的时候新建一个
        elementInstance.entryKey = []
      }
      if (elementInstance.entryKey.indexOf(elementKey) === -1) {
        elementInstance.entryKey.push(elementKey)
      }
      const elementKeyForList = this.get(elementKey)
      const selfId = localId()
      if (!elementInstance.selfId) {
        elementInstance.selfId = selfId;
      }
      const { shouldUpdateInstance, elementExist, currentElement } = this.getInstance(elementKeyForList, elementInstance, item, exposureSimilar)
      if (!elementKeyForList.length || (elementKeyForList.length && !elementExist)) {
        this.elementList.push({
          elementKey,
          elementInstance,
          isIntersecting: false,
          selfId,
          ...item
        })

      }
      if (elementKeyForList.length && (!shouldUpdateInstance && !elementExist)) return
      if (elementKeyForList.length && (shouldUpdateInstance && elementExist)) {
        if (currentElement.exposured && currentElement.cond.frequency === 'once') return
        this.exposureRemove(elementInstance)
      }
      this.exposureAdd(elementInstance)
    } catch (error) {
      console.error(error);
    }
  }
  get(elementKey: string | [string]) {
    const elementKeyForList = []
    if (this.elementList.length) {
      if (elementKey instanceof Array) {
        elementKey.forEach(key => {
          this.elementList.forEach(element => {
            if (element.elementKey === key) {
              elementKeyForList.push(element)
            }
          })
        })
      } else {
        this.elementList.forEach(element => {
          if (element.elementKey === elementKey) {
            elementKeyForList.push(element)
          }
        })
      }
    }
    return elementKeyForList
  }
  getInstance(elementKeyList: any, instance: any, configItem: any, exposureSimilar: boolean) {
    let shouldUpdateInstance = false
    let elementExist = false
    let currentElement = null
    elementKeyList.forEach(element => {
      if (element.uid === configItem.uid) {
        // key存在，但uid不一样，说明是不同的配置
        if (exposureSimilar) {
          if (element.selfId === instance.selfId) {
            if (element.elementInstance === instance) {
              elementExist = true
            } else {
              shouldUpdateInstance = true
              element.elementInstance = instance
              currentElement = element
            }
          }
        } else {
          elementExist = true
          // 已经有这个uid了，说明这个配置已经加入过了，但实例可能变化
          if (element.elementInstance !== instance) {
            shouldUpdateInstance = true
            element.elementInstance = instance
            currentElement = element
          }
        }
      }
    })
    return { shouldUpdateInstance, elementExist, currentElement }
  }
  set(elementKey: string, element: any, target: any) {
    const elementKeyList = this.get(elementKey)
    elementKeyList.filter(item => item.elementInstance === target).forEach(elementItem => {
      if (element.uid === elementItem.uid) {
        elementItem = Object.assign(elementItem, element)
      }
    })
  }
  observerCallback(entries: any) {
    try {
      entries.forEach(entry => {
        const { target, isIntersecting, time, intersectionRatio } = entry
        const url = new URL(this.visual.repalceUrl(window.location.href));
        const { pathname, search, hash } = url;
        const patternStr = pathname + search + hash;
        let entry_key = target.entryKey
        if (!isArray(entry_key) || !entry_key.length) return;
        const elementList = this.get(entry_key).filter(item => item.elementInstance === target)
        if (!elementList.length) return;
        elementList.forEach(element => {
          if (element.exportModel === 'once') return;
          if (!isIntersecting) {
            element.isIntersecting = false
            element.added = false
            element.exposured = false
            element.startTime = Date.now()
            this.set(entry_key, element, target)
            return
          }
          // url 不匹配
          if (!new RegExp(element.cond.url).test(patternStr)) return;
          if (!element.added) {
            // 发生交互了
            element.startTime = Date.now()
            // 元素交互时处理一次，为false或者undefined的代表没有处理过
            element.added = true
            element.isIntersecting = true
            if (element.cond.min_exposure_time) {
              const exposureTimeout = setTimeout(() => {
                if (element.isIntersecting && !element.exposured) {
                  this.handleExposure(entry_key, element, intersectionRatio, target)
                }
              }, element.cond.min_exposure_time)
              element.startWait = exposureTimeout
            }
            this.set(entry_key, element, target)
          }
          const endTime = Date.now()
          if (endTime - element.startTime >= element.cond.min_exposure_time && !element.exposured) {
            this.handleExposure(entry_key, element, intersectionRatio, target)
          }
        })
      })
    } catch (error) {
      console.error(error);
    }
  }
  handleExposure(entry_key: string, element: any, intersectionRatio: number, target: any) {
    try {
      // const element = this.elememtMap.get(entry_key)
      if (intersectionRatio * 100 < element.cond.min_exposure_ratio) return
      if (element.cond.text_pattern && !this.isMatchText(element, target)) return;
      // 符合最小的曝光时间
      if (element.cond.frequency === 'once') {
        // 只曝光一次
        if (element.startWait) {
          // 元素有定时器
          clearTimeout(element.startWait)
        }
        element.exportModel = 'once'
        this.reportEvent(element, target)
        // this.exposureRemove(element.elementInstance)
      } else {
        // 多次曝光
        element.exportModel = 'eveny'
        this.reportEvent(element, target)
      }
      element.startTime = Date.now()
      element.exposured = true
      // this.elememtMap.set(entry_key, element)
      this.set(entry_key, element, target)
    } catch (error) {
      console.error(error);
    }
  }

  /** 实际曝光的元素文本与配置下发的限定文本内容是否匹配 */
  isMatchText(config: any, target: HTMLElement) {
    const elementText = this.visual.getElementText(target);
    return elementText === config.element_config.text;
  }

  exposureAdd(elementInstance: any) {
    this.observerInstance.observe(elementInstance)
  }
  exposureRemove(elementInstance: any) {
    this.observerInstance.unobserve(elementInstance)
  }
  getElementPath(element: any) {
    const elementData: any = {}
    const { element_path, positions } = getXpath(element)
    const _position = positions.map(item => `${item}`)
    elementData.path = element_path
    elementData.positions = _position
    return elementData
  }
  reportEvent(element: any, target: HTMLElement) {
    try {
      // 解析出圈选标识命中的元素列表
      const nodeList = this.getElementsByElementConfig(element).elementInstanceList;
      const componentDomList = nodeList
        ? isArray(nodeList)
          ? Array.from(nodeList)
          : [nodeList]
        : [];
      const eventParams = element.params && element.params.length
        ? this.visual.calculateParams({
          configItem: element,
          target: target || null,
          componentDomList,
        })
        : {};
      this.visual.handleReport(element, { ...eventParams })
    } catch (error) {
      console.error(error);
    }
  }
}
export default Exposure
