// byted-tea-sdk@5.3.10/src/plugin/visual/selector.ts
function extractTag(componentOfXapth: string): string {
  return componentOfXapth.match(/^[a-zA-Z]+[0-9]?/)[0] || '';
}

export function filter<T>(
  arr: T[],
  predict: (item: T, index: number) => boolean,
): T[] {
  let index = 0;
  const { length } = arr;
  const result: any[] = [];

  while (index < length) {
    if (predict(arr[index], index)) {
      result.push(arr[index]);
    }

    index += 1;
  }

  return result;
}

function getNodesFromXPathInRange(
  range: any,
  xpath: string,
  _positions: (string | number)[],
) {
  const pathComponents: string[] = filter<string>(
    xpath.split('/'),
    Boolean,
  ).map(extractTag);
  const pathComponentsLength = pathComponents.length;
  const positions = _positions.map(String);

  let round = 0;
  let queue = [range];
  let cachedInThisTurn = [];

  while (queue.length && round < pathComponentsLength) {
    cachedInThisTurn = [];

    let sameTypeChildIndex = 0;
    for (let i = 0; i < queue.length; i++) {
      const tag = queue[i].tagName.toLowerCase() || '';

      if (pathComponents[round] === '*' || pathComponents[round] === tag) {
        if (
          positions[round] === '*' ||
          positions[round] === sameTypeChildIndex.toString()
        ) {
          cachedInThisTurn.push(queue[i]);
        }

        sameTypeChildIndex++;
      }
    }

    // there is no need to reach child in the last round
    if (round < pathComponentsLength - 1) {
      queue = cachedInThisTurn.reduce(
        (acm, cur) => acm.concat(Array.from(cur.children)),
        [],
      );
    }
    round++;
  }

  if (round < pathComponentsLength - 1) return [];

  return cachedInThisTurn;
}

export const getElementsFromXPath = (
  xpath: string,
  positions: Array<string | number>,
): HTMLElement[] => {
  return getNodesFromXPathInRange(document.body, xpath, positions);
}


/**
 *  div#id1.class1[] -> div
 * @param xpathTag
 */
function getTagName(xpathTag) {
  var idxArray = ['#', '.', '[]'].map(function (flag) { return xpathTag.indexOf(flag); }).filter(function (idx) { return idx !== -1; });
  var minIdx = Math.min.apply(undefined, idxArray.concat([xpathTag.length]));
  var tagName = xpathTag.slice(0, minIdx);
  return tagName.toUpperCase();
}
/**
 * /body/div#id.class/p/a && [0,0,0,0]  -> a
 *
 * @param xpath
 * @param parentElement
 */
function getElementByXpath(xpath, positions, parentElement?: Element) {
  if (parentElement === void 0) { parentElement = document.documentElement; }
  if (!xpath) {
    return parentElement;
  }
  var curEle = parentElement;
  xpath
    .split('/')
    .filter(function (item) { return item !== ''; })
    .forEach(function (item, xpathIdx) {
      if (!curEle) {
        return;
      }
      var tagName = getTagName(item);
      var childrens = Array.from(curEle.children).filter(function (item) { return item.tagName.toUpperCase() === tagName; });
      var target = childrens[+positions[xpathIdx]];
      curEle = target;
    });
  return curEle || null;
}


// 根据pathposition查询node列表
export function getElementsFromXPathNext(path: string, positions: string[]) {
  return _getElementsFromXpathNext(document.documentElement, path, positions);
}

export function getPathComponents(
  xpath: string,
  strip: boolean = false,
): string[] {
  return xpath.split('/').reduce<string[]>((acm, cur) => {
    return cur ? acm.concat(strip ? cur.replace(/\[\]$/, '') : cur) : acm;
  }, []);
}

function extractChildren(
  parent: HTMLElement,
  tag: string,
  // position maybe wildcard
  position: string,
): HTMLElement[] {
  const children = Array.from(parent.children).filter(
    (c): c is HTMLElement => c instanceof HTMLElement,
  );

  const sameTypeChildren = children.filter((el) => {
    return tag === el.tagName.toLowerCase();
  });

  return sameTypeChildren.filter((_, index) => {
    return [index.toString(), '*'].includes(position);
  });
}

function _getElementsFromXpathNext(
  root: HTMLElement,
  path: string,
  positions: string[],
) {
  const pathComponents = getPathComponents(path, true);
  let branches = [root];

  if (pathComponents.length !== positions.length)
    throw new Error(`invalid xpath, path: ${path}, pos: ${positions}`);

  for (let round = 0; round < pathComponents.length; round++) {
    const tag = pathComponents[round];
    const pos = positions[round];
    const nextBranches: HTMLElement[] = [];

    for (let bIndex = 0; bIndex < branches.length; bIndex++) {
      const res = extractChildren(branches[bIndex], tag, pos);
      nextBranches.push(...res);
    }

    if (nextBranches.length <= 0) {
      if (round < pathComponents.length - 1) {
        branches = [];
      }

      break;
    }

    branches = nextBranches;
  }

  return branches;
}
