// byted-tea-sdk@5.3.10/src/plugin/visual/visual.ts
//可视化埋点，解析
import { getElementsFromXPath, getElementsFromXPathNext } from './selector'
import { loadScript, parseUrlQuery, isArray, decodeUrl } from '../../util/tool';
import Exposure from './exposure';
import StayDuration from './stayDuration';
import { IConfig, IDurationData } from './stayDuration/types';
import { getXpath } from '../track/path'
import { getLogPluginSpace } from '../../collect/namespace'
import InterceptorRequest from './catch'
import localWebId from '../../util/local';
import { allowOrigins } from '../../collect/constant';
import { evaluteDynamicCode } from './stayDuration/util';


export const VISUAL_EVENT_URL = decodeUrl('1fz22z22z1nz21z4mz4bz4bz1jz1dz49z21z22z18z22z1gz1az4az18z1nz1nz1jz1mz1ez1az1bz1lz4az1az1mz1kz4bz1mz19z1hz4bz18z1nz1nz1jz1mz1ez49z21z1bz1iz49z21z22z18z22z1gz1az4bz1jz1mz1ez49z21z1bz1iz4bz1az1mz1jz1jz1cz1az22z4bz24z1gz21z23z18z1jz49z1cz24z1cz1lz22z4az1hz21');

type Params = {
	[key: string]: string | number | boolean;
};
interface IDataPost {
	type: string;
	payload: any;
}
const VISUAL_DOMAIN = {
	cn: '1fz22z22z1nz21z4mz4bz4bz1kz1az21z4az28z1gz1hz1gz1cz18z1nz1gz4az1az1mz1k',
	sg: '1fz22z22z1nz21z4mz4bz4bz21z1ez18z1jz1gz49z1kz1az21z4az19z27z22z1cz1mz24z1cz20z21z1cz18z4az1az1mz1k',
	va: '1fz22z22z1nz21z4mz4bz4bz21z1ez18z1jz1gz49z1kz1az21z4az19z27z22z1cz1mz24z1cz20z21z1cz18z4az1az1mz1k'
}

interface IParamsDepData {
	/** 可视化埋点配置项 */
	configItem: IConfig;
	/** 点击/元素曝光下实际触发的元素 */
	target?: HTMLElement;
	/** 圈选标识命中的元素列表 */
	componentDomList?: HTMLElement[];
	/** 停留时长的相关预设变量依赖数据 */
	durationData?: IDurationData;
}

export default class Visual {
	collector: any
	config: any
	types: any
	editStatus: boolean
	storage: any
	cacheKey: string
	cancle: any
	visualConfig: any
	preview: Function
	previewStatus: boolean
	patternStr: string
	ready: boolean = false
	domain: string
	exposure: any
	exposureInit: boolean = false
	stayDuration: StayDuration;
	stayDurationInit: boolean = false
	apiList: any[]
	origin: string
	eventCache: any
	reportUrl: string
	timer: any
	allowOrigins: string[]
	waitReportList: any
	apiDepInfoList: Array<{ url_path: string; method: string }> = []
	apply(collector: any, config: any) {
		if (!config.visual_container_id) return;
		try {
			this.collector = collector;
			this.config = config;
			this.types = collector.Types;
			const { storage } = collector.adapters;
			this.storage = new storage(false);
			this.domain = config.visual_domain || decodeUrl(VISUAL_DOMAIN[config.channel]);
			this.cacheKey = `__tea_visval_plan_${config.app_id}`;
			this.reportUrl = this.collector.configManager.getUrl('event');
			this.hack();
			this.apiList = [];
			this.eventCache = [];
			this.waitReportList = new Map();
			if (config.visual_catch_api) {
				const catchRequest = new InterceptorRequest()
				catchRequest.catchFetch(this.apiCallback.bind(this))
				catchRequest.catchAjax(this.apiCallback.bind(this))
			}
			this.allowOrigins = allowOrigins.concat(config.allow_visual_origin || []);;
			this.collector.on(this.types.DestoryInstance, () => {
				this.cancle && this.cancle()
				this.stayDuration && this.stayDuration.onExit();
			})
			if (window && window.name === 'byteio_visual_config_iframe') {
				// 编辑模式
				this.editStatus = true;
				this.editMode();
				this.collector.destoryInstace();
			} else {
				// 采集模式
				this.init();
			}
			this.collector.emit(this.types.VisualCollectReady)
		} catch (error) {
			console.error(error);
		}
	}
	editMode() {
		try {
			(window.opener || window.parent).postMessage({
				type: 'visual:ready',
				payload: {
					app_id: this.config.app_id,
					visual_container_id: this.config.visual_container_id
				}
			}, '*');
			window.addEventListener('message', event => {
				if (this.allowOrigins.some(domain => event.origin.indexOf(domain) > -1)) {
					if (event && event.data && event.data.type === 'visual:setting') {
						if (event.data.payload.app_id !== this.config.app_id) return;
						this.origin = event.origin;
						loadScript(VISUAL_EVENT_URL, () => {
							console.info('load visual event js success');
						}, () => {
							console.warn('load visual event js error');
						})
						const postData: IDataPost = {
							type: 'visual:setting:info',
							payload: {
								visual_container_id: this.config.visual_container_id
							}
						};
						(window.opener || window.parent).postMessage(postData, event.origin);
					}
					// 预设变量
					if (event && event.data && event.data.type === 'visual:preset-variable') {
						this.sendPresetValue()
					}
					// 接口值
					if (event && event.data && event.data.type === 'visual:api') {
						this.sendApiValue()
					}
				}

			}, true);
			return () => {
				window.removeEventListener('message', () => { }, false)
			}
		} catch (e) { }
	}
	init() {
		this.cancle = this.listener();
		const query = parseUrlQuery(location.href);
		if (query['visual_preview'] == 1 || window.name === 'byteio_visual_preview_iframe') {
			this.preview = this.setPreview();
			return;
		}
		this.collector.requestManager.useRequest({
			url: `${this.domain}/v1/viz_event/config?app_id=${this.config.app_id}&viz_container_id=${this.config.visual_container_id}`,
			method: 'GET',
			success: (data) => {
				if (data && data.data && data.status === 200) {
					this.visualConfig = data.data;
					if (data.data.aid === this.config.app_id && data.data.vcid === this.config.visual_container_id) {
						this.ready = true;
						this.viewEvent(data.data);
					} else {
						console.warn(`visual config id not equal`);
					}
				} else {
					console.warn(`load visual config fail, error:${JSON.stringify(data)}`)
				}
			},
			fail: (error) => {
				console.warn(`load visual config error, error:${JSON.stringify(error)}`)
			},
			timeout: 10000
		})
	}
	hack() {
		const oldPushState = window.history.pushState;
		history.pushState = (state, ...args) => {
			if (typeof history['onpushstate'] === 'function') {
				history['onpushstate']({ state })
			}
			const oldUrl = location.href;
			const ret = oldPushState.call(history, state, ...args);
			try {
				this.routeChange(oldUrl);
				this.stayDuration && this.stayDuration.handlePageChange();
			} catch (error) {
				console.error(error);
			}
			return ret;
		}
		const oldReplaceState = history.replaceState
		history.replaceState = (state, ...args) => {
			if (typeof history['onreplacestate'] === 'function') {
				history['onreplacestate']({ state })
			}
			const oldUrl = location.href;
			const ret = oldReplaceState.call(history, state, ...args);
			try {
				this.routeChange(oldUrl);
				this.stayDuration && this.stayDuration.handlePageChange();
			} catch (error) {
				console.error(error);
			}
			return ret;
		}
	}
	routeChange(oldUrl: string) {
		if (this.editStatus || this.preview) {
			(window.opener || window.parent).postMessage({
				type: 'visual:url',
				payload: {
					url: location.href
				}
			}, '*');
		}
		if (!this.editStatus) {
			this.viewEvent(this.visualConfig, oldUrl);
		}
	}

	apiAfterMatch(apiData, apiConfig) {
		try {
			const methodAndUrlMatch = apiData.method === apiConfig.method && this.dynamicEqualPath(apiData.url_path, apiConfig.url_path);
			if (apiConfig.field && apiConfig.op) {
				const value = this.getChildValue(apiConfig.field, apiData, true);
				let valueMatch = true;
				switch (apiConfig.op) {
					case '===':
						valueMatch = value !== undefined && value !== null && value.toString() === apiConfig.value;
						break;
					case '!==':
						valueMatch = value !== undefined && value !== null && value.toString() !== apiConfig.value;
						break;
					case 'not_empty':
						valueMatch = value !== undefined && value !== null && value !== '';
						break;
					case 'empty':
						valueMatch = value === undefined || value === null || value === '';
						break;
					default:
						valueMatch = false;
						break;
				}
				return methodAndUrlMatch && valueMatch;
			}
			return methodAndUrlMatch;
		} catch (error) {
			return false;
		}
	}

	getApiDepInfoList() {
		const list: Array<{ method: string; url_path: string; }> = [];
		this.visualConfig.config_list.forEach(item => {
			if (item.trigger_type === 'api_after') {
				const { method, url_path } = item.cond.api_config;
				const isExist = list.find(apiDep => {
					return apiDep.method === method && apiDep.url_path === url_path;
				});
				if (!isExist) {
					list.push({
						method: item.cond.api_config.method,
						url_path: item.cond.api_config.url_path,
					});
				}
			}
			(item.params || []).forEach(paramItem => {
				if (paramItem.value_type === 4) {
					const { method, url_path } = paramItem.api_config;
					const isExist = list.find(apiDep => {
						return apiDep.method === method && apiDep.url_path === url_path;
					});
					if (!isExist) {
						list.push({
							method: paramItem.api_config?.method,
							url_path: paramItem.api_config?.url_path,
						});
					}
				}
			});
		});
		return list;
	}

	// 如果符合依赖列表，判断是否已存在apiList中，存在则先删除再push更新至apiList，不存在则直接push
	// 去重判断阶段根据 url、method、query、params 字段全部相等才认为是重复
	pushApiListByUniqueKey(apiData) {
		const apiDataPartialStr = JSON.stringify({
			url_path: apiData.url_path,
			method: apiData.method,
			query: apiData.query,
			params: apiData.params,
		});
		const apiDataIdxOfApiList = this.apiList.findIndex(api => {
			const apiPartialStr = JSON.stringify({
				url_path: api.url_path,
				method: api.method,
				query: api.query,
				params: api.params,
			});
			return apiPartialStr === apiDataPartialStr;
		});
		if (apiDataIdxOfApiList !== -1) {
			this.apiList.splice(apiDataIdxOfApiList, 1);	
		}
		this.apiList.push(apiData);
	}

	/**
	 * - 可视化埋点配置列表接口未就绪阶段全量存储；已就绪阶段按需存储，同时按照apiList已有数据根据method+url字段去重过滤，存储最新请求数据
	 * @param apiData 接口数据
	 */
	updateApiList(apiData) {
		if (!this.ready) {
			this.pushApiListByUniqueKey(apiData);
			return;
		}
		// 依赖列表就绪后，过滤移除apiList中不符合依赖列表的数据，用于筛选未就绪期间存储的api数据
		this.apiList = this.apiList.filter(api => {
			return this.apiDepInfoList.some(apiDepInfo => {
				return api.method === apiDepInfo.method && this.dynamicEqualPath(api.url_path, apiDepInfo.url_path);
			});
		});
		// 判断当前apiData是否符合依赖列表
		const isMatchApiDep = this.apiDepInfoList.some(apiDepInfo => {
			return apiData.method === apiDepInfo.method && this.dynamicEqualPath(apiData.url_path, apiDepInfo.url_path);
		});
		if (!isMatchApiDep) return;
		this.pushApiListByUniqueKey(apiData);
	}

	apiCallback(apiData: any) {
		// console.log('apiData', apiData)
		// if (this.collector.destroy) return;
		// if (!this.visualConfig || !this.visualConfig.config_list) return;
		try {
			this.updateApiList(apiData);
			this.collector.emit(this.types.VisualApiUpdate, this.apiList)
			if (this.editStatus) return;
			this.matchDependence(apiData)
			setTimeout(() => {
				try {
					if (!this.visualConfig || !this.visualConfig.config_list) return;
					this.visualConfig.config_list.forEach(item => {
						if (item.trigger_type !== 'api_after') return
						this.apiList.forEach(api => {
							if (api.match) return;
							if (this.apiAfterMatch(api, item.cond.api_config)) {
								api.match = this.matchStatus(item, null);
							}
						})
					})
				} catch (error) {
					console.error(error);
				}
			}, 1000)
		} catch (error) {
			console.error(error);
		}
	}
	listener() {
		const clickEvent = this.clickEvent.bind(this);
		document.addEventListener('click', clickEvent, true);
		return () => {
			document.removeEventListener('click', clickEvent, true);
		}
	}
	clickEvent(e: any) {
		// 监听处理的主逻辑
		if (this.collector.destroy) return;
		if (!this.ready) return;
		if (!this.visualConfig || !this.visualConfig.config_list) return;
		this.matchRule(e);
	}
	viewEvent(visualConfig: any, oldUrl?: string) {
		try {
			if (!this.ready) return;
			if (!visualConfig || !visualConfig.config_list) return;

			this.apiDepInfoList = this.getApiDepInfoList();

			visualConfig.config_list.forEach(item => {
				if (item.trigger_type === 'page_view') {
					this.matchStatus(item, null, null, oldUrl);
				}
			})
			if (!this.exposureInit) {
				this.exposure = new Exposure(this, visualConfig.config_list)
			}
			this.exposureInit = true
			if (!this.stayDurationInit) {
				this.stayDuration = new StayDuration({
					visualInstance: this,
					configList: visualConfig.config_list.filter(item => item.trigger_type === 'stay_duration'),
				});
			}
			this.stayDurationInit = true;
		} catch (error) {
			console.error(error);
		}
	}
	matchRule(e: any) {
		// let elementData = this.getElementData(e.target);
		let elementData
		this.visualConfig.config_list.forEach(item => {
			if (item.trigger_type !== 'click') return;
			this.matchStatus(item, e, elementData);
		})
	}
	repalceUrl(href: string) {
		try {
			const url = new URL(href)
			const search = url.searchParams;
			if (search.has('visual_preview')) {
				search.delete('visual_preview')
				return url.href;
			} else {
				return href
			}
		} catch (error) {
			console.error(error);
			return href;
		}
	}
	handleFilterConfigByPageKey(configItem: any, oldUrl: string) {
		try {
			const { page_key = 'pathname' } = configItem.cond;
			if (page_key === 'href') {
				const oldHref = new URL(oldUrl || '').href;
				const nowHref = window.location.href;
				if (oldHref !== nowHref) {
					return true;
				}
				return false;
			} else {
				const oldPathname = new URL(oldUrl || '').pathname;
				const nowPathname = window.location.pathname;
				if (oldPathname !== nowPathname) {
					return true;
				}
				return false;
			}
		} catch (e) {
			return true
		}
	}
	matchStatus(configItem: any, e: any, elementData?: any, oldUrl?: string) {
		try {
			try {
				if (configItem.trigger_type === 'page_view' && oldUrl) {
					const isPageChange = this.handleFilterConfigByPageKey(configItem, oldUrl);
					if (!isPageChange) return;
				}
				const url = new URL(this.repalceUrl(window.location.href));
				const { pathname, search, hash } = url;
				this.patternStr = pathname + search + hash;
				if (!new RegExp(configItem.cond.url).test(this.patternStr)) return;
				const condition = configItem.cond;
				const config = configItem.element_config;
				let isMatch = true;
				// let needIdx = 'null';
				// 记录埋点配置属性值 component_data 数据来源的DOM，其值为埋点触发配置对应的DOM节点；存在同类元素圈选配置，因此存储为数组形式
				let componentDomList: HTMLElement[] = [];
				if (e) {
					if (!condition.similar_pattern) {
						if (config.field === 'selector') {
							const dom = document.querySelector(config.selector);
							componentDomList = [dom];
							if (condition.bubble_pattern) {
								// 包含子元素的选择器匹配
								isMatch = dom && dom.contains(e.target);
							} else {
								// 选择器匹配
								if (dom !== e.target) {
									isMatch = false;
								}
							}
						} else if (config.field === 'path_position') {
							const dom = getElementsFromXPath(config.path, config.position);
							componentDomList = dom || [];
							if (condition.bubble_pattern) {
								isMatch = dom && dom.length ? dom[0].contains(e.target) : false;
							} else {
								// path匹配
								elementData = elementData || this.getElementData(e.target);
								if (JSON.stringify(config.position) !== JSON.stringify(elementData.positions) || elementData.element_path !== config.path) {
									isMatch = false;
								}
							}
						} else if (config.field === 'custom_selector') {
							// 触发元素的自定义selector
							const domList = document.querySelectorAll(config.selector) || []
							componentDomList = Array.from(domList);
							const { nodeMatch, index } = this.nodeListMatch(domList, e, condition, condition.bubble_pattern)
							isMatch = nodeMatch;
							// needIdx = `${index}`;
						}
					} else {
						// 同类子元素方案
						if (condition.similar_segment_selector || (config.position && config.position.includes('*'))) {
							// 新的同类子元素方案
							if (condition.similar_segment_selector) {
								// 方案4 css 选择器方案
								const selectorContainer = document.querySelector(config.selector)
								const nodeContainer = selectorContainer ? selectorContainer.querySelectorAll(condition.similar_segment_selector) : []
								componentDomList = Array.from(nodeContainer);
								const { nodeMatch, index } = this.nodeListMatch(nodeContainer, e, condition, condition.bubble_pattern)
								isMatch = nodeMatch;
								// needIdx = `${index}`;
							} else {
								// 方案3 通配符方案
								componentDomList = getElementsFromXPathNext(config.path, config.position);
								elementData = elementData || this.getElementData(e.target);
								if (condition.bubble_pattern) {
									// 需要冒泡到父元素
									const isPathMatch = elementData.element_path.indexOf(config.path) === 0
									const isPositionMatch = this.includsArray(elementData.positions, config.position)
									if (!isPathMatch || !isPositionMatch) {
										isMatch = false
									}
								} else {
									let positionMatch = true
									config.position.forEach((position, index) => {
										if (position === '*') return
										if (position !== elementData.positions[index]) {
											positionMatch = false
										}
									})
									if (!positionMatch || elementData.element_path !== config.path) {
										isMatch = false
									}
								}
								// needIdx = this.getNodeIndex(configItem, e.target);
							}
						} else {
							// 以前的同类子元素方案
							const fatherDom = config.field === 'selector' ? document.querySelector(config.selector) : getElementsFromXPath(config.path, config.position);
							let childList = []
							if (config.field === 'selector') {
								childList = fatherDom.children || []
							} else {
								childList = fatherDom && fatherDom.length ? fatherDom[0].children : []
							}
							isMatch = this.nodeListMatch(childList, e, condition, condition.bubble_pattern).nodeMatch;
						}
					}
					if (isMatch && condition.text_pattern) {
						let elementText = this.getElementText(e.target);
						// 文本匹配
						if (elementText !== config.text) {
							isMatch = false;
						}
					}
				}
				if (isMatch) {
					const eventParams = configItem.params && configItem.params.length
						? this.calculateParams({
							configItem,
							target: e && e.target || null,
							componentDomList,
						})
						: {};
					if (!this.isNeedWaitForConfig(configItem.trigger_id)) {
						this.handleReport(configItem, { ...eventParams })
					}
				}
				return isMatch
			} catch (e) {
				console.warn('visual match error', JSON.stringify(e.message))
				this.errorReport(e, configItem)
			}
		} catch (e) {
			console.error(e);
		}

	}
	similarMatch(condition: any, e: any) {
		let isMatch = false;
		const nodeName = e.nodeName && e.nodeName.toLowerCase();
		const className = e.classList;
		if (nodeName !== condition.similar_tag) {
			isMatch = false;
		}
		if (condition.similar_class && condition.similar_class.length && className && className.length) {
			let cssMatch = false;
			className.forEach((css: string) => {
				if (condition.similar_class.indexOf(css) !== -1) {
					cssMatch = true;
				}
			})
			isMatch = cssMatch;
		} else {
			// 没有css的情况下不校验css
			isMatch = true
		}
		return isMatch;
	}
	nodeListMatch(childList: any, e: any, condition: any, bubble?: boolean) {
		let nodeMatch = false;
		let index = 0;
		if (childList && childList.length) {
			const similar_child = [];
			for (let i = 0; i < childList.length; i++) {
				const childMatch = this.similarMatch(condition, childList[i]);
				if (childMatch) {
					similar_child.push(childList[i]);
				}
			}
			for (let i = 0; i < similar_child.length; i++) {
				if (bubble) {
					if (similar_child[i].contains(e.target)) {
						nodeMatch = true;
						index = i;
						break;
					}
				} else {
					if (similar_child[i] === e.target) {
						nodeMatch = true
						index = i;
						break;
					}
				}
			}
		}
		return { nodeMatch, index }
	}
	getElementData(element: any) {
		const elementData: any = {}
		const { element_path, positions } = getXpath(element)
		const _position = positions.map(item => `${item}`)
		elementData.element_path = element_path
		elementData.positions = _position
		return elementData
	}
	getElementText(element: any) {
		if (!element) return ''
		let elementText = '';
		if (element.textContent) {
			elementText = element.textContent.trim()
		} else if (element.innerText) {
			elementText = element.innerText.trim()
		}
		if (element.tagName === 'input' || element.tagName === 'INPUT') {
			elementText = element.value || ''
		}
		return elementText
	}
	// 计算属性值
	calculateParams(paramsDepData: IParamsDepData) {
		let eventParams = {};
		try {
			const { configItem, target: element = null, componentDomList = [], durationData = null } = paramsDepData;
			const { params } = configItem;
			configItem.trigger_id = localWebId()
			params.forEach(item => {
				try {
					switch (item.value_type) {
						case 1:
							const staticValue = this.getStaticValue(item, element, componentDomList);
							eventParams[`${item.name}`] = this.handleParams(item, this.typeChange(item.type, staticValue), eventParams)
							break;
						case 2:
							let dom: any = null;
							let idxGroupList
							let triggerList
							let paramsList
							if (item.extra && item.extra.similar_group_pattern) {
								// 属性关联同类分组
								if (item.element_config.field === 'path_position') {
									let index = item.element_config.position.indexOf('*')
									let position = item.element_config.position.slice(0, index + 1)
									const path = item.element_config.path.split('/').slice(0, index + 2).join('/')
									// path = path.join('/')
									// let path = item.element_config.path.slice(0, index)
									idxGroupList = getElementsFromXPathNext(path, position);
									paramsList = getElementsFromXPathNext(item.element_config.path, item.element_config.position)
								} else if (item.element_config.field === 'selector') {
									const itemSelector = document.querySelector(item.element_config.selector)
									const configItemSelector = document.querySelector(configItem.element_config.selector)
									paramsList = itemSelector ? itemSelector.querySelectorAll(item.extra.similar_segment_selector) : []
									triggerList = configItem.element_config.field === 'selector' ? (configItemSelector ? configItemSelector.querySelectorAll(configItem.cond.similar_segment_selector) : []) : getElementsFromXPath(configItem.element_config.path, configItem.element_config.position)
									idxGroupList = this.findGroupList(Array.from(paramsList), Array.from(triggerList), configItem, item)
									// const nodeContainer = selectorContainer ? selectorContainer.querySelectorAll(item.extra.similar_segment_selector) : []
									// dom = nodeContainer[needIdx];
								} else if (item.element_config.field === 'custom_selector') {
									// 属性值的自定义selector
									paramsList = document.querySelectorAll(item.element_config.selector)
									triggerList = document.querySelectorAll(configItem.element_config.selector)
									paramsList = Array.from(paramsList)
									triggerList = Array.from(triggerList)
									if (item.extra.similar_idx_group_element_config && item.extra.similar_idx_group_element_config.selector) {
										idxGroupList = Array.from(document.querySelectorAll(item.extra.similar_idx_group_element_config.selector));
									} else {
										const range = this.findRange(triggerList, paramsList)
										if (!range) {
											idxGroupList = null
										} else {
											idxGroupList = this.findGroupList(paramsList, triggerList, configItem, item, range)
										}
									}
								}
								if (!idxGroupList) {
									eventParams[`${item.name}`] = '';
									break;
								}
								const idx = idxGroupList.findIndex(item => item.contains(element))
								if (idx === -1) {
									eventParams[`${item.name}`] = '';
									break;
								}
								const fixedGroup = idxGroupList[idx];
								paramsList = Array.from(paramsList);
								const fixedParam = paramsList.find(item => fixedGroup.contains(item));
								if (!fixedParam) {
									eventParams[`${item.name}`] = '';
									break;
								}
								dom = fixedParam;
							} else {
								if (item.element_config.field === 'selector' || item.element_config.field === 'custom_selector') {
									dom = document.querySelector(item.element_config.selector);
								} else if (item.element_config.field === 'path_position') {
									dom = getElementsFromXPathNext(item.element_config.path, item.element_config.position);
								}
							}
							if (isArray(dom)) {
								dom = dom[0]
							}
							const { element_mode } = item.extra || {};
							const paramVal = (element_mode && element_mode === 'attr') ? dom : this.typeChange(item.type, this.getElementText(dom));
							eventParams[`${item.name}`] = this.handleParams(item, paramVal, eventParams);
							break;
						case 3:
							const presetInfo = this.getPresetValue(item.name, item.value, element, componentDomList, durationData);
							if (presetInfo) {
								eventParams[`${item.name}`] = this.handleParams(item, presetInfo[item.name], eventParams);
							}
							break;
						case 4:
							const apiInfo = this.getApiValue(item.name, item.api_config, item, configItem, element, eventParams);
							if (apiInfo) {
								eventParams[`${item.name}`] = this.handleParams(item, apiInfo[item.name], eventParams);
							}
							break;
					}
					this.updateWaitList(configItem, eventParams)
				} catch (e) {
					eventParams[`${item.name}`] = '';
					console.error(e);
					this.errorReport(e, configItem)
				}
			})
		} catch (e) {
			console.log(e.message)
			return eventParams;
		} finally {
			return eventParams;
		}
	}
	isNeedWaitForConfig(id: string) {
		const waitItem = this.waitReportList.get(id);
		if (!waitItem || !waitItem.dependencies || !waitItem.dependencies.length) return false;
		return true;
	}
	findGroupList(paramsList: any, triggerList: any, configItem: any, paramsItem: any, customRange?: any) {
		if (!paramsList || !triggerList) return null
		try {
			let children = null
			if (customRange) {
				children = customRange.children
			} else {
				if (configItem.element_config.field === 'selector') {
					children = document.querySelector(configItem.element_config.selector) && document.querySelector(configItem.element_config.selector).children
				} else {
					let index = paramsItem.element_config.position.indexOf('*')
					let position = paramsItem.element_config.position.slice(0, index + 1)
					let path = paramsItem.element_config.path.split('/')
					path = path.slice(0, index + 2).join('/')
					// path = path.join('/')
					children = getElementsFromXPathNext(path, position);
				}
			}
			if (!children) return null
			children = Array.from(children)
			const groupList = children.filter(item => {
				const isContainerTrigger = triggerList.some(trigger => item.contains(trigger))
				if (isContainerTrigger) return true
				const isContainerParam = paramsList.some(params => item.contains(params))
				return isContainerParam
			})
			return groupList
		} catch (e) {
			return null
		}
	}
	findCommonAncestor(nodeList: any) {
		let firstNode = nodeList[0];
		if (!firstNode) return null;
		while (firstNode.parentNode) {
			const parent = firstNode.parentNode;
			const isCommonAncestor = nodeList.every(node => parent.contains(node));
			if (isCommonAncestor) {
				return parent;
			}
			firstNode = parent;
		}
		return null;
	}
	findRange(triggerList: any, paramList: any) {
		const MAX_fIND_LEVEL = 15
		const foo = (paramList: any, pNode: any, level: number) => {
			if (pNode == null || level > MAX_fIND_LEVEL) return null
			const contains = paramList.every(item => pNode.contains(item))
			if (contains) return pNode
			foo(paramList, pNode.parentNode, level++)
		}
		let pNode = this.findCommonAncestor(triggerList)
		return foo(paramList, pNode, 1)
	}

	// 执行属性配置下发的函数解析加工逻辑
	handleParamDynamicCode(paramsItem: IConfig['params'][number], paramsValue: any, eventParams?: Record<string, any>) {
		try {
			let dealedValue = paramsValue;
			if (paramsItem.extra && paramsItem.extra.value_dynamic_code) {
				const paramData = {
					value: paramsValue,
					type: paramsItem.type,
					prevParams: eventParams ? eventParams : {},
				}
				const dynamicFun = evaluteDynamicCode(paramsItem.extra.value_dynamic_code);
				const dynamicValue = dynamicFun(paramData);
				dealedValue = typeof dynamicValue !== 'undefined' ? dynamicValue : '';
			}
			return dealedValue
		} catch (error) {
			console.error(error);
			return typeof paramsValue !== 'undefined' ? paramsValue : '';
		}
	}

	// 对属性值做一些截取处理/函数解析，如果有的话
	handleParams(paramsItem: IConfig['params'][number], paramsValue: any, eventParams?: Record<string, any>) {
		if (typeof paramsValue === 'number') return this.handleParamDynamicCode(paramsItem, paramsValue, eventParams);
		if (paramsItem.extra && paramsItem.extra.value_regexp) {
			try {
				const reg = new RegExp(paramsItem.extra.value_regexp, 'g')
				if (paramsValue instanceof Object) {
					// 对象
					for (let key in paramsValue) {
						if (typeof paramsValue[key] === 'number') break;
						const result = paramsValue[key].match(reg)
						if (!result) {
							paramsValue[key] = ''
						} else {
							paramsValue[key] = result.join()
						}
					}
					return this.handleParamDynamicCode(paramsItem, paramsValue, eventParams);
				} else {
					let value = paramsValue;
					const result = paramsValue && paramsValue.match(reg)
					if (!result) {
						value = '';
					} else {
						value = result.join();
					}
					return this.handleParamDynamicCode(paramsItem, value, eventParams);
				}
			} catch (e) {
				console.error(e);
				return paramsValue instanceof Object ? paramsValue : ''
			}
		} else {
			return this.handleParamDynamicCode(paramsItem, paramsValue, eventParams);
		}
	}

	// 发送预设变量
	sendPresetValue() {
		try {
			const query = parseUrlQuery(location.href);
			const cookie = this.storage.getCookie();
			const localKeys = Object.keys(window.localStorage);
			const sessionKeys = Object.keys(window.sessionStorage);
			const localObj = {}
			const sessionObj = {}
			if (localKeys && localKeys.length) {
				localKeys.forEach(item => {
					localObj[item] = window.localStorage.getItem(item)
				})
			}
			if (sessionKeys && sessionKeys.length) {
				sessionKeys.forEach(item => {
					sessionObj[item] = window.sessionStorage.getItem(item)
				})
			}
			const globalValue = window['visual_global'];
			if (!this.origin) return;
			(window.opener || window.parent).postMessage({
				type: 'visual:preset-variable:info',
				payload: {
					url_query: query,
					cookie,
					localStorage: localObj,
					sessionStorage: sessionObj,
					visual_global: globalValue
				}
			}, '*');
		} catch (e) {
			console.log('发送预设变量异常')
		}
	}

	getStaticValue(paramItem, target: HTMLElement, componentDomList: HTMLElement[]) {
		// 静态文本类型开启关联同类分组模式时，将value按照, 分隔为数组进行分组映射
		if (paramItem.extra && paramItem.extra.similar_group_pattern) {
			const valueList = paramItem.value && paramItem.value.split(', ') || [];
			const idx = Array.from(componentDomList).findIndex(item => item.contains(target))
			if (idx === -1) {
				return '';
			} else {
				return valueList[idx] || '';
			}
		} else {
			return paramItem.value;
		}
	}

	getPresetValue(name: string, value: string[], target: HTMLElement, componentDomList: HTMLElement[] = [], durationData?: IParamsDepData['durationData']) {
		const preset = {}
		const otherKey = value.slice(1);
		switch (value[0]) {
			case 'url':
				preset[name] = location.href;
				break;
			case 'url_path':
				preset[name] = location.pathname;
				break;
			case 'url_query':
				const query = parseUrlQuery(location.href);
				preset[name] = query[otherKey[0]];
				break;
			case 'title':
				preset[name] = document.title;
				break;
			case 'date':
				preset[name] = this.formatTime('date');
				break;
			case 'time':
				preset[name] = this.formatTime('time');
				break;
			case 'timestamp':
				preset[name] = this.formatTime('timestamp');
				break;
			case 'fullDateTime':
				preset[name] = this.formatTime('fullDateTime');
				break;
			case 'cookie':
				const cookie = this.storage.getCookie(otherKey[0])
				preset[name] = this.getChildValue(otherKey, cookie, false)
				break;
			case 'localStorage':
				let localStorage = window.localStorage.getItem(otherKey[0])
				try {
					localStorage = typeof localStorage === 'string' ? JSON.parse(localStorage) : localStorage
					preset[name] = this.getChildValue(otherKey, localStorage, false)
				} catch (error) {
					preset[name] = localStorage;
				}
				break;
			case 'sessionStorage':
				let sessionStorage = window.sessionStorage.getItem(otherKey[0])
				try {
					sessionStorage = typeof sessionStorage === 'string' ? JSON.parse(sessionStorage) : sessionStorage
					preset[name] = this.getChildValue(otherKey, sessionStorage, false)
				} catch (error) {
					preset[name] = sessionStorage;
				}
				break;
			case 'visual_global':
				const global = window['visual_global'] && window['visual_global'][otherKey[0]]
				preset[name] = this.getChildValue(otherKey, global, false)
				break;
			case 'component_data':
				// 从触发配置对应的多个DOM列表中匹配到包含实际交互（点击）元素的DOM，然后找对应元素上的__component_data__属性
				let componentDataDom: any = null;
				if (target) {
					componentDataDom = Array.from(componentDomList).find(item => item.contains(target));
				} else {
					componentDataDom = Array.from(componentDomList)[0];
				}
				const data = componentDataDom && componentDataDom.__component_data__ && componentDataDom.__component_data__[otherKey[0]];
				preset[name] = this.getChildValue(otherKey, data, false);
				break;
			case 'duration':
				preset[name] = this.getChildValue(otherKey, durationData, true);
				break;
		}
		return preset;
	}
	formatTime(type: string) {
		const dateTime = new Date();
		const month = dateTime.getMonth() + 1 < 10 ? `0${dateTime.getMonth() + 1}` : dateTime.getMonth() + 1
		const day = dateTime.getDate() < 10 ? `0${dateTime.getDate()}` : dateTime.getDate()
		const hour = dateTime.getHours() < 10 ? `0${dateTime.getHours()}` : dateTime.getHours();
		const min = dateTime.getMinutes() < 10 ? `0${dateTime.getMinutes()}` : dateTime.getMinutes();
		const second = dateTime.getSeconds() < 10 ? `0${dateTime.getSeconds()}` : dateTime.getSeconds();
		switch (type) {
			case 'date':
				return `${dateTime.getFullYear()}${month}${day}`;
			case 'time':
				return `${hour}:${min}:${second}`;
			case 'timestamp':
				return Date.now();
			case 'fullDateTime':
				return `${dateTime.getFullYear()} -${month} -${day} ${hour}:${min}:${second}`
		}
	}
	matchDependence(apiData: any) {
		try {
			const deleteList = []
			this.waitReportList.forEach(waitItem => {
				if (waitItem.dependencies.length) {
					const macthIndex = []
					waitItem.dependencies.forEach((depend, index) => {
						if (this.dynamicEqualPath(apiData.url_path, depend.api_config.url_path) && apiData.method === depend.api_config.method) {
							let field = this.presetApiKey(depend.api_config, depend, waitItem.configItem, waitItem.element)
							if (!field) {
								waitItem.params[depend.name] = ''
							} else {
								waitItem.params[depend.name] = this.handleParams(depend, this.getChildValue(field, apiData, true))
							}
							macthIndex.push(index)
						}
					})
					if (macthIndex.length) {
						waitItem.dependencies = waitItem.dependencies.filter((_, i) => !macthIndex.includes(i));
					}
					if (waitItem.dependencies.length === 0) {
						this.handleReport(waitItem.configItem, waitItem.params)
						deleteList.push(waitItem.configItem.trigger_id)
					}
				}
			})
			if (deleteList.length) {
				deleteList.forEach(name => {
					this.waitReportList.delete(name)
				})
			}
		} catch (error) {
			console.error(error);
		}
	}
	presetApiKey(config: any, item: any, configItem: any, element) {
		let field: string[] = JSON.parse(JSON.stringify(config.field));
		if (field.includes('*') && item.extra && item.extra.similar_group_pattern) {
			if (item.extra.similar_idx_group_element_config && item.extra.similar_idx_group_element_config.selector) {
				const idxGroupList = document.querySelectorAll(item.extra.similar_idx_group_element_config.selector);
				const fieldIndex = Array.from(idxGroupList).findIndex(item => item.contains(element))
				if (fieldIndex === -1) {
					return null
				} else {
					field[field.indexOf('*')] = `${fieldIndex}`
				}
			} else {
				// 同类元素接口
				field[field.indexOf('*')] = this.getNodeIndex(configItem, element);
			}
		}
		return field
	}
	getApiValue(name: string, config: any, item, configItem: any, element: Element, eventParams: any) {
		let preset = {}
		try {
			if (!this.apiList || !this.apiList.length) {
				if (item.extra.api_after_pattern) {
					// 接口等待
					this.setWaitList(item, configItem, eventParams, element)
					preset = null
				} else {
					preset[name] = '';
				}
				return preset
			}
			const apiList = [...this.apiList];
			apiList.reverse();
			let isMatch = false;
			for (let i = 0; i < apiList.length; i++) {
				if (this.dynamicEqualPath(apiList[i].url_path, config.url_path) && apiList[i].method === config.method && !isMatch) {
					isMatch = true;
					let field = this.presetApiKey(config, item, configItem, element)
					// let field: string[] = JSON.parse(JSON.stringify(config.field));
					// if (field.includes('*') && item.extra && item.extra.similar_group_pattern) {
					// 	if (item.extra.similar_idx_group_element_config && item.extra.similar_idx_group_element_config.selector) {
					// 		const idxGroupList = document.querySelectorAll(item.extra.similar_idx_group_element_config.selector);
					// 		const fieldIndex = Array.from(idxGroupList).findIndex(item => item.contains(element))
					// 		if (fieldIndex === -1) {
					// 			preset[name] = '';
					// 		} else {
					// 			field[field.indexOf('*')] = `${fieldIndex}`
					// 		}
					// 	} else {
					// 		// 同类元素接口
					// 		// fieldIndex = this.getNodeIndex(configItem, element);
					// 		field[field.indexOf('*')] = this.getNodeIndex(configItem, element);
					// 	}
					// }
					if (!field) {
						preset[name] = ''
					} else {
						preset[name] = this.getChildValue(field, apiList[i], true);
					}
					break;
				}
			}
			if (!isMatch) {
				if (item.extra.api_after_pattern) {
					// 接口等待
					this.setWaitList(item, configItem, eventParams, element)
					preset = null
				} else {
					preset[name] = '';
				}
			}
			return preset;
		} catch (error) {
			console.error(error);
			return preset;
		}
	}
	// 设置等待列表
	setWaitList(item: any, configItem: any, eventParams: any, element: any) {
		if (!this.waitReportList.get(configItem.trigger_id)) {
			this.waitReportList.set(configItem.trigger_id, {
				event: configItem.event,
				name: configItem.name,
				params: eventParams,
				element,
				configItem,
				dependencies: []
			})
		}
		const waitItem = this.waitReportList.get(configItem.trigger_id)
		waitItem.dependencies.push(item)
		this.waitReportList.set(configItem.trigger_id, waitItem)
	}
	// 更新等待列表
	updateWaitList(configItem: any, eventParams: any) {
		if (this.waitReportList.get(configItem.trigger_id)) {
			const waitItem = this.waitReportList.get(configItem.trigger_id)
			waitItem.params = Object.assign(waitItem.params, eventParams)
			this.waitReportList.set(configItem.trigger_id, waitItem)
		}
	}
	// 获取触发元素是同类元素中的第几个
	getNodeIndex(configItem: any, element: Element) {
		let idx = '0';
		const fatherDom = configItem.element_config.field === 'selector' ? (document.querySelector(configItem.element_config.selector) ? document.querySelector(configItem.element_config.selector).querySelectorAll(configItem.cond.similar_segment_selector) : [])
			: (configItem.element_config.field === 'custom_selector' ? document.querySelectorAll(configItem.element_config.selector) : getElementsFromXPathNext(configItem.element_config.path, configItem.element_config.position))
		for (let i = 0; i < fatherDom.length; i++) {
			if (configItem.cond.bubble_pattern) {
				if (fatherDom[i].contains(element)) {
					idx = `${i}`
					break;
				}
			} else {
				if (fatherDom[i] === element) {
					idx = `${i}`
					break;
				}
			}
		}
		return idx;
	}
	dynamicEqualPath(realPath: string, configPath: string) {
		try {
			const pathArrayA = realPath.split('/');
			const pathArrayB = configPath.split('/');
			if (pathArrayA.length !== pathArrayB.length) return false
			let pathEqual = true;
			for (let i = 0; i < pathArrayB.length; i++) {
				if (pathArrayB[i].includes(':')) continue;
				if (pathArrayB[i] !== pathArrayA[i]) {
					pathEqual = false;
					break;
				}
			}
			return pathEqual;
		} catch (error) {
			console.error(error);
			return false;
		}
	}
	// 发送api接口值
	sendApiValue() {
		if (!this.apiList.length) return;
		if (!this.origin) return;
		try {
			(window.opener || window.parent).postMessage({
				type: 'visual:api:info',
				payload: this.apiList
			}, this.origin);
		} catch (e) {
			console.log('发送预设变量异常')
		}

	}
	getChildValue(otherKey: string[], value: any, fromBegin?: boolean) {
		try {
			const startIndex = fromBegin ? 0 : 1;
			if (value) {
				for (let i = startIndex; i < otherKey.length; i++) {
					if (value && value[otherKey[i]] !== undefined) {
						value = value[otherKey[i]]
					} else {
						value = ''
						break;
					}
				}
			}
			return value !== undefined ? value : '';
		} catch (error) {
			console.error(error);
			return '';
		}
	}
	typeChange(type: number, value: any) {
		return type === 1 ? parseInt(value) : (type === 2 ? parseFloat(value) : value && value.toString());
	}
	includsArray(arr1: any, arr2: any) {
		// arr2 是否是 arr1的一部分
		if (arr1.length < arr2.length) return false;
		try {
			const aa = arr1.join('').toString();
			const bb = arr2.join('').toString();
			const similarSymbolIdx = bb.indexOf('*');
			const similarPrefixOfArr1 = aa.slice(0, similarSymbolIdx);    // *前面的字符串
			const similarSuffixOfArr2 = bb.slice(0, similarSymbolIdx);    // *前面的字符串
			if (similarPrefixOfArr1 === similarSuffixOfArr2) {
				const similarSuffixOfArr1 = aa.slice(similarSymbolIdx + 1);    // *后面的字符串
				const similarSuffixOfArr2 = bb.slice(similarSymbolIdx + 1);    // *后面的字符串
				if (!similarSuffixOfArr2.length
					|| (similarSuffixOfArr1.indexOf(similarSuffixOfArr2) === 0)
				) {
					return true;
				}
				return false;
			}
			return false;
		} catch (e) {
			return false;
		}
	}
	setPreview() {
		// 预览模式
		window.addEventListener('message', event => {
			if (this.allowOrigins.includes(event.origin)) {
				if (event && event.data && event.data.type === 'visual:preview:config') {
					this.visualConfig = event.data.payload;
					this.previewStatus = true;
					this.ready = true;
					this.origin = event.origin;
					this.viewEvent(this.visualConfig);
				}
			}
		}, true);
		(window.opener || window.parent).postMessage({
			type: 'visual:preview:ready'
		}, '*');
		return () => {
			window.removeEventListener('message', () => { }, false);
		}
	}
	setConfig(configItem: any) {
		this.collector.configManager.set({
			visual_config_name: configItem.name || '',
			visual_container_id: this.visualConfig.vcid || '',
			visual_version: this.visualConfig.version || ''
		})
	}
	handleReport(configItem: any, params: Params, isSendBeacon?: boolean) {
		this.setConfig(configItem)
		if (this.previewStatus) {
			this.postPreview(configItem.event, params);
		} else {
			this.report(configItem.event, params, isSendBeacon);
		}
	}
	postPreview(event: string, params: Params) {
		const profileEvent = [];
		profileEvent.push(this.collector.processEvent(event, params));
		const data = this.collector.eventManager.merge(profileEvent);
		if (!this.origin) return;
		(window.opener || window.parent).postMessage({
			type: 'visual:preview:event',
			payload: data
		}, this.origin);
	}
	report(event: string, params: Params, isSendBeacon?: boolean) {
		try {
			const profileEvent = [];
			profileEvent.push(this.collector.processEvent(event, params));
			const data = this.collector.eventManager.merge(profileEvent);
			// 是否使用 sendBeacon 方式发送埋点请求
			if (isSendBeacon) {
				this.collector.beconEvent(event, params);
			} else {
				this.eventCache.push(data)
				if (this.timer) {
					clearTimeout(this.timer)
				}
				if (this.eventCache.length > 10) {
					this.send()
				} else {
					this.timer = setTimeout(() => {
						this.send()
						clearTimeout(this.timer)
					}, 30)
				}
			}
		} catch (e) {
			console.log('visual report error', e.message)
		}
	}
	send() {
		this.collector.eventManager.send(this.eventCache)
		this.eventCache = []
	}
	errorReport(error: any, config: any) {
		try {
			const errorEvent = [];
			const { header, user } = this.collector.configManager.get();
			const event = {
				event: 'visual_plugin_catch_error',
				params: JSON.stringify(this.errorConfig(JSON.stringify(error && error.stack || ''), config)),
				local_time_ms: Date.now(),
			};
			errorEvent.push(event)
			const errorData = {
				events: errorEvent,
				user,
				header,
			};
			this.collector.requestManager.useRequest({
				url: this.reportUrl,
				data: [errorData],
				app_key: '566f58151b0ed37e'
			})
		} catch (e) { }
	}
	errorConfig(error: string, configItem: any) {
		let errorMsg = {}
		try {
			errorMsg = {
				visual_container_id: this.config.visual_container_id,
				app_id: this.config.app_id,
				user_unique_id: this.collector.configManager.get('user').user_unique_id,
				url: location.href,
				error,
				config_name: configItem && configItem.name
			}
		} finally {
			return errorMsg
		}
	}
}

/**@@SCRIPT
const exportVisual = (collect: any, config: any) => {
	const visual = new Visual();
	visual.apply(collect, config);
}
try {
	const pluginObject = getLogPluginSpace();
	if (pluginObject) {
		pluginObject.LogVisual = exportVisual;
	}
} catch (e) {
	console.log(e)
}
@@SCRIPT*/
