// @ies/starling_client@3.6.6/example/eden.config.js
const createEdenConfig = require('@ies/create-eden-config');
const pkg = require('./package.json');

const CDN_DOMAIN = '';
const CDN_PREFIX = `/ies/${require('./package.json').name}`;

module.exports = createEdenConfig({
  projectType: 'spa',

  output: {
    publicPath() {
      return process.env.NODE_ENV === 'production' ? `${CDN_DOMAIN}${CDN_PREFIX}/` : '/';
    },
  },

  // 定义abilities Define abilities
  abilities: {
    react: true,
    semi: true,
    sass: {
      cssLoader: {
        modules: true,
        localIdentName: '[local]--[hash:base64:5]',
      },
    },
    babel: {
      polyfill: true,
      include: [
        // 进行babel转换时，将包含与此处定义的正则匹配的文件
        // Files match RegExps defined here will be included while babel transpiling
      ],
    },
  },
  raw(options) {
    return options;
  },

  other: {
    writeToDisk: true,
  },

  dev: {
    openBrowser: {
      enabled: true,
    },

    proxy: {
      urlRewrite: {
        'http://dev.eden.net': 'http://localhost:4000',
      },
    },
  },
});
