// @ies/intl-key-render-report-plugin@0.1.20/dist/es/index.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// ../../node_modules/.pnpm/idb-keyval@6.2.1/node_modules/idb-keyval/dist/index.js
function promisifyRequest(request) {
  return new Promise((resolve, reject) => {
    request.oncomplete = request.onsuccess = () => resolve(request.result);
    request.onabort = request.onerror = () => reject(request.error);
  });
}
function createStore(dbName, storeName) {
  const request = indexedDB.open(dbName);
  request.onupgradeneeded = () => request.result.createObjectStore(storeName);
  const dbp = promisifyRequest(request);
  return (txMode, callback) => dbp.then((db) => callback(db.transaction(storeName, txMode).objectStore(storeName)));
}
var defaultGetStoreFunc;
function defaultGetStore() {
  if (!defaultGetStoreFunc) {
    defaultGetStoreFunc = createStore("keyval-store", "keyval");
  }
  return defaultGetStoreFunc;
}
function get(key, customStore = defaultGetStore()) {
  return customStore("readonly", (store) => promisifyRequest(store.get(key)));
}
function set(key, value, customStore = defaultGetStore()) {
  return customStore("readwrite", (store) => {
    store.put(value, key);
    return promisifyRequest(store.transaction);
  });
}
function clear(customStore = defaultGetStore()) {
  return customStore("readwrite", (store) => {
    store.clear();
    return promisifyRequest(store.transaction);
  });
}

// src/utils.ts
function uniqArray(a, b = []) {
  const uniqueSet = /* @__PURE__ */ new Set([...a, ...b]);
  const result = [];
  uniqueSet.forEach((value) => {
    result.push(value);
  });
  return result;
}
function getTodayFirstSecond() {
  const now = /* @__PURE__ */ new Date();
  now.setHours(0);
  now.setMinutes(0);
  now.setSeconds(0);
  now.setMilliseconds(0);
  return now.getTime();
}
var storeKey = "STARLING_KEY_RENDER_STORE";
function splitArray(arr, limit = 20) {
  const result = [];
  for (let i = 0; i < arr.length; i += limit) {
    result.push(arr.slice(i, i + limit));
  }
  return result;
}
var DIVID_STR = "-S_DIVIDER-";
function isFallback(translator, key) {
  var _a, _b, _c, _d;
  let hitFallback = false;
  try {
    const locale = translator == null ? void 0 : translator.language;
    const transLocale = locale && ((_b = (_a = translator == null ? void 0 : translator.resourceStore) == null ? void 0 : _a.data) == null ? void 0 : _b[locale]);
    const ns = (_c = translator == null ? void 0 : translator.options) == null ? void 0 : _c.defaultNS.toString();
    const transValue = transLocale && ns && ((_d = transLocale == null ? void 0 : transLocale[ns]) == null ? void 0 : _d[key]);
    if (!transValue) {
      hitFallback = true;
    }
  } catch (error) {
  }
  return hitFallback;
}

// src/keyRenderReporter.ts
var KeyRenderReporter = class {
  constructor(pluginOptions) {
    __publicField(this, "keyRenderCache");
    __publicField(this, "timer");
    __publicField(this, "projectId");
    __publicField(this, "namespaceId");
    __publicField(this, "TEAChannelDomain");
    __publicField(this, "TEAChannelVersion");
    // v2是tob域名，参考https://bytedance.larkoffice.com/wiki/wikcnpYJnqdo7iSlhP9ufn6L58f
    __publicField(this, "logger");
    __publicField(this, "getProjectAndNs");
    __publicField(this, "sampleRate");
    __publicField(this, "strategy");
    __publicField(this, "reportFallback");
    this.keyRenderCache = {};
    this.timer = NaN;
    this.projectId = pluginOptions.projectId;
    this.namespaceId = pluginOptions.namespaceId;
    this.TEAChannelDomain = pluginOptions.TEAChannelDomain;
    this.TEAChannelVersion = pluginOptions.TEAChannelVersion || "v1";
    this.getProjectAndNs = pluginOptions.getProjectAndNs;
    this.logger = pluginOptions.logger || function() {
    };
    this.sampleRate = Math.min(pluginOptions.sampleRate || 1, 1);
    this.strategy = pluginOptions.strategy || "day";
    this._reportKeyRenderRecord();
    this.reportFallback = pluginOptions.reportFallback;
  }
  clear() {
    return __async(this, null, function* () {
      clear();
      this.keyRenderCache = {};
    });
  }
  _reportTea(store) {
    return __async(this, null, function* () {
      const events = Object.entries(store || this.keyRenderCache || {}).map(
        ([cachedKey, { keyList, extraInfo = {} }]) => {
          this.logger("cachedKey-info=", cachedKey, extraInfo);
          const [caller, projectId, namespaceId, hitFallback] = cachedKey.split(
            DIVID_STR
          );
          const p = {
            projectId,
            namespaceId,
            keyList,
            caller,
            callerType: "url",
            extraInfo
          };
          if (!extraInfo) {
            delete p.extraInfo;
          }
          return {
            event: "intl_key_render",
            params: JSON.stringify(p)
          };
        }
      );
      this.logger("events=", events);
      const eventArray = splitArray(events);
      const r = yield Promise.all(
        eventArray.map((ets) => {
          const protocol = /^https:\/\//.test(this.TEAChannelDomain || "") ? "" : "https://";
          return fetch(
            `${protocol}${this.TEAChannelDomain}/${this.TEAChannelVersion}${this.TEAChannelVersion === "v2" ? "/event" : ""}/list`,
            {
              method: "POST",
              headers: {
                "content-type": "application/json; charset=UTF-8"
              },
              body: JSON.stringify([
                {
                  user: { user_unique_id: window.location.host },
                  header: {
                    app_id: /zijieapi|bytedance|byted|dongchedi|bd\.feishu|volceapplog|czgts/.test(
                      this.TEAChannelDomain || ""
                    ) ? 4797 : 3297
                  },
                  events: ets,
                  verbose: 1
                }
              ])
            }
          ).then((r2) => r2.json());
        })
      ).catch((e) => {
        this.logger("出错了", e);
      });
      this.clear();
      this.logger("请求结果返回", r);
    });
  }
  getStore() {
    return __async(this, null, function* () {
      let record = (yield get(storeKey).catch((e) => void 0)) || { store: {}, firstSec: 0 };
      return record;
    });
  }
  print() {
    return __async(this, null, function* () {
      let r = (yield this.getStore()) || {};
      let store = Object.entries(r.store).reduce((prev, cur) => {
        const [k, v] = cur;
        const [caller, projectId, namespaceId, hitFallback] = k.split(DIVID_STR);
        prev[`${projectId}|${namespaceId}|${caller}${hitFallback ? "|" + hitFallback : ""}`] = v;
        return prev;
      }, {});
      console.log("store=", store);
      return r;
    });
  }
  _reportKeyRenderRecord() {
    return __async(this, null, function* () {
      if (typeof window === "undefined" || typeof fetch === "undefined" || typeof indexedDB === "undefined") {
        return;
      }
      this.logger("进入_reportKeyRenderRecord", this.TEAChannelDomain);
      if (!this.TEAChannelDomain) {
        return;
      }
      let r = yield this.getStore();
      this.logger(
        "进入_reportKeyRenderRecord,r=",
        r,
        this.TEAChannelDomain,
        this.strategy
      );
      if ((r == null ? void 0 : r.store) && typeof (r == null ? void 0 : r.store) === "object" && Object.keys(r == null ? void 0 : r.store).length && (r == null ? void 0 : r.firstSec)) {
        let shouldReport = false;
        switch (this.strategy) {
          case "immediately": {
            shouldReport = true;
            break;
          }
          case "hour": {
            shouldReport = (/* @__PURE__ */ new Date()).getTime() - (r == null ? void 0 : r.firstSec) > 1e3 * 60 * 60;
            break;
          }
          case "half_day": {
            shouldReport = (/* @__PURE__ */ new Date()).getTime() - (r == null ? void 0 : r.firstSec) > 1e3 * 60 * 60 * 12;
            break;
          }
          case "day": {
          }
          default: {
            shouldReport = (r == null ? void 0 : r.firstSec) < getTodayFirstSecond();
          }
        }
        this.logger("进入_reportKeyRenderRecord,shouldReport=", shouldReport);
        if (shouldReport) {
          if (Math.random() < this.sampleRate) {
            this.logger("将要 _reportTea");
            this._reportTea(r == null ? void 0 : r.store);
          } else {
            this.clear();
          }
        }
      }
    });
  }
  getFirstSecond() {
    if (this.strategy === "day") {
      return getTodayFirstSecond();
    }
    return (/* @__PURE__ */ new Date()).getTime();
  }
  storeKeyRenderRecord(keyRenderCache) {
    return __async(this, null, function* () {
      clearTimeout(this.timer);
      this.timer = window.setTimeout(() => __async(this, null, function* () {
        let record = yield this.getStore();
        const store = Object.entries(keyRenderCache).reduce((prev, cur) => {
          const [cachedKey, { keyList, extraInfo }] = cur;
          if (record.store[cachedKey]) {
            prev[cachedKey] = {
              keyList: uniqArray(record.store[cachedKey].keyList, keyList)
            };
          } else {
            prev[cachedKey] = {
              extraInfo,
              keyList: uniqArray(keyList)
            };
          }
          return prev;
        }, {});
        this.logger("将要写indexedDB2,store=", store);
        yield set(storeKey, {
          firstSec: record.firstSec || this.getFirstSecond(),
          store
        });
      }), 5e3);
    });
  }
  recordKeyRender({
    key,
    processArgs
  }) {
    var _a;
    if (typeof window === "undefined" || typeof fetch === "undefined" || typeof indexedDB === "undefined") {
      return;
    }
    let { projectId, namespaceId, TEAChannelDomain } = this;
    this.logger("原始check =", {
      projectId,
      namespaceId,
      TEAChannelDomain,
      key
    });
    if (!(projectId && namespaceId && TEAChannelDomain)) {
      const o = ((_a = processArgs == null ? void 0 : processArgs.translator) == null ? void 0 : _a.options) || {};
      if (o.projectId)
        projectId = o.projectId;
      if (o.namespaceId)
        namespaceId = o.namespaceId;
      if (o.TEAChannelDomain)
        TEAChannelDomain = o.TEAChannelDomain;
    }
    if (TEAChannelDomain && key) {
      if (!(projectId && namespaceId) && this.getProjectAndNs) {
        try {
          const res = this.getProjectAndNs(processArgs);
          projectId = res.projectId;
          namespaceId = res.namespaceId;
        } catch (error) {
          console.error("Error when excuting getProjectAndNs", error);
        }
      }
      this.logger("二次check =", {
        projectId,
        namespaceId,
        TEAChannelDomain,
        key
      });
      if (projectId && namespaceId) {
        let hitFallback = false;
        if (this.reportFallback) {
          if (typeof this.reportFallback === "boolean") {
            hitFallback = isFallback(processArgs == null ? void 0 : processArgs.translator, key.toString());
          } else if (this.reportFallback instanceof Function) {
            try {
              const tmpHitFallback = this.reportFallback(
                processArgs == null ? void 0 : processArgs.translator,
                key.toString()
              );
              if (typeof tmpHitFallback === "boolean" || typeof tmpHitFallback === "object") {
                hitFallback = tmpHitFallback;
              }
            } catch (error) {
            }
          }
        }
        let renderedKey = key.toString();
        if (Array.isArray(key) && key.length > 1) {
          renderedKey = key.join("|");
        }
        const caller = (window.location.host + window.location.pathname).replace(/\/\d+/g, "/_").replace(/\w{32}/, "____");
        let cachedKey = `${caller}${DIVID_STR}${projectId}${DIVID_STR}${namespaceId}`;
        if (hitFallback) {
          cachedKey = `${cachedKey}${DIVID_STR}hitFallback`;
        }
        if (!this.keyRenderCache[cachedKey]) {
          this.keyRenderCache[cachedKey] = {
            keyList: [renderedKey]
          };
        } else {
          this.keyRenderCache[cachedKey].keyList.push(renderedKey);
        }
        const o = this.keyRenderCache[cachedKey];
        this.logger("hitFallback,o=", o);
        if (hitFallback) {
          if (o.extraInfo) {
            o.extraInfo.hitFallback = true;
          } else {
            o.extraInfo = { hitFallback: true };
          }
          if (typeof hitFallback === "object" && hitFallback.extraInfo && typeof hitFallback.extraInfo === "object" && Object.values(hitFallback.extraInfo).every(
            (v) => typeof v === "string"
          )) {
            try {
              o.extraInfo = Object.assign(o.extraInfo, hitFallback.extraInfo);
            } catch (error) {
            }
          }
        }
        this.storeKeyRenderRecord(this.keyRenderCache);
      }
    }
  }
};

// src/index.ts
var KRRIMap = {};
var KRRList = [];
function registerKeyRenderReportPlugin(pluginOptions) {
  const krr = new KeyRenderReporter(pluginOptions || {});
  const _a = pluginOptions || {}, { getProjectAndNs, logger } = _a, simpleOptions = __objRest(_a, ["getProjectAndNs", "logger"]);
  KRRIMap[JSON.stringify(simpleOptions)] = krr;
  KRRList.push(krr);
  const renderPlugin = {
    name: "keyRender",
    type: "postProcessor",
    process(value, key, options, translator) {
      krr.recordKeyRender({
        key,
        processArgs: {
          value,
          key,
          options,
          translator
        }
      });
      return value;
    }
  };
  return renderPlugin;
}
export {
  DIVID_STR,
  KRRIMap,
  KRRList,
  KeyRenderReporter,
  registerKeyRenderReportPlugin
};
