// @rdservices/inject-runtime@0.0.17/dist/adapters/svelte/html-element-tree-node.js
import { getReferenceId } from "../../utils/get-reference-id.js";
class HtmlElementTreeNode {
    type = 'element';
    element;
    name;
    uniqueId;
    constructor(element){
        this.element = element;
        this.name = element.nodeName.toLowerCase();
        this.uniqueId = String(getReferenceId(element));
    }
    getBox() {
        return this.element.getBoundingClientRect();
    }
    getElement() {
        return this.element;
    }
    getChildren() {
        const children = Array.from(this.element.children);
        return children.map((child)=>{
            if (child instanceof HTMLElement) return new this.constructor(child);
            return null;
        }).filter((value)=>null != value);
    }
    getParent() {
        if (this.element.parentElement) return new this.constructor(this.element.parentElement);
        return null;
    }
    getSource() {
        throw new Error('Method not implemented.');
    }
    getComponent() {
        throw new Error('Method not implemented.');
    }
}
export { HtmlElementTreeNode };
