// @rdservices/inject-runtime@0.0.17/dist/adapters/svelte/svelte-adapter.js
import { HtmlElementTreeNode } from "./html-element-tree-node.js";
function getElementInfo(found) {
    if (found.__svelte_meta) {
        const { loc } = found.__svelte_meta;
        return {
            thisElement: {
                box: found.getBoundingClientRect(),
                label: found.nodeName.toLowerCase(),
                link: {
                    column: loc.column + 1,
                    line: loc.line + 1,
                    filePath: loc.file
                }
            },
            htmlElement: found,
            componentBox: found.getBoundingClientRect()
        };
    }
    return null;
}
class SvelteTreeNodeElement extends HtmlElementTreeNode {
    getSource() {
        const element = this.element;
        if (element.__svelte_meta) {
            const { loc } = element.__svelte_meta;
            return {
                fileName: loc.file,
                lineNumber: loc.line + 1,
                columnNumber: loc.column + 1
            };
        }
        return null;
    }
    getComponent() {
        return null;
    }
}
function getParentsPaths(element) {
    const path = [];
    let currentElement = element;
    let maxDepth = 10000;
    do {
        if (currentElement?.__svelte_meta) {
            const { loc } = currentElement.__svelte_meta;
            const isSameFileAsTheLastOne = loc.file === path[path.length - 1]?.link?.filePath;
            if (!isSameFileAsTheLastOne) path.push({
                title: currentElement.nodeName.toLowerCase(),
                link: {
                    column: loc.column + 1,
                    line: loc.line + 1,
                    filePath: loc.file
                }
            });
        }
        currentElement = currentElement.parentElement;
        maxDepth--;
        if (maxDepth < 0) break;
    }while (currentElement);
    return path;
}
const svelteAdapter = {
    getElementInfo,
    getParentsPaths
};
const svelte_adapter = svelteAdapter;
export { SvelteTreeNodeElement, svelte_adapter as default, getElementInfo };
