// @byted/hooks@2.61.4/es/useFullscreen/index.js
/* eslint no-empty: 0 */
import { useLayoutEffect, useRef } from 'react';
import screenfull from 'screenfull';
import useBoolean from '../useBoolean';
export default (function (el, options) {
  var _ref = options || {},
      onExitFull = _ref.onExitFull,
      onFull = _ref.onFull,
      defaultValue = _ref.defaultValue,
      isBrowserFullscreen = _ref.isBrowserFullscreen;

  var onExitFullRef = useRef(onExitFull);
  onExitFullRef.current = onExitFull;
  var onFullRef = useRef(onFull);
  onFullRef.current = onFull;
  var element = useRef();

  var _useBoolean = useBoolean(defaultValue),
      state = _useBoolean.state,
      toggle = _useBoolean.toggle,
      setTrue = _useBoolean.setTrue,
      setFalse = _useBoolean.setFalse;

  useLayoutEffect(function () {
    var passedInElement = typeof el === 'function' ? el() : el;
    var targetElement = passedInElement || element.current;

    if (!targetElement) {
      return;
    }

    var setFullStyle = function setFullStyle(val) {
      var getStyleElem = document.getElementById('hook--fullscreenStyle');

      if (getStyleElem) {
        if (!val) {
          getStyleElem.remove();
        }
      } else {
        if (val) {
          var stylElem = document.createElement('style');
          stylElem.setAttribute('id', 'hook--fullscreenStyle');
          stylElem.textContent = ".hook-fullscreen{position:fixed;left:0px;top:0px;right:0px;bottom:0px;z-index:9999;width:100%!important;height:100%!important;}";
          targetElement.appendChild(stylElem);
        }
      }
    };
    /* 非全屏时，不需要监听任何全屏事件 */


    if (!state) {
      if (isBrowserFullscreen) {
        targetElement.classList.remove('hook-fullscreen');
        setFullStyle(false);
      }

      return;
    } else {
      if (isBrowserFullscreen) {
        targetElement.classList.add('hook-fullscreen');
        setFullStyle(true);
        return;
      }
    }
    /* 监听退出 */


    var onChange = function onChange() {
      if (screenfull.isEnabled) {
        var isFullscreen = screenfull.isFullscreen;
        toggle(isFullscreen);
      }
    };

    if (screenfull.isEnabled) {
      try {
        screenfull.request(targetElement);
        setTrue();

        if (onFullRef.current) {
          onFullRef.current();
        }
      } catch (error) {
        setFalse();

        if (onExitFullRef.current) {
          onExitFullRef.current();
        }
      }

      screenfull.on('change', onChange);
    }
    /* state 从 true 变为 false，则关闭全屏 */


    return function () {
      if (screenfull.isEnabled) {
        try {
          screenfull.off('change', onChange);
          screenfull.exit();
        } catch (error) {}
      }

      if (onExitFullRef.current) {
        onExitFullRef.current();
      }
    };
  }, [state, typeof el === 'function' ? undefined : el]);

  var toggleFull = function toggleFull() {
    return toggle();
  };

  var result = {
    isFullscreen: !!state,
    setFull: setTrue,
    exitFull: setFalse,
    toggleFull: toggleFull
  };

  if (!el) {
    result.ref = element;
  }

  return result;
});
