// @rdservices/aime-code-inspector-webpack@1.0.17/dist/index.mjs
var _ = Object.defineProperty, L = Object.defineProperties;
var O = Object.getOwnPropertyDescriptors;
var C = Object.getOwnPropertySymbols;
var z = Object.prototype.hasOwnProperty, E = Object.prototype.propertyIsEnumerable;
var $ = (e, t, s) => t in e ? _(e, t, { enumerable: !0, configurable: !0, writable: !0, value: s }) : e[t] = s, c = (e, t) => {
  for (var s in t || (t = {}))
    z.call(t, s) && $(e, s, t[s]);
  if (C)
    for (var s of C(t))
      E.call(t, s) && $(e, s, t[s]);
  return e;
}, g = (e, t) => L(e, O(t));
var d = (e, t, s) => new Promise((n, a) => {
  var r = (o) => {
    try {
      u(s.next(o));
    } catch (f) {
      a(f);
    }
  }, i = (o) => {
    try {
      u(s.throw(o));
    } catch (f) {
      a(f);
    }
  }, u = (o) => o.done ? n(o.value) : Promise.resolve(o.value).then(r, i);
  u((s = s.apply(e, t)).next());
});
import { normalizePath as w, fileURLToPath as I, isNextjsProject as F, getCodeWithWebComponent as H } from "@rdservices/aime-code-inspector-core";
import y, { dirname as N } from "path";
function Q(e, t) {
  return d(this, null, function* () {
    if (!e || !t)
      return [];
    const s = typeof e == "function" ? yield e() : e, n = [];
    if (typeof s == "object" && !Array.isArray(s))
      for (const a of Object.keys(s)) {
        const r = s[a], i = r.import || r;
        W(n, i, t);
      }
    else
      W(n, s, t);
    return n.filter((a) => !!a);
  });
}
function W(e, t, s) {
  typeof t == "string" ? e.push(D(t, s)) : Array.isArray(t) && e.push(...t.map((n) => D(n, s)));
}
function D(e, t) {
  return y.isAbsolute(e) ? w(e) : e.startsWith(".") ? y.resolve(t, w(e)) : "";
}
let j = "";
typeof __dirname != "undefined" ? j = __dirname : j = N(I(import.meta.url));
let v = !0;
const R = (e, t) => {
  var r, i;
  if (!v)
    return;
  v = !1;
  const s = (t == null ? void 0 : t.compiler) || t, n = (r = s == null ? void 0 : s.options) == null ? void 0 : r.module;
  ((n == null ? void 0 : n.rules) || (n == null ? void 0 : n.loaders) || []).push(
    c({
      test: (i = e.match) != null ? i : /\.html$/,
      resourceQuery: /vue/,
      use: [
        {
          loader: y.resolve(j, "./loader.js"),
          options: e
        }
      ]
    }, e.enforcePre === !1 ? {} : { enforce: "pre" }),
    c({
      test: /\.(vue|jsx|tsx|js|ts|mjs|mts|svelte)$/,
      use: [
        {
          loader: y.resolve(j, "./loader.js"),
          options: e
        }
      ]
    }, e.enforcePre === !1 ? {} : { enforce: "pre" }),
    g(c({}, e.injectTo ? { resource: e.injectTo } : {
      test: /\.(jsx|tsx|js|ts|mjs|mts)$/,
      exclude: /node_modules/
    }), {
      use: [
        {
          loader: y.resolve(j, "./inject-loader.js"),
          options: e
        }
      ],
      enforce: F() ? "pre" : "post"
    })
  );
};
function k(e, t) {
  return H({
    options: g(c({}, e), { importClient: "code" }),
    file: "main.js",
    code: "",
    record: t,
    inject: !0
  });
}
function U(n) {
  return d(this, arguments, function* ({
    options: e,
    record: t,
    assets: s
  }) {
    const a = Object.keys(s).filter((r) => /\.html$/.test(r));
    if (a.length) {
      const r = yield k(e, t);
      a.forEach((i) => {
        var o, f;
        const u = (f = (o = s[i]) == null ? void 0 : o.source) == null ? void 0 : f.call(o);
        if (typeof u == "string") {
          const h = u.replace("<head>", '<head><script type="module">\n'.concat(r, "\n<\/script>"));
          s[i] = {
            source: () => h,
            size: () => h.length
          };
        }
      });
    }
  });
}
class J {
  constructor(t) {
    this.options = t;
  }
  apply(t) {
    return d(this, null, function* () {
      var n, a, r, i, u, o, f, h, x, A, p, P;
      v = !0;
      const s = {
        port: 0,
        entry: "",
        output: this.options.output,
        inputs: Q((n = t == null ? void 0 : t.options) == null ? void 0 : n.entry, (a = t == null ? void 0 : t.options) == null ? void 0 : a.context),
        envDir: (r = t == null ? void 0 : t.options) == null ? void 0 : r.context,
        root: (i = t == null ? void 0 : t.options) == null ? void 0 : i.context
      };
      if (((o = (u = t == null ? void 0 : t.options) == null ? void 0 : u.cache) == null ? void 0 : o.type) === "filesystem" && (this.options.cache ? k(this.options, s) : t.options.cache.version = "code-inspector-".concat(Date.now())), ((x = (h = (f = t == null ? void 0 : t.options) == null ? void 0 : f.experiments) == null ? void 0 : h.cache) == null ? void 0 : x.type) === "persistent" && (this.options.cache ? k(this.options, s) : t.options.experiments.cache.version = "code-inspector-".concat(Date.now())), R(g(c({}, this.options), { record: s }), t), (A = t == null ? void 0 : t.hooks) != null && A.emit && !((P = (p = this.options.skipSnippets) == null ? void 0 : p.includes) != null && P.call(p, "htmlScript"))) {
        const { options: S } = this;
        t.hooks.emit.tapAsync("WebpackCodeInspectorPlugin", (l, T) => d(this, null, function* () {
          let b = {};
          l.getAssets ? b = yield l.getAssets() : b = l.assets, yield U({
            options: S,
            record: s,
            assets: b
          }), T();
        }));
      }
    });
  }
}
export {
  J as default
};
