// byted-tea-sdk@5.3.10/src/plugin/track/dom.ts
import { isArray } from '../../util/tool'
export function isNeedElement(element: HTMLElement | null, type: string = 'list', container?: string): boolean {
  if (!element) return false
  if (type && type === 'list') {
    if (['LI', 'TR', 'DL'].includes(element.nodeName)) return true
    if (element.dataset && element.dataset.hasOwnProperty('teaIdx')) return true
    if (element.hasAttribute && element.hasAttribute('data-tea-idx')) return true
  } else {
    if (['A', 'BUTTON'].includes(element.nodeName)) return true
    if (element.dataset && element.dataset.hasOwnProperty('teaContainer')) return true
    if (element.hasAttribute && element.hasAttribute('data-tea-container')) return true
    if (container && element.hasAttribute && element.hasAttribute(container)) return true
  }

  return false
}
export const isAttrFilter = (element: HTMLElement, attrs: any) => {
  if (hasAttributes(element, attrs)) return true
  return false
}
export function getContainer(element: HTMLElement, options: any): HTMLElement {
  let current = element
  while (current && !isNeedElement(current, 'container', options.container_attr || '')) {
    if (current.nodeName === 'HTML' || current.nodeName === 'BODY') {
      return element
    }
    current = current.parentElement
  }
  return current || element
}

export function isContainer(element: HTMLElement, options: any): Boolean {
  let current = element
  let allLevelHasContainer = false
  while (current && !isNeedElement(current, 'container', options.container_attr || '')) {
    if (current.nodeName === 'HTML' || current.nodeName === 'BODY') {
      return allLevelHasContainer
    }
    current = current.parentElement
  }
  allLevelHasContainer = true
  return allLevelHasContainer
}

export function getNodeText(node: Node): string {
  let text = ''
  if (node.nodeType === 3) {
    text = node.textContent.trim()
  } else if (node['dataset'] && node['dataset'].hasOwnProperty('teaTitle')) {
    text = node['getAttribute']('data-tea-title')
  } else if (node['hasAttribute']('ata-tea-title')) {
    text = node['getAttribute']('data-tea-title')
  }
  // else if (node['hasAttribute']('title')) {
  //   text = node['getAttribute']('title')
  // } 
  else if (node.nodeName === 'INPUT' && ['button', 'submit'].includes(node['getAttribute']('type'))) {
    text = node['getAttribute']('value')
  } else if (node.nodeName === 'IMG' && node['getAttribute']('alt')) {
    text = node['getAttribute']('alt')
  }
  return text.slice(0, 200)
}

export function getText(element: HTMLElement, options: any): string[] {
  const ele = getContainer(element, options)
  const textArr = [];
  (function _get(node: Node) {
    const text = getNodeText(node)
    if (text && textArr.indexOf(text) === -1) {
      textArr.push(text)
    }
    if (node.childNodes.length > 0) {
      const { childNodes } = node
      for (let i = 0; i < childNodes.length; i++) {
        if (childNodes[i].nodeType !== 8) {
          _get(childNodes[i])
        }
      }
    }
  })(ele)
  return textArr
}

// export function getTextSingle(element: HTMLElement): string {
//   const ele = getContainer(element)
//   let text = '';
//   (function _get(node: Node) {
//     const _text = getNodeText(node)
//     if (_text) {
//       text = text + _text
//     }
//     if (node.childNodes.length > 0) {
//       const { childNodes } = node
//       for (let i = 0; i < childNodes.length; i++) {
//         if (childNodes[i].nodeType === 3) {
//           _get(childNodes[i])
//         }
//       }
//     }
//   })(ele)
//   return text
// }

export const getElementText = (element: any) => {
  if (!element) return ''
  let elementText = '';
  try {
    if (element.textContent) {
      elementText = element.textContent.trim().replace(/\s+/g, ",");
    } else if (element.innerText) {
      elementText = element.innerText.trim().replace(/\s+/g, ",");
    }
    if (element.tagName === 'input' || element.tagName === 'INPUT') {
      elementText = element.value || ''
    }
  } finally {
    return elementText
  }
}

export function ignore(element: any): boolean {
  try {
    let _element = element
    while (_element && _element.parentNode) {
      if (_element.hasAttribute('data-tea-ignore')) return true
      if (_element.classList && _element.classList.contains('applog-web-devtools-con')) return true
      if (_element.nodeName === 'HTML' || _element.nodeName === 'body') return false
      _element = _element.parentNode
    }
    return false
  } catch {
    return false
  }
}

export const hasAttribute = (ele: HTMLElement, attr: string) => {
  if (ele.hasAttribute) {
    return ele.hasAttribute(attr);
  } else if (ele.attributes) {
    return !!(ele.attributes[attr] && ele.attributes[attr].specified);
  }
}

export const hasAttributes = (ele: HTMLElement, attrs: any) => {
  if (typeof attrs === 'string') {
    return hasAttribute(ele, attrs);
  } else if (isArray(attrs)) {
    let result = false;
    for (let i = 0; i < attrs.length; i++) {
      let testResult = hasAttribute(ele, attrs[i]);
      if (testResult) {
        result = true;
        break;
      }
    }
    return result;
  }
}

export const getAttributes = (ele: HTMLElement, attrs: any) => {
  const result = {}
  if (typeof attrs === 'string') {
    if (hasAttribute(ele, attrs)) {
      result[attrs] = ele.getAttribute(attrs)
    }
  } else {
    if (isArray(attrs)) {
      for (let i = 0; i < attrs.length; i++) {
        let testResult = hasAttribute(ele, attrs[i]);
        if (testResult) {
          result[attrs[i]] = ele.getAttribute(attrs[i])
        }
      }
    }
  }
  return result
}
