// byted-tea-sdk@5.3.10/src/plugin/track/listener.ts
import { isTrack } from './node'
import { OptionsType, EventConfig } from './type'
import { beforePageUnload, isArray } from '../../util/tool'

export default class Listener {
  config: EventConfig
  options: OptionsType
  beatTime: number
  statistics: boolean = false
  eventHandel: Function
  collect: any
  pageView: boolean = false
  domContainer: any
  constructor(options: OptionsType, collect: any, Config: any) {
    this.config = Config.getConfig().eventConfig
    this.collect = collect
    this.options = options
    this.beatTime = options.beat
    this.domContainer = options.container || document
  }
  init(eventHandel: Function) {
    this.eventHandel = eventHandel
    const mode = this.config.mode
    this.addListener(mode)
  }
  addListener(mode: 'proxy-capturing') {
    // 注册事件捕获
    if (mode === 'proxy-capturing') {
      if (this.config.click) {
        if (isArray(this.domContainer)) {
          this.domContainer.forEach(container => {
            container.addEventListener('click', this.clickEvent, true)
          })
        } else {
          this.domContainer.addEventListener('click', this.clickEvent, true)
        }
      }
      // if (this.config.change) {
      //   this.domContainer.addEventListener('change', this.changeEvent, true)
      // }
      // if (this.config.submit) {
      //   this.domContainer.addEventListener('submit', this.submitEvent, true)
      // }
      if (this.config.pv) {
        this.collect.on('route-change', (info) => {
          const { config, name } = info
          this.getPageViewEvent(config, name)
          this.pageView = true;
        })
        const viewTime = setTimeout(() => {
          // 防止异步加载，route-change已经出发导致pv没有发出
          if (this.pageView) return
          this.getPageViewEvent(this.getDefaultConfig())
          this.pageView = true;
          clearTimeout(viewTime);
        }, 2000);
      }
      if (this.config.beat) {
        try {
          if (document.readyState === 'complete') {
            this.beatEvent(this.beatTime)
          } else {
            window.addEventListener('load', () => {
              this.beatEvent(this.beatTime)
            })
          }
          let t1 = 0;
          let t2 = 0;
          let timer = null; // 定时器
          window.addEventListener('scroll', () => {
            clearTimeout(timer);
            timer = setTimeout(isScrollEnd, 500);
            t1 = document.documentElement.scrollTop || document.body.scrollTop;
          })
          const isScrollEnd = () => {
            t2 = document.documentElement.scrollTop || document.body.scrollTop;
            if (t2 == t1) {
              this.eventHandel({ eventType: 'dom', eventName: 'beat' }, {
                beat_type: 1
              })
            }
          }
        } catch (e) { }
        try {
          var entryList = window.performance && window.performance.getEntriesByType('paint')
          if (entryList && entryList.length) {
            var observer = new PerformanceObserver((entryList) => {
              var entries = entryList.getEntries();
              var lastEntry = entries[entries.length - 1];
              var lcp = lastEntry['renderTime'] || lastEntry['loadTime'];
              if (!this.statistics) {
                this.getPageLoadEvent(lcp)
                this.statistics = true
              }
            });
            observer.observe({
              entryTypes: ['largest-contentful-paint']
            });
            // 没触发2S后强制触发
            const loadTime = setTimeout(() => {
              if (this.statistics) return
              this.getPageLoadEvent(entryList[0].startTime || 0)
              this.statistics = true
              clearTimeout(loadTime)
            }, 2000);
          } else {
            this.getPageLoadEvent(0)
          }
        } catch (e) {
          this.getPageLoadEvent(0)
        }
      }
    }
  }
  removeListener() {
    if (isArray(this.domContainer)) {
      this.domContainer.forEach(container => {
        container.removeEventListener('click', this.clickEvent, true)
      })
    } else {
      this.domContainer.removeEventListener('click', this.clickEvent, true)
    }
    // this.domContainer.removeEventListener('change', this.changeEvent, true)
    // this.domContainer.removeEventListener('submit', this.submitEvent, true)
  }
  clickEvent = (e: Object) => {
    if (isTrack(e['target'], this.options)) {
      this.eventHandel({ eventType: 'dom', eventName: 'click' }, e)
    }
  }
  changeEvent = (e: Object) => {
    this.eventHandel({ eventType: 'dom', eventName: 'change' }, e)
  }
  submitEvent = (e: Object) => {
    this.eventHandel({ eventType: 'dom', eventName: 'submit' }, e)
  }
  beatEvent(beatTime: number) {
    try {
      this.eventHandel({ eventType: 'dom', eventName: 'beat' }, {
        beat_type: 3
      })
      let beaInterval
      if (this.beatTime) {
        beaInterval = setInterval(() => {
          this.eventHandel({ eventType: 'dom', eventName: 'beat' }, {
            beat_type: 2
          })
        }, beatTime)
      }
      beforePageUnload(() => {
        this.eventHandel({ eventType: 'dom', eventName: 'beat', eventSend: 'becon' }, {
          beat_type: 0
        })
        if (this.beatTime) {
          clearInterval(beaInterval)
        }
      })
    } catch (e) { }
  }
  getPageViewEvent = (eventData: Object, name?: string) => {
    if (name && name === 'pushState') {
      this.eventHandel({ eventType: 'dom', eventName: 'beat' }, {
        beat_type: 0,
        ...eventData
      })
    }
    this.eventHandel({ eventType: 'dom', eventName: 'page_view' }, eventData)
  }
  getPageLoadEvent = (lcp: any) => {
    this.eventHandel({ eventType: 'dom', eventName: 'page_statistics' }, { lcp: lcp })
  }
  getDefaultConfig() {
    return {
      is_html: 1,
      url: location.href,
      referrer: document.referrer,
      page_key: location.href,
      refer_page_key: document.referrer,
      page_title: document.title || location.pathname,
      page_manual_key: '',
      refer_page_manual_key: '',
      refer_page_title: '',
      page_path: location.pathname,
      page_host: location.host,
    }
  }
}
