// @rdservices/inject-runtime@0.0.17/dist/components/runtime.js
import { createComponent, delegateEvents, effect, insert, memo, render, setAttribute, setStyleProperty, spread, style, template, use } from "solid-js/web";
import { Show, createEffect, createMemo, createSignal, onCleanup, onMount } from "solid-js";
import { MAX_ZINDEX } from "../index.js";
import { getElementInfo } from "../adapters/index.js";
import { v4 } from "uuid";
var _tmpl$ = /*#__PURE__*/ template('<svg width=1em height=1em viewBox="0 0 24 24"fill=currentColor><path d="M11 11V5h2v6h6v2h-6v6h-2v-6H5v-2h6z">'), _tmpl$2 = /*#__PURE__*/ template('<svg width=1em height=1em viewBox="0 0 24 24"fill=currentColor><path d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z">'), _tmpl$3 = /*#__PURE__*/ template('<div style="position:absolute;width:max-content;background:#33A033;color:#fff;font-size:12px;font-weight:500;border-radius:8px;display:flex;align-items:center;padding:3px 6px;gap:4px;border:1px solid rgba(29, 78, 216, 0.1);font-family:GeistMono, ui-monospace, SFMono-Regular, Roboto Mono, Menlo, Monaco, Liberation Mono, DejaVu Sans Mono, Courier New, monospace">'), _tmpl$4 = /*#__PURE__*/ template('<div style="background:rgb(209, 213, 219);width:1px;height:20px">'), _tmpl$5 = /*#__PURE__*/ template("<div style=cursor:pointer;pointer-events:auto>"), _tmpl$6 = /*#__PURE__*/ template('<div><div style="display:flex;align-items:center;border:1px solid rgba(77, 101, 148, 0.32);background:#fff;height:40px;border-radius:8px;padding:0 10px;pointer-events:auto;user-select:auto;gap:10px;box-shadow:0 16px 36px 0 rgba(0, 0, 0, 0.08), 0 8px 16px 0 rgba(0, 0, 0, 0.06)"><input style=border:none;width:300px;font-size:13px;outline:none;pointer-events:auto><div style=display:flex;align-items:center;gap:16px><div style=font-size:12px;border-radius:4px;display:inline-flex;align-items:center;padding:2px;pointer-events:auto>'), _tmpl$7 = /*#__PURE__*/ template("<div>");
const LabelHeight = 21;
const InputHeight = 40;
const InputWidth = 391;
const Gap = 6;
const IconPlus = (props)=>(()=>{
        var _el$ = _tmpl$();
        spread(_el$, props, true, true);
        return _el$;
    })();
const IconCode = (props)=>(()=>{
        var _el$2 = _tmpl$2();
        spread(_el$2, props, true, true);
        return _el$2;
    })();
function Runtime(props) {
    const [mouseOverElement, setMouseOverElement] = createSignal();
    const [query, setQuery] = createSignal('');
    let inputRef;
    let injectRef;
    const containerRef = document.querySelector('#aime-wrapper');
    const [lockedElement, setLockedElement] = createSignal();
    const [lockedElementInfo, setLockedElementInfo] = createSignal(null);
    const [lockedElementRect, setLockedElementRect] = createSignal();
    const mouseOverElementRect = createMemo(()=>mouseOverElement()?.getBoundingClientRect());
    const handleMouseOver = (e)=>{
        const target = e.target;
        if (injectRef?.contains(target) || mouseOverElement() && !mouseOverElement()?.isConnected) return void setMouseOverElement(void 0);
        if (target instanceof HTMLElement) setMouseOverElement(target);
    };
    const handleClick = (e)=>{
        const target = e.target;
        const currentEle = mouseOverElement();
        if (currentEle && !injectRef?.contains(target) && !containerRef?.contains(target)) {
            e.preventDefault();
            e.stopPropagation();
            setQuery('');
            setLockedElement(currentEle);
            setLockedElementRect(currentEle.getBoundingClientRect());
            setLockedElementInfo(getElementInfo(currentEle));
        }
    };
    const handleKeydown = (ev)=>{
        if ('Escape' === ev.key) setLockedElement(void 0);
    };
    const rectStyles = createMemo(()=>{
        const rect = mouseOverElementRect();
        return {
            position: 'fixed',
            top: `${rect?.y}px`,
            left: `${rect?.x}px`,
            width: `${rect?.width}px`,
            height: `${rect?.height}px`,
            'border-color': '#33A033',
            'border-width': '2px',
            'z-index': 999999999999,
            'pointer-events': 'none'
        };
    });
    const lockedRectStyle = createMemo(()=>{
        const lockedEl = lockedElement();
        const lockedEleRect = lockedElementRect();
        if (!lockedEl || !lockedEleRect) return;
        return {
            position: 'fixed',
            top: `${lockedEleRect.y}px`,
            left: `${lockedEleRect.x}px`,
            width: `${lockedEleRect.width}px`,
            height: `${lockedEleRect.height}px`,
            'border-color': '#33A033',
            'border-width': '1px',
            'z-index': 999999999998,
            'pointer-events': 'none'
        };
    });
    const labelPosition = createMemo(()=>{
        const lockedEl = lockedElement();
        const lockedEleRect = lockedElementRect();
        if (!lockedEl || !lockedEleRect) return;
        const left = window.innerWidth - lockedEleRect.left > InputWidth ? 0 : void 0;
        const right = window.innerWidth - lockedEleRect.left > InputWidth ? void 0 : 0;
        let top = '0px';
        const topLess = lockedEleRect.top < LabelHeight + Gap;
        const bottomLess = window.innerHeight - lockedEleRect.bottom < InputHeight + Gap;
        top = topLess && bottomLess ? `${Gap}px` : topLess ? 'calc(100% + 52px)' : bottomLess ? `-${Gap + Gap + InputHeight + LabelHeight}px` : `-${LabelHeight + Gap}px`;
        return {
            top,
            left,
            right
        };
    });
    const inputPosition = createMemo(()=>{
        const lockedEl = lockedElement();
        const lockedEleRect = lockedElementRect();
        if (!lockedEl || !lockedEleRect) return;
        const left = window.innerWidth - lockedEleRect.left > InputWidth ? 0 : void 0;
        const right = window.innerWidth - lockedEleRect.left > InputWidth ? void 0 : 0;
        const bottomLess = window.innerHeight - lockedEleRect.bottom < InputHeight + Gap;
        const topLess = lockedEleRect.top < Gap + Gap + InputHeight + LabelHeight;
        let top = '0px';
        top = bottomLess && topLess ? `${Gap + Gap + LabelHeight}px` : bottomLess || topLess ? topLess ? 'calc(100% + 6px)' : `-${Gap + InputHeight}px` : 'calc(100% + 6px)';
        return {
            position: 'absolute',
            top,
            left,
            right
        };
    });
    const handleMouseOverElementScroll = ()=>{
        if (mouseOverElement()) setMouseOverElement(void 0);
    };
    const handleLockElementScroll = ()=>{
        const lockedEl = lockedElement();
        if (lockedEl?.isConnected) return void setLockedElement(void 0);
        if (lockedEl) setLockedElementRect(lockedEl.getBoundingClientRect());
    };
    const handleResize = ()=>{
        handleMouseOverElementScroll();
        handleLockElementScroll();
    };
    const handleAddUIQuery = ()=>{
        if (!query()) return;
        const info = lockedElementInfo();
        window.parent.postMessage({
            type: '__aime_add_ui_query',
            payload: {
                file_name: info?.thisElement.link?.filePath,
                component_name: info?.thisElement.label,
                line: info?.thisElement.link?.line,
                column: info?.thisElement.link?.column,
                tag_name: lockedElement()?.tagName.toLowerCase(),
                query: query(),
                id: v4()
            }
        }, '*');
        setLockedElement(void 0);
    };
    const handleJumpCode = ()=>{
        const info = lockedElementInfo();
        window.parent.postMessage({
            type: '__aime_jump_code',
            payload: {
                file_name: info?.thisElement.link?.filePath,
                line: info?.thisElement.link?.line,
                column: info?.thisElement.link?.column
            }
        }, '*');
    };
    const handleInputKeydown = (ev)=>{
        if ('Enter' === ev.key && query()) handleAddUIQuery();
    };
    onMount(()=>{
        const roots = [
            document
        ];
        document.querySelectorAll('*').forEach((node)=>{
            if ('aime-wrapper' === node.id) return;
            if (node.shadowRoot) roots.push(node.shadowRoot);
        });
        for (const root of roots){
            root.addEventListener('mouseover', handleMouseOver, {
                capture: true
            });
            root.addEventListener('click', handleClick, {
                capture: true
            });
        }
        window.addEventListener('keydown', handleKeydown);
        window.addEventListener('resize', handleResize);
        onCleanup(()=>{
            for (const root of roots){
                root.removeEventListener('mouseover', handleMouseOver, {
                    capture: true
                });
                root.removeEventListener('click', handleClick, {
                    capture: true
                });
            }
            window.removeEventListener('keydown', handleKeydown);
            window.removeEventListener('resize', handleResize);
        });
    });
    createEffect(()=>{
        const parents = [];
        let element = mouseOverElement()?.parentElement;
        while(element){
            element.addEventListener('scroll', handleMouseOverElementScroll);
            parents.push(element);
            element = element.parentElement;
        }
        onCleanup(()=>{
            parents.forEach((p)=>p.removeEventListener('scroll', handleMouseOverElementScroll));
        });
    });
    createEffect(()=>{
        const parents = [];
        let element = lockedElement()?.parentElement;
        while(element){
            element.addEventListener('scroll', handleLockElementScroll);
            parents.push(element);
            element = element.parentElement;
        }
        onCleanup(()=>{
            parents.forEach((p)=>p.removeEventListener('scroll', handleLockElementScroll));
        });
    });
    createEffect(()=>{
        const disabledElement = [];
        document.addEventListener('focusin', (event)=>{
            if (!lockedElementInfo()) return;
            if (event.target?.id !== 'aime-wrapper') {
                event.target.setAttribute('disabled', 'true');
                disabledElement.push(event.target);
                inputRef?.focus();
            }
        }, true);
        onCleanup(()=>{
            disabledElement.forEach((p)=>p.removeAttribute('disabled'));
        });
    });
    return createComponent(Show, {
        get when () {
            return mouseOverElement() || lockedElement();
        },
        get children () {
            var _el$3 = _tmpl$7();
            var _ref$ = injectRef;
            "function" == typeof _ref$ ? use(_ref$, _el$3) : injectRef = _el$3;
            insert(_el$3, createComponent(Show, {
                get when () {
                    return lockedElement();
                },
                get children () {
                    var _el$4 = _tmpl$6(), _el$6 = _el$4.firstChild, _el$7 = _el$6.firstChild, _el$8 = _el$7.nextSibling, _el$9 = _el$8.firstChild;
                    insert(_el$4, createComponent(Show, {
                        get when () {
                            return !!lockedElementInfo();
                        },
                        get children () {
                            var _el$5 = _tmpl$3();
                            insert(_el$5, ()=>lockedElementInfo()?.thisElement?.label);
                            effect((_$p)=>style(_el$5, {
                                    ...labelPosition()
                                }, _$p));
                            return _el$5;
                        }
                    }), _el$6);
                    _el$7.$$input = (v)=>{
                        setQuery(v.currentTarget.value);
                    };
                    _el$7.$$keydown = handleInputKeydown;
                    var _ref$2 = inputRef;
                    "function" == typeof _ref$2 ? use(_ref$2, _el$7) : inputRef = _el$7;
                    _el$9.$$click = handleAddUIQuery;
                    insert(_el$9, createComponent(IconPlus, {}));
                    insert(_el$8, createComponent(Show, {
                        get when () {
                            return Boolean(props.canJumpCode && lockedElementInfo());
                        },
                        get children () {
                            return [
                                _tmpl$4(),
                                (()=>{
                                    var _el$1 = _tmpl$5();
                                    _el$1.$$click = handleJumpCode;
                                    insert(_el$1, createComponent(IconCode, {}));
                                    return _el$1;
                                })()
                            ];
                        }
                    }), null);
                    effect((_p$)=>{
                        var _v$ = lockedRectStyle(), _v$2 = {
                            ...inputPosition(),
                            'z-index': MAX_ZINDEX
                        }, _v$3 = lockedElementInfo() ? "\u8F93\u5165\u63CF\u8FF0\uFF0C\u8BA9 Aime \u5E2E\u4F60\u4FEE\u6539" : "\u8BE5\u5143\u7D20\u6682\u4E0D\u652F\u6301\u5708\u9009\uFF0C\u8BF7\u8054\u7CFB oncall", _v$4 = !lockedElementInfo(), _v$5 = query() ? 'pointer' : 'not-allowed', _v$6 = query() ? '#060e1f' : '#444c5c', _v$7 = query() ? '#fff' : '#c7cdd9';
                        _p$.e = style(_el$4, _v$, _p$.e);
                        _p$.t = style(_el$6, _v$2, _p$.t);
                        _v$3 !== _p$.a && setAttribute(_el$7, "placeholder", _p$.a = _v$3);
                        _v$4 !== _p$.o && (_el$7.disabled = _p$.o = _v$4);
                        _v$5 !== _p$.i && setStyleProperty(_el$9, "cursor", _p$.i = _v$5);
                        _v$6 !== _p$.n && setStyleProperty(_el$9, "background", _p$.n = _v$6);
                        _v$7 !== _p$.s && setStyleProperty(_el$9, "color", _p$.s = _v$7);
                        return _p$;
                    }, {
                        e: void 0,
                        t: void 0,
                        a: void 0,
                        o: void 0,
                        i: void 0,
                        n: void 0,
                        s: void 0
                    });
                    effect(()=>_el$7.value = query());
                    return _el$4;
                }
            }), null);
            insert(_el$3, createComponent(Show, {
                get when () {
                    return memo(()=>!!mouseOverElement())() && lockedElement() !== mouseOverElement();
                },
                get children () {
                    var _el$10 = _tmpl$7();
                    effect((_$p)=>style(_el$10, rectStyles(), _$p));
                    return _el$10;
                }
            }), null);
            return _el$3;
        }
    });
}
function App() {
    const [showUISelect, setShowUISelect] = createSignal(false);
    const [canJumpCode, setCanJumpCode] = createSignal(false);
    const handleMessage = (event)=>{
        if ('__aime_show_ui_select' === event.data.type) setShowUISelect(event.data?.payload?.showUISelect);
        if ('__aime_can_jump_code' === event.data.type) setCanJumpCode(event.data?.payload?.canJumpCode);
        if ('aime-reload-preview' === event.data.type) window.location.reload();
    };
    onMount(()=>{
        window.addEventListener('message', handleMessage);
        return ()=>{
            window.removeEventListener('message', handleMessage);
        };
    });
    return createComponent(Show, {
        get when () {
            return showUISelect();
        },
        get children () {
            return createComponent(Runtime, {
                get canJumpCode () {
                    return canJumpCode();
                }
            });
        }
    });
}
function initRender(solidLayer) {
    window.parent.postMessage({
        type: '__aime_can_select_ui',
        payload: {
            canSelectUI: true
        }
    }, '*');
    let previousUrl = window.location.href;
    const logUrlChange = ()=>{
        const currentUrl = window.location.href;
        if (currentUrl !== previousUrl) {
            previousUrl = currentUrl;
            window.parent.postMessage({
                type: '__aime_url_change',
                payload: currentUrl
            }, '*');
        }
    };
    window.addEventListener('hashchange', logUrlChange);
    const originalPushState = history.pushState;
    const originalReplaceState = history.replaceState;
    history.pushState = function(...args) {
        const result = originalPushState.apply(this, args);
        logUrlChange();
        return result;
    };
    history.replaceState = function(...args) {
        const result = originalReplaceState.apply(this, args);
        logUrlChange();
        return result;
    };
    window.addEventListener('popstate', logUrlChange);
    setInterval(logUrlChange, 3000);
    render(()=>createComponent(App, {}), solidLayer);
}
delegateEvents([
    "keydown",
    "input",
    "click"
]);
export { initRender };
