// @byted/hooks@2.61.4/lib/useRequest/baseCache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

// 将所有的缓存收敛到一个key下面，方便清理和压缩
var BYTED_HOOKS_REQUEST_CACHE_STORAGE = 'BYTED_HOOKS_REQUEST_CACHE_STORAGE';

var BaseCache = /*#__PURE__*/function () {
  // 缓存读写都放在内存，读写同时向Storage同步
  function BaseCache() {
    var _this = this,
        _window$localStorage$,
        _window$sessionStorag;

    _classCallCheck(this, BaseCache);

    this.memoryValue = void 0;
    this.localStorageValue = void 0;
    this.sessionStorageValue = void 0;

    this.getValue = function (key, cacheStoreType) {
      switch (cacheStoreType) {
        case 'memory':
          {
            return _this.memoryValue.get(key);
          }

        case 'sessionStorage':
          {
            return _this.sessionStorageValue[key];
          }

        case 'localStorage':
          {
            return _this.localStorageValue[key];
          }

        default:
          {
            return _this.memoryValue.get(key);
          }
      }
    };

    this.setValue = function (key, data, cacheStoreType) {
      switch (cacheStoreType) {
        case 'memory':
          {
            _this.memoryValue.set(key, data);

            break;
          }

        case 'sessionStorage':
          {
            _this.sessionStorageValue[key] = data;

            _this.syncStorage(sessionStorage, _this.sessionStorageValue);

            break;
          }

        case 'localStorage':
          {
            _this.localStorageValue[key] = data;

            _this.syncStorage(localStorage, _this.localStorageValue);

            break;
          }

        default:
          {
            _this.memoryValue.set(key, data);
          }
      }
    };

    this.removeItem = function (key, cacheStoreType) {
      switch (cacheStoreType) {
        case 'memory':
          {
            _this.memoryValue.delete(key);

            break;
          }

        case 'sessionStorage':
          {
            delete _this.sessionStorageValue[key];

            _this.syncStorage(sessionStorage, _this.sessionStorageValue);

            break;
          }

        case 'localStorage':
          {
            delete _this.localStorageValue[key];

            _this.syncStorage(localStorage, _this.localStorageValue);

            break;
          }

        default:
          {
            _this.memoryValue.delete(key);
          }
      }
    };

    this.memoryValue = new Map(); // 从Storage加载缓存，不再主动对过期缓存清理，仅在初始化以及达到存储限额时对缓存对象进行压缩

    this.localStorageValue = this.compressStorage(JSON.parse((_window$localStorage$ = window.localStorage.getItem(BYTED_HOOKS_REQUEST_CACHE_STORAGE)) !== null && _window$localStorage$ !== void 0 ? _window$localStorage$ : '{}'));
    this.sessionStorageValue = this.compressStorage(JSON.parse((_window$sessionStorag = window.sessionStorage.getItem(BYTED_HOOKS_REQUEST_CACHE_STORAGE)) !== null && _window$sessionStorag !== void 0 ? _window$sessionStorag : '{}'));
  }

  _createClass(BaseCache, [{
    key: "syncStorage",
    value:
    /**
     * 同步缓存到Storage
     * @param storage
     * @param data
     * @param isRetry
     */
    function syncStorage(storage, data) {
      var isRetry = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

      try {
        storage.setItem(BYTED_HOOKS_REQUEST_CACHE_STORAGE, JSON.stringify(data));
      } catch (_) {
        // 同步失败时对缓存进行压缩并重试
        if (!isRetry) {
          storage.removeItem(BYTED_HOOKS_REQUEST_CACHE_STORAGE);
          this.syncStorage(storage, this.compressStorage(data), true);
        }
      }
    }
    /**
     * 压缩缓存对象
     * @param data
     * @returns
     */

  }, {
    key: "compressStorage",
    value: function compressStorage(data) {
      var result = {};
      Object.entries(data).forEach(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2),
            k = _ref2[0],
            d = _ref2[1];

        if (d.expireTime === -1 || d.expireTime > Date.now()) {
          result[k] = d;
        }
      });
      return result;
    }
  }]);

  return BaseCache;
}();

var _default = BaseCache;
exports.default = _default;
