// byted-tea-sdk@5.3.10/src/plugin/degrade/index.ts
// web降级策略
import { getLogPluginSpace } from '../../collect/namespace'
import { LocalForage } from '../storage/storage'
import localId from '../../util/local'

interface DegradeInfo {
  status?: string;
  data?: any;
}
class Degrade {
  public collect: any
  public config: any
  private storage: any
  private eventKey: string
  private eventUrl: string
  private errorCache: Map<string, any> = new Map()
  private sleepTime: number = 1000 // sleep时间
  private maxTime: number = 16 * 1000  // sleep最大的时间
  private minTime: number = 1000 // sleep最小的时间
  private goodCount: number = 0 // 连续code 200的次数
  private maxGoodCount: number = 2 // 连续code 200的次数
  private maxSilceCount: number = 5 // 恢复时，一次性上报的最大数据量
  private sendStatus: boolean = false
  private firstTrigger: boolean = true
  public degradeStarted: boolean = false;
  public degradeRatioStarted: boolean = false;
  private backoff_ratio: number = 0;
  private eventSleepTimeKey: string;
  private tempId: string[] = [];
  private tempRatidId: string[] = [];
  private sendTempData: any[] = [];
  public sleepFn: any;
  private sendIds: string[] = []
  apply(collect: any, config: any) {
    try {
      if (collect.destroy) return;
      this.collect = collect;
      this.config = config;
      this.storage = new LocalForage(config.app_id, 'DEGRADE_EVENT_STORE');
      this.eventUrl = this.collect.configManager.getUrl('event');
      this.eventKey = '__applog_degrade_events';
      this.eventSleepTimeKey = '__applog_degrade_events_sleep_time';
      this.collect.on('http_code_503', async (degradeInfo: DegradeInfo) => {
        await this.triggerDegrade(degradeInfo)
      })
      this.collect.on('submit-degrade', async (eventData: any) => {
        console.log('submit-degrade')
        await this.setEventStorage(eventData);
      })
      this.collect.on('http_code_200', () => {
        this.reduceSleep();
      })
      this.collect.on('submit-degrade-ratio', (eventData: any) => {
        console.log('submit-degrade-ratio', JSON.stringify(eventData))
        this.checkRatioDegrade(eventData, false);
      })
      this.collect.on('backoff-ratio', (data) => {
        if (data && data.e === 0) {
          this.backoff_ratio = data.backoff_ratio || 0
          if (data.backoff_ratio) {
            this.degradeRatioStarted = true
            this.collect.degradeRatioStarted = true
          }
        }
      })
      this.tempId = [];
      this.tempRatidId = [];
      // this.getSleepTime();
    } catch (e) {
      console.log('degrade init error', JSON.stringify(e))
    }
  }
  async getSleepTime() {
    this.sleepTime = await this.storage.getItem(this.eventSleepTimeKey) || this.sleepTime;
    console.log(this.sleepTime)
  }
  /**
  * 首次初始化降级策略
  * @param eventData - 事件数据
  */
  async triggerDegrade(degradeInfo: DegradeInfo) {
    try {
      console.log('degradeInfo', degradeInfo)
      const { data, status } = degradeInfo
      if (status === 'degrade') {
        // 是sleep时间到上报的埋点再次触发了503后，只增加时间，不做其他改动
        this.addSleep();
        return;
      }
      this.degradeStarted = true;
      this.collect.degradeStarted = true;
      await this.setEventStorage(data);
      if (this.firstTrigger) {
        // 并发时，第一次触发retry，后续的只放DB
        this.sleepRetry();
      }
      // !this.firstTrigger && this.addSleep();
      this.firstTrigger = false;
    } catch (e) {
      console.log('trigger degrade error', e.message);
    }
  }
  /**
  * 每次重新打开页面时，也要检测下当前缓存中是否数据要发送
  */
  async checkStorage() {
    try {
      const eventMap = await this.storage.getItem(this.eventKey);
      if (!eventMap || !eventMap.size) {
        return;
      }
      this.degradeStarted = true;
      this.collect.degradeStarted = true;
      this.sleepRetry();
    } catch (e) {
      console.log('check storage error', e.message)
    }
  }
  /**
   * 将事件数据存储到本地存储中，每次存储前先取
   * @param event - 需要降级事件数据
   */
  async setEventStorage(eventData: any) {
    try {
      let eventCache = await this.storage.getItem(this.eventKey);
      if (!eventCache) {
        eventCache = new Map();
      }
      const eventId = localId();
      console.log('eventS', JSON.stringify(eventData))
      eventCache.set(eventId, { eventData, id: eventId });
      this.setDataToDB(eventCache);
    } catch (e) {
      console.warn('set DB storage error', e.message);
    }
  }
  /**
  * 从本地存储中获取一部分事件数据
  * @returns 返回一个包含事件数据的数组
  */
  async getStorageformDBToSlice(isRatio: boolean) {
    try {
      const eventMap = await this.storage.getItem(this.eventKey);
      console.log('eventMap from DB', eventMap)
      let tempMap = new Map();
      // const sendTempData = [];
      if (eventMap.size <= this.maxSilceCount) {
        tempMap = eventMap;
      } else {
        for (let [id, eventData] of eventMap) {
          if (tempMap.size < this.maxSilceCount) {
            // 如果新的Map的长度小于50，就把当前项添加进去
            tempMap.set(id, eventData);
          } else {
            // 否则，就跳出循环
            break;
          }
        }
      }
      tempMap.forEach(item => {
        if (isRatio) {
          if (!this.tempRatidId.includes(item.id)) {
            this.tempRatidId.push(item.id)
            this.sendTempData.push(item.eventData[0]);
          }
        } else {
          if (!this.tempId.includes(item.id)) {
            this.tempId.push(item.id);
            this.sendTempData.push(item.eventData[0]);
          }
        }
      })
      console.log('这次取值id', this.tempRatidId)
      // return this.sendTempData;
      return { sendData: this.sendTempData, tempRatidId: this.tempRatidId, tempId: this.tempId }
    } catch (e) {
      return { sendData: [], tempRatidId: [], tempId: [] }
    }
  }
  setDataToDB(data: any) {
    try {
      this.storage.setItem(this.eventKey, data);
    } catch (e) {
      // 设置到DB内出错了，那就丢弃当前数据
      console.log('set data to DB error', e.message)
      this.dropEvents(data);
    }
  }
  // 存储超限后，需要丢弃历史埋点
  dropEvents(data: any) {
    try {
      const { user, header } = this.collect.configManager.get();
      header.app_id = 1338;
      const storageFull = [];
      data.forEach(event => {
        const { event: eventName, local_time_ms } = event;
        storageFull.push({
          eventName,
          eventTime: local_time_ms
        })
      })
      const event = {
        event: 'sdk_monitor',
        params: JSON.stringify({
          storage_fails: storageFull
        }),
        local_time_ms: Date.now(),
      };
      const sendData = {
        events: [event],
        user,
        header,
      };
      this.collect.requestManager.useRequest({
        url: this.eventUrl,
        data: sendData,
        timeout: 30000,
        type: 'event',
      })
    } catch (e) {
      // 设置到DB内出错了，那就丢弃当前数据
      console.log('drop data error', JSON.stringify(e))
    }
  }
  sleepRetry() {
    if (this.sendStatus) return;
    this.sendStatus = true;
    try {
      this.sleepFn = setTimeout(async () => {
        const { sendData } = await this.getStorageformDBToSlice(false);
        if (!sendData.length) {
          clearTimeout(this.sleepFn);
          return;
        }
        // 主动降级
        if (this.backoff_ratio !== 0 && this.backoff_ratio) {
          if (this.random() < this.backoff_ratio) {
            clearTimeout(this.sleepFn);
            return;
          }
        }
        this.fetchData(sendData);
      }, this.sleepTime)
    } catch (error) {
      console.warn('error check storage', error.message);
    }
  }
  // 主动降级的判断
  async checkRatioDegrade(data, isRetry: boolean, currendSendId?: string[]) {
    if (this.backoff_ratio !== 0 && this.backoff_ratio) {
      const rand = this.random()
      if (rand < this.backoff_ratio) {
        // 命中ratio
        console.log('backoff_ratio check', rand, this.backoff_ratio)
        if (!isRetry) {
          console.log('set data to db', JSON.stringify(data))
          await this.setEventStorage(data)
        }
        this.triggerDegradeRatio()
      } else {
        console.log('未命中ratio，直接发送数据', JSON.stringify(data))
        this.sendRatioData(data, isRetry, currendSendId)
        // this.fetchData(data)
      }
    } else {
      console.log('没有ratio，直接发送数据', JSON.stringify(data))
      this.sendRatioData(data, isRetry, currendSendId)
      // this.fetchData(data)
    }
  }
  // 等待1S后重新发送
  triggerDegradeRatio() {
    const ratioFn = setTimeout(async () => {
      const { sendData, tempRatidId } = await this.getStorageformDBToSlice(true);
      if (!sendData.length) {
        clearTimeout(ratioFn);
        return;
      }
      console.log('这次取的数据', JSON.stringify(sendData))
      this.checkRatioDegrade(sendData, true, tempRatidId)
      clearTimeout(ratioFn)
      // this.fetchData(sendData);
    }, 1000)
  }
  /**
   * 增加每次上报的延迟时间 
  */
  addSleep() {
    this.sendStatus = false;
    this.sleepTime = this.sleepTime * 2;
    if (this.sleepTime > this.maxTime) {
      this.sleepTime = this.maxTime;
    }
    this.storage.setItem(this.eventSleepTimeKey, this.sleepTime);
    clearTimeout(this.sleepFn);
    // 继续尝试发送
    this.sleepRetry();
  }
  /**
  * 减少延迟时间，并检查缓存里还有没有剩下的埋点
  */
  async reduceSleep() {
    try {
      if (!this.degradeStarted) return;
      this.sendStatus = false;
      this.goodCount += 1;
      clearTimeout(this.sleepFn);
      await this.deleteDataFromDB()
      // this.storage.setItem(this.eventKey, this.errorCache);
      if (this.goodCount === this.maxGoodCount) {
        // 连续2次200后，sleep时间减半
        this.goodCount = 0;
        this.sleepTime = this.sleepTime / 2;
      }
      // sleep时间降到最低了就恢复实时上报
      if (this.sleepTime < this.minTime) {
        // 恢复实时上报
        this.sleepTime = this.minTime;
        this.collect.degradeStarted = false;
      }
      this.storage.setItem(this.eventSleepTimeKey, this.sleepTime);
      await this.reportRest();
    } catch (e) {
      console.warn('reduceSleep error', e.message)
    }
  }
  async deleteDataFromDB() {
    console.log('成功后，删除已经发送的数据')
    const eventCache = await this.storage.getItem(this.eventKey);
    if (this.tempId.length) {
      if (eventCache) {
        this.tempId.forEach(id => {
          eventCache.delete(id);
        })
        this.tempId = [];
      }
    }
    if (this.tempRatidId.length) {
      if (eventCache) {
        const ids = []
        this.tempRatidId.forEach(id => {
          if (eventCache.has(id)) {
            eventCache.delete(id);
            ids.push(id)
          }
        })
        console.log('成功后，删除的id', ids)
        this.tempRatidId = [];
      }
    }
    if (this.tempId.length === 0 && this.tempRatidId.length === 0) {
      this.sendTempData = []
      this.sendIds = []
    }
    this.setDataToDB(eventCache);
  }
  /**
  * 继续上报剩余的数据
  * @returns 返回一个 Promise 对象
  */
  async reportRest() {
    try {
      const eventCache = await this.storage.getItem(this.eventKey);
      if (!eventCache || !eventCache.size) return;
      this.sleepRetry();
    } catch (e) {
      console.log(e.message)
    }
  }
  random() {
    return Math.floor(Math.random() * 9900);
  }
  // 发送由等比例降级存储的数据
  sendRatioData(data: any, isRetry: boolean, currendSendId?: string[]) {
    // if (currendSendId.length) {
    //   currendSendId.forEach(ids => {})
    // }
    if (currendSendId) {
      console.log('currendSendId', currendSendId)
      console.log('sendIds', this.sendIds)
      if (currendSendId.length && currendSendId === this.sendIds) return
      this.sendIds = currendSendId

    }
    this.fetchData(data, async () => {
      try {
        // 发送剩余数据
        console.log('数据发送成功', JSON.stringify(data))
        if (isRetry) {
          await this.deleteDataFromDB()
          console.log('删除后，发送剩余数据')
        }
        const eventCache = await this.storage.getItem(this.eventKey);
        console.log('剩余数据长度', eventCache?.size)
        if (!eventCache || !eventCache?.size) return;
        console.log('继续触发定时器')
        this.triggerDegradeRatio();
      } catch (e) {
        console.log('err', e.message)
      }
    })
  }
  fetchData(data: any, callback?: Function) {
    this.collect.requestManager.useRequest({
      url: this.eventUrl,
      data: data,
      timeout: 30000,
      type: 'event',
      status: 'degrade',
      success: (data) => {
        if (data && data.e === 0) {
          this.backoff_ratio = data.backoff_ratio || 0
        }
        callback && callback()
      }
    })
  }
  destroy() {
    this.collect.off('http_code_503');
    this.collect.off('http_code_200');
  }
}

/**@@SCRIPT
const exportDegrade = (collect: any, config: any) => {
  const retry = new Degrade()
  retry.apply(collect, config)
}
try {
  const pluginObject = getLogPluginSpace()
  if (pluginObject) {
    pluginObject.LogDegrade = exportDegrade
  }
} catch (e) {
  console.log(e)
}
@@SCRIPT*/

export default Degrade
