// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/element.ts
import { isVisibleByStyling, transformHumoToLine } from './util';
import { ElementInterface, ElementOptions, IConfigPage, DurationEvent, IDurationData, IConfig } from './types';
import Visual from '../visual';
import PageTracer from './pageTracer';

enum MARK_STATE {
  VISIBLE = "visible",
  INVISIBLE = "INVISIBLE",
  // ACTIVE = "active",
  // INACTIVE = 'inactive',
  EXIT = 'exit',

}
class MonitorElement implements ElementInterface {
  private key: string;
  private node: HTMLElement;
  private page: PageTracer;
  public configItem: IConfigPage;
  public isReported: boolean;
  private visualInstance: Visual;

  private observer: IntersectionObserver;

  private ratio: number;
  private isVisible: boolean;
  private enterTime: number;
  private leaveTime: number;
  private lastVisibleTime: number;
  private visibleDuration: number;
  private visibleTimer: number;

  constructor(options: ElementOptions) {
    const { key, node, configItem, page, isReported, visualInstance } = options;
    this.key = key;
    this.node = node;
    this.configItem = configItem;
    this.page = page;
    this.isReported = isReported;
    this.visualInstance = visualInstance;
    
    this.ratio = 0;
    this.isVisible = false;
    this.visibleDuration = 0;

    this.monitorIntersection();
  }

  public updatePage(pageIns: PageTracer) {
    this.page = pageIns;
  }

  private monitorIntersection() {
    try {
      const minExposureRatio = this.configItem.cond.min_exposure_ratio / 100;

      this.observer = new IntersectionObserver((entries) => {
        this.ratio = entries[0] && entries[0].intersectionRatio;
        this.checkStatus(true, entries[0] && entries[0].target as HTMLElement);
      }, {
        threshold: minExposureRatio,
      });
      this.observer.observe(this.node);
    } catch (error) {
      console.error(error);
    }
  }

  public checkStatus(lazy: boolean, node: HTMLElement) {
    if (this.configItem.cond.text_pattern && !this.isMatchText(node)) return;
    const isVisible = this.checkVisibility();
    if (isVisible) {
      this.onVisible(lazy);
    } else {
      this.onInVisible(lazy);
    }
  }

  public checkVisibility() {
    if (!isVisibleByStyling(this.node)) {
      return false;
    }
    if (this.ratio >= (this.configItem.cond.min_exposure_ratio / 100)) {
      return true;
    }
    return false;
  }

  public onVisible(lazy: boolean) {
    this.clearLazyTimer();
    if (this.isVisible) {
      return this.report;
    }
    if (lazy && this.configItem.cond.min_exposure_time) {
      this.startVisibleTimer(MARK_STATE.VISIBLE, this.configItem.cond.min_exposure_time);
    } else {
      this.mark(MARK_STATE.VISIBLE);
    }
    return this.report;
  }

  public onInVisible(lazy: boolean) {
    this.clearLazyTimer();
    if (!this.isVisible) {
      return this.report;
    }
    this.mark(MARK_STATE.INVISIBLE);
    // if (lazy && this.configItem.cond.min_exposure_time) {
    //   this.startVisibleTimer(MARK_STATE.INVISIBLE, this.configItem.cond.min_exposure_time);
    // } else {
    //   this.mark(MARK_STATE.INVISIBLE);
    // }
    return this.report;
  }

  /**
   * 
   * @param visible 设置的状态
   * @param action 页面的动作，可用于resetTime 逻辑时区分元素不可见 还是开启关联页面活跃时时页面非活跃
   * @param lazy 是否设关联关联最小可见时长 setTimeout
   */
  public setVisibleState(visible: boolean, action: 'visible' | 'invisible' | 'active' | 'inactive', lazy: boolean) {
    try {
      if (this.configItem.cond.text_pattern && !this.isMatchText(this.node)) return;
      if (visible && this.checkVisibility()) {
        this.onVisible(lazy);
      } else {
        this.onInVisible(lazy);
      }
    } catch (error) {
      console.error(error);
    }
  }

  public exit() {
    this.observer && this.observer.disconnect();
    this.mark(MARK_STATE.EXIT);
    return this.report;
  }

  public mark(state: MARK_STATE, time = Date.now()) {
    try {
      switch (state) {
        case MARK_STATE.VISIBLE: {
          if (this.isVisible) return;
          this.isVisible = true;
          this.lastVisibleTime = time;
          if (!this.enterTime) {
            this.enterTime = time;
          }
          break;
        }
        case MARK_STATE.INVISIBLE: {
          if (!this.isVisible) return;
          this.isVisible = false;
          this.visibleDuration += time - this.lastVisibleTime;
          this.leaveTime = time;
          this.generateDurationEvent();
          break;
        }
        case MARK_STATE.EXIT: {
          if (this.isVisible) {
            this.isVisible = false;
            this.visibleDuration += time - this.lastVisibleTime;
          }
          this.leaveTime = time;
          this.generateDurationEvent();
          break;
        }
        default:
          break;
      }
    } catch (error) {
      console.error(error);
    }
  }

  private startVisibleTimer(state: MARK_STATE, time: number) {
    this.clearLazyTimer();
    // const now = Date.now();
    this.visibleTimer = window.setTimeout(() => {
      // now 计算放置于定时器里面，因为超过最小可见时长时才开始计算可见时长，需满足最少可见时长条件时才标记为可见
      const now = Date.now();
      this.mark(state, now);
    }, time);
  }

  private clearLazyTimer() {
    window.clearTimeout(this.visibleTimer);
  }

  private generateDurationEvent() {
    // 该元素已经上报过埋点且对应埋点配置的上报频率为首次曝光则不再进行上报
    if (this.isReported && this.configItem.cond.frequency === 'once') {
      return;
    }
    this.isReported = true;
    const durationData = this.report;
    this.resetTimeAfterGenerate();
    this.reportEvent(transformHumoToLine(durationData) as IDurationData);
  }

  reportEvent(durationPresetData: IDurationData) {
    try {
      const eventParams = this.configItem.params && this.configItem.params.length
        ? this.visualInstance.calculateParams({
          configItem: this.configItem,
          durationData: durationPresetData,
          target: this.node,
        })
        : {};
      this.visualInstance.handleReport(this.configItem, { ...eventParams }, true);
    } catch (error) {
      console.error(error);
    }
  }

  /**
   * 用于针对不同触发时机上报埋点后，重置相关时长统计，避免下次触发上报时造成上次上报完时间未重置时间累加的情况
   * @param trigger 埋点配置的触发时机
   */
  public resetTimeAfterGenerate() {
    this.visibleDuration = 0;
    this.lastVisibleTime = undefined;
    this.enterTime = undefined;
    this.leaveTime = undefined;
  }

  /** 实际曝光的元素文本与配置下发的限定文本内容是否匹配 */
  isMatchText(target: HTMLElement) {
    const elementText = this.visualInstance.getElementText(target);
    return elementText === this.configItem.element_config.text;
  }

  get report() {
    return {
      currUrl: this.page && this.page.curPage && this.page.curPage.currUrl,
      prevUrl: this.page && this.page.curPage && this.page.curPage.prevUrl,
      pathname: this.page && this.page.curPage && this.page.curPage.curPath,
      title: document.title,
      enterTime: this.enterTime,
      leaveTime: this.leaveTime,
      visibleDuration: this.visibleDuration,
    };
  }
}

export default MonitorElement;
