// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/elementTracer.ts
import MonitorElement from './element';
import { getElementsFromXPath, getElementsFromXPathNext } from '../selector'
import localId from '../../../util/local'
import { ElementTracerOptions, ElementInfo, IConfigPage } from './types';
import PageTracer from './pageTracer';
import Visual from '../visual';

class Element {
  /** 每个埋点圈选配置会对应多个target元素（勾选同类 / selector命中多个） */
  public monitorElements: ElementInfo[];
  private configItem: IConfigPage;
  private visualInstance: Visual;
  private page: PageTracer;
  constructor(options: ElementTracerOptions) {
    try {
      this.configItem = options.config;
      this.page = options.page;
      this.visualInstance = options.visualInstance;
      this.monitorElements = [];
      this.initElementsByConfig();
    } catch (error) {
      console.error(error);
    }
  }

  public setPageOptions(pageIns: PageTracer) {
    this.page = pageIns;
    this.monitorElements.forEach(item => {
      item.observer.updatePage(pageIns);
    });
  }

  public initElementsByConfig() {
    /**
     * 1. 根据配置获取 nodes
     * 2. 遍历node，如果 monitorElements 中不存在该 node，则添加至数组中（绑定intersection）；如果存在，则检查对应status
     * 3. 根据 page onActive / onVisible / onInactive / onInvisible 事件，设置 element 的 visible 状态
     */
    if (!this.isMatchConfigUrl()) return;
    const nodes = this.getNodesByElementConfig();
    nodes.map(this.checkElement.bind(this));
  }
  //    TODO: 可以增加判断，如果已经存在element，可以先 disconnect 对应每个intersectionObserver

  /**
   * element.keys: 配置和元素属于多对多关系，一个配置对应多个元素（同类元素），一个元素上绑定了多个配置（不同配置圈选同一个元素）
   * 元素上的keys数组字段记录不同配置的key内容，用于匹配当前配置是否绑定监听逻辑
   */
  private checkElement(element) {
    try {
      if (!element.keys) {
        element.keys = [];
      }
      // 判断 element.keys 中的某一项 是否已经存在于 monitorElements 数组中
      const currentMonitorEle = this.monitorElements.find(
        (item) => element.keys.includes(item.key),
      );
      if (!currentMonitorEle) {
        const key = localId() as string;
        element.keys.push(key);
        this.addElement(element, key);
      } else {
        currentMonitorEle && currentMonitorEle.observer && currentMonitorEle.observer.checkStatus && currentMonitorEle.observer.checkStatus(true, currentMonitorEle.node);
      }
    } catch (error) {
      console.error(error);
    }
  }

  private addElement(element: HTMLElement, key: string) {
    const info = {
      key,
      node: element,
    }
    this.monitorElements.push({
      ...info,
      observer: new MonitorElement({
        ...info,
        configItem: this.configItem,
        page: this.page,
        isReported: false,
        visualInstance: this.visualInstance,
      }),
    })
  }

  public onExitElements() {
    this.monitorElements.forEach((item) => {
      item.observer && item.observer.exit();
    });
  }

  private getNodesByElementConfig(): HTMLElement[] {
      let exposureSimilar = false
      let elementInstanceList: HTMLElement[] = [];
      try {
        const { element_config, cond } = this.configItem;
        const fieldSelector = element_config.field.includes('selector') ? 'selector' : 'path';
        if (fieldSelector === 'selector') {
          if (element_config.field === 'selector') {
            if (cond.similar_pattern) {
              exposureSimilar = true
              const querySelect = document.querySelector(element_config.selector)
              elementInstanceList = querySelect ? Array.from(querySelect.querySelectorAll(cond.similar_segment_selector)) : []
            } else {
              const node = document.querySelector(element_config.selector) as HTMLElement;
              elementInstanceList = node ? [node] : [];
            }
          } else {
            // custom_selector
            exposureSimilar = true
            elementInstanceList = Array.from(document.querySelectorAll(element_config.selector))
          }
        } else {
          if (cond.similar_pattern) {
            exposureSimilar = true
            elementInstanceList = getElementsFromXPathNext(element_config.path, element_config.position)
          } else {
            elementInstanceList = getElementsFromXPath(this.configItem.element_config.path, this.configItem.element_config.position)
          }
        }
      } catch (error) {
        console.error(error);
      } finally {
        return elementInstanceList;
      }
  }

  private replaceUrl(href: string): string {
		try {
      const url = new URL(href)
      const search = url.searchParams;
      if (search.has('visual_preview')) {
        search.delete('visual_preview')
        return url.href;
      } else {
        return href;
      }
    } catch (error) {
      console.error(error);
      return href;
    }
	}

	private isMatchConfigUrl(url: string = location.href) {
		try {
      const { pathname, search, hash } = new URL(this.replaceUrl(url));
      const patternStr = pathname + search + hash;
      if (!this.configItem.cond.url || new RegExp(this.configItem.cond.url).test(patternStr)) {
        return true;
      }
      return false;
    } catch (error) {
      console.error(error);
      return false;
    }
	}
}

export default Element;
