// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/index.ts
import { StayDurationInterface, StayDurationOptions, UrlChangeEvent, UrlObj, IConfig, IConfigPage, VISUAL_COND_DURATION_TYPE, ElementInfo, LifecycleState } from './types';
import { parseQueryObject, debounce } from './util';
import { TriggeringMethod, TRIGGER_TYPE, ActivityEvent, DEBOUNCE_TIMEOUT } from './const';
import PageTracer from './pageTracer';
import ElementTracer from './elementTracer';
import Lifecycle from './lifecycle';

class StayDuration implements StayDurationInterface {
  /** 是否销毁 */
  private destroy: boolean;
  /** 用于记录每次路由更新时上次的url信息，用于同最新的url信息比较 */
  private curUrl: string;
  private visualInstance: StayDurationOptions['visualInstance'];
  private configPageList: Array<IConfigPage>;
  private lifecycle: Lifecycle;
  private options: StayDurationOptions;
  private mutationObserverInstance: MutationObserver;
  private activityEventsListenerDeb: any;

  constructor(options: StayDurationOptions) {
    try {
      this.options = options;
      this.visualInstance = options.visualInstance;
      this.configPageList = [];
      this.curUrl = window.location.href;
      this.activityEventsListenerDeb = debounce(this.activityEventsListener.bind(this), DEBOUNCE_TIMEOUT);
      // 存在停留时长类型的埋点配置时再进行相关初始化逻辑
      if (this.options.configList.length) {
        this.initLoadListener();
      }
    } catch (error) {
      console.error(error);
    }
  }

    /**
   * 添加对页面初始化 load 事件的监听
   */
    private initLoadListener() {
     try {
      // 初始化 PageLifecycle 事件监听
      const lifecycleHandle = (event: Event, state: LifecycleState) => {
        this.configPageList.forEach(item => {
        item.page && item.page.handleListenLifecycle(event, state);
        });
      }
      this.lifecycle = new Lifecycle({ lifecycleHandle });

      // 统计页面停留时长
      this.initConfigPageList();

      // 处理元素停留时长
      this.mutationObserverInstance = this.createMutationObserver();
     } catch (error) {
      console.error(error);
     }
  }

  /**
   * 根据 configList 关联 Page 信息生成 configPageList
   * 1. 统计类型为element时，同时绑定element & page，page用于element时设置关联页面活跃
   */
  initConfigPageList(): void {
    try {
      this.configPageList = this.options.configList
        .filter(item => item.trigger_type === TRIGGER_TYPE.STAY_DURATION)
        .map(item => {
          if (item.cond.duration_type === VISUAL_COND_DURATION_TYPE.ELEMENT) {
            const withElementConfig = this.buildElementInstance(item, null);
            return {
              ...withElementConfig,
              page: new PageTracer({
                config: item,
                visualInstance: this.visualInstance,
                element: withElementConfig.element,
              }),
            }
          }
          return {
            ...item,
            page: new PageTracer({
              config: item,
              visualInstance: this.visualInstance,
            }),
          }
        });
      
      if (this.configPageList.length) {
        this.addMonitorActivityEventsListener();
      }
    } catch (error) {
      console.error(error);
    }
  }

  /**
   * 处理监听 history pushState / replaceState 事件执行后的逻辑
   */
  handlePageChange(): void {
    try {
      if (this.destroy) return;
      const event = this.buildUrlChangeEvent();
      this.updateConfigPageList(event);
    } catch (error) {
      console.error(error);
    }
  }

  /**
   * 处理页面类型停留时长统计
   * 初始化 + 单页面路由发生变化时遍历列表，更新每个配置项的 page 信息
   */
  updateConfigPageList(event?: UrlChangeEvent): void {
    this.configPageList.forEach(item => {
      item.page && item.page.handlePageChange(item, event);
    })
  }

  createMutationObserver(): MutationObserver {
    const observer = new MutationObserver(
      debounce(function(mutations) {
        if (!mutations.length) return;
        this.initConfigPageListOfElement();
      }.bind(this)),
    );

    observer.observe(document.documentElement, {
      childList: true,
      subtree: true,
      attributes: true,
    });
    return observer;
  }

  /**
   * 1. 筛选过滤 DURATION_TYPE.ELEMENT
   * 2. 如果不存在 element，则初始化 ElementTracer 实例
   * 3. 如果已存在 element（后续 MutationObserver 回调触发），则执行 ElementTracer initElementsByConfig方法，用于更新configItem 对应的 monitorElements
   */
  initConfigPageListOfElement() {
    try {
      this.configPageList = this.configPageList.map(item => {
        if (item.cond.duration_type === VISUAL_COND_DURATION_TYPE.ELEMENT) {
          return this.buildElementInstance(item, item.page);
        }
        return item;
      });
    } catch (error) {
      console.error(error);
    }
  }

  buildElementInstance(item: IConfigPage, page: PageTracer) {
    if (!item.element) {
      return {
        ...item,
        element: new ElementTracer({
          config: item,
          visualInstance: this.visualInstance,
          page,
        })
      }
    } else {
      item.element.initElementsByConfig();
      return item;
    }
  }

  private activityEventsListener() {
    try {
      this.configPageList.forEach(item => {
        item.page && item.page.handleMonitorActivityEvents();
      });
    } catch (error) {
      console.error(error);
    }
  }

  /**
   * 监控用户活跃事件（鼠标、键盘及触摸事件）
   * @private
   */
  private addMonitorActivityEventsListener() {
    Object.values(ActivityEvent).forEach((value) => {
      window.addEventListener(
        value,
        this.activityEventsListenerDeb,
      );
    });
  }

  private removeMonitorActivityEventsListener() {
    Object.values(ActivityEvent).forEach((value) => {
      window.removeEventListener(
        value,
        this.activityEventsListenerDeb,
      );
    });
  }

  /** 移除 mutationObserver 以及 元素 intersectionObserver监听 */
  public onExit(): void {
    try {
      this.destroy = true;
      this.lifecycle.removePageEventListener();
      this.removeMonitorActivityEventsListener();
      this.mutationObserverInstance && this.mutationObserverInstance.disconnect();
      this.configPageList
        .filter(item => item.cond.duration_type === VISUAL_COND_DURATION_TYPE.ELEMENT)
        .forEach(item => {
          if (item.element) {
            item.element.onExitElements();
          }
        })
    } catch (error) {
      console.error(error);
    }
  }

  buildUrlChangeEvent(): UrlChangeEvent | undefined {
    let urlChangeEvent: UrlChangeEvent | undefined;
    let newUrlObject: URL | undefined;
    let oldUrlObject: URL | undefined;

    try {
      newUrlObject = new URL(window.location.href);
      oldUrlObject = new URL(this.curUrl || '');
      if (newUrlObject.href !== oldUrlObject.href) {
        const isHashChanged: boolean = newUrlObject.hash !== oldUrlObject.hash;
        const isPathChanged: boolean = newUrlObject.pathname !== oldUrlObject.pathname;

        const triggeredBy: TriggeringMethod =
        isHashChanged && isPathChanged
          ? TriggeringMethod.BOTH
          : isHashChanged
          ? TriggeringMethod.HASH
          : TriggeringMethod.HISTORY;

        urlChangeEvent = {
          curUrl: this.buildCustomUrlObject(newUrlObject),
          prevUrl: this.buildCustomUrlObject(oldUrlObject),
          triggeredBy,
        }
      } else {
        urlChangeEvent = {
          curUrl: this.buildCustomUrlObject(newUrlObject),
          prevUrl: this.buildCustomUrlObject(oldUrlObject),
        }
      }
    } catch (error) {
      console.error(
        `[VisualStayDurationError] error parsing previous URL ${this.curUrl} or current URL ${window.location.href}`,
      );
    } finally {
      this.curUrl = window.location.href;
    }
    return urlChangeEvent;
  }

  buildCustomUrlObject(originalObj: URL): UrlObj {
    return {
      href: originalObj.href,
      hash: originalObj.hash,
      pathname: originalObj.pathname,
      host: originalObj.host,
      port: originalObj.port,
      pathWithHash: originalObj.pathname + originalObj.hash,
      query: parseQueryObject(originalObj.search),
    }
  }
}

export default StayDuration;
