// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/pageTracer.ts
import Page from './page';
import { parseQueryObject, transformHumoToLine } from './util';
import { PAGE_STATS_CACHE_KEY, INACTIVE_TIMEOUT } from './const';
import { PageTracerInterface, PageTracerOptions, UrlChangeEvent, IConfig, VISUAL_COND_DURATION_TRIGGER, DurationEvent, VISUAL_COND_PAGE_KEY, IConfigPage, VISUAL_COND_DURATION_TYPE, IDurationData, LifecycleState } from './types';
import Visual from '../visual';
import ElementTracer from './elementTracer';

class Tracer implements PageTracerInterface {
	public curPage?: Page;
	private configItem: IConfig;
  private element: ElementTracer;

	private inactiveTimerHandle: number;
	private visualInstance: Visual;

	constructor(options: PageTracerOptions) {
		this.configItem = options.config;
    	this.element = options.element;
		this.element && this.element.setPageOptions(this);
		this.visualInstance = options.visualInstance;
		this.handlePageChange(options.config, options.event);
	}

	 /**
   * 处理监控用户活跃事件（鼠标、键盘及触摸事件）
   */
	 public handleMonitorActivityEvents() {
    // 如果活跃事件被触发，则清除当前的不活跃定时器并开启一个新的不活跃定时
		window.clearTimeout(this.inactiveTimerHandle);
		this.startInactiveTimer();
		this.curPage && this.curPage.onActive();
  }

	/**
   * 处理 PageLifecycle 事件监听
	 * A、B页面，若当前视图在A页面时进行关闭A，会触发 HIDDEN、TERMINATED；若当前视图在B页面时进行关闭A，会触发 TERMINATED
   * @private
   */
  public handleListenLifecycle (event: Event, state: LifecycleState) {
		switch (state) {
			case LifecycleState.HIDDEN:
				if (this.curPage) {
					this.handleHiddenState();
				}
				break;
			case LifecycleState.TERMINATED:
				if (this.curPage) {
					this.handleTerminatedState();
				}
				break;
			case LifecycleState.ACTIVE:
			case LifecycleState.PASSIVE:
				this.handlePageVisibleState(state);
				break;
		}
	}


	public handlePageChange(configItem: IConfigPage, event?: UrlChangeEvent): void {
		try {
			const pageKey = configItem.cond.page_key || VISUAL_COND_PAGE_KEY.PATHNAME;
			// 如果上次和当前对应的pageKey没有发生变化，则无需操作
			if (this.curPage && this.curPage.currUrl && event && event.curUrl && this.curPage.currUrl[pageKey] === event.curUrl[pageKey]) {
				return;
			}
			
			// 若存在上一页面，将其置为 terminated 状态，然后触发新页面的 enter 事件
			if (this.curPage) {
				this.handleHiddenState();
				this.handleTerminatedState();
			}
			this.handleCacheByPageUpdate();

			// 如果config设置了限定url条件，则判断当前url是否匹配限定url
			if (this.isMatchConfigUrl(event && event.curUrl && event.curUrl.href)) {
				this.onPageEnter(event);
			}
		} catch (error) {
			console.error(error);
		}
	}

	/** 
	 * 处理页面history更新后上报并清空缓存的埋点数据(不需要判断限定条件URL)
	 */
	private handleCacheByPageUpdate() {
		try {
			const savedData = localStorage.getItem(PAGE_STATS_CACHE_KEY);
			if (savedData) {
				this.triggerPageDurationEvent(JSON.parse(savedData));
			}
		} catch (error) {
			console.error(`[visualStayDuration] onPageEnter error: `, error);
		}

    	this.removePageStatsCache();
	}

	/**
	 * 处理新页面进入时更新Page信息
	 * @param event 
	 */
	private onPageEnter(event?: UrlChangeEvent): void {
		this.curPage = event
      ? event.curUrl &&
        new Page({
          pathname: event.curUrl.pathname,
          url: event.curUrl,
          prevUrl: event.prevUrl,
					visualInstance: this.visualInstance,
          element: this.element,
        })
      : new Page({
          pathname: window.location.pathname,
          url: {
            pathname: window.location.pathname,
            href: window.location.href,
            hash: window.location.hash,
            host: window.location.host,
            port: window.location.port,
            query: parseQueryObject(window.location.search),
            pathWithHash: window.location.pathname + window.location.hash,
          },
		  visualInstance: this.visualInstance,
          element: this.element,
        });
		this.curPage && this.curPage.onEnter();
		this.startInactiveTimer();
	}

	/**
   * 开启config对应页面非活跃定时器
   * @private
   */
	private startInactiveTimer(): void {
		// 非活跃定时器仅在页面可见时触发
		if (document.visibilityState === 'visible') {
			window.clearTimeout(this.inactiveTimerHandle);
			this.inactiveTimerHandle = window.setTimeout(() => {
				this.curPage && this.curPage.onInactive();
				if (this.configItem.cond.duration_trigger === VISUAL_COND_DURATION_TRIGGER.INACTIVE) {
          			this.triggerPageDurationEvent();
        		}
			}, this.configItem.cond.inactive_timeout || INACTIVE_TIMEOUT)
		}
	}

	  /**
   * 处理页面可见的生命周期 (passive 和 active)
   * @private
   */
		private handlePageVisibleState(state: LifecycleState.ACTIVE | LifecycleState.PASSIVE) {
			// 页面可见
			this.curPage && this.curPage.onVisible();

			if (state === LifecycleState.ACTIVE) {
				this.curPage && this.curPage.onActive();
				this.startInactiveTimer();
			}
	
			// 并且由于页面仍活跃，所以移除 localStorage 缓存
			this.removePageStatsCache();
		}

	/**
	 * 处理页面 hidden状态
	 * 网页不可见，但未被浏览器冻结，一般由用户切换到别的 tab 或最小化浏览器触发
	 */
	private handleHiddenState() {
		window.clearTimeout(this.inactiveTimerHandle);

		this.curPage && this.curPage.onInactive();
		this.curPage && this.curPage.onInvisible();

		// hidden 状态可能由于页面被回收或用户清理浏览器进程触发，所以需要将当前页面信息缓存下来，防止丢失
    this.cachePageStats();

		if (
      [VISUAL_COND_DURATION_TRIGGER.HIDDEN, VISUAL_COND_DURATION_TRIGGER.INACTIVE].includes(
        this.configItem.cond.duration_trigger,
      )
    ) {
      this.triggerPageDurationEvent();
    }
	}

	/** 
	 * 处理页面 terminated 状态
	 * 网页被浏览器卸载并从内存中清理。一般用户主动将网页关闭时触发此状态
	*/
	private handleTerminatedState() {
		const triggerList = [VISUAL_COND_DURATION_TRIGGER.EXIT];

		if (triggerList.includes(this.configItem.cond.duration_trigger)) {
			this.triggerPageDurationEvent();
		}

		delete this.curPage;

		// 如果进入 terminated 状态，说明网页正常退出，无需缓存，因此清除 localStorage
    this.removePageStatsCache();
	}

	/**
   * 触发 pageview 停留时长上报 事件
   * @param event pageview 事件
   * @see PVEvent
   * @private
   */
  private triggerPageDurationEvent(data?: { page: DurationEvent; configItem: IConfig }) {
    const { duration_trigger } = this.configItem.cond;
    
    let durationInfo: DurationEvent = null;
		let cacheConfig = null;
    if (data) {
      durationInfo = data.page;
			cacheConfig = data.configItem;
    } else {
      durationInfo = this.curPage.generateDurationEvent();
      this.curPage.resetTimeAfterGenerate(duration_trigger);
    }
    window.clearTimeout(this.inactiveTimerHandle);
    if (this.configItem.cond.duration_type === VISUAL_COND_DURATION_TYPE.PAGE) {
      this.reportEvent(transformHumoToLine(durationInfo) as IDurationData, cacheConfig);
    }
  }

  reportEvent(durationPresetData: IDurationData, cacheConfig?: IConfig) {
    const eventParams = this.configItem.params && this.configItem.params.length
      ? this.visualInstance.calculateParams({
        configItem: cacheConfig || this.configItem,
        durationData: durationPresetData,
      })
      : {};
    this.visualInstance.handleReport(this.configItem, { ...eventParams }, true);
  }

	/** 将页面数据存入 localstorage */
	private cachePageStats() {
		try {
			const value = this.curPage && this.curPage.generateDurationEvent();
			localStorage.setItem(PAGE_STATS_CACHE_KEY, JSON.stringify({ page: value, configItem: this.configItem }));
		} catch (error) {
			console.error(`[visualStayDuration] cachePageStats error: `, error);
		}
	}

	/**
   * 移除当前存在 localStorage 中的页面数据缓存
   * @private
   */
  private removePageStatsCache() {
    localStorage.removeItem(PAGE_STATS_CACHE_KEY);
  }

	private replaceUrl(href: string): string {
		const url = new URL(href)
		const search = url.searchParams;
		if (search.has('visual_preview')) {
			search.delete('visual_preview')
			return url.href;
		} else {
			return href
		}
	}

	private isMatchConfigUrl(url: string = location.href) {
		const { pathname, search, hash } = new URL(this.replaceUrl(url));
		const patternStr = pathname + search + hash;
		if (!this.configItem.cond.url || new RegExp(this.configItem.cond.url).test(patternStr)) {
			return true;
		}
		return false;
	}
}

export default Tracer;
