// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/page.ts
import Visual from '../visual';
import ElementTracer from './elementTracer';
import { PageInterface, PageOptions, UrlObj, QueryObj, DurationEvent, ElementInfo, VISUAL_COND_DURATION_TRIGGER } from './types';
import { debounce, parseQueryObject } from './util';

class Page implements PageInterface {
  private readonly pathname: string;
  public readonly prevUrl?: UrlObj;
  private readonly url: UrlObj;

  private isVisible: boolean;
  private isActive: boolean;

  private enterTime: Date;
  private leaveTime?: Date;

  private lastActiveTime?: Date;
  /** 用于标记页面标记visible状态时的时间 或 每次上报后的时间用于下次区间计算 */
  private lastVisibleTime?: Date;

  private activeDuration: number;
  private visibleDuration: number;

  private options: PageOptions;
  private visualInstance: Visual;
  private element: ElementTracer;

  constructor(options: PageOptions) {
    const { pathname, url, prevUrl } = options;
    this.pathname = pathname;
    this.url = url;
    this.prevUrl = prevUrl;
    this.options = options;
    this.visualInstance = options.visualInstance;
    this.element = options.element;
    this.initPageStats();
  }


  private initPageStats(): void {
    this.isActive = false;
    this.isVisible = false;

    // this.enterTime = new Date();
    
    this.activeDuration = 0;
    this.visibleDuration = 0;
  }

  public onEnter(): void {
    if (document.visibilityState === 'visible') {
      this.onVisible();
      this.onActive();
    }
  }

  public onActive() {
    if (this.isActive) return;
    const time = new Date();
    this.isActive = true;
    this.lastActiveTime = time;
    if (!this.enterTime) {  // 用于记录上报时机为页面不活跃情况下，页面第二次活跃时enter_time字段（每次上报都会reset time）
      this.enterTime = time;
    }
    // console.log(`%cvisible: ${new Date()}`, 'color: blue');
    this.leaveTime = undefined;
    this.toogleElementsVisible(true, true, "active");
  }

  public onInactive() {
    if (!this.isActive) return;
    const time = new Date();
    this.getLatestActiveDuration();

    this.isActive = false;
    this.leaveTime = time;  // 用于区分上报时机页面不活跃时上报埋点记录的leave_time字段为页面不活跃时间戳
    this.toogleElementsVisible(false, false, "inactive");
  }

  public onVisible() {
    if (this.isVisible) return;
    const time = new Date();
    this.isVisible = true;
    this.lastVisibleTime = time;
    if (!this.enterTime) {
      this.enterTime = time;
    }
    // console.log(`%cvisible: ${new Date()}`, 'color: blue');
    this.leaveTime = undefined;
    this.toogleElementsVisible(true, true, "visible");
  }

  public onInvisible() {
    if (!this.isVisible) return;
    const time = new Date();
    this.isVisible = false;
    this.getLatestVisibleDuration();
    this.leaveTime = time;
    this.toogleElementsVisible(false, false, "invisible");
  }

  public toogleElementsVisible(visible: boolean, lazy: boolean, action: "visible" | 'invisible' | 'active' | 'inactive') {
    try {
      this.element &&  this.element.monitorElements && this.element.monitorElements.forEach(item => {
        const bindActivityPattern = item.observer.configItem.cond.bind_activity_pattern;
        // 开启元素可见性关联页面活跃状态时，匹配 action为 active / inactive时的动作
        if (bindActivityPattern && ['active', 'inactive'].includes(action)) {
          item.observer.setVisibleState(visible, action, lazy);
        }
        // 未开启时，匹配 页面 visible / invisible 动作
        if (!bindActivityPattern && ['visible', 'invisible'].includes(action)) {
          item.observer.setVisibleState(visible, action, lazy);
        }
      });
    } catch (error) {
      console.error(error);
    }
  }

  private getLatestActiveDuration() {
    const now = new Date();
    this.activeDuration += this.isActive
      ? this.lastActiveTime
        ? now.getTime() - (this.lastActiveTime && this.lastActiveTime.getTime() || 0)
        : 0
      : 0;
    this.lastActiveTime = now;
  }

  private getLatestVisibleDuration() {
    const now = new Date();
    this.visibleDuration += this.lastVisibleTime
      ? now.getTime() - this.lastVisibleTime.getTime()
      : 0;
    this.lastVisibleTime = now;
  }

  public generateDurationEvent(): DurationEvent {
    // 用于解决上报时机：页面非活跃，页面标记为非活跃时解决visibleDuration因未触发onVisible导致为0问题
    this.getLatestActiveDuration();
    this.getLatestVisibleDuration();
  
    return {
      currUrl: this.url,
      prevUrl: this.prevUrl,
      pathname: this.pathname,
      query: this.query,
      title: document.title,
      enterTime: this.enterTime && this.enterTime.getTime(),
      leaveTime: this.leaveTime && this.leaveTime.getTime(),
      activeDuration: this.activeDuration,
      visibleDuration: this.visibleDuration,
    };
  }

  /**
   * @param trigger 埋点配置的触发时机
   */
  public resetTimeAfterGenerate(trigger: VISUAL_COND_DURATION_TRIGGER) {
    this.visibleDuration = 0;
    this.activeDuration = 0;
    // this.lastVisibleTime = undefined;  // 用于上报情况：页面非活跃，二次非活跃上报时，因为页面不触发onVisible，即 lastVisibleTime 不会重新赋值，影响上报时 duration 计算，因此不重置 lastVisibleTime
    // this.lastActiveTime = undefined;
    this.enterTime = undefined;
    this.leaveTime = undefined;
  }

  get curPath(): string {
    return this.pathname;
  }

  get query(): QueryObj {
    return parseQueryObject(location.search);
  }

  get currUrl(): UrlObj {
    return this.url;
  }
}

export default Page;
