// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/types.ts
import { TriggeringMethod, TRIGGER_TYPE } from './const';
import Visual from '../visual';
import Page from './page';
import PageTracer from './pageTracer';
import ElementTracer from './elementTracer';
import Element from './element';

export interface IConfig {
  [key: string]: any;
  event: string;
  trigger_type: TRIGGER_TYPE;
  name: string;
  element_config: {
    field: 'path_position' | 'selector' | 'custom_selector';
    path: string;
    position: string[];
    selector: string;
    text: string;
    tag: string;
  };
  cond: {
    url: string;
    similar_pattern?: boolean;
    frequency?: 'once' | 'every';
    min_exposure_ratio?: number;
    min_exposure_time?: number;
    page_key?: VISUAL_COND_PAGE_KEY;
    inactive_timeout?: number;
    duration_type?: VISUAL_COND_DURATION_TYPE;
    duration_trigger?: VISUAL_COND_DURATION_TRIGGER;
    /**
     * duration_type === element 时，
     * 是否将元素可见时长统计与页面活跃状态绑定，若开启，页面处于不活跃状态时，停止计算页面元素的可见时长
    */
    bind_activity_pattern?: boolean;
    [key: string]: any;
  }
  params: Array<{
    name: string;
    type: VISUAL_PARAM_TYPE
    value_type: VISUAL_PARAM_VALUE_TYPE;
    value: any;
    extra: Record<string, any>;
    element_config: {
      field: 'path_position' | 'selector' | 'custom_selector';
      path: string;
      position: string[];
      selector: string;
      text: string;
      tag: string;
    };
    api_config: {
      [key: string]: any;
      method: string;
      url_path: string;
    };
  }>;

}

export interface IConfigPage extends IConfig {
  /**
   * 每个config配置携带page页面信息，包含可见时长 / 活跃时长
   * （可见时长可以共用统一一个page实例信息，但是活跃时长取决于config对应的非活跃倒计时配置，因此将Page信息关联到每个config上）
   */
  page?: PageTracer;
  element?: ElementTracer;
}

export enum VISUAL_COND_PAGE_KEY {
  PATHNAME = 'pathname',
  HREF = 'HREF',
}

export enum VISUAL_PARAM_VALUE_TYPE {
  static = 1,
  picker = 2,
  preset_variable = 3,
  api = 4,
}

export enum VISUAL_PARAM_TYPE {
  integer = 1,
  float = 2,
  string = 3,
}

export enum VISUAL_COND_DURATION_TYPE {
  PAGE = 'page',
  ELEMENT = 'element'
}

export enum VISUAL_COND_DURATION_TRIGGER {
  HIDDEN = 'hidden',
  EXIT = 'exit',
  INACTIVE = 'inactive',
}

export type QueryObj = Record<string, string>;

export type UrlObj = {
  href: string;
  hash: string;
  pathname: string;
  host: string;
  port: string;
  pathWithHash: string;
  query: QueryObj;
};

export type UrlChangeEvent = {
  curUrl: UrlObj | undefined;
  prevUrl?: UrlObj;
  origEvent?: Event;
  triggeredBy?: TriggeringMethod;
};

export interface StayDurationInterface {
  handlePageChange(): void;
}

export type StayDurationOptions = {
  visualInstance: Visual;
  configList: Array<IConfigPage>;
};

export interface PageTracerInterface {

}

export interface PageTracerOptions {
  config: IConfig;
  event?: UrlChangeEvent;
  visualInstance: Visual;
  element?: ElementTracer;
}

export interface ElementTracerOptions {
  config: IConfigPage;
  visualInstance: Visual;
  page: PageTracer;
}


export interface PageInterface {
  onEnter(): void;
  onVisible(): void;
  onInvisible(): void;
  onActive(): void;
  onInactive(): void;
  generateDurationEvent(): DurationEvent;
}

export interface PageOptions {
  pathname: string;
  url: UrlObj;
  prevUrl?: UrlObj;
  visualInstance: Visual;
  element: ElementTracer;
}

export type DurationEvent = {
  currUrl: UrlObj;
  prevUrl?: UrlObj;
  pathname: string;
  title: string;
  query?: QueryObj;
  enterTime: number;
  leaveTime?: number;
  activeDuration?: number;
  visibleDuration?: number;
}

export type ElementInfo = {
  key: string;
  node: HTMLElement;
  observer: Element;
}

export interface ElementInterface {

}

export interface ElementOptions {
  key: string;
  node: HTMLElement;
  configItem: IConfigPage;
  page: PageTracer;
  /** 记录当前元素target是否已经根据对应的配置信息上报过埋点信息 */
  isReported: boolean;
  visualInstance: Visual;
}

export interface IDurationData {
  curr_url: Record<string, string>;
  prev_url?: Record<string, string>;
  title: string;
  pathname: string;
  query: Record<string, string>;
  enter_time: number;
  leave_time: number;
  visible_duration: number;
  active_duration?: number;
}
export interface LifecycleOption {
  lifecycleHandle: (event: Event, state: LifecycleState) => void;
}

export enum LifecycleState {
  ACTIVE = 'active',
  PASSIVE = 'passive',
  HIDDEN = 'hidden',
  FROZEN = 'frozen',
  TERMINATED = 'terminated',
}
