// byted-tea-sdk@5.3.10/src/plugin/visual/stayDuration/util.ts
import { QueryObj } from './types';
/**
 * Parses query object from a query string
 *
 * @param queryStr the query string being parsed
 * @return the query object with key-value pairs
 */
export const parseQueryObject = (queryStr: string) => {
    const result: QueryObj = {};
    queryStr &&
      queryStr
        .substring(1)
        .split('&')
        .forEach((pairStr) => {
          const pair = pairStr.split('=');
          pair[0] &&
            (result[decodeURIComponent(pair[0])] = pair[1]
              ? decodeURIComponent(pair[1])
              : '');
        });
  
    return result;
  };


/**
 * debounce function from ts-debounce
 * @see https://github.com/chodorowicz/ts-debounce
 */
export type Procedure = (...args: any[]) => void;

export type Options = {
  isImmediate: boolean;
};

export interface DebouncedFunction<F extends Procedure> {
  (this: any, ...args: Parameters<F>): void;
  cancel: () => void;
}

export function debounce<F extends Procedure>(
  func: F,
  waitMilliseconds = 50,
  options: Options = {
    isImmediate: false,
  },
): DebouncedFunction<F> {
  let timeoutId: ReturnType<typeof setTimeout> | undefined;

  const debouncedFunction = function (
    this: any,
    ...args: Parameters<F>
  ) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this;

    const doLater = function () {
      timeoutId = undefined;
      if (!options.isImmediate) {
        func.apply(context, args);
      }
    };

    const shouldCallNow = options.isImmediate && timeoutId === undefined;

    if (timeoutId !== undefined) {
      clearTimeout(timeoutId);
    }

    timeoutId = setTimeout(doLater, waitMilliseconds);

    if (shouldCallNow) {
      func.apply(context, args);
    }
  };

  debouncedFunction.cancel = function () {
    if (timeoutId !== undefined) {
      clearTimeout(timeoutId);
    }
  };

  return debouncedFunction;
}

export const computedStyle = (element: Element) => {
  return window.getComputedStyle(element, null);
};

export const styleProperty = (style: CSSStyleDeclaration, property: string) => {
  return style.getPropertyValue(property);
};

export const isElement = (value: Node) => {
  return (value && value.nodeType === 1) || false;
};

export const isDisplayed = (
  element: Element,
  style?: CSSStyleDeclaration,
): any => {
  if (!style) {
    style = computedStyle(element);
  }

  const display = styleProperty(style, 'display');
  if (display === 'none') {
    return false;
  }

  const parent = element.parentNode;
  return parent && isElement(parent) ? isDisplayed(parent as Element) : true;
};

export const isVisibleByStyling = (element: Element | Document) => {
  if (element == window.document) {
    return true;
  }

  if (!element || !element.parentNode) {
    return false;
  }

  const style = computedStyle(element as Element);

  const visibility = styleProperty(style, 'visibility');
  if (visibility === 'hidden' || visibility === 'collapse') {
    return false;
  }

  const opacity = styleProperty(style, 'opacity');
  if (opacity === '0') {
    return false;
  }

  return isDisplayed(element as Element, style);
};

export const transformHumoToLine = (obj: Record<string, any>) => {
  const newObj = {};
  Object.keys(obj).forEach(key => {
    const lineKey = key.replace(/([A-Z])/g,"_$1").toLowerCase();
    if (obj[key] && typeof newObj[key] === 'object') {
      newObj[lineKey] = transformHumoToLine(obj[key]);
    } else {
      newObj[lineKey] = obj[key];
    }
  });
  return newObj;
}

export const evaluteDynamicCode = (code: string) => {
  const sandbox = Object.create(null);
  const fn = new Function('sandbox', `with(sandbox){${code}}`);
  const proxy = new Proxy(sandbox, {
    has(target, key) {
      // 让动态执行的代码认为属性已存在
      return true; 
    },
    get(target, key) {
        if (key === Symbol.unscopables) return undefined
        return target[key]
    }
  });
  return fn(proxy);
}
