// @webspatial/vite-plugin@1.0.1/dist/index.js
// src/index.ts
import { mergeConfig } from "vite";
import {
  AVP,
  DEFAULT_BASE,
  getDefineByMode,
  getDefineXrEnvBase,
  getEnv,
  getFinalBase,
  getFinalOutdir,
  getReactSDKAliasReplacementByMode
} from "@webspatial/shared";
function getFinalBaseVite(...args) {
  const base = getFinalBase(...args);
  return base === DEFAULT_BASE ? `${base}/` : base;
}
function index_default(options = {}) {
  let mode = options?.mode ?? getEnv();
  let outputDir = options?.outputDir;
  console.log("\u{1F680} ~ mode:", mode);
  return [
    {
      name: "vite-plugin-webspatial-common",
      config: (config) => {
        const myConfig = {
          esbuild: {
            jsxImportSource: mode === "avp" ? "@webspatial/react-sdk/default" : "@webspatial/react-sdk/web"
          }
        };
        const finalConfig = mergeConfig(config, myConfig);
        return finalConfig;
      }
    },
    {
      name: "vite-plugin-webspatial-serve",
      apply: "serve",
      config: (userCfg) => {
        const userBase = userCfg.base;
        const finalBase = getFinalBaseVite(userBase, mode, outputDir);
        console.log("\u{1F680} ~ finalBase:", finalBase);
        const userOutDir = userCfg.build?.outDir;
        const finalOutdir = getFinalOutdir(userOutDir, mode, outputDir);
        const config = {
          define: {
            // Define environment variables for both Node and browser
            ...getDefineByMode(mode),
            ...getDefineXrEnvBase(finalBase)
          },
          resolve: {
            alias: [getReactSDKAliasReplacementByMode(mode)]
          },
          build: {
            // Set output directory
            outDir: finalOutdir
          },
          base: finalBase
        };
        console.log("\u{1F680} ~ config:", config);
        return config;
      }
    },
    {
      name: "vite-plugin-webspatial-build",
      apply: "build",
      config: (config, { command }) => {
        const userOutDir = config.build?.outDir;
        const xrEnv = getEnv();
        const userBase = config.base;
        const finalBase = getFinalBaseVite(userBase, mode, outputDir);
        const finalOutdir = getFinalOutdir(userOutDir, xrEnv, outputDir);
        return {
          base: finalBase,
          resolve: {
            alias: [getReactSDKAliasReplacementByMode(mode)]
          },
          define: {
            // Define environment variables for both Node and browser
            ...getDefineByMode(mode),
            ...getDefineXrEnvBase(finalBase)
          },
          build: {
            // Set output directory
            outDir: finalOutdir,
            // Do not empty the output directory for AVP build
            emptyOutDir: xrEnv !== AVP
            // Remove custom rollup naming logic; use Vite defaults
          }
        };
      }
    }
  ];
}
export {
  index_default as default
};
