// @rdservices/inject-runtime@0.0.17/dist/index.js
import external_generated_styles_js_default from "./_generated_styles.js";
import { fontFamily } from "./consts.js";
import { initRender } from "./components/runtime.js";
import { getElementInfo } from "./adapters/index.js";
const MAX_ZINDEX = 2147483647;
function initRuntime() {
    if ('undefined' == typeof window || 'undefined' == typeof document) return;
    if (document.getElementById('aime-wrapper')) return;
    const style = document.createElement('style');
    style.id = 'aime-style';
    style.innerHTML = `
      #aime-layer {
        all: initial;
        pointer-events: none;
        font-family: ${fontFamily};
      }
      #aime-layer * {
        box-sizing: border-box;
      }
      #aime-labels-wrapper {
        display: flex;
        gap: 8px;
      }
      .aime-tree-node:hover {
        background-color: #eee;
      }
      ${external_generated_styles_js_default}
    `;
    const globalStyle = document.createElement('style');
    globalStyle.id = 'aime-global-style';
    globalStyle.innerHTML = `
      #aime-wrapper {
        z-index: ${MAX_ZINDEX};
        // pointer-events: none;
        position: fixed;
      }
      .aime-active-pointer * {
        cursor: pointer !important;
      }
    `;
    const wrapper = document.createElement('div');
    wrapper.setAttribute('id', 'aime-wrapper');
    wrapper.setAttribute('tabindex', '-1');
    const shadow = wrapper.attachShadow({
        mode: 'open'
    });
    const layer = document.createElement('div');
    layer.setAttribute('id', 'aime-layer');
    shadow.appendChild(style);
    shadow.appendChild(layer);
    document.body.appendChild(wrapper);
    document.head.appendChild(globalStyle);
    initRender(layer);
}
window.___aime_get_element_info_by_pos = (pos)=>{
    const { x = 0, y = 0 } = pos ?? {};
    const element = document.elementFromPoint(x, y);
    const elementInfo = element ? getElementInfo(element) : null;
    return {
        x,
        y,
        elementInfo: elementInfo ? {
            thisElement: elementInfo.thisElement,
            componentBox: element?.getBoundingClientRect()?.toJSON(),
            htmlElement: {
                tagName: elementInfo.htmlElement.tagName.toLowerCase(),
                id: elementInfo.htmlElement.id,
                className: elementInfo.htmlElement.className
            }
        } : null
    };
};
export { MAX_ZINDEX, initRuntime };
