// @webspatial/core-sdk@1.7.0/dist/index.js

    (function(){
      if(typeof window === 'undefined') return;
      if(!window.__webspatialsdk__) window.__webspatialsdk__ = {}
      window.__webspatialsdk__['core-sdk-version'] = "1.7.0"
  })()
    
var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/runtime/userAgent.ts
function parseShellToken(ua) {
  const ws = /\bWSAppShell\/([\w.-]+)/i.exec(ua);
  if (ws?.[1]) return { version: ws[1], source: "wsapp" };
  const pico = /\bPicoWebApp\/([\w.-]+)/i.exec(ua);
  if (pico?.[1]) return { version: pico[1], source: "picoapp" };
  return { version: null, source: null };
}
function inferPicoOs(ua) {
  return /\bPicoWebApp\//i.test(ua) || /\bPicoBrowser\b/i.test(ua);
}
function inferVisionOsFromUa(ua) {
  return /Mac OS X/i.test(ua);
}
function computeRuntimeFromUserAgent(userAgent) {
  if (userAgent === void 0 || userAgent === "") {
    return { type: null, shellVersion: null };
  }
  if (userAgent.includes("Puppeteer")) {
    const { version: version2 } = parseShellToken(userAgent);
    return { type: "puppeteer", shellVersion: version2 };
  }
  const { version, source } = parseShellToken(userAgent);
  if (!version) {
    return { type: null, shellVersion: null };
  }
  if (source === "picoapp" || inferPicoOs(userAgent)) {
    return { type: "picoos", shellVersion: version };
  }
  if (source === "wsapp") {
    if (inferVisionOsFromUa(userAgent)) {
      return { type: "visionos", shellVersion: version };
    }
    return { type: null, shellVersion: version };
  }
  return { type: null, shellVersion: version };
}

// src/runtime/jsbAdapterPlatform.ts
function getWebSpatialVersion(ua) {
  const match = ua.match(/WebSpatial\/(\d+)\.(\d+)\.(\d+)/);
  if (!match) {
    return null;
  }
  return [Number(match[1]), Number(match[2]), Number(match[3])];
}
function isVersionGreater(a, b) {
  if (!a) {
    return false;
  }
  for (let index = 0; index < 3; index += 1) {
    const diff = a[index] - b[index];
    if (diff > 0) {
      return true;
    }
    if (diff < 0) {
      return false;
    }
  }
  return false;
}
function resolveJsbAdapterPlatform(userAgent) {
  const rt = computeRuntimeFromUserAgent(userAgent);
  if (rt.type === "puppeteer") {
    return "puppeteer";
  }
  const webSpatialVersion = getWebSpatialVersion(userAgent);
  if (userAgent.includes("PicoWebApp") && isVersionGreater(webSpatialVersion, [0, 0, 1])) {
    return "picoos";
  }
  return "visionos";
}

// src/ssr-polyfill.ts
var isSSR = typeof window === "undefined";
var isSSREnv = () => isSSR;

// src/platform-adapter/CommandResultUtils.ts
function CommandResultSuccess(data) {
  return {
    success: true,
    data,
    errorCode: "",
    errorMessage: ""
  };
}
function CommandResultFailure(errorCode, errorMessage = "") {
  return {
    success: false,
    data: void 0,
    errorCode,
    errorMessage
  };
}

// src/platform-adapter/spatialSceneQuery.ts
function buildSpatialSceneQuery(url, config) {
  const params = { url, config };
  return Object.keys(params).map((key) => {
    const value = params[key];
    const finalValue = typeof value === "object" ? JSON.stringify(value) : value;
    return `${key}=${encodeURIComponent(finalValue)}`;
  }).join("&");
}

// src/SpatialWebEvent.ts
var SpatialWebEvent = class _SpatialWebEvent {
  static eventReceiver = {};
  static init() {
    window.__SpatialWebEvent = ({ id, data }) => {
      _SpatialWebEvent.eventReceiver[id]?.(data);
    };
  }
  static addEventReceiver(id, callback) {
    _SpatialWebEvent.eventReceiver[id] = callback;
  }
  static removeEventReceiver(id) {
    delete _SpatialWebEvent.eventReceiver[id];
  }
};

// src/platform-adapter/pico-os/PicoOSPlatform.ts
var requestId = 0;
var MAX_ID = 1e5;
var DEFAULT_JSB_ERROR_CODE = "E_PICO_JSB";
var DEFAULT_JSB_ERROR_MESSAGE = "Pico JSB execution failed";
function nextRequestId() {
  requestId = (requestId + 1) % MAX_ID;
  return `rId_${requestId}`;
}
var PicoOSPlatform = class {
  async callJSB(cmd, msg) {
    return new Promise((resolve) => {
      try {
        const rId = nextRequestId();
        SpatialWebEvent.addEventReceiver(rId, (result) => {
          SpatialWebEvent.removeEventReceiver(rId);
          resolve(this.toCommandResult(result));
        });
        const ans = window.webspatialBridge.postMessage(rId, cmd, msg);
        if (ans !== "") {
          SpatialWebEvent.removeEventReceiver(rId);
          resolve(this.parseBridgeResponse(ans));
        }
      } catch (error) {
        console.error(`SwanPlatform cmd: ${cmd}, msg: ${msg} error: ${error}`);
        const { code, message } = this.parseJSBError(error);
        resolve(CommandResultFailure(code, message));
      }
    });
  }
  openSpatialSceneSync(url, config, target, features) {
    const query = buildSpatialSceneQuery(url, config);
    const { spatialId: id = "", windowProxy } = this.openWindow(
      "createSpatialScene",
      query,
      target,
      features
    );
    return CommandResultSuccess({ windowProxy, id });
  }
  createNativeSpatialDiv() {
    return this.waitForRidProtocolAsync("createSpatialized2DElement");
  }
  createNativeAttachment(_options) {
    return this.waitForRidProtocolAsync("createAttachment");
  }
  /**
   * Async path for createNativeSpatialDiv / createNativeAttachment: open webspatial URL
   * with rid= correlation (Pico OS 6).
   */
  waitForRidProtocolAsync(command) {
    return new Promise((resolve) => {
      const createdId = nextRequestId();
      try {
        let windowProxy = null;
        SpatialWebEvent.addEventReceiver(
          createdId,
          (result) => {
            resolve(
              CommandResultSuccess({
                windowProxy,
                id: result.spatialId
              })
            );
            SpatialWebEvent.removeEventReceiver(createdId);
          }
        );
        windowProxy = this.openWindow(command, "rid=" + createdId).windowProxy;
      } catch (error) {
        const { code, message } = this.parseJSBError(error);
        SpatialWebEvent.removeEventReceiver(createdId);
        resolve(CommandResultFailure(code, message));
      }
    });
  }
  openWindow(command, query, target, features) {
    const url = query ? `webspatial://${command}?${query}` : `webspatial://${command}`;
    const windowProxy = window.open(url, target, features);
    return { spatialId: "", windowProxy };
  }
  parseBridgeResponse(ans) {
    try {
      const result = JSON.parse(ans);
      return this.toCommandResult(result);
    } catch {
      return CommandResultFailure(
        DEFAULT_JSB_ERROR_CODE,
        "Invalid Pico JSB response payload"
      );
    }
  }
  toCommandResult(result) {
    if (result.success) {
      return CommandResultSuccess(result.data);
    }
    const { code, message } = this.parseJSBError(result.data);
    return CommandResultFailure(code, message);
  }
  parseJSBError(error) {
    return {
      code: error?.code ?? DEFAULT_JSB_ERROR_CODE,
      message: error?.message ?? (typeof error === "string" ? error : DEFAULT_JSB_ERROR_MESSAGE)
    };
  }
};

// src/platform-adapter/puppeteer/PuppeteerPlatform.ts
var PuppeteerPlatform = class {
  // store iframe instance
  iframeRegistry = /* @__PURE__ */ new Map();
  constructor() {
  }
  callJSB(cmd, msg) {
    return new Promise((resolve) => {
      try {
        if (window.__handleJSBMessage) {
          try {
            console.log(` core-sdk Puppeteer Platform: callJSB: ${cmd}::${msg}`);
            const result = window.__handleJSBMessage(`${cmd}::${msg}`);
            console.log(
              ` core-sdk Puppeteer Platform callJSB result: ${result}`
            );
            resolve(CommandResultSuccess(result));
          } catch (err) {
            resolve(CommandResultFailure("500", "JSB execution error"));
          }
        } else {
          resolve(CommandResultSuccess("ok"));
        }
      } catch (error) {
        console.error(
          `PuppeteerPlatform cmd Error: ${cmd}, msg: ${msg} error: ${error}`
        );
        resolve(CommandResultFailure("500", "Internal error"));
      }
    });
  }
  /**
   * Synchronously create Spatialized2DElement to Puppeteer Runner
   */
  createSpatializedElementSync(spatialId, webspatialUrl) {
    try {
      console.log(
        `[Puppeteer Platform] Creating spatialized element sync with id: ${spatialId}, url: ${webspatialUrl}`
      );
      const win = window;
      if (win.__handleJSBMessage) {
        const createCommand = {
          id: spatialId,
          url: webspatialUrl
        };
        win.__handleJSBMessage(
          `CreateSpatialized2DElement::${JSON.stringify(createCommand)}`
        );
      }
    } catch (error) {
      console.error("Error creating spatialized element sync:", error);
    }
  }
  openSpatialSceneSync(url, config, target, features) {
    const query = buildSpatialSceneQuery(url, config);
    return this.runProtocolSync("createSpatialScene", query, target, features);
  }
  createNativeSpatialDiv() {
    return this.runProtocolAsync(
      "createSpatialized2DElement",
      "",
      void 0,
      void 0
    );
  }
  createNativeAttachment(_options) {
    return this.runProtocolAsync("createAttachment", "", void 0, void 0);
  }
  runProtocolAsync(command, query, target, features) {
    console.log(
      `PuppeteerPlatform: Calling webspatial protocol: webspatial://${command}${query ? `?${query}` : ""}`
    );
    return new Promise((resolve) => {
      try {
        const webspatialUrl = `webspatial://${command}${query ? `?${query}` : ""}`;
        const { spatialId, iframe, windowProxy } = this.createIframeWindow(
          webspatialUrl,
          target,
          features
        );
        if (command === "createSpatialized2DElement") {
          this.createSpatializedElementSync(spatialId, webspatialUrl);
        }
        console.log(
          `[Puppeteer Platform] iframe created with spatialId: ${spatialId}`
        );
        this.iframeRegistry.set(spatialId, iframe);
        resolve(CommandResultSuccess({ windowProxy, id: spatialId }));
      } catch (error) {
        console.error("Error calling webspatial protocol:", error);
        resolve(
          CommandResultFailure("500", "Failed to call webspatial protocol")
        );
      }
    });
  }
  runProtocolSync(command, query, target, features) {
    try {
      const webspatialUrl = `webspatial://${command}${query ? `?${query}` : ""}`;
      console.log(`Calling webspatial protocol sync: ${webspatialUrl}`);
      const { spatialId, iframe, windowProxy } = this.createIframeWindow(
        webspatialUrl,
        target,
        features
      );
      if (command === "createSpatialized2DElement") {
        this.createSpatializedElementSync(spatialId, webspatialUrl);
      }
      this.iframeRegistry.set(spatialId, iframe);
      return CommandResultSuccess({ windowProxy, id: spatialId });
    } catch (error) {
      console.error("Error calling webspatial protocol sync:", error);
      return CommandResultFailure(
        "500",
        "Failed to call webspatial protocol sync"
      );
    }
  }
  /**
   * Synchronously create iframe-based window
   */
  createIframeWindow(url, target, features) {
    const iframe = document.createElement("iframe");
    iframe.style.border = "none";
    iframe.style.display = "none";
    iframe.style.width = "100%";
    iframe.style.height = "100%";
    const spatialId = this.generateUUID();
    iframe.spatialId = spatialId;
    iframe.id = `spatial-iframe-${spatialId}`;
    const featuresObj = this.parseFeatures(features || "");
    if (featuresObj.width) {
      iframe.style.width = featuresObj.width;
    }
    if (featuresObj.height) {
      iframe.style.height = featuresObj.height;
    }
    if (featuresObj.left) {
      iframe.style.left = featuresObj.left;
      iframe.style.position = "absolute";
    }
    if (featuresObj.top) {
      iframe.style.top = featuresObj.top;
      iframe.style.position = "absolute";
    }
    document.body.appendChild(iframe);
    const windowProxy = this.createEnhancedWindowProxy(iframe, url, spatialId);
    iframe.src = "about:blank";
    console.log(
      `PuppeteerPlatform created iframe window with spatialId: ${spatialId}, URL: ${url}`
    );
    this.initializeIframeContent(iframe, url, spatialId);
    return { spatialId, iframe, windowProxy };
  }
  /**
   * create enhanced windowProxy object
   */
  createEnhancedWindowProxy(iframe, url, spatialId) {
    return {
      // basic properties
      location: {
        href: url,
        toString: () => url,
        reload: () => {
          if (iframe.contentWindow) {
            iframe.contentWindow.location.reload();
          }
        }
      },
      navigator: {
        userAgent: `Mozilla/5.0 (WebKit) SpatialId/${spatialId}`
      },
      // methods
      close: () => {
        console.log(`Closing iframe with spatialId: ${spatialId}`);
        iframe.remove();
        this.iframeRegistry.delete(spatialId);
      },
      // document access
      document: iframe.contentDocument || {},
      contentWindow: iframe.contentWindow || {},
      // add message communication method
      postMessage: (message, targetOrigin) => {
        if (iframe.contentWindow) {
          iframe.contentWindow.postMessage(message, targetOrigin || "*");
        }
      },
      // add event listener method
      addEventListener: (type, listener) => {
        if (iframe.contentWindow) {
          iframe.contentWindow.addEventListener(type, listener);
        }
      },
      removeEventListener: (type, listener) => {
        if (iframe.contentWindow) {
          iframe.contentWindow.removeEventListener(type, listener);
        }
      },
      // execute JavaScript
      executeScript: (code) => {
        if (iframe.contentWindow) {
          try {
            const win = iframe.contentWindow;
            return win.eval(code);
          } catch (error) {
            console.error(
              `Error executing script in iframe ${spatialId}:`,
              error
            );
            return null;
          }
        }
        return null;
      },
      // get iframe reference
      getIframe: () => iframe,
      // get spatialId
      getSpatialId: () => spatialId
    };
  }
  /**
   * initialize iframe content
   */
  initializeIframeContent(iframe, url, spatialId) {
    try {
      iframe.onload = () => {
        try {
          const iframeContent = `
            // inject communication script
            window.webSpatialId = '${spatialId}';
            window.SpatialId = '${spatialId}';
            
            // override window.open to support webspatial protocol
            const originalOpen = window.open;
            window.open = function(url, target, features) {
              if (url && url.startsWith('webspatial://')) {
                // handle webspatial protocol through windowProxy
                const windowProxy = new Proxy({}, {
                  get: function(target, prop) {
                    if (prop === 'toString') {
                      return function() { return url; };
                    }
                    return undefined;
                  }
                });
                return windowProxy;
              }
              return originalOpen.call(window, url, target, features);
            };
            
            // set navigator.userAgent to identify webspatial environment
            Object.defineProperty(navigator, 'userAgent', {
              value: 'WebSpatial/1.0 ' + navigator.userAgent,
              configurable: true
            });
            
            // send loaded message
            window.parent.postMessage({
              type: 'iframe_loaded',
              spatialId: '${spatialId}',
              url: '${url}'
            }, '${window.location.origin}');
            
            // set message handler
            window.addEventListener('message', (event) => {
              if (event.origin !== window.parent.location.origin) return;
              
              const data = event.data;
              if (data && data.type === 'webspatial_command') {
                // handle command from parent window
                console.log('Received command in iframe from parent:', data.command);
                // add command handling logic here
              }
            });
          `;
          const doc = iframe.contentDocument;
          if (doc) {
            doc.open();
            doc.write(`
              <!DOCTYPE html>
              <html>
              <head>
                <title>Spatial Iframe - ${spatialId}</title>
                <meta charset="UTF-8">
                <style>
                  body {
                    margin: 0;
                    padding: 0;
                    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
                  }
                </style>
              </head>
              <body>
                <script>${iframeContent}</script>
              </body>
              </html>
            `);
            doc.close();
          }
        } catch (error) {
          console.error("Error initializing iframe content:", error);
        }
      };
    } catch (error) {
      console.error("Error setting up iframe:", error);
    }
  }
  /**
   * parse features string to object
   */
  parseFeatures(features) {
    const result = {};
    const pairs = features.split(",");
    pairs.forEach((pair) => {
      const [key, value] = pair.split("=").map((s) => s.trim());
      if (key && value) {
        result[key] = value;
      }
    });
    return result;
  }
  /**
   * send message to iframe with specified spatialId
   */
  sendMessageToIframe(spatialId, message) {
    const iframe = this.iframeRegistry.get(spatialId);
    if (iframe && iframe.contentWindow) {
      iframe.contentWindow.postMessage(message, window.location.origin);
      return true;
    }
    return false;
  }
  /**
   * get all active iframes
   */
  getAllActiveIframes() {
    const result = [];
    this.iframeRegistry.forEach((iframe, spatialId) => {
      result.push({ spatialId, iframe });
    });
    return result;
  }
  /**
   * dispose all active iframes
   */
  dispose() {
    this.iframeRegistry.forEach((iframe, spatialId) => {
      console.log(`Disposing iframe with spatialId: ${spatialId}`);
      iframe.remove();
    });
    this.iframeRegistry.clear();
  }
  // generate UUID function
  generateUUID() {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
      /[xy]/g,
      function(c) {
        const r = Math.random() * 16 | 0;
        const v = c === "x" ? r : r & 3 | 8;
        return v.toString(16).toUpperCase();
      }
    );
  }
};

// src/platform-adapter/ssr/SSRPlatform.ts
var SSRPlatform = class {
  callJSB(cmd, msg) {
    return Promise.resolve({
      success: true,
      data: void 0,
      errorCode: void 0,
      errorMessage: void 0
    });
  }
  openSpatialSceneSync(_url, _config, _target, _features) {
    return {
      success: true,
      data: void 0,
      errorCode: void 0,
      errorMessage: void 0
    };
  }
  createNativeSpatialDiv() {
    return Promise.resolve({
      success: true,
      data: void 0,
      errorCode: void 0,
      errorMessage: void 0
    });
  }
  createNativeAttachment(_options) {
    return Promise.resolve({
      success: true,
      data: void 0,
      errorCode: void 0,
      errorMessage: void 0
    });
  }
};

// src/platform-adapter/vision-os/VisionOSPlatform.ts
var UUID_RE = /\b([0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12})\b/gi;
var VisionOSPlatform = class {
  async callJSB(cmd, msg) {
    try {
      const result = await window.webkit.messageHandlers.bridge.postMessage(
        `${cmd}::${msg}`
      );
      return CommandResultSuccess(result);
    } catch (error) {
      const { code, message } = this.parseJSBError(error);
      return CommandResultFailure(code, message);
    }
  }
  openSpatialSceneSync(url, config, target, features) {
    const query = buildSpatialSceneQuery(url, config);
    const { spatialId: id = "", windowProxy } = this.openWindow(
      "createSpatialScene",
      query,
      target,
      features
    );
    return CommandResultSuccess({ windowProxy, id });
  }
  createNativeSpatialDiv() {
    return this.openProtocolAsync("createSpatialized2DElement");
  }
  createNativeAttachment(_options) {
    return this.openProtocolAsync("createAttachment");
  }
  async openProtocolAsync(command) {
    const { spatialId: id = "", windowProxy } = this.openWindow(
      command,
      "",
      void 0,
      void 0
    );
    return CommandResultSuccess({ windowProxy, id });
  }
  openWindow(command, query, target, features) {
    const windowProxy = window.open(
      `webspatial://${command}?${query || ""}`,
      target,
      features
    );
    const ua = windowProxy?.navigator.userAgent;
    const spatialId = ua?.match(UUID_RE)?.[0];
    return { spatialId, windowProxy };
  }
  parseJSBError(error) {
    try {
      const parsed = JSON.parse(error.message ?? "");
      return {
        code: parsed.code ?? "E_VISIONOS_JSB",
        message: parsed.message ?? "VisionOS JSB execution failed"
      };
    } catch {
      return {
        code: "E_VISIONOS_JSB",
        message: error?.message ?? "VisionOS JSB execution failed"
      };
    }
  }
};

// src/platform-adapter/createPlatformSync.ts
function assertNever(_) {
  throw new Error("Unhandled jsb adapter platform kind");
}
function createPlatformSync() {
  if (isSSREnv()) {
    return new SSRPlatform();
  }
  const userAgent = window.navigator.userAgent;
  const kind = resolveJsbAdapterPlatform(userAgent);
  switch (kind) {
    case "puppeteer":
      return new PuppeteerPlatform();
    case "picoos":
      return new PicoOSPlatform();
    case "visionos":
      return new VisionOSPlatform();
    default:
      return assertNever(kind);
  }
}

// src/platform-runtime.ts
var platformResolved;
function getPlatformSync() {
  platformResolved ??= createPlatformSync();
  return platformResolved;
}
async function getPlatform() {
  return getPlatformSync();
}

// src/utils.ts
function parseBorderRadius(borderProperty, width) {
  if (borderProperty === "") {
    return 0;
  }
  if (borderProperty.endsWith("%")) {
    return width * parseFloat(borderProperty) / 100;
  }
  return parseFloat(borderProperty);
}
function parseCornerRadius(computedStyle) {
  const width = parseFloat(computedStyle.getPropertyValue("width"));
  const topLeftPropertyValue = computedStyle.getPropertyValue(
    "border-top-left-radius"
  );
  const topRightPropertyValue = computedStyle.getPropertyValue(
    "border-top-right-radius"
  );
  const bottomLeftPropertyValue = computedStyle.getPropertyValue(
    "border-bottom-left-radius"
  );
  const bottomRightPropertyValue = computedStyle.getPropertyValue(
    "border-bottom-right-radius"
  );
  const cornerRadius = {
    topLeading: parseBorderRadius(topLeftPropertyValue, width),
    bottomLeading: parseBorderRadius(bottomLeftPropertyValue, width),
    topTrailing: parseBorderRadius(topRightPropertyValue, width),
    bottomTrailing: parseBorderRadius(bottomRightPropertyValue, width)
  };
  return cornerRadius;
}
function composeSRT(position, rotation, scale) {
  const { x: px, y: py, z: pz } = position;
  const { x: rx, y: ry, z: rz } = rotation;
  const { x: sx, y: sy, z: sz } = scale;
  let m = new DOMMatrix();
  m = m.translate(px, py, pz);
  m = m.rotate(rx, ry, rz);
  m = m.scale(sx, sy, sz);
  return m;
}
function deepCloneJSON(value) {
  const sc = globalThis.structuredClone;
  if (typeof sc === "function") {
    try {
      return sc(value);
    } catch {
    }
  }
  return JSON.parse(JSON.stringify(value));
}

// src/JSBCommand.ts
var JSBCommand = class {
  commandType = "";
  async execute() {
    const param = this.getParams();
    const msg = param ? JSON.stringify(param) : "";
    const platform = await getPlatform();
    return platform.callJSB(this.commandType, msg);
  }
};
var UpdateEntityPropertiesCommand = class extends JSBCommand {
  constructor(entity, properties) {
    super();
    this.entity = entity;
    this.properties = properties;
  }
  commandType = "UpdateEntityProperties";
  getParams() {
    const transform = composeSRT(
      this.properties.position ?? this.entity.position,
      this.properties.rotation ?? this.entity.rotation,
      this.properties.scale ?? this.entity.scale
    ).toFloat64Array();
    return {
      entityId: this.entity.id,
      transform
    };
  }
};
var UpdateEntityEventCommand = class extends JSBCommand {
  constructor(entity, type, isEnable) {
    super();
    this.entity = entity;
    this.type = type;
    this.isEnable = isEnable;
  }
  commandType = "UpdateEntityEvent";
  getParams() {
    return {
      type: this.type,
      entityId: this.entity.id,
      isEnable: this.isEnable
    };
  }
};
var UpdateSpatialSceneProperties = class extends JSBCommand {
  properties;
  commandType = "UpdateSpatialSceneProperties";
  constructor(properties) {
    super();
    this.properties = properties;
  }
  getParams() {
    return this.properties;
  }
};
var UpdateSceneConfig = class extends JSBCommand {
  config;
  commandType = "UpdateSceneConfig";
  constructor(config) {
    super();
    this.config = config;
  }
  getParams() {
    return { config: this.config };
  }
};
var FocusScene = class extends JSBCommand {
  constructor(id) {
    super();
    this.id = id;
  }
  commandType = "FocusScene";
  getParams() {
    return { id: this.id };
  }
};
var GetSpatialSceneState = class extends JSBCommand {
  commandType = "GetSpatialSceneState";
  constructor() {
    super();
  }
  getParams() {
    return {};
  }
};
var SpatializedElementCommand = class extends JSBCommand {
  constructor(spatialObject) {
    super();
    this.spatialObject = spatialObject;
  }
  getParams() {
    const extraParams = this.getExtraParams();
    return { id: this.spatialObject.id, ...extraParams };
  }
};
var UpdateSpatialized2DElementProperties = class extends SpatializedElementCommand {
  properties;
  commandType = "UpdateSpatialized2DElementProperties";
  constructor(spatialObject, properties) {
    super(spatialObject);
    this.properties = properties;
  }
  getExtraParams() {
    return this.properties;
  }
};
var UpdateSpatializedDynamic3DElementProperties = class extends SpatializedElementCommand {
  properties;
  commandType = "UpdateSpatializedDynamic3DElementProperties";
  constructor(spatialObject, properties) {
    super(spatialObject);
    this.properties = properties;
  }
  getExtraParams() {
    return {
      id: this.spatialObject.id,
      ...this.properties
    };
  }
};
var UpdateUnlitMaterialProperties = class extends SpatializedElementCommand {
  properties;
  commandType = "UpdateUnlitMaterialProperties";
  constructor(spatialObject, properties) {
    super(spatialObject);
    this.properties = properties;
  }
  getExtraParams() {
    return this.properties;
  }
};
var UpdateSpatializedElementTransform = class extends SpatializedElementCommand {
  matrix;
  commandType = "UpdateSpatializedElementTransform";
  constructor(spatialObject, matrix) {
    super(spatialObject);
    this.matrix = matrix;
  }
  getExtraParams() {
    return { matrix: Array.from(this.matrix.toFloat64Array()) };
  }
};
var UpdateSpatializedStatic3DElementProperties = class extends SpatializedElementCommand {
  properties;
  commandType = "UpdateSpatializedStatic3DElementProperties";
  constructor(spatialObject, properties) {
    super(spatialObject);
    this.properties = properties;
  }
  getExtraParams() {
    return this.properties;
  }
};
var AddSpatializedElementToSpatialized2DElement = class extends SpatializedElementCommand {
  commandType = "AddSpatializedElementToSpatialized2DElement";
  spatializedElement;
  constructor(spatialObject, spatializedElement) {
    super(spatialObject);
    this.spatializedElement = spatializedElement;
  }
  getExtraParams() {
    return { spatializedElementId: this.spatializedElement.id };
  }
};
var AddSpatializedElementToSpatialScene = class extends JSBCommand {
  commandType = "AddSpatializedElementToSpatialScene";
  spatializedElement;
  constructor(spatializedElement) {
    super();
    this.spatializedElement = spatializedElement;
  }
  getParams() {
    return {
      spatializedElementId: this.spatializedElement.id
    };
  }
};
var CreateSpatializedStatic3DElementCommand = class extends JSBCommand {
  constructor(modelURL, sources, loading = "eager") {
    super();
    this.modelURL = modelURL;
    this.sources = sources;
    this.loading = loading;
    this.modelURL = modelURL;
    this.sources = sources;
    this.loading = loading;
  }
  commandType = "CreateSpatializedStatic3DElement";
  getParams() {
    return {
      modelURL: this.modelURL,
      sources: this.sources,
      loading: this.loading
    };
  }
};
var CreateSpatializedDynamic3DElementCommand = class extends JSBCommand {
  getParams() {
    return { test: true };
  }
  commandType = "CreateSpatializedDynamic3DElement";
};
var CreateSpatialEntityCommand = class extends JSBCommand {
  constructor(name) {
    super();
    this.name = name;
  }
  getParams() {
    return { name: this.name };
  }
  commandType = "CreateSpatialEntity";
};
var CreateModelComponentCommand = class extends JSBCommand {
  constructor(options) {
    super();
    this.options = options;
  }
  getParams() {
    let geometryId = this.options.mesh.id;
    let materialIds = this.options.materials.map((material) => material.id);
    return { geometryId, materialIds };
  }
  commandType = "CreateModelComponent";
};
var CreateSpatialModelEntityCommand = class extends JSBCommand {
  constructor(options) {
    super();
    this.options = options;
  }
  getParams() {
    return this.options;
  }
  commandType = "CreateSpatialModelEntity";
};
var CreateModelAssetCommand = class extends JSBCommand {
  constructor(options) {
    super();
    this.options = options;
  }
  getParams() {
    return { url: this.options.url };
  }
  commandType = "CreateModelAsset";
};
var CreateSpatialGeometryCommand = class extends JSBCommand {
  constructor(type, options = {}) {
    super();
    this.type = type;
    this.options = options;
  }
  getParams() {
    return { type: this.type, ...this.options };
  }
  commandType = "CreateGeometry";
};
var CreateSpatialUnlitMaterialCommand = class extends JSBCommand {
  constructor(options) {
    super();
    this.options = options;
  }
  getParams() {
    return this.options;
  }
  commandType = "CreateUnlitMaterial";
};
var AddComponentToEntityCommand = class extends JSBCommand {
  constructor(entity, comp) {
    super();
    this.entity = entity;
    this.comp = comp;
  }
  getParams() {
    return {
      entityId: this.entity.id,
      componentId: this.comp.id
    };
  }
  commandType = "AddComponentToEntity";
};
var RemoveComponentFromEntityCommand = class extends JSBCommand {
  constructor(entity, comp) {
    super();
    this.entity = entity;
    this.comp = comp;
  }
  getParams() {
    return {
      entityId: this.entity.id,
      componentId: this.comp.id
    };
  }
  commandType = "RemoveComponentFromEntity";
};
var SetMaterialsOnEntityCommand = class extends JSBCommand {
  constructor(entityId, materials) {
    super();
    this.entityId = entityId;
    this.materials = materials;
  }
  getParams() {
    return {
      entityId: this.entityId,
      materialIds: this.materials.map((m) => m.id)
    };
  }
  commandType = "SetMaterialsOnEntity";
};
var SetParentForEntityCommand = class extends JSBCommand {
  // childId, parentId
  constructor(childId, parentId) {
    super();
    this.childId = childId;
    this.parentId = parentId;
  }
  getParams() {
    return {
      childId: this.childId,
      parentId: this.parentId
    };
  }
  commandType = "SetParentToEntity";
};
var ConvertFromEntityToEntityCommand = class extends JSBCommand {
  constructor(fromEntityId, toEntityId, fromPosition) {
    super();
    this.fromEntityId = fromEntityId;
    this.toEntityId = toEntityId;
    this.fromPosition = fromPosition;
  }
  getParams() {
    return {
      fromEntityId: this.fromEntityId,
      toEntityId: this.toEntityId,
      position: this.fromPosition
    };
  }
  commandType = "ConvertFromEntityToEntity";
};
var ConvertFromEntityToSceneCommand = class extends JSBCommand {
  constructor(fromEntityId, position) {
    super();
    this.fromEntityId = fromEntityId;
    this.position = position;
  }
  getParams() {
    return {
      fromEntityId: this.fromEntityId,
      position: this.position
    };
  }
  commandType = "ConvertFromEntityToScene";
};
var ConvertFromSceneToEntityCommand = class extends JSBCommand {
  //  let entityId: String
  // let position:Vec3
  constructor(entityId, position) {
    super();
    this.entityId = entityId;
    this.position = position;
  }
  getParams() {
    return {
      entityId: this.entityId,
      position: this.position
    };
  }
  commandType = "ConvertFromSceneToEntity";
};
var ConvertCoordinateCommand = class extends JSBCommand {
  constructor(position, fromId, toId) {
    super();
    this.position = position;
    this.fromId = fromId;
    this.toId = toId;
  }
  getParams() {
    return {
      position: this.position,
      fromId: this.fromId,
      toId: this.toId
    };
  }
  commandType = "ConvertCoordinate";
};
var CreateTextureCommand = class extends JSBCommand {
  constructor(url) {
    super();
    this.url = url;
  }
  getParams() {
    return {
      url: this.url
    };
  }
  commandType = "CreateTexture";
};
var UpdateTexturePropertiesCommand = class extends SpatializedElementCommand {
  properties;
  commandType = "UpdateTextureProperties";
  constructor(spatialObject, properties) {
    super(spatialObject);
    this.properties = properties;
  }
  getExtraParams() {
    return this.properties;
  }
};
var InspectCommand = class extends JSBCommand {
  constructor(id = "") {
    super();
    this.id = id;
  }
  commandType = "Inspect";
  getParams() {
    return this.id ? { id: this.id } : { id: "" };
  }
};
var DestroyCommand = class extends JSBCommand {
  constructor(id) {
    super();
    this.id = id;
  }
  commandType = "Destroy";
  getParams() {
    return { id: this.id };
  }
};
var InitializeAttachmentCommand = class extends JSBCommand {
  constructor(attachmentId, options) {
    super();
    this.attachmentId = attachmentId;
    this.options = options;
  }
  commandType = "InitializeAttachment";
  getParams() {
    return {
      id: this.attachmentId,
      parentEntityId: this.options.parentEntityId,
      position: this.options.position ?? [0, 0, 0],
      size: this.options.size,
      ownerViewId: this.options.ownerViewId
    };
  }
};
var UpdateAttachmentEntityCommand = class extends JSBCommand {
  constructor(attachmentId, options) {
    super();
    this.attachmentId = attachmentId;
    this.options = options;
  }
  commandType = "UpdateAttachmentEntity";
  getParams() {
    return {
      id: this.attachmentId,
      ...this.options
    };
  }
};

// src/SpatialObject.ts
var SpatialObject = class {
  /** @hidden */
  constructor(id) {
    this.id = id;
  }
  name;
  isDestroyed = false;
  async inspect() {
    const ret = await new InspectCommand(this.id).execute();
    if (ret.success) {
      return ret.data;
    }
    throw new Error(ret.errorMessage);
  }
  async destroy() {
    if (this.isDestroyed) {
      return;
    }
    const ret = await new DestroyCommand(this.id).execute();
    if (ret.success) {
      this.onDestroy();
      this.isDestroyed = true;
      return ret.data;
    } else if (this.isDestroyed) {
      return;
    }
    throw new Error(ret.errorMessage);
  }
  // override this method to do some cleanup
  onDestroy() {
  }
};

// src/spatial-host.ts
async function createNativeSpatialDiv() {
  const platform = await getPlatform();
  return platform.createNativeSpatialDiv();
}
async function createNativeAttachment(options) {
  const platform = await getPlatform();
  return platform.createNativeAttachment(options);
}
function openSpatialSceneSync(url, config, target, features) {
  return getPlatformSync().openSpatialSceneSync(url, config, target, features);
}

// src/SpatialScene.ts
var instance;
var SpatialScene = class _SpatialScene extends SpatialObject {
  /**
   * Gets the singleton instance of the SpatialScene.
   * Creates a new instance if one doesn't exist yet.
   * @returns The singleton SpatialScene instance
   */
  static getInstance() {
    if (!instance) {
      instance = new _SpatialScene("");
    }
    return instance;
  }
  async convertCoordinate(position, fromId, toId) {
    try {
      const ret = await new ConvertCoordinateCommand(
        position,
        fromId,
        toId
      ).execute();
      return ret?.data ?? position;
    } catch (error) {
      console.warn("SpatialScene.convertCoordinate error:", error);
      throw error;
    }
  }
  /**
   * Updates the properties of the spatial scene.
   * This can include background settings, lighting, and other scene-wide properties.
   * @param properties Partial set of properties to update
   * @returns Promise resolving when the update is complete
   */
  async updateSpatialProperties(properties) {
    return new UpdateSpatialSceneProperties(properties).execute();
  }
  /**
   * Adds a spatialized element to the scene.
   * This makes the element visible and interactive in the spatial environment.
   * @param element The SpatializedElement to add to the scene
   * @returns Promise resolving when the element is added
   */
  async addSpatializedElement(element) {
    return new AddSpatializedElementToSpatialScene(element).execute();
  }
  /**
   * Updates the scene creation configuration.
   * This allows changing scene parameters after initial creation.
   * @param config The new scene creation configuration
   * @returns Promise resolving when the update is complete
   */
  async updateSceneCreationConfig(config) {
    return new UpdateSceneConfig(config).execute();
  }
  /**
   * Gets the current state of the spatial scene.
   * This includes information about active elements and scene configuration.
   * @returns Promise resolving to the current SpatialSceneState
   */
  async getState() {
    return (await new GetSpatialSceneState().execute()).data.name;
  }
};

// src/types/types.ts
var SpatializedElementType = /* @__PURE__ */ ((SpatializedElementType2) => {
  SpatializedElementType2[SpatializedElementType2["Spatialized2DElement"] = 0] = "Spatialized2DElement";
  SpatializedElementType2[SpatializedElementType2["SpatializedStatic3DElement"] = 1] = "SpatializedStatic3DElement";
  SpatializedElementType2[SpatializedElementType2["SpatializedDynamic3DElement"] = 2] = "SpatializedDynamic3DElement";
  return SpatializedElementType2;
})(SpatializedElementType || {});
var BaseplateVisibilityValues = [
  "automatic",
  "visible",
  "hidden"
];
function isValidBaseplateVisibilityType(type) {
  return BaseplateVisibilityValues.includes(type);
}
var WorldScalingValues = ["automatic", "dynamic"];
function isValidWorldScalingType(type) {
  return WorldScalingValues.includes(type);
}
var WorldAlignmentValues = [
  "adaptive",
  "automatic",
  "gravityAligned"
];
function isValidWorldAlignmentType(type) {
  return WorldAlignmentValues.includes(type);
}
var SpatialSceneValues = ["window", "volume"];
function isValidSpatialSceneType(type) {
  return SpatialSceneValues.includes(type);
}
function isValidSceneUnit(val) {
  if (typeof val === "number") {
    return val >= 0;
  }
  if (typeof val === "string") {
    if (val.endsWith("px")) {
      if (isNaN(Number(val.slice(0, -2)))) {
        return false;
      }
      return Number(val.slice(0, -2)) >= 0;
    }
    if (val.endsWith("m")) {
      if (isNaN(Number(val.slice(0, -1)))) {
        return false;
      }
      return Number(val.slice(0, -1)) >= 0;
    }
  }
  return false;
}
var SpatialSceneState = /* @__PURE__ */ ((SpatialSceneState2) => {
  SpatialSceneState2["idle"] = "idle";
  SpatialSceneState2["pending"] = "pending";
  SpatialSceneState2["willVisible"] = "willVisible";
  SpatialSceneState2["visible"] = "visible";
  SpatialSceneState2["fail"] = "fail";
  return SpatialSceneState2;
})(SpatialSceneState || {});
var CubeInfo = class {
  constructor(size, origin) {
    this.size = size;
    this.origin = origin;
    this.size = size;
    this.origin = origin;
  }
  get x() {
    return this.origin.x;
  }
  get y() {
    return this.origin.y;
  }
  get z() {
    return this.origin.z;
  }
  get width() {
    return this.size.width;
  }
  get height() {
    return this.size.height;
  }
  get depth() {
    return this.size.depth;
  }
  get left() {
    return this.x;
  }
  get top() {
    return this.y;
  }
  get right() {
    return this.x + this.width;
  }
  get bottom() {
    return this.y + this.height;
  }
  get back() {
    return this.z;
  }
  get front() {
    return this.z + this.depth;
  }
};

// src/physicalMetrics.ts
var physicalMetrics_exports = {};
__export(physicalMetrics_exports, {
  getValue: () => getValue,
  physicalToPoint: () => physicalToPoint,
  pointToPhysical: () => pointToPhysical,
  subscribe: () => subscribe
});
var snapshot = {
  meterToPtUnscaled: 1360,
  meterToPtScaled: 1360
};
function getWorldScalingCompensation(options) {
  return options?.worldScalingCompensation ?? "scaled";
}
function pointToPhysical(point, options) {
  updateValue();
  const compensation = getWorldScalingCompensation(options);
  if (compensation === "unscaled") {
    return point / snapshot.meterToPtUnscaled;
  }
  return point / snapshot.meterToPtScaled;
}
function physicalToPoint(physical, options) {
  updateValue();
  const compensation = getWorldScalingCompensation(options);
  if (compensation === "unscaled") {
    return physical * snapshot.meterToPtUnscaled;
  }
  return physical * snapshot.meterToPtScaled;
}
function updateValue() {
  if (typeof window === "undefined") return;
  const src = window.__webspatialsdk__?.physicalMetrics;
  if (!src) return;
  const next = {
    meterToPtScaled: src.meterToPtScaled ?? snapshot.meterToPtScaled,
    meterToPtUnscaled: src.meterToPtUnscaled ?? snapshot.meterToPtUnscaled
  };
  if (next.meterToPtScaled !== snapshot.meterToPtScaled || next.meterToPtUnscaled !== snapshot.meterToPtUnscaled) {
    snapshot = next;
  }
}
function getValue() {
  updateValue();
  return snapshot;
}
function subscribe(cb) {
  if (typeof window === "undefined") return () => {
  };
  const handler = () => {
    cb();
  };
  SpatialWebEvent.addEventReceiver("window", handler);
  return () => {
    SpatialWebEvent.removeEventReceiver("window");
  };
}

// src/scene-polyfill.ts
var defaultSceneConfig = {
  defaultSize: {
    width: 1280,
    height: 720
  }
};
var defaultSceneConfigVolume = {
  defaultSize: {
    width: "0.94m",
    height: "0.94m",
    depth: "0.94m"
  }
};
var xr_window_defaults = {
  ...defaultSceneConfig
};
var xr_volume_defaults = {
  ...defaultSceneConfigVolume
};
var INTERNAL_SCHEMA_PREFIX = "webspatial://";
function deepMergePlain(base, over) {
  if (!over) return { ...base || {} };
  const out = { ...base || {} };
  for (const k of Object.keys(over)) {
    const bv = out[k];
    const ov = over[k];
    if (ov && typeof ov === "object" && !Array.isArray(ov) && bv && typeof bv === "object" && !Array.isArray(bv)) {
      out[k] = deepMergePlain(bv, ov);
    } else {
      out[k] = ov;
    }
  }
  return out;
}
function normalizeXRDefaultsToSceneOptions(src) {
  const out = { ...src || {} };
  const ds = src.defaultSize !== void 0 ? src.defaultSize : src.default_size;
  if (ds !== void 0) {
    out.defaultSize = ds;
  }
  if ("default_size" in out) {
    delete out.default_size;
  }
  return out;
}
var SceneManager = class _SceneManager {
  originalOpen;
  static instance;
  manifestReady = null;
  static getInstance() {
    if (!_SceneManager.instance) {
      _SceneManager.instance = new _SceneManager();
    }
    return _SceneManager.instance;
  }
  init(window2) {
    this.manifestReady = this.setupManifest();
    this.originalOpen = window2.open.bind(window2);
    window2.open = this.open;
  }
  // Stores the latest formatted config used by the platform (per scene name).
  // This object contains normalized values and is safe for internal consumption.
  configMap = {};
  // Stores the raw callback return value (per scene name) to feed into the next initScene call as `pre`.
  // We keep this unformatted so developers receive exactly what they last returned.
  callbackReturnMap = {};
  getConfig(name) {
    if (name === void 0 || !this.configMap[name]) return void 0;
    return this.configMap[name];
  }
  waitManifest() {
    return this.manifestReady ?? Promise.resolve();
  }
  // Ensure URL is absolute; only convert when a relative path is provided
  // - Keep external and special schemes untouched (http, https, data, blob, about, file, mailto, etc.)
  // - Handle protocol-relative URLs (//example.com/path)
  // - Resolve relative paths against document.baseURI (respects <base href>)
  ensureAbsoluteUrl(raw) {
    if (!raw) return raw;
    if (/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(raw)) {
      return raw;
    }
    if (raw.startsWith("//")) {
      return `${window.location.protocol}${raw}`;
    }
    try {
      return new URL(raw, document.baseURI).toString();
    } catch {
      return raw;
    }
  }
  async setupManifest() {
    const manifest = await this.getPWAManifest();
    try {
      const xr = manifest?.xr_spatial_scene;
      if (!xr || typeof xr !== "object") return;
      const { overrides, ...topLevel } = xr;
      const windowRaw = deepMergePlain(topLevel, overrides?.window_scene);
      const volumeRaw = deepMergePlain(topLevel, overrides?.volume_scene);
      const windowNext = normalizeXRDefaultsToSceneOptions(windowRaw);
      const volumeNext = normalizeXRDefaultsToSceneOptions(volumeRaw);
      if (windowNext && Object.keys(windowNext).length > 0) {
        xr_window_defaults = windowNext;
      }
      if (volumeNext && Object.keys(volumeNext).length > 0) {
        xr_volume_defaults = volumeNext;
      }
    } catch (error) {
      console.warn(
        "SceneManager.setupManifest failed; using built-in defaults.",
        error?.message || error
      );
    }
  }
  open = (url, target, features) => {
    if (url?.startsWith(INTERNAL_SCHEMA_PREFIX)) {
      if (url.includes("createSpatialized2DElement") || url.includes("createAttachment")) {
        const token = (
          //@ts-ignore
          (window.webSpatial || window.__webspatialShell__)?.genToken?.()
        );
        if (token) {
          const command = url.includes("createAttachment") ? "createAttachment" : "createSpatialized2DElement";
          const host = window.location.host;
          const protocol = window.location.protocol;
          const finalURL = `${protocol}//${host}/${token}/?command=${command}`;
          const rid = new URL(url).searchParams.get("rid");
          const final = new URL(finalURL);
          if (rid) final.searchParams.set("rid", rid);
          return this.originalOpen(final.toString(), target, features);
        }
      }
      return this.originalOpen(url, target, features);
    }
    url = this.ensureAbsoluteUrl(url);
    if (target === "_self" || target === "_parent" || target === "_top") {
      const newWindow = this.originalOpen(url, target, features);
      return newWindow;
    }
    let cfg = target ? this.getConfig(target) : void 0;
    if (cfg === void 0) {
      const preFormatted = deepCloneJSON(getSceneDefaultConfig("window"));
      const [ans] = formatSceneConfig(preFormatted, "window");
      cfg = { ...ans, type: "window" };
    }
    const result = openSpatialSceneSync(
      url,
      cfg,
      target,
      features
    );
    const id = result.data?.id;
    if (id) {
      let focusCmd = new FocusScene(id);
      focusCmd.execute();
    }
    return result.data?.windowProxy;
  };
  initScene(name, callback, options) {
    const sceneType = options?.type ?? "window";
    const defaultConfigRaw = getSceneDefaultConfig(sceneType);
    const previousOrDefault = this.callbackReturnMap[name] ?? (() => {
      const cloned = deepCloneJSON(defaultConfigRaw);
      return cloned;
    })();
    const rawReturnVal = callback(previousOrDefault);
    const sanitizedReturnVal = sanitizeSceneOptionsUnits(
      deepCloneJSON(rawReturnVal)
    );
    const clonedForFormat = deepCloneJSON(sanitizedReturnVal);
    const baseDefaults = deepCloneJSON(getSceneDefaultConfig(sceneType));
    const mergedForFormat = deepMergePlain(baseDefaults, clonedForFormat);
    const [formattedConfig, errors] = formatSceneConfig(
      mergedForFormat,
      sceneType
    );
    if (errors.length > 0) {
      console.warn(`initScene ${name} with errors: ${errors.join(", ")}`);
    }
    this.callbackReturnMap[name] = sanitizedReturnVal;
    this.configMap[name] = {
      ...formattedConfig,
      type: sceneType
    };
  }
  /**
   * Resolve and load a PWA manifest as JSON:
   * 1) Determine href:
   *    - Prefer explicit manifestUrl if provided;
   *    - Fallback to <link rel="manifest">, preferring the raw attribute over computed href.
   * 2) Normalize href to absolute using ensureAbsoluteUrl (respects <base href>).
   * 3) Handle data URLs inline:
   *    - data:...;base64,... → atob then JSON.parse
   *    - data:...,... → decodeURIComponent then JSON.parse
   * 4) Fetch with credentials same-origin first; if that fails (e.g., CORS), attempt unauthenticated fetch.
   * 5) Parse as JSON; if response body is text, parse the text as JSON.
   */
  async getPWAManifest(manifestUrl) {
    let href = manifestUrl;
    if (!href) {
      const el = document.querySelector(
        'link[rel="manifest"]'
      );
      href = el?.getAttribute("href") || el?.href;
    }
    if (!href) return;
    href = this.ensureAbsoluteUrl(href);
    if (!href) return;
    if (href.startsWith("data:")) {
      try {
        const comma = href.indexOf(",");
        if (comma < 0) return;
        const meta = href.slice(5, comma);
        const data = href.slice(comma + 1);
        const isBase64 = /;base64/i.test(meta);
        const decoded = isBase64 ? atob(data) : decodeURIComponent(data);
        return JSON.parse(decoded);
      } catch {
        return;
      }
    }
    try {
      const res = await fetch(href, { credentials: "same-origin" });
      if (!res.ok) throw new Error(String(res.status));
      try {
        return await res.json();
      } catch {
        const t = await res.text();
        return JSON.parse(t);
      }
    } catch {
      try {
        const res = await fetch(href);
        if (!res.ok) return;
        try {
          return await res.json();
        } catch {
          const t = await res.text();
          return JSON.parse(t);
        }
      } catch {
        return;
      }
    }
  }
};
function sanitizeSceneOptionsUnits(val) {
  if (val?.defaultSize) {
    const keys = ["width", "height", "depth"];
    for (const k of keys) {
      if (k in val.defaultSize && !isValidSceneUnit(val.defaultSize[k])) {
        delete val.defaultSize[k];
      }
    }
  }
  if (val?.resizability) {
    const keys = ["minWidth", "minHeight", "maxWidth", "maxHeight"];
    for (const k of keys) {
      if (k in val.resizability && !isValidSceneUnit(val.resizability[k])) {
        delete val.resizability[k];
      }
    }
  }
  return val;
}
function pxToMeter(px) {
  return pointToPhysical(px);
}
function meterToPx(meter) {
  return physicalToPoint(meter);
}
function formatToNumber(str, targetUnit, defaultUnit) {
  if (typeof str === "number") {
    if (defaultUnit === "px" && targetUnit === "px" || defaultUnit === "m" && targetUnit === "m") {
      return str;
    }
    if (defaultUnit === "px" && targetUnit === "m") {
      return pxToMeter(str);
    } else if (defaultUnit === "m" && targetUnit === "px") {
      return meterToPx(str);
    }
    return str;
  }
  if (targetUnit === "m") {
    if (str.endsWith("m")) {
      return Number(str.slice(0, -1));
    } else if (str.endsWith("px")) {
      return pxToMeter(Number(str.slice(0, -2)));
    } else {
      throw new Error("formatToNumber: invalid str");
    }
  } else if (targetUnit === "px") {
    if (str.endsWith("px")) {
      return Number(str.slice(0, -2));
    } else if (str.endsWith("m")) {
      return meterToPx(Number(str.slice(0, -1)));
    } else {
      throw new Error("formatToNumber: invalid str");
    }
  } else {
    throw new Error("formatToNumber: invalid targetUnit");
  }
}
function formatSceneConfig(config, sceneType) {
  const errors = [];
  const isWindow = sceneType === "window";
  if (!isValidSpatialSceneType(sceneType)) {
    errors.push(`sceneType`);
  }
  if (config.defaultSize) {
    const iterKeys = ["width", "height", "depth"];
    for (let k of iterKeys) {
      if (!(k in config.defaultSize)) continue;
      if (isValidSceneUnit(config.defaultSize[k])) {
        ;
        config.defaultSize[k] = formatToNumber(
          config.defaultSize[k],
          isWindow ? "px" : "m",
          "px"
        );
      } else {
        delete config.defaultSize[k];
        errors.push(`defaultSize.${k}`);
      }
    }
  }
  if (config.resizability) {
    const iterKeys = ["minWidth", "minHeight", "maxWidth", "maxHeight"];
    for (let k of iterKeys) {
      if (!(k in config.resizability)) continue;
      if (isValidSceneUnit(config.resizability[k])) {
        ;
        config.resizability[k] = formatToNumber(
          config.resizability[k],
          "px",
          "px"
        );
      } else {
        delete config.resizability[k];
        errors.push(`resizability.${k}`);
      }
    }
  }
  if (config.worldScaling) {
    if (!isValidWorldScalingType(config.worldScaling)) {
      config.worldScaling = "automatic";
      errors.push("worldScaling");
    }
  }
  if (config.worldAlignment) {
    if (!isValidWorldAlignmentType(config.worldAlignment)) {
      config.worldAlignment = "automatic";
      errors.push("worldAlignment");
    }
  }
  if (config.baseplateVisibility) {
    if (!isValidBaseplateVisibilityType(config.baseplateVisibility)) {
      config.baseplateVisibility = "automatic";
      errors.push("baseplateVisibility");
    }
  }
  return [config, errors];
}
function initScene(name, callback, options) {
  return SceneManager.getInstance().initScene(name, callback, options);
}
function hijackWindowOpen(window2) {
  SceneManager.getInstance().init(window2);
}
function hijackWindowATag(openedWindow) {
  openedWindow.document.onclick = function(e) {
    let element = e.target;
    while (element) {
      if (element.tagName == "A") {
        if (handleATag(e, element)) {
          return false;
        }
        return true;
      }
      if (element.parentElement) {
        element = element.parentElement;
      } else {
        break;
      }
    }
  };
}
function handleATag(event, link) {
  if (event.defaultPrevented) return false;
  const target = link.target;
  const url = link.href;
  if (target && target !== "_self") {
    event.preventDefault();
    window.open(url, target);
    return true;
  }
}
function getSceneDefaultConfig(sceneType) {
  return sceneType === "window" ? xr_window_defaults || defaultSceneConfig : xr_volume_defaults || defaultSceneConfigVolume;
}
async function injectScenePolyfill() {
  if (!window.opener) return;
  const state = await SpatialScene.getInstance().getState();
  if (state !== "pending" /* pending */) return;
  function onContentLoaded(callback) {
    if (document.readyState === "interactive" || document.readyState === "complete") {
      callback();
    } else {
      document.addEventListener("DOMContentLoaded", callback);
    }
  }
  onContentLoaded(async () => {
    await SceneManager.getInstance().waitManifest();
    const sceneType = window.xrCurrentSceneType ?? "window";
    const rawDefault = getSceneDefaultConfig(sceneType);
    const pre = deepCloneJSON(rawDefault);
    let cfg = pre;
    if (typeof window.xrCurrentSceneDefaults === "function") {
      try {
        cfg = await window.xrCurrentSceneDefaults?.(pre);
      } catch (error) {
        console.error(error);
      }
    }
    await new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(null);
      }, 1e3);
    });
    const mergedCfg = deepMergePlain(deepCloneJSON(rawDefault), cfg);
    const [formattedConfig, errors] = formatSceneConfig(mergedCfg, sceneType);
    if (errors.length > 0) {
      console.warn(
        `window.xrCurrentSceneDefaults with errors: ${errors.join(", ")}`
      );
    }
    const finalCfg = {
      ...formattedConfig,
      type: sceneType
    };
    await SpatialScene.getInstance().updateSceneCreationConfig(finalCfg);
  });
}
function injectSceneHook() {
  hijackWindowOpen(window);
  hijackWindowATag(window);
  injectScenePolyfill();
}

// src/SpatialWebEventCreator.ts
function createSpatialEvent(type, detail) {
  return new CustomEvent(type, {
    bubbles: true,
    cancelable: false,
    detail
  });
}

// src/SpatializedElement.ts
var SpatializedElement = class extends SpatialObject {
  /**
   * Creates a new spatialized element with the specified ID.
   * Registers the element to receive spatial events.
   * @param id Unique identifier for this element
   */
  constructor(id) {
    super(id);
    this.id = id;
    SpatialWebEvent.addEventReceiver(id, this.onReceiveEvent.bind(this));
  }
  /**
   * Updates the transformation matrix of this element in 3D space.
   * This affects the position, rotation, and scale of the element.
   * @param matrix The new transformation matrix
   * @returns Promise resolving when the transform is updated
   */
  async updateTransform(matrix) {
    return new UpdateSpatializedElementTransform(this, matrix).execute();
  }
  /**
   * Information about the element's bounding cube.
   * Used for spatial calculations and hit testing.
   */
  _cubeInfo;
  /**
   * Gets the current cube information for this element.
   * @returns The current CubeInfo or undefined if not set
   */
  get cubeInfo() {
    return this._cubeInfo;
  }
  /**
   * The current transformation matrix of this element.
   */
  _transform;
  /**
   * The inverse of the current transformation matrix.
   * Used for converting world coordinates to local coordinates.
   */
  _transformInv;
  /**
   * Gets the current transformation matrix.
   * @returns The current transformation matrix or undefined if not set
   */
  get transform() {
    return this._transform;
  }
  /**
   * Gets the inverse of the current transformation matrix.
   * @returns The inverse transformation matrix or undefined if not set
   */
  get transformInv() {
    return this._transformInv;
  }
  /**
   * Processes events received from the WebSpatial environment.
   * Handles various spatial events like transforms, gestures, and interactions.
   * @param data The event data received from the WebSpatial system
   */
  onReceiveEvent(data) {
    const { type } = data;
    if (type === "objectdestroy" /* objectdestroy */) {
      this.isDestroyed = true;
    } else if (type === "spatialtap" /* spatialtap */) {
      const event = createSpatialEvent(
        "spatialtap" /* spatialtap */,
        data.detail
      );
      this._onSpatialTap?.(event);
    } else if (type === "spatialdragstart" /* spatialdragstart */) {
      const dragStartEvent = createSpatialEvent(
        "spatialdragstart" /* spatialdragstart */,
        data.detail
      );
      this._onSpatialDragStart?.(dragStartEvent);
    } else if (type === "spatialdrag" /* spatialdrag */) {
      const event = createSpatialEvent(
        "spatialdrag" /* spatialdrag */,
        data.detail
      );
      this._onSpatialDrag?.(event);
    } else if (type === "spatialdragend" /* spatialdragend */) {
      const event = createSpatialEvent(
        "spatialdragend" /* spatialdragend */,
        data.detail
      );
      this._onSpatialDragEnd?.(event);
    } else if (type === "spatialrotate" /* spatialrotate */) {
      const event = createSpatialEvent(
        "spatialrotate" /* spatialrotate */,
        data.detail
      );
      this._onSpatialRotate?.(event);
    } else if (type === "spatialrotateend" /* spatialrotateend */) {
      const event = createSpatialEvent(
        "spatialrotateend" /* spatialrotateend */,
        data.detail
      );
      this._onSpatialRotateEnd?.(event);
    } else if (type === "spatialmagnify" /* spatialmagnify */) {
      const event = createSpatialEvent(
        "spatialmagnify" /* spatialmagnify */,
        data.detail
      );
      this._onSpatialMagnify?.(event);
    } else if (type === "spatialmagnifyend" /* spatialmagnifyend */) {
      const event = createSpatialEvent(
        "spatialmagnifyend" /* spatialmagnifyend */,
        data.detail
      );
      this._onSpatialMagnifyEnd?.(event);
    }
  }
  _onSpatialTap;
  set onSpatialTap(value) {
    this._onSpatialTap = value;
    this.updateProperties({
      enableTapGesture: value !== void 0
    });
  }
  _onSpatialDragStart;
  set onSpatialDragStart(value) {
    this._onSpatialDragStart = value;
    this.updateProperties({
      enableDragStartGesture: this._onSpatialDragStart !== void 0
    });
  }
  _onSpatialDrag;
  set onSpatialDrag(value) {
    this._onSpatialDrag = value;
    this.updateProperties({
      enableDragGesture: this._onSpatialDrag !== void 0
    });
  }
  _onSpatialDragEnd;
  set onSpatialDragEnd(value) {
    this._onSpatialDragEnd = value;
    this.updateProperties({
      enableDragEndGesture: value !== void 0
    });
  }
  _onSpatialRotate;
  set onSpatialRotate(value) {
    this._onSpatialRotate = value;
    this.updateProperties({
      enableRotateGesture: this._onSpatialRotate !== void 0
    });
  }
  _onSpatialRotateEnd;
  set onSpatialRotateEnd(value) {
    this._onSpatialRotateEnd = value;
    this.updateProperties({
      enableRotateEndGesture: value !== void 0
    });
  }
  _onSpatialMagnify;
  set onSpatialMagnify(value) {
    this._onSpatialMagnify = value;
    this.updateProperties({
      enableMagnifyGesture: value !== void 0
    });
  }
  _onSpatialMagnifyEnd;
  set onSpatialMagnifyEnd(value) {
    this._onSpatialMagnifyEnd = value;
    this.updateProperties({
      enableMagnifyEndGesture: value !== void 0
    });
  }
  /**
   * Cleans up resources when this element is destroyed.
   * Removes event receivers to prevent memory leaks.
   */
  onDestroy() {
    SpatialWebEvent.removeEventReceiver(this.id);
  }
};

// src/Spatialized2DElement.ts
var Spatialized2DElement = class extends SpatializedElement {
  /**
   * Creates a new spatialized 2D element.
   * @param id Unique identifier for this element
   * @param windowProxy Reference to the window object containing the 2D content
   */
  constructor(id, windowProxy) {
    super(id);
    this.windowProxy = windowProxy;
    hijackWindowATag(windowProxy);
  }
  /**
   * Updates the properties of this 2D element.
   * This can include size, position, background, and other visual properties.
   * @param properties Partial set of properties to update
   * @returns Promise resolving when the update is complete
   */
  async updateProperties(properties) {
    return new UpdateSpatialized2DElementProperties(this, properties).execute();
  }
  /**
   * Adds a child spatialized element to this 2D element.
   * This allows for creating hierarchical structures of spatial elements.
   * @param element The child element to add
   * @returns Promise resolving when the element is added
   */
  async addSpatializedElement(element) {
    return new AddSpatializedElementToSpatialized2DElement(
      this,
      element
    ).execute();
  }
};

// src/SpatializedStatic3DElement.ts
var SpatializedStatic3DElement = class extends SpatializedElement {
  /**
   * Creates a new spatialized static 3D element with the specified ID and URL.
   * Registers the element to receive spatial events.
   * @param id Unique identifier for this element
   * @param modelURL URL of the 3D model
   * @param sources Optional fallback model sources
   * @param loading Initial loading mode (`'eager'` by default)
   */
  constructor(id, modelURL, sources, loading = "eager") {
    super(id);
    this.modelURL = modelURL;
    this.sources = sources;
    this._loading = loading;
  }
  /**
   * Promise resolver for the ready state.
   * Used to resolve the ready promise when the model is loaded.
   */
  _readyResolve;
  /**
   * Caches the last model URL to detect changes.
   * Used to reset the ready promise when the model URL changes.
   */
  modelURL;
  // @TODO: Deprecate modelURL property from Web and Swift code
  get modelUrl() {
    return this.modelURL;
  }
  /**
   * Caches the last sources array to detect changes.
   */
  sources;
  /**
   * The model URL that was successfully loaded by the native runtime.
   */
  _currentSrc = "";
  get currentSrc() {
    return this._currentSrc;
  }
  /**
   * Creates a new promise for tracking the ready state of the model.
   * @returns Promise that resolves when the model is loaded (true) or fails to load (false)
   */
  createReadyPromise() {
    this._readyResolve?.(false);
    return new Promise((resolve) => {
      this._readyResolve = resolve;
    });
  }
  /**
   * Promise that resolves when the model is loaded.
   * Resolves to true on successful load, false on failure.
   */
  ready = this.createReadyPromise();
  /**
   * Updates the properties of this static 3D element.
   * Handles special case for modelURL changes by resetting the ready promise.
   * @param properties Partial set of properties to update
   * @returns Promise resolving when the update is complete
   */
  async updateProperties(properties) {
    let needsReadyReset = false;
    if (properties.modelURL !== void 0) {
      if (this.modelURL !== properties.modelURL) {
        this.modelURL = properties.modelURL;
        needsReadyReset = true;
      }
    }
    if (properties.sources !== void 0) {
      const prevJson = JSON.stringify(this.sources);
      const nextJson = JSON.stringify(properties.sources);
      if (prevJson !== nextJson) {
        this.sources = properties.sources;
        needsReadyReset = true;
      }
    }
    if (needsReadyReset) {
      this.ready = this.createReadyPromise();
    }
    if (properties.autoplay !== void 0) {
      this._autoplay = properties.autoplay;
    }
    if (properties.loop !== void 0) {
      this._loop = properties.loop;
    }
    if (properties.playbackRate !== void 0) {
      if (!this._paused) {
        this._currentTime = this.currentTime;
        this._anchorTimestamp = Date.now();
      }
      this._playbackRate = properties.playbackRate;
    }
    if (properties.currentTime !== void 0) {
      this._currentTime = properties.currentTime;
      this._anchorTimestamp = Date.now();
    }
    if (properties.loading !== void 0) {
      this._loading = properties.loading;
    }
    return new UpdateSpatializedStatic3DElementProperties(
      this,
      properties
    ).execute();
  }
  /**
   * Total animation duration in seconds, synced from native.
   */
  _duration = 0;
  /**
   * Returns the total animation duration in seconds.
   */
  get duration() {
    return this._duration;
  }
  /**
   * Playback speed multiplier.
   */
  _playbackRate = 1;
  /**
   * Returns the current playback rate.
   */
  get playbackRate() {
    return this._playbackRate;
  }
  /**
   * Sets the playback rate and sends it to native.
   */
  set playbackRate(value) {
    this.updateProperties({ playbackRate: value });
  }
  /**
   * Last playback position sampled from native (seconds).
   */
  _currentTime = 0;
  /**
   * Unix epoch time (ms) corresponding to `_currentTime`. Sourced from the
   * native `timestamp` on samples, or `Date.now()` on local seeks/transitions.
   */
  _anchorTimestamp = 0;
  clampTime(time) {
    if (!Number.isFinite(time) || time < 0) return 0;
    if (time > this._duration) {
      return this.loop && this.duration > 0 ? time % this._duration : this.duration;
    }
    return time;
  }
  /**
   * Returns the current (un-scaled) playback position in seconds. While
   * playing, the value is extrapolated from the last anchor using
   * `playbackRate` and clamped to `[0, duration]`; while paused it returns
   * the anchor directly.
   */
  get currentTime() {
    if (this._paused) return this._currentTime;
    const elapsed = (Date.now() - this._anchorTimestamp) / 1e3;
    return this.clampTime(this._currentTime + elapsed * this._playbackRate);
  }
  /**
   * Seeks the animation to `value` seconds (clamped to `[0, duration]`) and
   * forwards the request to native.
   */
  set currentTime(value) {
    const time = Number.isNaN(value) ? 0 : clamp(value, 0, this.duration);
    this.updateProperties({ currentTime: time });
  }
  /**
   * Whether the animation is currently paused.
   */
  _paused = true;
  /**
   * Returns whether the animation is currently paused.
   */
  get paused() {
    return this._paused;
  }
  /**
   * Callback for animation state changes.
   */
  _onAnimationStateChangeCallback;
  /**
   * Sets the callback for animation state changes.
   */
  set onAnimationStateChangeCallback(callback) {
    this._onAnimationStateChangeCallback = callback;
  }
  /**
   * Starts or resumes animation playback.
   * @returns Promise resolving when the command is sent
   */
  async play() {
    if (this._paused) {
      this._anchorTimestamp = Date.now();
    }
    this._paused = false;
    await this.updateProperties({ animationPaused: false });
  }
  /**
   * Pauses animation playback.
   * @returns Promise resolving when the command is sent
   */
  async pause() {
    if (!this._paused) {
      this._currentTime = this.currentTime;
      this._anchorTimestamp = Date.now();
    }
    this._paused = true;
    await this.updateProperties({ animationPaused: true });
  }
  /**
   * Processes events received from the WebSpatial environment.
   * Handles model loading events in addition to base spatial events.
   * @param data The event data received from the WebSpatial system
   */
  onReceiveEvent(data) {
    if (data.type === "modelloaded" /* modelloaded */) {
      this._currentSrc = data.detail?.src ?? this.modelURL ?? "";
      this._onLoadCallback?.();
      this._readyResolve?.(true);
    } else if (data.type === "modelloadfailed" /* modelloadfailed */) {
      this._onLoadFailureCallback?.();
      this._readyResolve?.(false);
    } else if (data.type === "animationstatechange" /* animationstatechange */) {
      this._paused = data.detail.paused;
      this._duration = data.detail.duration;
      this._currentTime = data.detail.currentTime ?? 0;
      this._anchorTimestamp = data.detail.timestamp ?? Date.now();
      this._onAnimationStateChangeCallback?.(data.detail);
    } else {
      super.onReceiveEvent(data);
    }
  }
  /**
   * Whether the model should automatically play its first animation on load.
   */
  _autoplay = false;
  /**
   * Returns whether autoplay is enabled for this element.
   */
  get autoplay() {
    return this._autoplay;
  }
  /**
   * Asset fetch policy. `'lazy'` defers fetching until the host signals the
   * element is in view; `'eager'` fetches immediately.
   */
  _loading = "eager";
  get loading() {
    return this._loading;
  }
  /**
   * Whether the model animation should loop continuously.
   */
  _loop = false;
  /**
   * Returns whether loop is enabled for this element.
   */
  get loop() {
    return this._loop;
  }
  /**
   * Callback function for successful model loading.
   */
  _onLoadCallback;
  /**
   * Sets the callback function for successful model loading.
   * @param callback Function to call when the model is loaded successfully
   */
  set onLoadCallback(callback) {
    this._onLoadCallback = callback;
  }
  /**
   * Callback function for model loading failure.
   */
  _onLoadFailureCallback;
  /**
   * Sets the callback function for model loading failure.
   * @param callback Function to call when the model fails to load
   */
  set onLoadFailureCallback(callback) {
    this._onLoadFailureCallback = callback;
  }
  updateModelTransform(transform) {
    const modelTransform = Array.from(transform.toFloat64Array());
    this.updateProperties({ modelTransform });
  }
};
function clamp(num, min, max) {
  return num <= min ? min : num >= max ? max : num;
}

// src/SpatializedDynamic3DElement.ts
var SpatializedDynamic3DElement = class extends SpatializedElement {
  children = [];
  events = {};
  constructor(id) {
    super(id);
  }
  async addEntity(entity) {
    const ans = new SetParentForEntityCommand(entity.id, this.id).execute();
    this.children.push(entity);
    entity.parent = this;
    return ans;
  }
  addEvent(type, callback) {
    this.events[type] = callback;
  }
  removeEvent(eventName) {
    if (this.events[eventName]) {
      delete this.events[eventName];
    }
  }
  dispatchEvent(evt) {
    this.events[evt.type]?.(evt);
  }
  async updateProperties(properties) {
    return new UpdateSpatializedDynamic3DElementProperties(
      this,
      properties
    ).execute();
  }
};

// src/SpatializedElementCreator.ts
async function createSpatialized2DElement() {
  const result = await createNativeSpatialDiv();
  if (!result.success) {
    throw new Error("createSpatialized2DElement failed");
  } else {
    const { id, windowProxy } = result.data;
    windowProxy.document.head.innerHTML = `<meta name="viewport" content="width=device-width, initial-scale=1">
      <base href="${document.baseURI}">`;
    return new Spatialized2DElement(id, windowProxy);
  }
}
async function createSpatializedStatic3DElement(modelURL, sources, loading = "eager") {
  const result = await new CreateSpatializedStatic3DElementCommand(
    modelURL,
    sources,
    loading
  ).execute();
  if (!result.success) {
    throw new Error("createSpatializedStatic3DElement failed");
  } else {
    const { id } = result.data;
    return new SpatializedStatic3DElement(id, modelURL, sources, loading);
  }
}
async function createSpatializedDynamic3DElement() {
  const result = await new CreateSpatializedDynamic3DElementCommand().execute();
  if (!result.success) {
    throw new Error("createSpatializedDynamic3DElement failed");
  } else {
    const { id } = result.data;
    return new SpatializedDynamic3DElement(id);
  }
}

// src/reality/Attachment.ts
var Attachment = class extends SpatialObject {
  constructor(id, windowProxy, options) {
    super(id);
    this.windowProxy = windowProxy;
    this.options = options;
  }
  getContainer() {
    return this.windowProxy.document.body;
  }
  getWindowProxy() {
    return this.windowProxy;
  }
  async update(options) {
    if (this.isDestroyed) return;
    if (options.position) this.options.position = options.position;
    if (options.size) this.options.size = options.size;
    return new UpdateAttachmentEntityCommand(this.id, options).execute();
  }
};
async function createAttachmentEntity(options) {
  const result = await createNativeAttachment(options);
  if (!result.success) {
    throw new Error("createAttachmentEntity failed: " + result?.errorMessage);
  }
  const { id, windowProxy } = result.data;
  await new InitializeAttachmentCommand(id, options).execute();
  return new Attachment(id, windowProxy, options);
}

// src/reality/entity/SpatialEntity.ts
var SpatialEntity = class extends SpatialObject {
  constructor(id, userData) {
    super(id);
    this.userData = userData;
    SpatialWebEvent.addEventReceiver(id, this.onReceiveEvent);
  }
  position = { x: 0, y: 0, z: 0 };
  rotation = { x: 0, y: 0, z: 0 };
  scale = { x: 1, y: 1, z: 1 };
  events = {};
  children = [];
  parent = null;
  _enableInput = false;
  get enableInput() {
    return this._enableInput;
  }
  set enableInput(value) {
    if (this._enableInput === value) return;
    this._enableInput = value;
    void this.updateEntityEvent("spatialtap", value).catch((err) => {
      console.error("enableInput updateEntityEvent failed", "spatialtap", err);
      if (this._enableInput === value) {
        this._enableInput = !value;
      }
    });
  }
  async addComponent(component) {
    return new AddComponentToEntityCommand(this, component).execute();
  }
  async removeComponent(component) {
    return new RemoveComponentFromEntityCommand(this, component).execute();
  }
  async setPosition(position) {
    return this.updateTransform({ position });
  }
  async setRotation(rotation) {
    return this.updateTransform({ rotation });
  }
  async setScale(scale) {
    return this.updateTransform({ scale });
  }
  async addEntity(ent) {
    const ans = await new SetParentForEntityCommand(ent.id, this.id).execute();
    this.children.push(ent);
    ent.parent = this;
    return ans;
  }
  async removeFromParent() {
    const ans = await new SetParentForEntityCommand(
      this.id,
      void 0
    ).execute();
    if (this.parent) {
      this.parent.children = this.parent.children.filter(
        (child) => child.id !== this.id
      );
      this.parent = null;
    }
    return ans;
  }
  async updateTransform(properties) {
    this.position = properties.position ?? this.position;
    this.rotation = properties.rotation ?? this.rotation;
    this.scale = properties.scale ?? this.scale;
    return new UpdateEntityPropertiesCommand(this, properties).execute();
  }
  async addEvent(type, callback) {
    if (this.events[type]) {
      this.events[type] = callback;
    } else {
      try {
        await this.updateEntityEvent(type, true);
        this.events[type] = callback;
      } catch (error) {
        console.error("addEvent failed", type);
      }
    }
  }
  async removeEvent(eventName) {
    if (this.events[eventName]) {
      delete this.events[eventName];
      try {
        await this.updateEntityEvent(eventName, false);
      } catch (error) {
        console.error("removeEvent failed", eventName);
      }
    }
  }
  async updateEntityEvent(eventName, isEnable) {
    return new UpdateEntityEventCommand(this, eventName, isEnable).execute();
  }
  onReceiveEvent = (data) => {
    const { type } = data;
    if (type === "objectdestroy" /* objectdestroy */) {
      this.isDestroyed = true;
    } else if (type === "spatialtap" /* spatialtap */) {
      const evt = createSpatialEvent("spatialtap" /* spatialtap */, data.detail);
      this.dispatchEvent(evt);
    } else if (type === "spatialdragstart" /* spatialdragstart */) {
      const evt = createSpatialEvent(
        "spatialdragstart" /* spatialdragstart */,
        data.detail
      );
      this.dispatchEvent(evt);
    } else if (type === "spatialdrag" /* spatialdrag */) {
      const evt = createSpatialEvent("spatialdrag" /* spatialdrag */, data.detail);
      this.dispatchEvent(evt);
    } else if (type === "spatialdragend" /* spatialdragend */) {
      const evt = createSpatialEvent(
        "spatialdragend" /* spatialdragend */,
        data.detail
      );
      this.dispatchEvent(evt);
    } else if (type === "spatialrotate" /* spatialrotate */) {
      const evt = createSpatialEvent(
        "spatialrotate" /* spatialrotate */,
        data.detail
      );
      this.dispatchEvent(evt);
    } else if (type === "spatialrotateend" /* spatialrotateend */) {
      const evt = createSpatialEvent(
        "spatialrotateend" /* spatialrotateend */,
        data.detail
      );
      this.dispatchEvent(evt);
    } else if (type === "spatialmagnify" /* spatialmagnify */) {
      const evt = createSpatialEvent(
        "spatialmagnify" /* spatialmagnify */,
        data.detail
      );
      this.dispatchEvent(evt);
    } else if (type === "spatialmagnifyend" /* spatialmagnifyend */) {
      const evt = createSpatialEvent(
        "spatialmagnifyend" /* spatialmagnifyend */,
        data.detail
      );
      this.dispatchEvent(evt);
    }
  };
  dispatchEvent(evt) {
    if (!evt.__origin) {
      Object.defineProperty(evt, "__origin", { value: this, enumerable: false });
    }
    this.events[evt.type]?.(evt);
    if (evt.bubbles && !evt.cancelBubble) {
      if (this.parent) {
        this.parent.dispatchEvent(evt);
      }
    }
  }
  onDestroy() {
    SpatialWebEvent.removeEventReceiver(this.id);
    this.children.forEach((child) => {
      child.parent = null;
    });
    this.children = [];
    if (this.parent) {
      this.parent.children = this.parent.children.filter(
        (child) => child.id !== this.id
      );
      this.parent = null;
    }
  }
  // onUpdate(properties: SpatialEntityProperties) {
  //   this.position = properties.position
  //   this.rotation = properties.rotation
  //   this.scale = properties.scale
  // }
  async convertFromEntityToEntity(fromEntityId, toEntityId, position) {
    return new ConvertFromEntityToEntityCommand(
      fromEntityId,
      toEntityId,
      position
    ).execute();
  }
  async convertFromEntityToScene(fromEntityId, position) {
    return new ConvertFromEntityToSceneCommand(fromEntityId, position).execute();
  }
  async convertFromSceneToEntity(entityId, position) {
    return new ConvertFromSceneToEntityCommand(entityId, position).execute();
  }
};

// src/reality/entity/SpatialModelEntity.ts
var SpatialModelEntity = class extends SpatialEntity {
  constructor(id, options, userData) {
    super(id, userData);
    this.id = id;
    this.options = options;
    this.userData = userData;
  }
  async setMaterials(materials) {
    return new SetMaterialsOnEntityCommand(this.id, materials).execute();
  }
};

// src/reality/component/SpatialComponent.ts
var SpatialComponent = class extends SpatialObject {
  constructor(id) {
    super(id);
    SpatialWebEvent.addEventReceiver(id, this.onReceiveEvent);
  }
  onReceiveEvent = (data) => {
    const { type } = data;
    if (type === "objectdestroy" /* objectdestroy */) {
      this.isDestroyed = true;
    }
  };
};

// src/reality/component/ModelComponent.ts
var ModelComponent = class extends SpatialComponent {
  constructor(id, options) {
    super(id);
    this.options = options;
  }
};

// src/reality/material/SpatialMaterial.ts
var SpatialMaterial = class extends SpatialObject {
  constructor(id, type) {
    super(id);
    this.id = id;
    this.type = type;
    this.type = type;
  }
};

// src/reality/material/SpatialUnlitMaterial.ts
var SpatialUnlitMaterial = class extends SpatialMaterial {
  constructor(id, options) {
    super(id, "unlit");
    this.id = id;
    this.options = options;
  }
  updateProperties(properties) {
    return new UpdateUnlitMaterialProperties(this, properties).execute();
  }
};

// src/reality/resource/SpatialModelAsset.ts
var SpatialModelAsset = class extends SpatialObject {
  constructor(id, options) {
    super(id);
    this.id = id;
    this.options = options;
  }
};

// src/reality/resource/SpatialTextureResource.ts
var SpatialTextureResource = class extends SpatialObject {
  constructor(id, options) {
    super(id);
    this.id = id;
    this.options = options;
  }
  updateProperties(properties) {
    return new UpdateTexturePropertiesCommand(this, properties).execute();
  }
};

// src/reality/realityCreator.ts
async function createSpatialEntity(userData) {
  const result = await new CreateSpatialEntityCommand(userData?.name).execute();
  if (!result.success) {
    throw new Error("createSpatialEntity failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new SpatialEntity(id, userData);
  }
}
async function createSpatialGeometry(ctor, options) {
  const result = await new CreateSpatialGeometryCommand(
    ctor.type,
    options
  ).execute();
  if (!result.success) {
    throw new Error("createSpatialGeometry failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new ctor(id, options);
  }
}
async function createSpatialUnlitMaterial(options) {
  const result = await new CreateSpatialUnlitMaterialCommand(options).execute();
  if (!result.success) {
    throw new Error("createSpatialUnlitMaterial failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new SpatialUnlitMaterial(id, options);
  }
}
async function createModelComponent(options) {
  const result = await new CreateModelComponentCommand(options).execute();
  if (!result.success) {
    throw new Error("createModelComponent failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new ModelComponent(id, options);
  }
}
async function createSpatialModelEntity(options, userData) {
  const result = await new CreateSpatialModelEntityCommand(options).execute();
  if (!result.success) {
    throw new Error("createSpatialModelEntity failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new SpatialModelEntity(id, options, userData);
  }
}
async function createModelAsset(options) {
  const result = await new CreateModelAssetCommand(options).execute();
  if (!result.success) {
    throw new Error("createModelAsset failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new SpatialModelAsset(id, options);
  }
}
async function createSpatialTexture(options) {
  const result = await new CreateTextureCommand(options.url).execute();
  if (!result.success) {
    throw new Error("createSpatialTexture failed:" + result?.errorMessage);
  } else {
    const { id } = result.data;
    return new SpatialTextureResource(id, options);
  }
}

// src/reality/geometry/SpatialGeometry.ts
var SpatialGeometry = class extends SpatialObject {
  constructor(id, options) {
    super(id);
    this.id = id;
    this.options = options;
  }
  static type;
};

// src/reality/geometry/SpatialBoxGeometry.ts
var SpatialBoxGeometry = class extends SpatialGeometry {
  constructor(id, options) {
    super(id, options);
    this.id = id;
    this.options = options;
  }
  static type = "BoxGeometry";
};

// src/reality/geometry/SpatialSphereGeometry.ts
var SpatialSphereGeometry = class extends SpatialGeometry {
  constructor(id, options) {
    super(id, options);
    this.id = id;
    this.options = options;
  }
  static type = "SphereGeometry";
};

// src/reality/geometry/SpatialCylinderGeometry.ts
var SpatialCylinderGeometry = class extends SpatialGeometry {
  constructor(id, options) {
    super(id, options);
    this.id = id;
    this.options = options;
  }
  static type = "CylinderGeometry";
};

// src/reality/geometry/SpatialPlaneGeometry.ts
var SpatialPlaneGeometry = class extends SpatialGeometry {
  constructor(id, options) {
    super(id, options);
    this.id = id;
    this.options = options;
  }
  static type = "PlaneGeometry";
};

// src/reality/geometry/SpatialConeGeometry.ts
var SpatialConeGeometry = class extends SpatialGeometry {
  constructor(id, options) {
    super(id, options);
    this.id = id;
    this.options = options;
  }
  static type = "ConeGeometry";
};

// src/SpatialSession.ts
var SpatialSession = class {
  /**
   * Gets the singleton instance of the spatial scene.
   * The spatial scene is the root container for all spatial elements.
   * @returns The SpatialScene singleton instance
   */
  getSpatialScene() {
    return SpatialScene.getInstance();
  }
  /**
   * Creates a new 2D element that can be spatialized in the 3D environment.
   * 2D elements represent HTML content that can be positioned in 3D space.
   * @returns Promise resolving to a new Spatialized2DElement instance
   */
  createSpatialized2DElement() {
    return createSpatialized2DElement();
  }
  /**
   * Creates a new static 3D element with an optional model URL.
   * Static 3D elements represent pre-built 3D models that can be loaded from a URL.
   * @param modelURL Optional URL to the 3D model to load
   * @param sources Optional list of fallback model sources
   * @param loading Whether the asset should fetch eagerly or be deferred (`'lazy'`)
   * @returns Promise resolving to a new SpatializedStatic3DElement instance
   */
  createSpatializedStatic3DElement(modelURL, sources, loading = "eager") {
    return createSpatializedStatic3DElement(modelURL, sources, loading);
  }
  /**
   * Initializes the spatial scene with custom configuration.
   * This is a reference to the initScene function from scene-polyfill.
   */
  initScene = initScene;
  /**
   * Creates a new dynamic 3D element that can be manipulated at runtime.
   * Dynamic 3D elements allow for programmatic creation and modification of 3D content.
   * @returns Promise resolving to a new SpatializedDynamic3DElement instance
   */
  createSpatializedDynamic3DElement() {
    return createSpatializedDynamic3DElement();
  }
  /**
   * Creates a new spatial entity with an optional name.
   * Entities are the basic building blocks for creating custom 3D content.
   * @param name Optional name for the entity
   * @returns Promise resolving to a new SpatialEntity instance
   */
  createEntity(userData) {
    return createSpatialEntity(userData);
  }
  /**
   * Creates a box geometry with optional configuration.
   * @param options Configuration options for the box geometry
   * @returns Promise resolving to a new SpatialBoxGeometry instance
   */
  createBoxGeometry(options = {}) {
    return createSpatialGeometry(SpatialBoxGeometry, options);
  }
  /**
   * Creates a plane geometry with optional configuration.
   * @param options Configuration options for the plane geometry
   * @returns Promise resolving to a new SpatialPlaneGeometry instance
   */
  createPlaneGeometry(options = {}) {
    return createSpatialGeometry(SpatialPlaneGeometry, options);
  }
  /**
   * Creates a sphere geometry with optional configuration.
   * @param options Configuration options for the sphere geometry
   * @returns Promise resolving to a new SpatialSphereGeometry instance
   */
  createSphereGeometry(options = {}) {
    return createSpatialGeometry(SpatialSphereGeometry, options);
  }
  /**
   * Creates a cone geometry with the specified configuration.
   * @param options Configuration options for the cone geometry
   * @returns Promise resolving to a new SpatialConeGeometry instance
   */
  createConeGeometry(options) {
    return createSpatialGeometry(SpatialConeGeometry, options);
  }
  /**
   * Creates a cylinder geometry with the specified configuration.
   * @param options Configuration options for the cylinder geometry
   * @returns Promise resolving to a new SpatialCylinderGeometry instance
   */
  createCylinderGeometry(options) {
    return createSpatialGeometry(SpatialCylinderGeometry, options);
  }
  /**
   * Creates a model component with the specified configuration.
   * Model components are used to add 3D model rendering capabilities to entities.
   * @param options Configuration options for the model component
   * @returns Promise resolving to a new ModelComponent instance
   */
  createModelComponent(options) {
    return createModelComponent(options);
  }
  /**
   * Creates an unlit material with the specified configuration.
   * Unlit materials don't respond to lighting in the scene.
   * @param options Configuration options for the unlit material
   * @returns Promise resolving to a new SpatialUnlitMaterial instance
   */
  createUnlitMaterial(options) {
    return createSpatialUnlitMaterial(options);
  }
  /**
   * Creates a texture resource from the specified image URL.
   * Texture resources can be referenced by materials and other spatial content.
   * @param options Configuration options for the texture resource
   * @returns Promise resolving to a new SpatialTextureResource instance
   */
  createTexture(options) {
    return createSpatialTexture(options);
  }
  /**
   * Creates a model asset with the specified configuration.
   * Model assets represent 3D model resources that can be used by entities.
   * @param options Configuration options for the model asset
   * @returns Promise resolving to a new SpatialModelAsset instance
   */
  createModelAsset(options) {
    return createModelAsset(options);
  }
  /**
   * Creates a spatial model entity with the specified configuration.
   * This is a convenience method for creating an entity with a model component.
   * @param options Configuration options for the spatial model entity
   * @returns Promise resolving to a new SpatialModelEntity instance
   */
  createSpatialModelEntity(options, userData) {
    return createSpatialModelEntity(options, userData);
  }
  /**
   * Creates an attachment entity that renders 2D HTML content as a child
   * of a 3D entity in the scene graph.
   * @param options Configuration options including parent entity ID, position, and size
   * @returns Promise resolving to a new Attachment instance
   */
  createAttachmentEntity(options) {
    return createAttachmentEntity(options);
  }
};

// src/Spatial.ts
var Spatial = class {
  /**
   * Requests a spatial session object from the browser.
   * This is the primary method to initialize spatial functionality.
   * @returns The SpatialSession instance or null if not available in the current browser
   * [TODO] discuss implications of this not being async
   */
  requestSession() {
    if (this.runInSpatialWeb()) {
      SpatialWebEvent.init();
      return new SpatialSession();
    } else {
      return null;
    }
  }
  /**
   * Checks if the current page is running in a spatial web environment.
   * This method detects if the application is running in a WebSpatial-compatible browser.
   * @returns True if running in a spatial web environment, false otherwise
   */
  runInSpatialWeb() {
    if (navigator.userAgent.indexOf("WebSpatial/") >= 0) {
      return true;
    }
    return false;
  }
  /** @deprecated
   * Checks if WebSpatial is supported in the current environment.
   * Verifies compatibility between native and client versions.
   * @returns True if web spatial is supported by this webpage
   */
  isSupported() {
    return true;
  }
  /** @deprecated
   * Gets the native WebSpatial version from the browser environment.
   * The version format follows semantic versioning (x.x.x).
   * @returns Native version string in format "x.x.x"
   */
  getNativeVersion() {
    if (window.__WebSpatialData && window.__WebSpatialData.getNativeVersion) {
      return window.__WebSpatialData.getNativeVersion();
    }
    return window.WebSpatailNativeVersion === "WS_SHELL_VERSION" ? this.getClientVersion() : window.WebSpatailNativeVersion;
  }
  /** @deprecated
   * Gets the client SDK version.
   * The version format follows semantic versioning (x.x.x).
   * @returns Client SDK version string in format "x.x.x"
   */
  getClientVersion() {
    return "1.7.0";
  }
};

// src/runtime/WebSpatialRuntimeError.ts
var WebSpatialRuntimeError = class extends Error {
  capability;
  constructor(capability, message) {
    super(
      message ?? `Capability "${capability}" is not supported in this WebSpatial runtime`
    );
    this.name = "WebSpatialRuntimeError";
    this.capability = capability;
  }
};

// src/runtime/keys.ts
var COMPONENT_KEYS = [
  "Model",
  "Reality",
  "Entity",
  "BoxEntity",
  "SphereEntity",
  "ConeEntity",
  "CylinderEntity",
  "PlaneEntity",
  "SceneGraph",
  "ModelAsset",
  "ModelEntity",
  "UnlitMaterial",
  "Material",
  "AttachmentAsset",
  "AttachmentEntity"
];
var CSS_KEYS = [
  "-xr-background-material",
  "-xr-back",
  "-xr-depth",
  "-xr-transform"
];
var GESTURE_KEYS = [
  "SpatialTapEvent",
  "SpatialDragStartEvent",
  "SpatialDragEvent",
  "SpatialDragEndEvent",
  "SpatialRotateEvent",
  "SpatialRotateEndEvent",
  "SpatialMagnifyEvent",
  "SpatialMagnifyEndEvent"
];
var JS_SCENE_KEYS = [
  "useMetrics",
  "convertCoordinate",
  "initScene",
  "WindowScene",
  "VolumeScene"
];
var ELEMENT_DOM_DEPTH_KEYS = ["xrClientDepth", "xrOffsetBack"];
var WINDOW_DOM_DEPTH_KEYS = ["xrInnerDepth", "xrOuterDepth"];
var DOM_DEPTH_KEYS = [
  ...ELEMENT_DOM_DEPTH_KEYS,
  ...WINDOW_DOM_DEPTH_KEYS
];
var TOP_LEVEL_KEYS = [
  ...COMPONENT_KEYS,
  ...CSS_KEYS,
  ...GESTURE_KEYS,
  ...JS_SCENE_KEYS,
  ...DOM_DEPTH_KEYS
];
var ALIAS_TO_CANONICAL = {
  Box: "BoxEntity",
  Sphere: "SphereEntity",
  Cone: "ConeEntity",
  Cylinder: "CylinderEntity",
  Plane: "PlaneEntity",
  World: "SceneGraph"
};
function normalizeCapabilityName(name) {
  return ALIAS_TO_CANONICAL[name] ?? name;
}
var SUB_TOKENS_BY_NAME = {
  Material: ["unlit"],
  WindowScene: ["defaultSize", "resizability"],
  VolumeScene: [
    "defaultSize",
    "resizability",
    "worldScaling",
    "worldAlignment",
    "baseplateVisibility"
  ],
  SpatialRotateEvent: ["constrainedToAxis"],
  Model: [
    "autoplay",
    "loop",
    "stagemode",
    "poster",
    "loading",
    "source",
    "ready",
    "currentSrc",
    "entityTransform",
    "paused",
    "duration",
    "playbackRate",
    "play",
    "pause",
    "currentTime"
  ]
};
function isKnownTopLevel(name) {
  return TOP_LEVEL_KEYS.includes(name);
}
function isKnownSubToken(name, token) {
  const allowed = SUB_TOKENS_BY_NAME[name];
  return allowed !== void 0 && allowed.includes(token);
}

// src/runtime/capability-data.ts
function baseTrueFlags() {
  const flags = {};
  for (const k of TOP_LEVEL_KEYS) {
    flags[k] = true;
  }
  for (const [name, toks] of Object.entries(SUB_TOKENS_BY_NAME)) {
    for (const t of toks) {
      flags[`${name}:${t}`] = true;
    }
  }
  return flags;
}
function matrixVision_1_5_0_Flags() {
  const flags = baseTrueFlags();
  const modelNo = [
    "autoplay",
    "loop",
    "stagemode",
    "poster",
    "loading",
    "source",
    "paused",
    "duration",
    "playbackRate",
    "play",
    "pause",
    "currentTime"
  ];
  for (const t of modelNo) {
    flags[`Model:${t}`] = false;
  }
  flags["SpatialRotateEvent:constrainedToAxis"] = true;
  return flags;
}
function matrixVision_1_6_0_Flags() {
  const flags = baseTrueFlags();
  for (const t of ["stagemode", "poster", "loading", "currentTime"]) {
    flags[`Model:${t}`] = false;
  }
  flags["SpatialRotateEvent:constrainedToAxis"] = true;
  return flags;
}
function matrixPico_0_1_1_Flags() {
  const flags = matrixVision_1_5_0_Flags();
  flags.xrInnerDepth = false;
  flags.xrOuterDepth = false;
  return flags;
}
function matrixPico_0_1_2_Flags() {
  const flags = matrixVision_1_6_0_Flags();
  flags.xrInnerDepth = false;
  flags.xrOuterDepth = false;
  return flags;
}
function visionOsRow_1_5_0() {
  return { version: "1.5.0", flags: matrixVision_1_5_0_Flags() };
}
function visionOsRow_1_6_0() {
  return { version: "1.6.0", flags: matrixVision_1_6_0_Flags() };
}
function picoOsRow_0_1_1() {
  return { version: "0.1.1", flags: matrixPico_0_1_1_Flags() };
}
function picoOsRow_0_1_2() {
  return { version: "0.1.2", flags: matrixPico_0_1_2_Flags() };
}
var CAPABILITY_TABLE = {
  visionos: [visionOsRow_1_5_0(), visionOsRow_1_6_0()],
  picoos: [picoOsRow_0_1_1(), picoOsRow_0_1_2()]
};

// src/runtime/semver.ts
function compareSemver(a, b) {
  const pa = parseSemverParts(a);
  const pb = parseSemverParts(b);
  if (!pa || !pb) {
    return String(a).localeCompare(String(b));
  }
  const len = Math.max(pa.length, pb.length);
  for (let i = 0; i < len; i++) {
    const da = pa[i] ?? 0;
    const db = pb[i] ?? 0;
    if (da !== db) return da - db;
  }
  return 0;
}
function parseSemverParts(v) {
  const m = /^(\d+)(?:\.(\d+)(?:\.(\d+))?)?/.exec(v.trim());
  if (!m) return null;
  const major = Number(m[1]);
  const minor = m[2] !== void 0 ? Number(m[2]) : 0;
  const patch = m[3] !== void 0 ? Number(m[3]) : 0;
  if ([major, minor, patch].some((n) => Number.isNaN(n))) return null;
  return [major, minor, patch];
}
function parseSemverOrNull(v) {
  const m = /^(\d+(?:\.\d+){0,2})/.exec(v.trim());
  return m ? m[1] : null;
}

// src/runtime/supports.ts
var VISIONOS_DEBUG_SHELL_VERSION_PLACEHOLDER = "WS_SHELL_VERSION";
var runtimeCache;
function resetRuntimeCacheForTests() {
  runtimeCache = void 0;
}
function getRuntime() {
  if (runtimeCache !== void 0) return runtimeCache;
  if (typeof navigator === "undefined") {
    runtimeCache = { type: null, shellVersion: null };
    return runtimeCache;
  }
  runtimeCache = computeRuntimeFromUserAgent(navigator.userAgent);
  return runtimeCache;
}
function selectRow(type, shellVersion) {
  const norm = parseSemverOrNull(shellVersion);
  if (!norm) return null;
  const rows = CAPABILITY_TABLE[type];
  if (!rows.length) return null;
  const sorted = [...rows].sort((a, b) => compareSemver(a.version, b.version));
  const minV = sorted[0].version;
  if (compareSemver(norm, minV) < 0) return null;
  let chosen = null;
  for (const row of sorted) {
    if (compareSemver(row.version, norm) <= 0) {
      chosen = row;
    } else {
      break;
    }
  }
  return chosen;
}
function supports(name, tokens) {
  if (typeof name !== "string") return false;
  const canonical = normalizeCapabilityName(name);
  if (!isKnownTopLevel(canonical)) return false;
  const tokList = tokens === void 0 ? [] : Array.isArray(tokens) ? [...tokens] : [];
  if (tokList.some((t) => typeof t !== "string")) return false;
  for (const t of tokList) {
    if (!isKnownSubToken(canonical, t)) return false;
  }
  const rt = getRuntime();
  if (rt.type === "puppeteer") {
    return true;
  }
  if (rt.type === null) return false;
  if (rt.shellVersion === null) return false;
  if (rt.type === "visionos" && rt.shellVersion === VISIONOS_DEBUG_SHELL_VERSION_PLACEHOLDER) {
    return true;
  }
  const parsedShell = parseSemverOrNull(rt.shellVersion);
  if (!parsedShell) return false;
  if (rt.type !== "visionos" && rt.type !== "picoos") return false;
  const row = selectRow(rt.type, parsedShell);
  if (!row) return false;
  if (tokList.length === 0) {
    return row.flags[canonical] === true;
  }
  if (row.flags[canonical] !== true) return false;
  return tokList.every((t) => row.flags[`${canonical}:${t}`] === true);
}

// src/spatial-window-polyfill.ts
var spatial = new Spatial();
var session = void 0;
var SpatialGlobalCustomVars = {
  backgroundMaterial: "--xr-background-material"
};
var htmlBackgroundMaterial = "";
function setCurrentWindowStyle(backgroundMaterial) {
  if (backgroundMaterial !== htmlBackgroundMaterial) {
    session?.getSpatialScene()?.updateSpatialProperties({
      material: backgroundMaterial
    });
    htmlBackgroundMaterial = backgroundMaterial;
  }
}
function checkHtmlBackgroundMaterial() {
  const computedStyle = getComputedStyle(document.documentElement);
  const backgroundMaterial = computedStyle.getPropertyValue(
    SpatialGlobalCustomVars.backgroundMaterial
  );
  setCurrentWindowStyle(backgroundMaterial || "none");
}
var htmlCornerRadius = {
  topLeading: 0,
  bottomLeading: 0,
  topTrailing: 0,
  bottomTrailing: 0
};
function checkCornerRadius() {
  const computedStyle = getComputedStyle(document.documentElement);
  const cornerRadius = parseCornerRadius(computedStyle);
  setCornerRadius(cornerRadius);
}
function setCornerRadius(cornerRadius) {
  if (htmlCornerRadius.topLeading !== cornerRadius.topLeading || htmlCornerRadius.bottomLeading !== cornerRadius.bottomLeading || htmlCornerRadius.topTrailing !== cornerRadius.topTrailing || htmlCornerRadius.bottomTrailing !== cornerRadius.bottomTrailing) {
    session?.getSpatialScene()?.updateSpatialProperties({
      cornerRadius
    });
    htmlCornerRadius.topLeading = cornerRadius.topLeading;
    htmlCornerRadius.bottomLeading = cornerRadius.bottomLeading;
    htmlCornerRadius.topTrailing = cornerRadius.topTrailing;
    htmlCornerRadius.bottomTrailing = cornerRadius.bottomTrailing;
  }
}
function setOpacity(opacity) {
  session?.getSpatialScene().updateSpatialProperties({
    opacity
  });
}
function checkOpacity() {
  const computedStyle = getComputedStyle(document.documentElement);
  const opacity = parseFloat(computedStyle.getPropertyValue("opacity"));
  setOpacity(opacity);
}
function hijackDocumentElementStyle() {
  const rawDocumentStyle = document.documentElement.style;
  const styleProxy = new Proxy(rawDocumentStyle, {
    set: function(target, key, value) {
      const ret = Reflect.set(target, key, value);
      if (key === SpatialGlobalCustomVars.backgroundMaterial) {
        setCurrentWindowStyle(value);
      }
      if (key === "border-radius" || key === "borderRadius" || key === "border-top-left-radius" || key === "borderTopLeftRadius" || key === "border-top-right-radius" || key === "borderTopRightRadius" || key === "border-bottom-left-radius" || key === "borderBottomLeftRadius" || key === "border-bottom-right-radius" || key === "borderBottomRightRadius") {
        checkCornerRadius();
      }
      if (key === "opacity") {
        checkOpacity();
      }
      return ret;
    },
    get: function(target, prop) {
      if (typeof target[prop] === "function") {
        return function(...args) {
          if (prop === "setProperty") {
            const [property, value] = args;
            if (property === SpatialGlobalCustomVars.backgroundMaterial) {
              setCurrentWindowStyle(value);
            }
          } else if (prop === "removeProperty") {
            const [property] = args;
            if (property === SpatialGlobalCustomVars.backgroundMaterial) {
              setCurrentWindowStyle("none");
            }
          }
          return target[prop](
            ...args
          );
        };
      }
      return Reflect.get(target, prop);
    }
  });
  Object.defineProperty(document.documentElement, "style", {
    get: function() {
      return styleProxy;
    }
  });
}
function monitorExternalStyleChange() {
  const headObserver = new MutationObserver(checkCSSProperties);
  headObserver.observe(document.head, { childList: true, subtree: true });
}
function checkCSSProperties() {
  checkHtmlBackgroundMaterial();
  checkCornerRadius();
  checkOpacity();
}
function monitorHTMLAttributeChange() {
  const observer = new MutationObserver((mutations) => {
    mutations.forEach((mutation) => {
      if (mutation.type === "attributes" && mutation.attributeName) {
        checkCSSProperties();
      }
    });
  });
  observer.observe(document.documentElement, {
    attributes: true,
    attributeFilter: ["style", "class"]
  });
}
async function spatialWindowPolyfill() {
  if (!spatial.runInSpatialWeb()) {
    return;
  }
  session = await spatial.requestSession();
  if (document.readyState === "complete") {
    checkCSSProperties();
  } else {
    window.addEventListener("load", () => {
      checkCSSProperties();
    });
  }
  hijackDocumentElementStyle();
  monitorExternalStyleChange();
  monitorHTMLAttributeChange();
}

// src/index.ts
if (!isSSREnv() && navigator.userAgent.indexOf("WebSpatial/") > 0) {
  injectSceneHook();
  spatialWindowPolyfill();
}
export {
  Attachment,
  BaseplateVisibilityValues,
  CAPABILITY_TABLE,
  COMPONENT_KEYS,
  CSS_KEYS,
  CubeInfo,
  DOM_DEPTH_KEYS,
  ELEMENT_DOM_DEPTH_KEYS,
  GESTURE_KEYS,
  JS_SCENE_KEYS,
  ModelComponent,
  physicalMetrics_exports as PhysicalMetrics,
  SUB_TOKENS_BY_NAME,
  Spatial,
  SpatialBoxGeometry,
  SpatialComponent,
  SpatialConeGeometry,
  SpatialCylinderGeometry,
  SpatialEntity,
  SpatialGeometry,
  SpatialMaterial,
  SpatialModelAsset,
  SpatialModelEntity,
  SpatialObject,
  SpatialPlaneGeometry,
  SpatialScene,
  SpatialSceneState,
  SpatialSceneValues,
  SpatialSession,
  SpatialSphereGeometry,
  SpatialTextureResource,
  SpatialUnlitMaterial,
  Spatialized2DElement,
  SpatializedDynamic3DElement,
  SpatializedElement,
  SpatializedElementType,
  SpatializedStatic3DElement,
  TOP_LEVEL_KEYS,
  VISIONOS_DEBUG_SHELL_VERSION_PLACEHOLDER,
  WINDOW_DOM_DEPTH_KEYS,
  WebSpatialRuntimeError,
  WorldAlignmentValues,
  WorldScalingValues,
  compareSemver,
  computeRuntimeFromUserAgent,
  createAttachmentEntity,
  getRuntime,
  isSSREnv,
  isValidBaseplateVisibilityType,
  isValidSceneUnit,
  isValidSpatialSceneType,
  isValidWorldAlignmentType,
  isValidWorldScalingType,
  parseSemverOrNull,
  parseShellToken,
  resetRuntimeCacheForTests,
  resolveJsbAdapterPlatform,
  supports
};
