// @ies/starling_client@3.6.6/rollup.config.js
import typescript from 'rollup-plugin-typescript2';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import { terser } from 'rollup-plugin-terser';
import json from '@rollup/plugin-json';
import filesize from 'rollup-plugin-filesize';
import path from 'path';
import fs from 'fs';
import pkg from './package.json';

function deleteall(path) {
  let files = [];
  if (fs.existsSync(path)) {
    files = fs.readdirSync(path);
    files.forEach(function (file) {
      const curPath = path + '/' + file;
      if (fs.statSync(curPath).isDirectory()) {
        deleteall(curPath);
      } else {
        fs.unlinkSync(curPath);
      }
    });
    fs.rmdirSync(path);
  }
}

deleteall(path.resolve(__dirname, './lib'));

const web = './src/index.ts';
const webBase = './src/base.ts';

export default [
  {
    input: web,
    output: [
      {
        file: pkg.module,
        format: 'es',
      },
      {
        file: pkg.main,
        format: 'cjs',
      },
      {
        file: pkg.unpkg,
        format: 'umd',
        name: 'StarlingClient',
        plugins: [terser()],
      },
    ],
    plugins: [
      json(),
      filesize(),
      resolve({
        browser: true,
      }),
      typescript(),
      commonjs(),
    ],
  },
  {
    input: webBase,
    output: [
      {
        file: pkg.module.replace('starling_client.esm', 'base'),
        format: 'es',
      },
      {
        file: pkg.main.replace('starling_client', 'base'),
        format: 'cjs',
      },
      {
        file: pkg.unpkg.replace('starling_client', 'base'),
        format: 'umd',
        name: 'StarlingClient',
        plugins: [terser()],
      },
    ],
    plugins: [
      json(),
      filesize(),
      resolve({
        browser: true,
      }),
      typescript(),
      commonjs(),
    ],
  },
];
