// byted-tea-sdk@5.3.10/src/plugin/request/request.ts
// 通用请求管理
import { IInitParam, EventParams } from '../../../types/types'
import Collector from '../../collect/collect';
import pako from 'pako'
import Types, { DebuggerMesssge } from "../../collect/hooktype";



const isSupportBeacon = () => {
  if (window.navigator && window.navigator.sendBeacon) {
    return true
  } else {
    return false
  }
}
interface RequestOptions {
  url: string,
  data?: any,
  method?: string,
  success?: Function,
  fail?: Function,
  timeout?: number,
  useBeacon?: boolean,
  withCredentials?: boolean,
  app_key?: string,
  forceXhr?: boolean,
  status?: string,
  type?: string,
  zip?: boolean
}
class RequestManager {
  collector: Collector
  config: IInitParam
  errorCode: Record<string, number>
  errorMessage: Record<string, string>
  requestType: string = ''
  supportBeacon: boolean
  customHeader: EventParams
  supportZip: boolean
  constructor(collector: Collector, config: IInitParam) {
    this.collector = collector;
    this.config = config;
    this.requestType = config.request_type;
    this.supportBeacon = isSupportBeacon();
    this.errorCode = {
      NO_URL: 4001,
      IMG_ON: 4000,
      IMG_CATCH: 4002,
      BEACON_FALSE: 4003,
      XHR_ERROR: 5002,
      TIMEOUT: 5005,
      FETCH_ERROR: 5006,
    }
    this.errorMessage = {
      BEACON_FALSE: 'xhr sendbeacon error',
      XHR_ERROR: 'xhr request error',
      BEACON: 'xhr sendbeacon error',
      TIMEOUT: 'xhr request timeout',
      FETCH_ERROR: 'fetch network error',
      FETCH_RESPONSE_ERROR: 'fetch response error',
    }
    this.customHeader = config.custom_request_header || {};
    this.collector.on(Types.CustomHeader, (info: EventParams) => {
      this.setRequestHeader(info)
    })
    this.supportZip = this.config.channel_domain ? false : true;
  }

  setRequestHeader(info: EventParams) {
    this.customHeader = Object.assign(this.customHeader, info);
  }

  useFetch(requestOptions: RequestOptions) {
    // 自定义场景，使用fetch
    const { url, data, method, success, fail, zip, timeout, useBeacon } = requestOptions;
    if (window.fetch && typeof Request !== 'undefined') {
      if (useBeacon && !('keepalive' in Request.prototype)) {
        // 支持fetch但不支持keepalive时，离开类的请求使用sendBeacon，且忽略压缩
        this.useBeacon(requestOptions);
        return;
      }
      const headers = {
        'Content-Type': 'application/json; charset=utf-8'
      }
      if (Object.keys(this.customHeader).length) {
        for (const key in this.customHeader) {
          headers[key] = this.customHeader[key]
        }
      }
      let controller = null;
      if (typeof AbortController !== 'undefined') {
        controller = new AbortController();
      } else {
        controller = { signal: null, abort: () => { } }
      }
      const timeoutId = setTimeout(() => {
        controller.abort();
      }, timeout);
      let requestBody = JSON.stringify(data);
      if (this.supportZip && zip && this.config.need_zip) {
        const textEncoder = new TextEncoder();
        const strToUint8 = textEncoder.encode(requestBody);
        requestBody = pako.gzip(strToUint8);
        headers['Content-Encoding'] = 'gzip';
      }
      const startTime = Date.now();
      try {
        fetch(url, {
          method: method || 'POST',
          headers,
          body: requestBody,
          keepalive: true,
          signal: controller.signal
        }).then(response => {
          if (!response.ok) {
            fail && fail(data, response.status, startTime, response.statusText || this.errorMessage.FETCH_RESPONSE_ERROR);
          }
          return response.json();
        })
          .then(res => {
            clearTimeout(timeoutId);
            success && success(res, data, startTime, 200);
          })
          .catch(error => {
            clearTimeout(timeoutId);
            fail && fail(data, this.errorCode.FETCH_ERROR, startTime, error.message || this.errorMessage.FETCH_ERROR);
          })
      } catch (e) {
        this.useXhr(requestOptions);
        this.collector.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK useFetch 发送请求发生了一些问题', level: 'error', time: Date.now(), data: e.message || 'SDK useFetch has some errors', infoType: 'sdk' });
      }
      
    } else {
      if (useBeacon) {
        this.useBeacon(requestOptions);
      } else {
        this.requestType = '';
        this.useXhr(requestOptions);
      }
      console.log('your brwoser not support fetch, use xhr');
    }
  }
  useBeacon(requestOptions: RequestOptions) {
    // 使用window.sendbeacon，主要在跳转场景下
    const { url, data, success, fail } = requestOptions;
    const startTime = Date.now();
    const reponseStatus = window.navigator.sendBeacon(url, JSON.stringify(data));
    if (reponseStatus) {
      success && success({ e: 0 }, data, startTime, 200);
    } else {
      fail && fail(data, this.errorCode.BEACON_FALSE, startTime, this.errorMessage.BEACON_FALSE);
    }
  }
  useXhr(requestOptions: RequestOptions) {
    // 常规场景，使用xmlhttprequest
    const { url, data, method, success, fail, timeout, withCredentials, app_key, zip } = requestOptions;
    try {
      const xhr = new XMLHttpRequest()
      const requestMethod = method || 'POST';
      const startTime = Date.now();
      xhr.open(requestMethod, `${url}`, true);
      xhr.setRequestHeader('Content-Type', 'application/json; charset=utf-8');
      if (app_key) {
        xhr.setRequestHeader('X-MCS-AppKey', `${app_key}`);
      }
      if (Object.keys(this.customHeader).length) {
        for (const key in this.customHeader) {
          xhr.setRequestHeader(key, this.customHeader[key]);
        }
      }
      if (withCredentials) {
        xhr.withCredentials = true;
      }
      if (timeout) {
        xhr.timeout = timeout;
        xhr.ontimeout = () => {
          fail && fail(data, this.errorCode.TIMEOUT, startTime, this.errorMessage.TIMEOUT);
        }
      }
      xhr.onload = () => {
        if (success) {
          var res = null
          if (xhr.responseText) {
            try {
              res = JSON.parse(xhr.responseText);
            } catch (e) {
              res = {}
            }
            success(res, data, startTime, xhr.status);
          }
        }
      }
      xhr.onerror = () => {
        xhr.abort()
        fail && fail(data, this.errorCode.XHR_ERROR, startTime, this.errorMessage.XHR_ERROR);
      }
      // try {
      //   // 使用 pako 进行 gzip 压缩
      //   if (zip && !this.config.disable_zip) {
      //     const textEncoder = new TextEncoder();
      //     const strToUint8 = textEncoder.encode(JSON.stringify(data));
      //     const zip = pako.gzip(strToUint8);
      //     xhr.setRequestHeader('Content-Encoding', 'gzip');
      //     xhr.send(zip)
      //   } else {
      //     xhr.send(JSON.stringify(data))
      //   }
      // } catch (e) {
      //   xhr.send(JSON.stringify(data))
      //   this.collector.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK Pako Zip Something Error', logType: 'fetch', level: 'error', data: e.message, time: Date.now(), infoType: 'sdk' })
      // }
      // xhr.send(JSON.stringify(data))
      try {
        const sendData = JSON.stringify(data);
        // 使用 pako 进行 gzip 压缩
        if (this.config.need_zip && zip) {
          if (this.supportZip) {
            const textEncoder = new TextEncoder();
            const strToUint8 = textEncoder.encode(sendData);
            const compressData = pako.gzip(strToUint8);
            xhr.setRequestHeader('Content-Encoding', 'gzip');
            xhr.send(compressData);
          } else {
            xhr.send(sendData);
            this.collector.logger.warn('your channel do not support gzip, use normal data!');
          }
        } else {
          // xhr.setRequestHeader('Content-Encoding', '');
          xhr.send(sendData);
        }
      } catch (e) {
        xhr.send(JSON.stringify(data));
        this.collector.emit('sdk-inner-error', { type: 'sdk_error_send', eventData: data, errorCode: 1000, msg: e.message || e.stack || 'sdk_error_send' });
      }
    } catch (e) {
      this.collector.emit('sdk-inner-error', { type: 'sdk_error_request', eventData: data, errorCode: 1001, msg: e.message || e.stack || 'sdk_error_request' });
      this.collector.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK useXhr has some error', level: 'error', time: Date.now(), data: e.message, infoType: 'sdk' });
    }
  }
  useRequest(requestOptions: RequestOptions) {
    const { useBeacon } = requestOptions;
    if (this.config.transport && this.collector.configManager.getEnableClientTransport()) {
      this.useCustomRequest(requestOptions);
      return;
    }
    if (this.requestType && this.requestType === 'xhr') {
      this.useXhr(requestOptions);
      return;
    }

    if (useBeacon && this.supportBeacon) {
      this.useBeacon(requestOptions)
      return;
    }

    if (this.requestType && this.requestType === 'fetch') {
      this.useFetch(requestOptions)
      return;
    }

    this.useXhr(requestOptions);

    // if (this.requestType === 'xhr') {
    //   this.useXhr(requestOptions)
    // } else {
    //   if (useBeacon && this.supportBeacon) {
    //     this.useBeacon(requestOptions)
    //     return;
    //   }
    //   if (this.requestType === 'fetch') {
    //     this.useFetch(requestOptions)
    //     return;
    //   }
    //   this.useXhr(requestOptions)
    // }
  }
  useCustomRequest(requestOptions: RequestOptions) {
    const { transport } = this.config;
    const { url, data, success, fail } = requestOptions;
    if (transport) {
      try {
        transport.post({
          url: url,
          data: JSON.stringify(data),
          success: (res) => {
            success && success(res, data);
          },
          fail: (err) => {
            fail && fail(err, data);
          },
        });
      } catch (e) {
        this.collector.configManager.setEnableClientTransport(false);
        this.useRequest(requestOptions);
        this.collector.emit('sdk-inner-error', { type: 'sdk_transport_error', eventData: data, errorCode: 1003, msg: e.message || e.stack || 'sdk_transport_error' });
      }
    }
  }
}

export default RequestManager
