// @ies/starling_intl@1.5.12/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var i18next = _interopDefault(require('i18next'));
var ICU = require('@ies/intl-icu-plugin');
var ICU__default = _interopDefault(ICU);
var intlKeyRenderReportPlugin = require('@ies/intl-key-render-report-plugin');
var InContext = _interopDefault(require('@ies/intl-incontext-plugin'));

function _arrayLikeToArray(r, a) {
  (null == a || a > r.length) && (a = r.length);
  for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e];
  return n;
}
function _arrayWithoutHoles(r) {
  if (Array.isArray(r)) return _arrayLikeToArray(r);
}
function _classCallCheck(a, n) {
  if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(e, r) {
  for (var t = 0; t < r.length; t++) {
    var o = r[t];
    o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o);
  }
}
function _createClass(e, r, t) {
  return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", {
    writable: !1
  }), e;
}
function _createForOfIteratorHelper(r, e) {
  var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"];
  if (!t) {
    if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) {
      t && (r = t);
      var n = 0,
        F = function () {};
      return {
        s: F,
        n: function () {
          return n >= r.length ? {
            done: !0
          } : {
            done: !1,
            value: r[n++]
          };
        },
        e: function (r) {
          throw r;
        },
        f: F
      };
    }
    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }
  var o,
    a = !0,
    u = !1;
  return {
    s: function () {
      t = t.call(r);
    },
    n: function () {
      var r = t.next();
      return a = r.done, r;
    },
    e: function (r) {
      u = !0, o = r;
    },
    f: function () {
      try {
        a || null == t.return || t.return();
      } finally {
        if (u) throw o;
      }
    }
  };
}
function _defineProperty(e, r, t) {
  return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
    value: t,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : e[r] = t, e;
}
function _iterableToArray(r) {
  if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r);
}
function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _toConsumableArray(r) {
  return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread();
}
function _toPrimitive(t, r) {
  if ("object" != typeof t || !t) return t;
  var e = t[Symbol.toPrimitive];
  if (void 0 !== e) {
    var i = e.call(t, r || "default");
    if ("object" != typeof i) return i;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return ("string" === r ? String : Number)(t);
}
function _toPropertyKey(t) {
  var i = _toPrimitive(t, "string");
  return "symbol" == typeof i ? i : i + "";
}
function _unsupportedIterableToArray(r, a) {
  if (r) {
    if ("string" == typeof r) return _arrayLikeToArray(r, a);
    var t = {}.toString.call(r).slice(8, -1);
    return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0;
  }
}

/**
 * Currency code list
 * https://www.currency-iso.org/en/home/tables/table-a1.html
 */
var currency = ['AFN', 'EUR', 'ALL', 'DZD', 'USD', 'AOA', 'XCD', 'ARS', 'AMD', 'AWG', 'AUD', 'AZN', 'BSD', 'BHD', 'BDT', 'BBD', 'BYN', 'BZD', 'XOF', 'BMD', 'INR', 'BTN', 'BOB', 'BOV', 'BAM', 'BWP', 'NOK', 'BRL', 'BND', 'BGN', 'BIF', 'CVE', 'KHR', 'XAF', 'CAD', 'KYD', 'CLP', 'CLF', 'CNY', 'COP', 'COU', 'KMF', 'CDF', 'NZD', 'CRC', 'HRK', 'CUP', 'CUC', 'ANG', 'CZK', 'DKK', 'DJF', 'DOP', 'EGP', 'SVC', 'ERN', 'ETB', 'FKP', 'FJD', 'XPF', 'GMD', 'GEL', 'GHS', 'GIP', 'GTQ', 'GBP', 'GNF', 'GYD', 'HTG', 'HNL', 'HKD', 'HUF', 'ISK', 'IDR', 'XDR', 'IRR', 'IQD', 'ILS', 'JMD', 'JPY', 'JOD', 'KZT', 'KES', 'KPW', 'KRW', 'KWD', 'KGS', 'LAK', 'LBP', 'LSL', 'ZAR', 'LRD', 'LYD', 'CHF', 'MOP', 'MKD', 'MGA', 'MWK', 'MYR', 'MVR', 'MRO', 'MUR', 'XUA', 'MXN', 'MXV', 'MDL', 'MNT', 'MAD', 'MZN', 'MMK', 'NAD', 'NPR', 'NIO', 'NGN', 'OMR', 'PKR', 'PAB', 'PGK', 'PYG', 'PEN', 'PHP', 'PLN', 'QAR', 'RON', 'RUB', 'RWF', 'SHP', 'WST', 'STD', 'SAR', 'RSD', 'SCR', 'SLL', 'SGD', 'XSU', 'SBD', 'SOS', 'SSP', 'LKR', 'SDG', 'SRD', 'SZL', 'SEK', 'CHE', 'CHW', 'SYP', 'TWD', 'TJS', 'TZS', 'THB', 'TOP', 'TTD', 'TND', 'TRY', 'TMT', 'UGX', 'UAH', 'AED', 'USN', 'UYU', 'UYI', 'UZS', 'VUV', 'VEF', 'VND', 'YER', 'ZMW', 'ZWL', 'XBA', 'XBB', 'XBC', 'XBD', 'XTS', 'XXX', 'XAU', 'XPD', 'XPT', 'XAG'];
var numberFormat = {};
for (var i = 0; i < currency.length; i++) {
  numberFormat[currency[i]] = {
    style: 'currency',
    currency: currency[i]
  };
}
var defaultFormats = {
  number: numberFormat
};

var types = {
  LOGGER: 'logger',
  LAMGUAGE_DETECTOR: 'languageDetector',
  LAMGUAGE_TRANSFORMER: 'languageTransformer',
  BACKEND: 'backend'
};
function isTypes(type) {
  return function (itm) {
    return itm.type === type;
  };
}

function formatLang(lng, plugins) {
  var formatLang = lng;
  (plugins || []).map(function (plugin) {
    formatLang = plugin.process(lng) || formatLang;
  });
  return formatLang;
}

var defaultFallbackLanguage = 'zh-CN';
var defaultConfig = {
  lng: defaultFallbackLanguage,
  // 如果使用了 Language Detector，i18next 底层 lng 的权重是大于插件的
  fallbackLng: ['en-US'],
  inContext: true
};
// 默认开启ICU插值解析
var defaultPluginMap = {
  icu: ICU__default,
  inContext: InContext,
  renderReport: intlKeyRenderReportPlugin.registerKeyRenderReportPlugin
};
var defaultPlugins = [defaultPluginMap.icu, defaultPluginMap.inContext, defaultPluginMap.renderReport];
/**
 * I18n内核
 * 安全校验
 */
var I18next = /*#__PURE__*/function () {
  function I18next(config, plugins) {
    _classCallCheck(this, I18next);
    _defineProperty(this, "_waitingToAddResourceBundle", []);
    if (config !== null && config !== void 0 && config.copiedI18nextInstance) {
      // just clone instance
      this.instance = config.copiedI18nextInstance;
      return;
    }
    this._handlePlugins(plugins);
    this._handleConfigs(config);
    this.instance = i18next.createInstance();
    this.instance.isInitialized = false;
    this.KRRList = intlKeyRenderReportPlugin.KRRList;
  }
  return _createClass(I18next, [{
    key: "_handlePlugins",
    value: function _handlePlugins(plugins) {
      var _this$config, _this$config$icuConfi;
      var processedPlugins = [].concat(defaultPlugins);
      // 关闭ICU插值解析
      if (((_this$config = this.config) === null || _this$config === void 0 ? void 0 : (_this$config$icuConfi = _this$config.icuConfig) === null || _this$config$icuConfi === void 0 ? void 0 : _this$config$icuConfi.open) === false) {
        processedPlugins = processedPlugins.filter(function (p) {
          return p.name !== ICU.pluginName;
        });
      }
      this.plugins = Array.isArray(plugins) ? plugins.concat(processedPlugins) : processedPlugins;
    }
  }, {
    key: "_handleConfigs",
    value: function _handleConfigs(config) {
      this.userLng = (config === null || config === void 0 ? void 0 : config.lng) || null; // 用户自己设定的 lng
      this.config = Object.assign({}, defaultConfig, config || {});
    }
  }, {
    key: "language",
    get: function get() {
      return (this.instance || {}).language;
    }
  }, {
    key: "createInstance",
    value: function createInstance() {
      var _this = this;
      return new Promise(function (resolve, reject) {
        // fixed: 非浏览器环境不支持incontext
        // fixed: Cocos 有 window 但不使用 in-context
        // feat: 支持手动关闭inContext
        if (typeof window !== 'undefined' && !window.CocosEngine && _this.config.inContext) {
          _this.config.postProcess = [].concat(['inContext', 'keyRender'], _this.config.postProcess || []);
          window['__STARLING_DEVTOOLS_I18N'] = _this.instance;
        }
        _this.plugins.map(function (p) {
          if (p.name === 'IES_INTL_ICU_PLUGIN') {
            p.staticInit(_this.instance);
          }
          if (p === defaultPluginMap.renderReport) {
            _this.instance.use(p(_this.config.keyRenderReportConfig));
            // registerKeyRenderReportPlugin
          } else {
            _this.instance.use(p);
          }
        });
        var config = _this.config;
        _this.config.formats = Object.assign({}, _this.config.formats, defaultFormats);
        var formatLng = formatLang(config.lng, _this.plugins.filter(isTypes(types.LAMGUAGE_TRANSFORMER)));
        _this.instance.init(Object.assign({}, config, {
          lng: formatLng,
          i18nFormat: Object.assign({}, config.i18nFormat || {}, {
            formats: _this.config.formats
          })
        }), function (err, t) {
          // 初始化好了
          try {
            // 把等待添加的东西都加进去
            var _iterator = _createForOfIteratorHelper(_this._waitingToAddResourceBundle),
              _step;
            try {
              for (_iterator.s(); !(_step = _iterator.n()).done;) {
                var _this$instance;
                var item = _step.value;
                (_this$instance = _this.instance).addResourceBundle.apply(_this$instance, _toConsumableArray(item));
              }
            } catch (err) {
              _iterator.e(err);
            } finally {
              _iterator.f();
            }
            _this._waitingToAddResourceBundle = [];
            // eslint-disable-next-line no-empty
          } catch (err) {}
          if (!err) {
            _this._updateLanguages();
            resolve({
              t: t,
              err: err
            });
          }
          _this.init = true;
          reject({
            t: t,
            err: err
          });
        });
      });
    }
  }, {
    key: "_updateLanguages",
    value: function _updateLanguages() {
      this.languages = this.instance ? Array.from(new Set([this.instance.language].concat(_toConsumableArray(this.instance.languages)))) : null;
    }
  }, {
    key: "getLanguages",
    value: function getLanguages() {
      return this.languages;
    }
  }, {
    key: "changeLanguage",
    value: function changeLanguage(lng, callback) {
      var _this2 = this;
      this.config.lng = lng;
      this.instance.changeLanguage(lng, function (err, t) {
        if (!err) {
          _this2._updateLanguages();
        }
        callback && callback(err, t);
      });
    }
  }, {
    key: "changeLanguageWithPromise",
    value: function changeLanguageWithPromise(lng) {
      var _this3 = this;
      return new Promise(function (resolve, reject) {
        _this3.config.lng = lng;
        _this3.instance.changeLanguage(lng, function (err, t) {
          if (err) {
            reject({
              err: err,
              t: t
            });
          }
          _this3._updateLanguages();
          resolve({
            err: err,
            t: t
          });
        });
      });
    }
  }, {
    key: "getDir",
    value: function getDir(lng) {
      return this.instance.dir(lng);
    }
  }, {
    key: "t",
    value: function t(keys, options, fallbackText) {
      var separatorMock = Array.isArray(keys) ? Array.from(keys).map(function () {
        return ' ';
      }).join('') : Array(keys.length).fill(' ');
      // fixed: 去除默认lngs，有lngs i18next就会忽略lng
      var opt = Object.assign({
        keySeparator: separatorMock,
        nsSeparator: separatorMock
      }, options);
      if ('thirdParamFallback' in this.config && this.config.thirdParamFallback) {
        opt.thirdParamFallback = true;
      }
      var result = this.instance.t(keys, fallbackText, opt);
      if (Array.isArray(result) && this.config.reactKeyAdder) {
        return this.config.reactKeyAdder(result);
      }
      return result;
    }
  }, {
    key: "addResourceBundle",
    value: function addResourceBundle(lng, ns, resources, deep, overwrite) {
      if (this.instance.isInitialized) {
        return this.instance.addResourceBundle(lng, ns, resources, deep, overwrite);
      }
      // 还没初始化好
      this._waitingToAddResourceBundle.push([lng, ns, resources, !!deep, !!overwrite]);
      return this.instance;
    }
  }, {
    key: "cloneInstance",
    value: function cloneInstance(options, callback) {
      var nextI18nextInstance = this.instance.cloneInstance(options, callback);
      var nextCore = new I18next({
        copiedI18nextInstance: nextI18nextInstance
      });
      nextCore.t = this.t.bind(nextCore);
      nextCore.config = this.config;
      nextCore.plugins = this.plugins;
      nextCore.languages = this.languages;
      nextCore.init = nextI18nextInstance.isInitialized || Boolean(this.init);
      nextCore.userLng = this.userLng;
      nextI18nextInstance.on('initialized', function () {
        nextCore.init = true;
      });
      return nextCore;
    }
  }]);
}();

var PREFIX = 'Starling_intl:';
var logger = {
  warn: function warn() {
    var _console;
    var prefix = "".concat(PREFIX, " [WARNING]");
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    (_console = console).warn.apply(_console, [prefix].concat(args));
  },
  error: function error() {
    var _console2;
    var prefix = "".concat(PREFIX, " [ERROR]");
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }
    (_console2 = console).error.apply(_console2, [prefix].concat(args));
  }
};

function promiseQueue(executors) {
  return new Promise(function (resolve, reject) {
    if (!Array.isArray(executors)) {
      executors = Array.from(executors);
    }
    if (executors.length <= 0) {
      return resolve([]);
    }
    var res = [];
    executors = executors.map(function (x, i) {
      return function () {
        var p = typeof x === 'function' ? new Promise(x) : Promise.resolve(x);
        p.then(function (response) {
          res[i] = response;
          if (i === executors.length - 1) {
            resolve(res);
          } else {
            executors[i + 1]();
          }
        }, reject);
      };
    });
    executors[0]();
  });
}

function _await(value, then, direct) {
  if (direct) {
    return then ? then(value) : value;
  }
  if (!value || !value.then) {
    value = Promise.resolve(value);
  }
  return then ? value.then(then) : value;
}
function _catch(body, recover) {
  try {
    var result = body();
  } catch (e) {
    return recover(e);
  }
  if (result && result.then) {
    return result.then(void 0, recover);
  }
  return result;
}
function _empty() {}
function _awaitIgnored(value, direct) {
  if (!direct) {
    return value && value.then ? value.then(_empty) : Promise.resolve();
  }
}
function _async(f) {
  return function () {
    for (var args = [], i = 0; i < arguments.length; i++) {
      args[i] = arguments[i];
    }
    try {
      return Promise.resolve(f.apply(this, args));
    } catch (e) {
      return Promise.reject(e);
    }
  };
}
var IntlModuleType;
(function (IntlModuleType) {
  IntlModuleType["intl3rdParty"] = "intl3rdParty";
  IntlModuleType["backend"] = "backend";
  IntlModuleType["logger"] = "logger";
  IntlModuleType["languageDetector"] = "languageDetector";
  IntlModuleType["postProcessor"] = "postProcessor";
  IntlModuleType["i18nFormat"] = "i18nFormat";
  IntlModuleType["3rdParty"] = "3rdParty";
})(IntlModuleType || (IntlModuleType = {}));
var IntlInstance = null;
/**
 * I18n实例
 * 自定义配置
 */
var Intl = /*#__PURE__*/function () {
  function Intl(opts) {
    var _opts$i18nInstance;
    _classCallCheck(this, Intl);
    this.plugins = [];
    this.i18nInstance = (_opts$i18nInstance = opts === null || opts === void 0 ? void 0 : opts.i18nInstance) !== null && _opts$i18nInstance !== void 0 ? _opts$i18nInstance : new I18next();
  }
  return _createClass(Intl, [{
    key: "use",
    value: function use(plugin) {
      if (!this.plugins.includes(plugin)) {
        this.plugins.push(plugin);
        return this;
      }
      logger.warn("Plugin: ".concat(plugin.name, " has beeninit installed."));
      return this;
    }
  }, {
    key: "init",
    value: function init(config, initCallback) {
      try {
        var _this = this;
        // ignore warning
        _this.ignoreWarning = config.ignoreWarning;
        _this.i18nInstance._handleConfigs(config);
        _this.i18nInstance._handlePlugins(_this.plugins);
        return _await(_catch(function () {
          return _await(_this.i18nInstance.createInstance(), function (_ref) {
            var err = _ref.err,
              t = _ref.t;
            // 接入 Intl thirdParty 类型 Plugin
            var intlThirdPartyArray = _this.plugins.filter(function (p) {
              return p.type === IntlModuleType.intl3rdParty;
            });
            return _await(promiseQueue(intlThirdPartyArray.map(_async(function (intlThirdParty) {
              if (!intlThirdParty || typeof intlThirdParty.init !== 'function') {
                return;
              }
              var initFunction = intlThirdParty.init instanceof Promise ? intlThirdParty.init : _async(function (context) {
                return Promise.resolve(intlThirdParty.init(context));
              });
              return _awaitIgnored(initFunction(_this));
            }))), function () {
              typeof initCallback === 'function' && initCallback(err, t);
              return {
                err: err,
                t: t
              };
            });
          });
        }, function (err) {
          logger.error("The instance init error: ".concat(err));
          return {
            err: err,
            t: function t(key) {
              return key;
            }
          };
        }));
      } catch (e) {
        return Promise.reject(e);
      }
    }
  }, {
    key: "language",
    get: function get() {
      return (this.i18nInstance || {}).language;
    }
  }, {
    key: "setLang",
    value: function setLang(lng, callback) {
      var formatLng = formatLang(lng, this.plugins.filter(isTypes(types.LAMGUAGE_TRANSFORMER)));
      this.i18nInstance.changeLanguage(formatLng, callback);
    }
  }, {
    key: "getLanguages",
    value: function getLanguages() {
      return this.i18nInstance.getLanguages();
    }
  }, {
    key: "setLangWithPromise",
    value: function setLangWithPromise(lng) {
      var formatLng = formatLang(lng, this.plugins.filter(isTypes(types.LAMGUAGE_TRANSFORMER)));
      return this.i18nInstance.changeLanguageWithPromise(formatLng);
    }
  }, {
    key: "dir",
    value: function dir(lng) {
      return this.i18nInstance.getDir(lng);
    }
  }, {
    key: "t",
    value: function t(keys, options, fallbackText) {
      var that = null;
      if (typeof this === 'undefined') {
        that = IntlInstance;
      } else {
        that = this;
      }
      if (!that.ignoreWarning && (!that.i18nInstance || !that.i18nInstance.init)) {
        try {
          var msg = "The instance has not been initialized: i18n.t('".concat(keys, "'").concat(options ? ',' + JSON.stringify(options) : '', ");");
          logger.error(msg);
        } catch (error) {}
        return fallbackText !== null && fallbackText !== void 0 ? fallbackText : Array.isArray(keys) ? keys[0] : keys;
      }
      // 有人给 key 传空字符串?
      if (!keys || typeof keys === 'string' && !keys.trim()) {
        return '';
      }
      return that.i18nInstance.t(keys, options, fallbackText);
    }
  }, {
    key: "addResourceBundle",
    value: function addResourceBundle(lng, ns, resources, deep, overwrite) {
      // to to something validate
      return this.i18nInstance.addResourceBundle(lng, ns, resources, deep, overwrite);
    }
  }, {
    key: "cloneInstance",
    value: function cloneInstance(options) {
      var cloneInstance = this.i18nInstance.cloneInstance(options);
      var cloneIntl = new Intl({
        i18nInstance: cloneInstance
      });
      // todo:: 先考虑 clone t 一个函数，因为 intl-react-plugin 会覆盖
      // todo:: 是否需要全量 copy？包括所有的函数
      cloneIntl.t = this.t.bind(cloneIntl);
      cloneIntl.plugins = this.plugins;
      cloneIntl.ignoreWarning = this.ignoreWarning;
      return cloneIntl;
    }
  }]);
}();
IntlInstance = new Intl();

var I18n = IntlInstance;
I18n.t = I18n.t.bind(I18n);
var I18nConstructor = Intl;

Object.defineProperty(exports, 'IntlMessageFormat', {
  enumerable: true,
  get: function () {
    return ICU.IntlMessageFormat;
  }
});
exports.I18n = I18n;
exports.I18nConstructor = I18nConstructor;
exports.I18nCore = I18next;
exports.Intl = Intl;
exports.default = I18n;
