// @ies/starling_client_core@1.0.5/__tests__/request.ts
class StarlingWebError extends Error {
  constructor(
    message: string,
    public url?: string,
    public code?: string,
    public request?: XMLHttpRequest,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public response?: { data: { [keyText: string]: string } },
    public useWebCache?: boolean
  ) {
    super(message)
    this.url = url
    this.code = code
    this.request = request
    this.response = response
    this.useWebCache = !!useWebCache
  }
}

/**
 * 安全的 JSON.parse，如果 parse 失败，则会返回空 Object
 * @param data 将要被 parse 的 string 数据
 */
function _safeParse(data: string) {
  try {
    return JSON.parse(data)
  } catch (error) {
    return {}
  }
}

type GetTextResponse = {
  message: { [k: string]: string }
  status: number
}

/**
 * 检验拉取文案数据，应当为 "{"data": {"message": {"keyText": "TargetText"}}}" 格式
 * 即 typeof _.data.message === 'object' 且不为 undefined
 */
function _checkTextData(_: GetTextResponse): GetTextResponse | null {
  if (!_.message || typeof _.message !== 'object') {
    return null
  }
  return _
}

/**
 * 响应错误处理
 */
function _responseErrHandler(
  request: XMLHttpRequest,
  reject: (_?: any) => void,
  url: string,
  uncheckedResponseData: any
) {
  // 如果 uncheckedResponseData 里面存在失败原因（message），则也需要吐回去
  let errorMessage: string = (uncheckedResponseData || {}).message || ''
  errorMessage = errorMessage ? ', reason is: '.concat(errorMessage) : ''

  reject(
    new StarlingWebError(
      'Request failed with status code ' + request.status + errorMessage,
      url,
      null,
      request,
      { data: {} }, // todo:: 失败了需要返回 localStorage 缓存？
      true
    )
  )
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export default function request(url: string, options: { timeout: number }): Promise<{ data: any }> {
  const { timeout } = options
  return new Promise(function (resolve, reject) {
    let request = new XMLHttpRequest()

    request.open('GET', url)
    request.overrideMimeType('application/json; charset=utf-8')

    request.withCredentials = false

    request.onreadystatechange = function handleLoad() {
      if (!request || request.readyState !== 4) {
        return
      }

      // The request errored out and we didn't get a response, this will be
      // handled by onerror instead
      // With one exception: request that using file: protocol, most browsers
      // will return status as 0 even though it's a successful request
      if (
        request.status === 0 &&
        !(request.responseURL && request.responseURL.indexOf('file:') === 0)
      ) {
        return
      }

      // 没经过规范校验的 response，可能是 {"message":"none project match this app_key", "status":1000} 格式
      const uncheckedResponseData = _safeParse(request.responseText)
      // 经过规范校验的 response，可能是 null 或者 message 为文案 Object
      const checkedResponseData = _checkTextData(uncheckedResponseData)

      if (request.status < 200 || request.status >= 300 || !checkedResponseData) {
        _responseErrHandler(request, reject, url, uncheckedResponseData)
        request = null
        return
      }

      const response = {
        data: checkedResponseData,
        status: request.status,
        statusText: request.statusText,
        request
      }
      resolve(response)

      // Clean up request
      request = null
    }

    // Handle low level network errors
    request.onerror = function handleError() {
      // Real errors are hidden from us by the browser
      // onerror should only fire if it's a network error
      // 就算是断网了，也得去看看 cache 里面有没有
      reject(new StarlingWebError('Network Error', url, null, request, { data: {} }, true))

      // Clean up request
      request = null
    }

    // Handle timeout
    request.timeout = timeout
    request.ontimeout = function handleTimeout() {
      reject(
        new StarlingWebError(
          'timeout of ' + timeout + 'ms exceeded',
          url,
          'ECONNABORTED',
          request,
          { data: {} },
          true
        )
      )

      // Clean up request
      request = null
    }
    request.send()
  })
}
