// byted-tea-sdk@5.3.10/src/entry/web-process.ts
import { isArray } from '../util/tool'
import { getNameSpace } from '../collect/namespace'

const asyncApi = ['getConfig', 'getAbSdkVersion']
const beforeApi = ['init', 'usePlugin', 'autoInitializationRangers', 'on']
const beforeApiSet = new Set(['init', 'usePlugin', 'autoInitializationRangers', 'on'])
interface Task {
  scope: string;
  cmd: string;
  params: any[];
}
export default class Process {
  task: any[] = []
  newProcessTask: any = []
  collect: any
  Collector: any
  collectMap: any
  processMap: any = false
  taskMap: any
  constructor(Collector: any) {
    this.Collector = Collector
    this.collectMap = new Map()
    this.processMap = new Map()
    this.taskMap = new Map()
  }
  processTask() {
    if (!Array.isArray(this.task)) {
      console.warn('Expected an array for this.task, but got:', this.task);
      return;
    }
    try {
      if (this.task.length) {
        this.task.forEach(tasks => {
          const { scope, cmd } = this.processScope(tasks[0])
          if (!this.taskMap.get(scope)) {
            this.taskMap.set(scope, {
              taskLisMap: [],
              scope,
              inited: false
            })
          }
          const scopeTask = this.taskMap.get(scope)
          scopeTask.taskLisMap.push({
            cmd, params: tasks
          })
          this.taskMap.set(scope, scopeTask)
        })
        if (this.taskMap.size) {
          this.taskMap.forEach((cmdMapValue, cmdMapKeys) => {
            const executeQueue = (queue: Task[]) => {
              for (let i = queue.length - 1; i >= 0; i--) {
                // 不是可以提前执行的api，且当前scope实例还没有初始化
                if (!beforeApiSet.has(queue[i].cmd) && !this.processMap.get(cmdMapKeys)) {
                  continue;
                }
                if (beforeApiSet.has(queue[i].cmd)) {
                  this.processMap.set(cmdMapKeys, true)
                  if (queue[i].cmd === 'init') {
                    cmdMapValue.inited = true
                  }
                }
                this.processCmd(...queue[i].params)
                const sliceQueue = this.taskMap.get(cmdMapKeys)
                sliceQueue.taskLisMap.splice(i, 1)
                this.taskMap.set(cmdMapKeys, sliceQueue)
              }
              const leftQueue = this.taskMap.get(cmdMapKeys).taskLisMap
              if (leftQueue.length && cmdMapValue.inited) {
                executeQueue(leftQueue.reverse());
              }
            }
            const _queue = this.taskMap.get(cmdMapKeys).taskLisMap
            if (_queue.length) {
              executeQueue(_queue.reverse() as Task[]);
            }
          })
        }
      }
      this.task = [];
    } catch (e) {
      console.error('An error occurred while processing tasks:', e);
    }
    // if (this.task.length) {
    //   let initIndex = this.findIndex(this.task, beforeApi, '0')
    //   if (initIndex !== -1) {
    //     // 第一个执行的命令必须是beforeApi
    //     let { scope } = this.processScope(this.task[initIndex][0])
    //     this.processMap.set(scope, true)
    //     this.processCmd(...this.task[initIndex])
    //     this.newProcessTask = this.task.filter(item => {
    //       if (scope === 'default') {
    //         if (item[0].indexOf(`.`) === -1) {
    //           return item
    //         }
    //       } else {
    //         if (item[0].indexOf(`${scope}.`) !== -1) {
    //           return item
    //         }
    //       }
    //     })
    //     const newProcessIndex = this.findIndex(this.newProcessTask, beforeApi, '0')
    //     this.newProcessTask.forEach((argCmd: any, index: number) => {
    //       if (index !== newProcessIndex) {
    //         this.processCmd(...argCmd)
    //       }
    //     })
    //     // this.task = [];
    //     this.newProcessTask = []
    //   }
    // }
  }
  processCmd(...args) {
    if (isArray(args)) {
      let { scope, cmd } = this.processScope(args[0])
      if (!this.processMap.get(scope)) {
        // 没有自己的scope，说明自己的实例还没有初始化
        this.task.push(args);
        this.processTask();
        return;
      }
      args.splice(0, 1)
      if (asyncApi.indexOf(cmd) !== -1) {
        return this.callCmd(scope, cmd, args)
      } else {
        this.callCmd(scope, cmd, args)
      }
    }
  }
  callCmd(scope: string, cmd: string, params?: any) {
    try {
      if (beforeApi.indexOf(cmd) !== -1) {
        if (!this.collectMap.get(scope)) {
          this.collect = new this.Collector(scope)
          this.collectMap.set(scope, this.collect)
        }
      }
      if (!this.collectMap.get(scope)) {
        console.log(`init or use must be first call, cmd:${cmd}`)
        return
      }
      const collectInstance = this.collectMap.get(scope)
      if (collectInstance.__proto__.hasOwnProperty(cmd)) {
        if (cmd === 'event') {
          collectInstance.event(cmd, ...params)
        } else {
          if (asyncApi.indexOf(cmd) !== -1) {
            return collectInstance[cmd](...params)
          } else {
            collectInstance[cmd](...params)
          }
        }
      } else {
        collectInstance.event(cmd, ...params)
      }
    } catch (e) { }
  }
  findIndex(array: any, val: any, key: string) {
    let itemIndex = -1
    try {
      array.forEach((i, index) => {
        const value = typeof key !== 'undefined' ? i[key] : i;
        const { cmd } = this.processScope(value)
        if (val.indexOf(cmd) !== -1) {
          itemIndex = index
        }
      })
    } catch (e) {
      return itemIndex
    }
    return itemIndex
  }
  processScope(cmd: string) {
    let scope = 'default'
    try {
      if (typeof cmd === 'string' && cmd.indexOf('.') > -1) {
        var tmp = cmd.split('.')
        scope = tmp[0]
        cmd = tmp[1]
      }
    } catch (e) {
      return { scope, cmd }
    }
    return { scope, cmd }
  }
}

export const transferAsync = (newProcess: any) => {
  if (typeof window === 'undefined') return
  const nameSpace = getNameSpace()
  if (!nameSpace || !window[nameSpace]) {
    return
  }
  const client = window[nameSpace]
  if (client['sdkLoad']) {
    console.log('you page has load sdk file already, do not load another sdk file')
    return
  }
  newProcess.task = client['q'] || []
  window[nameSpace] = newProcess.processCmd.bind(newProcess)
  window[nameSpace]['sdkLoad'] = true
  newProcess.processTask()
}
