// byted-tea-sdk@5.3.10/src/plugin/logsetting/logsetting.ts
import {
  getLogPluginSpace,
} from '../../collect/namespace';
import Collector from "../../collect/collect";
import { IInitParam } from '../../../types/types';
import Types from "../../collect/hooktype";
import { DebuggerMesssge } from "../../collect/hooktype";
import { LocalForage } from '../../plugin/storage/storage'



type SettingConfigT = {
  params_block: Record<string, any>
  header_custom_allow: [string]
}
type SettingT = {
  config: SettingConfigT
  magic_tag: Record<string, any>
  server_time: number
  block_event?: []
}
type SettingCacheT = {
  timestamp: number
  settingConfig: SettingT
}
export default class LogSetting {
  private setUrl: string = '/service/2/log_settings'
  public collect: Collector
  public config: IInitParam
  private setKey: string
  private cacheStorgae: any
  private expireTime: number = 6 * 3600 * 1000
  private logSettingConfig: SettingT
  private DBStorage: any
  private uniqueDataKey: string
  public apply(collect: Collector, config: IInitParam) {
    if (!config.enable_logsetting) {
      collect.emit(Types.LogSettingReady);
      return;
    };
    this.collect = collect;
    this.config = config;
    const { storage } = collect.adapters;
    this.setKey = `applog_web_logsetting_${config.app_id}`;
    this.uniqueDataKey = 'web_logsetting_data';
    this.DBStorage = new LocalForage(this.setKey, 'LOGSETTING_DATA_STORE')
    this.cacheStorgae = new storage(false);
    this.check();
  }
  private async check() {
    try {
      const cacheFromStorage = await this.getLocalData();
      if (cacheFromStorage && Object.keys(cacheFromStorage).length) {
        if (cacheFromStorage?.timestamp < (Date.now() - this.expireTime)) {
          // 缓存过期
          this.fetchSetting();
        } else {
          this.setSettingData(cacheFromStorage.settingConfig)
        }
      } else {
        this.fetchSetting();
      }
    } catch (e) {
      this.collect.emit(Types.LogSettingReady);
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 检测logSetting缓存失败', logType: 'sdk', level: 'error', time: Date.now(), infoType: 'sdk' })
    }
  }
  private fetchSetting(ready?: boolean) {
    if (ready) {
      this.collect.emit(Types.LogSettingReady);
      return;
    }
    const header = this.collect.configManager.get('header')
    header.custom = JSON.stringify(header.custom)
    const user = this.collect.configManager.get('user')
    const body = {
      event_filter: 0,
      header: {
        aid: this.config.app_id,
        ...header,
        ...user
      },
      options: {
        params_block_enable: this.config.enable_logsetting_params ? 1 : 0,
        header_custom_allow_enable: this.config.enable_logsetting_header_custom ? 1 : 0
      }
    }
    this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 发起LogSetting请求', logType: 'fetch', level: 'info', time: Date.now(), data: body, infoType: 'sdk' })
    this.collect.requestManager.useRequest({
      url: `${this.collect.configManager.getDomain()}${this.setUrl}`,
      data: body,
      timeout: 5000,
      success: (response) => {
        if (response?.config) {
          this.setLocalData(response.config);
          this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'LogSetting请求成功', logType: 'fetch', level: 'info', time: Date.now(), data: response, infoType: 'sdk' })
        } else {
          this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'LogSetting请求返回值异常', logType: 'fetch', level: 'error', time: Date.now(), infoType: 'sdk' })
        }
        this.collect.emit(Types.LogSettingReady);
      },
      fail: () => {
        this.collect.emit(Types.LogSettingReady);
        this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'LogSetting请求网络异常', logType: 'fetch', level: 'error', time: Date.now(), infoType: 'sdk' })
      }
    })
  }
  private async getLocalData(): Promise<SettingCacheT> {
    // return this.cacheStorgae.getItem(this.setKey);
    try {
      const cacheFromDB = await this.DBStorage.getItem(this.uniqueDataKey)
      return cacheFromDB
    } catch (e) {
      return null
    }
  }
  private setLocalData(settingData: SettingT) {
    try {
      const settingCache = {
        timestamp: Date.now(),
        settingConfig: settingData
      }
      this.setSettingData(settingData);
      this.cacheStorgae.setItem(this.setKey, settingCache);
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 更新LogSetting缓存', logType: 'info', level: 'info', time: Date.now(), infoType: 'sdk', data: settingData })
    } catch (e) {
      this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: 'SDK 更新LogSetting缓存出错', logType: 'info', level: 'error', time: Date.now(), infoType: 'sdk' })
    }
  }
  public getSettingData(key?: string): any {
    if (key) {
      return this.logSettingConfig[key]
    }
    return this.logSettingConfig;
  }
  public setSettingData(data: SettingT) {
    this.logSettingConfig = data;
  }
  public getPredefineEventList() {
    return new Set(['predefine_pageview', 'onload', 'bav2b_click', 'bav2b_page', 'bav2b_beat', 'bav2b_page_statistics', 'bav2b_exposure', 'predefine_page_alive', 'predefine_page_close', '_be_active', '__bav_click', '__bav_page', '__bav_beat', '__bav_page_statistics', '__bav_page_exposure'])
  }
  public getPredefineParamsList() {
    return new Set(['event_index'])
  }
}

/**@@SCRIPT
try {
  const exportLogSetting = (collect: any, config: any) => {
    const logSetting = new LogSetting()
    logSetting.apply(collect, config)
  }
  const pluginObject = getLogPluginSpace()
  if (pluginObject) {
    pluginObject.LogSetting = exportLogSetting
  }
} catch (e) {
  console.log(e)
}
@@SCRIPT*/
