// byted-tea-sdk@5.3.10/src/plugin/monitor/monitor.ts
import { SDK_TYPE } from '../../collect/constant';
import Types from '../../collect/hooktype';
import { IInitParam } from '../../../types/types';
import localId from '../../util/local';
import { decodeUrl } from '../../util/tool';
import { LOG_URL } from '../../collect/constant';
import Storage from '../storage/storage';

interface InitT {
  time?: number,
  status?: string,
  type?: string,
  message?: string
}

interface FailT {
  message?: string
}

interface RequestT {
  type?: string,
  eventData?: any,
  errorCode?: number,
  errorMessage?: string,
  response?: any,
  startTime?: number,
  endTime?: number
  status?: number
}

interface ErrorT {
  time?: number,
  status?: string,
  type?: string,
  error?: string
}

interface ReportDataT {
  type: string,
  session_id?: string,
  start_time?: number,
  end_time?: number,
  [key: string]: any;
}

export default class Monitor {
  private config: IInitParam
  private collect: any
  private reportUrl: string
  private reportDomain: string
  private initData: Object = {}
  private requestData: Object = {}
  private errorData: Object = {}
  private timeout: number = 30000
  private intervalTime: number = 1 * 60 * 1000
  private timer: any
  private startTime: number
  private requestIndex: number // 已经统计的区间次数
  private preEvent: number // 已经上报的事件数量
  private requestRangerStartTime: number
  private errorRangerStartTime: number
  private errorLimitCount: number = 3 // error数据触发上报的最大条数
  private errorSampling: number = 0.9 // error数据触发上报的最大条数
  private randomUser: string[]
  private randomSUser: string[]
  private reportId: number
  private sampling: number = 0.9
  private retryLimitCount: number = 3 // 最大重试次数
  private retryMap: any
  private currentUser: string
  private currentSUser: string
  private idRange: number = 4
  private idPrefix: string = 'sdk_solid_user_'
  private remove: any
  private enableMonitor: boolean
  private cacheStorgae: any
  private monitorAppid: number
  constructor(collect: any, config: IInitParam) {
    this.collect = collect;
    this.config = config;
    this.cacheStorgae = new Storage();
    this.getMonitorEnable();
    this.monitorAppid = Number(config.app_id);
    this.reportUrl = '1fz22z22z1nz21z4mz4bz4bz21z22z18z1ez1gz1lz1ez49z1az1jz1mz23z1bz23z19z18z4az19z27z22z1cz1bz18z1lz1az1cz4az1lz1cz22z4bz1jz1gz21z22';
    this.reportDomain = decodeUrl('1fz22z22z1nz21z4mz4bz4bz21z22z18z1ez1gz1lz1ez49z1az1jz1mz23z1bz23z19z18z4az19z27z22z1cz1bz18z1lz1az1cz4az1lz1cz22');
    this.reportId = 20000050;
    this.startTime = Date.now();
    this.requestRangerStartTime = this.startTime;
    this.errorRangerStartTime = this.startTime;
    this.requestIndex = 1;
    this.preEvent = 0;
    this.retryMap = new Map();
    this.randomUser = ['sdk_solid_user_1', 'sdk_solid_user_2', 'sdk_solid_user_3', 'sdk_solid_user_4'];
    this.randomSUser = ['1738231126490767749', '1738231126490767757', '1738231126490767758', '1738231126490767760'];
    const currentUserIndex = Math.floor(Math.random() * 1000);
    // this.currentUser = this.randomUser[currentUserIndex];
    this.currentSUser = this.randomSUser[currentUserIndex];
    this.currentUser = `${this.idPrefix}${Math.ceil(Math.random() * this.idRange)}`;
    this.collect.on(Types.LogSettingReady, () => {
      const observe_appid = this.collect.configManager.getSettingData('observe_appid');
      const observe_enable = this.collect.configManager.getSettingData('observe_enable');
      if (observe_appid) {
        if (observe_enable === 0) {
          this.clearMonitor();
        } else {
          this.monitorAppid = observe_appid;
          this.enableMonitor = true;
        }
      }
    })
    this.collect.on(Types.Init, () => {
      this.addInitData({
        time: Date.now(),
        status: 'run_start',
        type: 'init'
      })
    });
    this.collect.on(Types.MonitorConfig, ({ initConfig }) => {
      this.addConfig(initConfig);
    })
    this.collect.on(Types.Ready, () => {
      this.addInitData({
        time: Date.now(),
        status: 'run_end',
        type: 'init'
      })
    });
    this.collect.on(Types.SDKFail, ({ message }: FailT) => {
      this.addInitData({
        time: Date.now(),
        status: 'run_fail',
        type: 'init',
        message
      })
    });
    this.collect.on(Types.XhrBeforeSend, (requestData: RequestT) => {
      this.addRequestData(requestData);
    });
    this.collect.on(Types.SubmitSuccess, (requestData: RequestT) => {
      this.addRequestData(requestData);
    });
    this.collect.on(Types.SubmitError, (requestData: RequestT) => {
      this.addRequestData(requestData);
    });
    this.collect.on(Types.XhrCurrendSend, (requestData: RequestT) => {
      this.addRequestData(requestData);
    })
    this.collect.on(Types.SDKException, (exception: ErrorT) => {
      this.addErrorData({
        time: Date.now(),
        status: 'run_fail',
        type: 'init',
        error: exception.error
      })
    });
    this.collect.on(Types.SessionResetInner, () => {
      this.reportImmediately()
    });
    this.collect.on('send-sdk-monitor-request', () => {
      this.reportRequestData();
    })
    this.collect.on('send-sdk-monitor-error', () => {
      this.reportErrorData();
    })
    this.remove = this.addListener();
    this.ready('monitor');
  }

  private ready(name: string) {
    this.collect.set(name);
    if (this.collect.hook._hooksCache.hasOwnProperty(name)) {
      const emits = this.collect.hook._hooksCache[name];
      if (!Object.keys(emits).length) return;
      for (let key in emits) {
        if (emits[key].length) {
          emits[key].forEach(item => {
            this.collect.hook.emit(key, item);
          })
        }
      }
    }
  }

  private getMonitorEnable() {
    if (this.config.enable_logsetting) {
      // 开启云控，则以云控为准
      const key = `__applog_web_logsetting_${this.config.app_id}`;
      const settings = this.cacheStorgae.getLocalItem(key);
      if (settings?.observe_appid) {
        if (settings?.observe_enable == 1) {
          this.enableMonitor = true;
          this.monitorAppid = settings.observe_appid;
        } else {
          this.enableMonitor = false;
        }
      } else {
        // 没读到，以本地开关为主
        this.enableMonitor = this.config.enable_observe || false;
      }
    } else {
      this.enableMonitor = this.config.enable_observe || false;
    }

  }

  private addConfig(initConfig: IInitParam) {
    if (!this.enableMonitor) return;
    const deviceConfig = {};
    for (let key in initConfig) {
      const deviceKey = `config_${key}`;
      deviceConfig[deviceKey] = initConfig[key];
    }
    if (initConfig.channel_domain) {
      if (initConfig.channel_domain === 'https://gator.volces.com') {
        this.reportId = 20000050;
      } else {
        this.reportId = 666666;
      }
      this.reportDomain = initConfig.channel_domain;
    } else {
      if (initConfig.channel === 'cn') {
        this.reportId = 2174;
      } else {
        this.reportId = 183136;
      }
      this.reportDomain = decodeUrl(LOG_URL[initConfig.channel]);
    }
    this.initData = {
      ...this.initData,
      ...deviceConfig
    };
  }

  /**
   * SDK初始化状态
   */
  private addInitData(initData: InitT) {
    if (!this.enableMonitor) return;
    const { status } = initData;
    if (status === 'run_start') {
      // const deviceConfig = {};
      // for (let key in this.config) {
      //   const deviceKey = `config_${key}`;
      //   deviceConfig[deviceKey] = this.config[key];
      // }
      this.initData = {
        ...this.initData,
        ...{
          run_start: initData.time,
          type: 'init',
          status: 0,
          sdk_channel: SDK_TYPE
        },
        // ...deviceConfig
      };
    } else if (status === 'run_end') {
      const runStartTime = this.getData('init', 'run_start') || this.startTime;
      const initSpendTime = runStartTime ? initData.time - runStartTime : 0;
      const envConfig = this.collect.configManager.get('header');
      const sessionId = this.collect.sessionManager.getSessionId();
      const deviceConfig = {};
      deviceConfig['device_platform'] = envConfig['device_platform'];
      deviceConfig['device_language'] = envConfig['platform'];
      deviceConfig['device_network'] = envConfig['network'];
      deviceConfig['sdk_version'] = envConfig['sdk_version'];
      deviceConfig['sdk_lib'] = envConfig['sdk_lib'];
      deviceConfig['device_browser'] = envConfig['browser'];
      deviceConfig['device_os'] = envConfig['os_name'];
      this.initData = {
        ...this.initData,
        ...{
          run_duration: initSpendTime,
          run_end: initData.time,
          status: 1,
          session_id: sessionId,
          stage: "init_post",
        },
        ...deviceConfig
      };
      this.reportInitData();
    } else if (status === 'run_fail') {
      this.initData = {
        ...this.initData,
        ...{
          run_fail: initData.message,
          status: 2
        }
      };
      this.reportInitData();
    }
  }
  /**
   * SDK请求数据
   */
  private addRequestData(requestData: RequestT) {
    if (!this.enableMonitor) return;
    const { type, eventData, response } = requestData;
    let previousNetCount = this.getData('report', 'net') || 0;
    let previousFEventCount = this.getData('report', 'f_event') || 0;
    const sessionId = this.collect.sessionManager.getSessionId();
    this.setRequestData('session_id', sessionId);
    let currentRequestIncludeEventCount = 0;
    if (eventData && eventData.length > 0) {
      const events = eventData[0].events;
      currentRequestIncludeEventCount = events && events.length || 0;
      // this.preEvent = this.preEvent + currentRequestIncludeEventCount;
      // this.setRequestData('pre_event', this.preEvent);
    }
    if (type === 'f_net') {
      // 网络失败
      const previousFNetCount = this.getData('report', 'f_net') || 0;
      const currentFNetCount = previousFNetCount + 1;
      const currentFEventCount = previousFEventCount + currentRequestIncludeEventCount;
      this.setRequestData('f_net', currentFNetCount);
      this.setRequestData('f_net_event', currentFEventCount);
    } else if (type === 'f_data') {
      // 数据异常失败
      const previousFDataCount = this.getData('report', 'f_data') || 0;
      const previousFEventDataCount = this.getData('report', 'f_data_event') || 0;
      const currentFDataCount = previousFDataCount + 1;
      this.setRequestData('f_data', currentFDataCount);
      if (response?.tc) {
        const sc = response.sc || 0;
        const currentFEventDataCount = previousFEventDataCount + (response.tc - sc);
        this.setRequestData('f_data_event', currentFEventDataCount);
      }
    } else if (type === 'make_event') {
      // 触发的事件量
      const previousMakeEventCount = this.getData('report', 'make_event') || 0;
      const currentMakeEventCount = previousMakeEventCount + currentRequestIncludeEventCount;
      this.setRequestData('make_event', currentMakeEventCount);
    } else if (type === 'net_event') {
      // 请求携带的事件量
      const previousNetEventCount = this.getData('report', 'net_event') || 0;
      const currentMakeEventCount = previousNetEventCount + currentRequestIncludeEventCount;
      this.setRequestData('net_event', currentMakeEventCount);
    } else {
      // 请求成功，只看200的情况
      const previousEventCount = this.getData('report', 'event') || 0;
      const currentEventCount = previousEventCount + currentRequestIncludeEventCount;
      this.setRequestData('event', currentEventCount);
    }
    const currentNetCount = previousNetCount + 1;
    this.setRequestData('net', currentNetCount);
    // 采样失败请求
    if (type === 'f_net' || type === 'f_data') {
      const { startTime, endTime, status, errorCode, errorMessage } = requestData;
      const currentStatus = status || errorCode;
      if (currentStatus && currentStatus !== 200 && currentStatus.toString()) {
        let failStatus = '';
        if (currentStatus.toString().startsWith('4')) {
          failStatus = 'f_4xx';
        } else if (currentStatus.toString().startsWith('5')) {
          failStatus = 'f_5xx';
        }
        const previousFailStatus = this.getData('report', failStatus) || 0;
        const currentFailStatus = previousFailStatus + 1;
        this.setRequestData(failStatus, currentFailStatus);
      }
      const previousErrorSampling = this.getData('report', 'sampling') || [];
      if (Math.random() < this.sampling) {
        const httpStatus = status || errorCode;
        const duration = endTime - startTime || 0;
        const size = JSON.stringify(eventData);
        previousErrorSampling.push({
          request_start: startTime,
          request_end: endTime,
          size: size.length,
          duration,
          http_status: httpStatus,
          error: response?.m || response?.message || errorMessage || 'request default error'
        })
        this.setRequestData('sampling', previousErrorSampling);
      }
    }
  }

  /**
   * SDK异常监控
   */
  private addErrorData(errorData: ErrorT) {
    if (!this.enableMonitor) return;
    const previousErrorCount = this.getData('exception', 'count') || 0;
    const currentErrorCount = previousErrorCount + 1;
    const previousErrorSampling = this.getData('exception', 'sampling') || [];
    previousErrorSampling.push({
      timestamp: errorData.time,
      error: errorData.error
    })
    this.setErrorData('sampling', previousErrorSampling);
    this.setErrorData('count', currentErrorCount);
    if (currentErrorCount >= this.errorLimitCount) {
      this.reportErrorData();
    }
  }

  /**
   * 获取已经存储的值
   */
  private getData(type: string, key?: string) {
    let currentParent = {};
    if (type === 'init') {
      currentParent = this.initData;
    } else if (type === 'report') {
      currentParent = this.requestData;
    } else {
      currentParent = this.errorData;
    }
    if (!key) {
      return currentParent;
    }
    return currentParent[key] || null;
  }

  private setRequestData(key: string, value: any) {
    if (!key) return;
    this.requestData[key] = value;
  }

  private setErrorData(key: string, value: any) {
    if (!key) return;
    this.errorData[key] = value;
  }

  private addListener() {
    const reportRightNow = () => {
      this.reportImmediately();
    }
    window.addEventListener('pagehide', reportRightNow, false);
    window.addEventListener('beforeunload', reportRightNow, false)
    this.timer = setInterval(() => {
      this.reportRequestData();
    }, this.intervalTime);
    return () => {
      clearInterval(this.timer);
      this.timer = null;
      window.removeEventListener('pagehide', reportRightNow, false);
      window.removeEventListener('beforeunload', reportRightNow, false);
    }
  }

  private reportRequestData() {
    const current = JSON.parse(JSON.stringify(this.requestData));
    // 没有请求的话，就不触发上报
    if (current.net === 0 || current.event === 0) return;
    if (!current.net || !current.event) return;
    const currentData = {
      start_time: this.requestRangerStartTime,
      end_time: Date.now(),
      type: 'report',
      index: this.requestIndex,
      pre_event: this.preEvent,
      ...current
    };
    const sendData = this.composeData(currentData);
    this.report(sendData, this.resetRequestData.bind(this));
    this.preEvent = this.preEvent + current.event;
  }

  private reportInitData() {
    const current = JSON.parse(JSON.stringify(this.initData));
    const sendData = this.composeData(current);
    this.report(sendData, this.resetInitData.bind(this));
  }

  private reportErrorData() {
    const current = JSON.parse(JSON.stringify(this.errorData));
    const currentData = {
      start_time: this.errorRangerStartTime,
      session_id: this.collect.sessionManager.getSessionId(),
      end_time: Date.now(),
      type: 'exception',
      ...current
    };
    const sendData = this.composeData(currentData);
    this.report(sendData, this.resetErrorData.bind(this));
  }

  private composeData(current: ReportDataT) {
    const observe_enable = this.collect.configManager.getSettingData('observe_enable');
    if (observe_enable && observe_enable === 0) {
      this.clearMonitor();
      return null;
    }
    const observe_appid = this.collect.configManager.getSettingData('observe_appid');
    const { header, user } = this.collect.configManager.get();
    const { web_id, user_unique_id } = user;
    const monitorEvent = {
      event: '$finder_sdk_monitor',
      params: JSON.stringify(current),
      local_time_ms: Date.now(),
    };
    header.app_id = this.reportId || observe_appid;
    header['custom']['client_webid'] = web_id || '';
    header['custom']['client_uuid'] = user_unique_id || '';
    header['custom']['client_appid'] = this.collect.configManager.get('app_id');
    header['custom'] = JSON.stringify(header['custom']);
    const monitorData = {
      events: [monitorEvent],
      user: {
        user_unique_id: this.currentUser,
      },
      header,
    };
    return [monitorData];
  }

  private resetRequestData() {
    this.requestRangerStartTime = Date.now();
    this.requestIndex++;
    this.requestData = {};
  }

  private resetInitData() {
    // this.initData = {};
  }

  private resetErrorData() {
    this.errorData = {};
  }

  private clearMonitor() {
    this.enableMonitor = false;
    this.initData = {};
    this.errorData = {};
    this.requestData = {};
    this.requestRangerStartTime = Date.now();
    this.requestIndex = 0;
    this.remove();
  }

  /**
   * 发生sessionid变化，或者离开页面理解上报所有数据
   */
  private reportImmediately() {
    this.reportRequestData();
    this.reportInitData();
    this.reportErrorData();
  }

  private report(data: any, callback: Function, eventId?: string) {
    if (!data) return;
    this.collect.requestManager.useRequest({
      url: `${this.reportDomain}/list`,
      data,
      success: (res, data, statusType, startTime, status) => {
        if (status !== 200 && !eventId) {
          this.retry(data);
        }
        callback && callback(status);
      },
      fail: () => {
        this.retry(data, eventId);
        callback && callback();
      },
      timeout: this.timeout,
      status: 'monitor'
    })
  }

  private retry(data: any, id?: string) {
    let retryData = null;
    let retryId = null;
    if (id) {
      const currentRetry = this.retryMap.get(id);
      if (!currentRetry) return;
      if (currentRetry.count >= this.retryLimitCount) {
        this.retryMap.delete(id);
        return;
      }
      retryId = id;
      retryData = currentRetry.data
      currentRetry.count++;
      this.retryMap.set(id, currentRetry);
    } else {
      retryData = data;
      retryId = localId();
      this.retryMap.set(retryId, {
        data,
        count: 1
      });
    }
    this.report(retryData, (status?: number) => {
      if (status && status === 200) {
        this.retryMap.delete(id);
      } else {
        this.retry(retryData, retryId);
      }
    }, retryId)
  }
}
