// @rdservices/aime-code-inspector-vite@1.0.17/dist/index.mjs
var w = Object.defineProperty, j = Object.defineProperties;
var P = Object.getOwnPropertyDescriptors;
var v = Object.getOwnPropertySymbols;
var W = Object.prototype.hasOwnProperty, q = Object.prototype.propertyIsEnumerable;
var h = (e, n, i) => n in e ? w(e, n, { enumerable: !0, configurable: !0, writable: !0, value: i }) : e[n] = i, u = (e, n) => {
  for (var i in n || (n = {}))
    W.call(n, i) && h(e, i, n[i]);
  if (v)
    for (var i of v(n))
      q.call(n, i) && h(e, i, n[i]);
  return e;
}, m = (e, n) => j(e, P(n));
var f = (e, n, i) => new Promise((t, a) => {
  var c = (l) => {
    try {
      p(i.next(l));
    } catch (r) {
      a(r);
    }
  }, s = (l) => {
    try {
      p(i.throw(l));
    } catch (r) {
      a(r);
    }
  }, p = (l) => l.done ? t(l.value) : Promise.resolve(l.value).then(c, s);
  p((i = i.apply(e, n)).next());
});
import { isExcludedFile as I, getCodeWithWebComponent as k, normalizePath as C, getMappingFilePath as T, isJsTypeFile as b, transformCode as R } from "@rdservices/aime-code-inspector-core";
import d from "chalk";
const x = "@code-inspector/vite", S = [
  {
    name: "vite:react-babel",
    package: "@vitejs/plugin-react"
  },
  {
    name: "vite:react-swc",
    package: "@vitejs/plugin-react-swc"
  },
  {
    name: "vite:react-oxc:config",
    package: "@vitejs/plugin-react-oxc"
  },
  {
    name: "solid",
    package: "vite-plugin-solid"
  },
  {
    name: "vite-plugin-qwik",
    package: "qwikVite"
  },
  {
    name: "vite-plugin-qwik-city",
    package: "qwikCity"
  },
  {
    name: "vite-plugin-qwik-react",
    package: "qwikReact"
  },
  {
    name: "vite:preact-jsx",
    package: "@preact/preset-vite"
  },
  {
    name: "vite-plugin-svelte",
    package: "@sveltejs/vite-plugin-svelte"
  }
], _ = ["isJsx", "isTsx", "lang.jsx", "lang.tsx"];
function D(e = []) {
  const n = e.findIndex((i) => i.name === x);
  S.forEach((i) => {
    const t = e.findIndex((a) => a.name === i.name);
    if (t !== -1 && t < n) {
      const a = [
        d.yellow("[WARNING]"),
        "You need to put",
        d.green("code-inspector-plugin"),
        "before",
        d.green(i.package),
        "in the vite config file."
      ];
      console.log(a.join(" "));
    }
  });
}
function E(e) {
  const n = {
    port: 0,
    entry: "",
    output: e.output,
    envDir: ""
  };
  return m(u({
    name: x
  }, e.enforcePre === !1 ? {} : { enforce: "pre" }), {
    apply(t, { command: a }) {
      return !e.close;
    },
    configResolved(t) {
      n.envDir = t.envDir || t.root, n.root = t.root;
    },
    transform(t, a) {
      return f(this, null, function* () {
        if (I(a, e))
          return t;
        t = yield k({
          options: e,
          file: a,
          code: t,
          record: n
        });
        const { escapeTags: c = [], mappings: s } = e, [p, l] = a.split("?", 2);
        let r = C(p);
        r = T(r, s);
        const o = new URLSearchParams(l);
        if (e.match && !e.match.test(r))
          return t;
        let g = "";
        return b(r) || r.endsWith(".vue") && (_.some((y) => o.get(y) !== null) || o.get("lang") === "tsx" || o.get("lang") === "jsx") ? g = "jsx" : r.endsWith(".html") && o.get("type") === "template" && o.has("vue") || r.endsWith(".vue") && o.get("type") !== "style" && o.get("raw") === null ? g = "vue" : r.endsWith(".svelte") && (g = "svelte"), g ? R({
          content: t,
          filePath: r,
          fileType: g,
          escapeTags: c,
          pathType: e.pathType
        }) : t;
      });
    },
    // 追加到 html 中，适配 MPA 项目
    transformIndexHtml(t) {
      return f(this, null, function* () {
        var c, s;
        if ((s = (c = e.skipSnippets) == null ? void 0 : c.includes) != null && s.call(c, "htmlScript"))
          return t;
        const a = yield k({
          options: m(u({}, e), { importClient: "code" }),
          file: "main.js",
          code: "",
          record: n,
          inject: !0
        });
        return t.replace("<head>", '<head><script type="module">\n'.concat(a, "\n<\/script>"));
      });
    },
    configureServer(t) {
      const a = t.config.logger.info;
      t.config.logger.info = function(c, s) {
        a.call(this, c, s), D(t.config.plugins), t.config.logger.info = a;
      };
    }
  });
}
export {
  E as ViteCodeInspectorPlugin
};
