// byted-tea-sdk@5.3.10/src/plugin/native/index.ts
//@ts-nocheck
import { IInitParam } from '../../../types/types'

class AppBridge {
  config: IInitParam
  configManager: any
  currentUserUniqueId: string
  currentUserUniqueIdType: string
  native: boolean
  os: string
  constructor(config: IInitParam, configManager: any) {
    this.configManager = configManager;
    this.native = config.enable_native || config['evitaN'.split('').reverse().join('')];
    this.os = configManager.get('os_name');
    this.currentUserUniqueId = '';
    this.currentUserUniqueIdType = '';
  }
  error(type: string, message: string) {
    console.error(`${type}, msg: ${message || ''}`)
  }
  bridgeInject() {
    try {
      if (!this.native) return false;
      if (AppLogBridge) {
        console.log(`AppLogBridge is injected`)
        if (AppLogBridge.osName) {
          this.os = AppLogBridge.osName();
        } else {
          this.os = this.os;
        }
        return true;
      } else {
        console.log('AppLogBridge is not inject');
        return false;
      }
    } catch (e) {
      console.log('AppLogBridge is not inject');
      return false;
    }
  }
  bridgeReady() {
    return new Promise((resolve, reject) => {
      try {
        if (this.bridgeInject()) {
          AppLogBridge.hasStarted(start => {
            console.log(`AppLogBridge is started? : ${start}`)
            if (start) {
              resolve(true)
            } else {
              reject(false)
            }
          })
        } else {
          reject(false)
        }
      } catch (e) {
        console.log(`AppLogBridge, error:${JSON.stringify(e.stack)}`)
        reject(false)
      }
    })
  }
  setNativeAppId(appId: number) {
    try {
      AppLogBridge.setNativeAppId(JSON.stringify(appId))
      console.log(`change bridge appid, event report with appid: ${appId}`);
    } catch (e) {
      this.error('setNativeAppId error', e.message);
    }
  }
  setConfig(info: Record) {
    try {
      const { user_unique_id, user_unique_id_type, ...otherKeys } = info;
      if (user_unique_id && user_unique_id_type) {
        this.setUserUniqueIdWithType(user_unique_id, user_unique_id_type);
      } else if (user_unique_id) {
        this.setUserUniqueId(user_unique_id);
      } else if (user_unique_id_type) {
        this.validateAndSetUserIdType(user_unique_id_type);
      }
      // 处理其他键值对
      for (const [key, value] of Object.entries(otherKeys)) {
        if (value) {
          this.addHeaderInfo(key, value);
        } else {
          this.removeHeaderInfo(key);
        }
      }
    } catch (e) {
      this.error('setConfig error', e.message);
    }
  }
  setUserUniqueId(uuid: string) {
    try {
      this.currentUserUniqueId = uuid;
      this.currentUserIdType = ''; // 重置user_id_type
      // AppLogBridge.setUserUniqueId(uuid);
    } catch (e) {
      this.error('setUserUniqueId error', e.message);
    }
  }
  validateAndSetUserIdType(type: string) {
    if (this.currentUserUniqueId) {
      this.setUserUniqueIdWithType(this.currentUserUniqueId, type);
    } else {
      console.warn('please set user_unique_id and user_unique_id_type together');
    }
  }
  setUserUniqueIdWithType(uuid: string, type: string) {
    try {
      this.currentUserUniqueId = uuid;
      this.currentUserUniqueIdType = type;
      AppLogBridge.setUserUniqueIdWithType(uuid, type);
    } catch (e) {
      this.error('setUserUniqueType error', e.message);
    }
  }
  addHeaderInfo(key: string, value: any) {
    try {
      AppLogBridge.addHeaderInfo(key, value)
    } catch (e) {
      this.error('addHeaderInfo error', e.message);
    }
  }
  setHeaderInfo(map: Record<string, any>) {
    try {
      AppLogBridge.setHeaderInfo(JSON.stringify(map))
    } catch (e) {
      this.error('setHeaderInfo error', e.message);
    }
  }
  removeHeaderInfo(key: string) {
    try {
      AppLogBridge.removeHeaderInfo(key)
    } catch (e) {
      this.error('removeHeaderInfo error', e.message);
    }
  }
  reportPv(params: Record<string, any>) {
    this.onEventV3('predefine_pageview', params)
  }
  onEventV3(event: string, params: Record<string, any>) {
    try {
      AppLogBridge.onEventV3(event, params)
    } catch (e) {
      this.error('onEventV3 error', e.message);
    }
  }
  profileSet(profile: Record<string, any>) {
    try {
      AppLogBridge.profileSet(profile)
    } catch (e) {
      this.error('profileSet error', e.message);
    }
  }
  profileSetOnce(profile: Record<string, any>) {
    try {
      AppLogBridge.profileSetOnce(profile)
    } catch (e) {
      this.error('profileSetOnce error', e.message);
    }
  }
  profileIncrement(profile: Record<string, any>) {
    try {
      AppLogBridge.profileIncrement(profile)
    } catch (e) {
      this.error('profileIncrement error', e.message);
    }
  }
  profileUnset(key: string) {
    try {
      AppLogBridge.profileUnset(key)
    } catch (e) {
      this.error('profileUnset error', e.message);
    }
  }
  profileAppend(profile: Record<string, any>) {
    try {
      AppLogBridge.profileAppend(profile)
    } catch (e) {
      this.error('profileAppend error', e.message);
    }
  }
  // AB
  setExternalAbVersion(vid) {
    try {
      AppLogBridge.setExternalAbVersion(vid)
    } catch (e) {
      this.error('setExternalAbVersion error', e.message);
    }
  }
  getVar(key: string, defaultValue: any, callback: Function) {
    try {
      if (this.os === 'harmonyos') {
        callback(AppLogBridge.getABTestConfigValueForKey(key, defaultValue));
      } else if (this.os === 'android') {
        callback(AppLogBridge.getABTestConfigValueForKey(key, defaultValue));
      } else {
        AppLogBridge.getABTestConfigValueForKey(key, defaultValue, (res) => {
          callback(res);
        });
      }
    } catch (e) {
      this.error('getVar error', e.message);
      callback(defaultValue);
    }
  }
  getAllVars(callback: Function) {
    try {
      if (this.os === 'harmonyos') {
        callback(AppLogBridge.getAllAbTestConfigs());
      } else if (this.os === 'android') {
        callback(AppLogBridge.getAllAbTestConfigs());
      } else {
        AppLogBridge.getAllAbTestConfigs((res) => {
          callback(res);
        });
      }
    } catch (e) {
      this.error('getAllVars error', e.message);
      callback(null);
    }
  }
  getAbSdkVersion(callback: Function) {
    try {
      if (this.os === 'harmonyos') {
        callback(AppLogBridge.getAbSdkVersion());
      } else if (this.os === 'android') {
        callback(AppLogBridge.getAbSdkVersion());
      } else {
        AppLogBridge.getAbSdkVersion((res) => {
          callback(res);
        });
      }
    } catch (e) {
      this.error('getAbSdkVersion error', e.message);
      callback('');
    }
  }
}

export default AppBridge
