// byted-tea-sdk@5.3.10/src/plugin/cep/cep.ts

// import { SDK_VERSION } from '../../collect/constant'
// import { getLogPluginSpace } from '../../collect/namespace'
// import dayjs from 'dayjs'
// import duration from 'dayjs/plugin/duration'
// import { DebuggerMesssge } from '../../collect/hooktype';
// dayjs.extend(duration)

// const ignoreList = [
//   '__cep_sdk_onboard',
//   '__cep_trigger_sdk',
//   '_be_active',
// ]
// const moreOP = [
//   'and',
//   'or',
//   'not'
// ]
// export default class CepRule {
//   collect: any
//   config: any
//   storage: any
//   key: string
//   url: string
//   configManager: any
//   ruleType: number
//   support: boolean = true
//   stateArray: any
//   waitArray: any
//   matchResult: any
//   isWait: boolean
//   reportUrl: string
//   ruleData: any
//   planData: any
//   currentPlan: number = 1
//   currentPlanId: string
//   quotasKey: string
//   shouldSendEvent: boolean = false
//   waitFnQueue: any
//   planMap: any
//   staing: boolean = false
//   test: boolean = false
//   cepReady: boolean = false
//   eventCache: any = []
//   types: any
//   apply(collect: any, config: any) {
//     if (!config.cep) return;
//     this.collect = collect
//     this.config = config
//     const { storage } = collect.adapters
//     const { Types } = collect
//     this.types = Types
//     this.storage = new storage(false)
//     this.configManager = collect.configManager
//     this.url = config.cep_url ? `${config.cep_url}` : `${collect.configManager.getDomain()}/service/2/cep_settings/`
//     this.reportUrl = collect.configManager.getUrl('event')
//     this.key = `__tea_cep_plan_${config.app_id}`
//     this.quotasKey = `__tea_cep_plan_quotas_${config.app_id}`
//     this.isWait = false
//     this.eventCache = []
//     this.test = collect.configManager.staging
//     this.collect.on(Types.SubmitAfter, (events: any) => {
//       this.matchEvent(events)
//     });
//     this.init()
//   }
//   init() {
//     let shouldUpdate = true
//     let last_update_time = 0
//     let fetch_interval = 0
//     let config_version = 0
//     let hasLocal = false
//     const cep_cache = this.storage.getItem(this.key) || {}
//     if (cep_cache && Object.keys(cep_cache).length) {
//       hasLocal = true
//       last_update_time = cep_cache.last_update_time
//       fetch_interval = cep_cache.fetch_interval
//       config_version = cep_cache.config_version || 0
//       shouldUpdate = Date.now() - last_update_time > fetch_interval * 1000
//     }
//     if (shouldUpdate) {
//       const { header, user } = this.configManager.get()
//       this.collect.requestManager.useRequest({
//         url: this.url,
//         data: {
//           header,
//           user,
//           local_time: Date.now(),
//           config_version
//         },
//         success: (data) => {
//           if (data && data.code === 0) {
//             cep_cache.fetch_interval = data.fetch_interval
//             cep_cache.data = data.data
//             cep_cache.last_update_time = Date.now()
//             cep_cache.config_version = data.config_version
//             this.clearQuotas()
//             this.initState(data.data)
//             this.storage.setItem(this.key, cep_cache)
//             data.data.forEach(item => {
//               this.event('__cep_sdk_onboard', item.plan_id)
//             })
//           } else if ((data.code === 304 || data.code === 400) && hasLocal) {
//             console.log(`code: ${data.code}, use old rules~`)
//             cep_cache.last_update_time = Date.now()
//             this.storage.setItem(this.key, cep_cache)
//             this.initState(cep_cache.data)
//           } else {
//             console.log(`code: ${data.code}, request error，please try leater`)
//           }
//         },
//         fail: () => {
//           if (hasLocal) {
//             this.planData = cep_cache.data
//             this.initState(cep_cache.data)
//           } else {
//             this.support = false
//             console.log(`request error，please try leater`)
//           }
//         },
//         timeout: 5000
//       })
//     } else {
//       cep_cache.last_update_time = Date.now()
//       this.storage.setItem(this.key, cep_cache)
//       console.log(`rule still work, use cache`)
//       this.initState(cep_cache.data)
//     }
//   }
//   initState(data: any) {
//     if (!data || !data.length) return
//     this.stateArray = new Map()
//     this.waitArray = new Map()
//     this.planData = data
//     data.forEach((item: any) => {
//       let patternState = new Map()
//       if (item.pattern && item.pattern.events) {
//         const singleRule = item.pattern.events.length === 1
//         item.pattern.events.forEach((child: any) => {
//           patternState.set(child.id, {
//             id: child.id,
//             isMatch: false,
//             event: null,
//             after: child.after || null,
//             connection: child.connection || null,
//             window: child.window || null,
//             matchTime: 0,
//             initTime: Date.now(),
//             singleRule
//           })
//         })
//         this.stateArray.set(item.plan_id, {
//           patternState,
//           singleRule,
//           pattern: {
//             aid: item.aid,
//             biz_id: item.biz_id,
//             end_time: item.end_time,
//             start_time: item.start_time,
//             web_sdk_version: item.web_sdk_version,
//             stage: item.stage
//           },
//           quotas: item.quotas
//         })
//       }
//     })
//     this.cepReady = true
//     this.dealCache()
//     this.collect.emit(this.types.CepReady)
//     // console.log(this.stateArray)
//   }
//   check(planData: any) {
//     const date = Date.now()
//     if (Date.now() > planData.end_time * 1000) {
//       console.log(`event rules:${planData.plan_id} is over`)
//       return false
//     }
//     if (date < planData.start_time * 1000 || date > planData.end_time * 1000) {
//       console.log(`rule:${planData.plan_id} time error`)
//       return false
//     }
//     this.staing = planData.staing === 'testing'
//     if (SDK_VERSION < planData.web_sdk_version) {
//       this.support = false
//       console.log(`you sdk version is old，please update to ${planData.web_sdk_version} at least`)
//       return false
//     }
//     return true
//   }
//   dealCache() {
//     if (this.eventCache.length) {
//       let count = 1
//       this.eventCache.forEach(item => {
//         this.matchEvent(item)
//         count++
//       })
//       if (count === this.eventCache.length) {
//         this.eventCache = []
//       }
//     }
//   }
//   matchEvent(event: any) {
//     if (!this.support) return
//     if (this.staing && !this.test) return
//     if (!event || !event.length) return
//     try {
//       if (!this.cepReady) {
//         this.eventCache.push(event)
//         return
//       }
//       const { header, user, events } = event[0]
//       if (events.length) {
//         events.reverse()
//       }
//       for (let x = 0; x < events.length; x++) {
//         let { event } = events[x]
//         if (ignoreList.indexOf(event) !== -1) return false
//         this.matchRule(header, user, events[x])
//       }
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return
//     }
//   }
//   matchRule(header: any, user: any, events: any) {
//     try {
//       for (let i = 0; i < this.planData.length; i++) {
//         const currentPlan = this.planData[i]
//         const currentPlanId = currentPlan.plan_id
//         if (!this.check(currentPlan)) continue
//         if (!this.checkQuotas(currentPlanId)) continue
//         if (currentPlan.pattern && currentPlan.pattern.events) {
//           const currentEvents = currentPlan.pattern.events
//           for (let x = 0; x < currentEvents.length; x++) {
//             let isMatch = this.rule(header, user, events, currentEvents[x].condition)
//             if (!isMatch) continue
//             let currentRule = this.stateArray.get(currentPlanId).patternState
//             let currentCondition = currentRule.get(currentEvents[x].id)
//             let isSingle = this.stateArray.get(currentPlanId).singleRule
//             if (isSingle) {
//               this.changeStatus(currentPlanId, currentEvents[x].id, events)
//               this.publish([events], currentPlanId, currentEvents[x].id)
//               break;
//             } else {
//               // more
//               if (!currentCondition.isMatch) {
//                 if (currentCondition.after && currentRule.get(currentCondition.after).isMatch) {
//                   // 后置条件
//                   this.changeStatus(currentPlanId, currentEvents[x].id, events)
//                 } else if (!currentCondition.after) {
//                   // 前置条件
//                   this.changeStatus(currentPlanId, currentEvents[x].id, events)
//                 }
//                 break
//               }
//               continue;
//             }
//           }
//         }
//       }
//       this.statusCheck()
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return
//     }
//   }
//   changeStatus(planId: string, ruleId: string, events: any) {
//     const planRule = this.stateArray.get(planId)
//     const status = planRule.patternState.get(ruleId)
//     status.isMatch = true
//     status.event = events
//     status.matchTime = events.local_time_ms || Date.now()
//     planRule.patternState.set(ruleId, status)
//     this.stateArray.set(planId, planRule)
//   }
//   statusCheck() {
//     for (let [planKey, planObj] of this.stateArray) {
//       let plan = planObj.patternState
//       if (planObj.isSingle) break
//       for (let [ruleKey, rule] of plan) {
//         if (!rule.after || (rule.after && !plan.get(rule.after).isMatch)) continue
//         if (rule.connection === 'NOT_FOLLOWED_BY') {
//           // N时间内不要触发 NOT_FOLLOWED_BY
//           if (!rule.isMatch) {
//             if (Date.now() - plan.get(rule.after).matchTime > rule.window) {
//               this.publish([plan.get(rule.after).event, rule.event], planKey, ruleKey)
//             } else {
//               this.startWait(planKey, ruleKey, rule)
//             }
//           } else {
//             // after触发了，同时时间小于window，则重置则规则
//             this.reset(planKey, ruleKey)
//           }
//           break
//         } else {
//           // N时间内要触发 FOLLOWED_BY
//           if (rule.isMatch) {
//             if (rule.matchTime - plan.get(rule.after).matchTime <= rule.window) {
//               // after触发了，同时时间小于window N时间事件，则立即publish
//               this.publish([plan.get(rule.after).event, rule.event], planKey, ruleKey)
//             } else {
//               // after触发了，同时时间大于window N时间事件，则立即重置此规则
//               this.reset(planKey, ruleKey)
//             }
//           } else {
//             // 还没匹配
//             if (Date.now() - plan.get(rule.after).matchTime > rule.window) {
//               // after没触发，同时时间已经超过window N时间，则立即重置此规则
//               this.reset(planKey, ruleKey)
//             }
//             else {
//               this.startWait(planKey, ruleKey, rule)
//             }
//           }
//           break
//         }
//       }
//     }
//   }
//   startWait(planKey: string, ruleKey: string, rule: any) {
//     try {
//       if (!this.waitArray.get(planKey)) {
//         this.waitArray.set(planKey, new Map())
//       }
//       const planWait = this.waitArray.get(planKey)
//       if (!planWait.get(ruleKey)) {
//         const waitFn = setTimeout(() => {
//           const currentPlan = this.stateArray.get(planKey).patternState
//           const currentRule = currentPlan.get(ruleKey)
//           if (currentRule.connection === 'NOT_FOLLOWED_BY') {
//             if ((!currentRule.isMatch && currentPlan.get(currentRule.after).isMatch) || (currentRule.matchTime - currentPlan.get(currentRule.after).matchTime > rule.window)) {
//               const ruleWatiStatus = this.waitArray.get(planKey).get(ruleKey)
//               ruleWatiStatus.triggerWait = true
//               this.waitArray.set(planKey, this.waitArray.get(planKey).set(ruleKey, ruleWatiStatus))
//               this.publish([currentPlan.get(currentRule.after).event], planKey, ruleKey)
//             } else {
//               this.reset(planKey, ruleKey)
//             }
//           } else if (currentRule.connection === 'FOLLOWED_BY') {
//             if ((currentRule.isMatch && currentPlan.get(currentRule.after).isMatch)) {
//               const ruleWatiStatus = this.waitArray.get(planKey).get(ruleKey)
//               ruleWatiStatus.triggerWait = true
//               this.waitArray.set(planKey, this.waitArray.get(planKey).set(ruleKey, ruleWatiStatus))
//               this.publish([currentPlan.get(currentRule.after).event], planKey, ruleKey)
//             } else {
//               this.reset(planKey, ruleKey)
//             }
//           }

//         }, rule.window)
//         planWait.set(ruleKey, { waitFn, triggerWait: false })
//         this.waitArray.set(planKey, planWait)
//       }
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message, e.stack)
//     }

//   }
//   rule(header: any, user: any, _event: any, rule: any) {
//     try {
//       let { event, params } = _event
//       params = JSON.parse(params)
//       if (moreOP.indexOf(rule.op) !== -1) {
//         if (rule.conditions && rule.conditions.length) {
//           const conditions = rule.conditions
//           let conditionsKey = 0
//           let resultMap = new Map()
//           // 利用一个map来存储每个条件的结果
//           resultMap.set(conditionsKey, {
//             op: rule.op,
//             result: undefined,
//             resultList: [],
//             resultKey: [],
//             resultLength: conditions.length
//           })
//           const recursion = (conditions: any, fatherKey: number) => {
//             for (let x = 0; x < conditions.length; x++) {
//               conditionsKey++
//               if (moreOP.indexOf(conditions[x].op) !== -1 && conditions[x].hasOwnProperty('conditions')) {
//                 // 叉节点的结果
//                 resultMap.set(conditionsKey, {
//                   op: conditions[x].op,
//                   result: undefined,
//                   fatherKey,
//                   resultList: [],
//                   resultKey: [],
//                   resultLength: conditions[x].conditions.length
//                 })
//                 recursion(conditions[x].conditions, conditionsKey)
//               } else {
//                 const { scope, key, key2 } = this.scope(conditions[x].key)
//                 let valueA = scope ? (scope === 'header' ? (key2 ? JSON.parse(header[key2])[key] : header[key]) : (scope === 'user' ? user[key] : params[key])) : event
//                 const calculateResult = this.calculate(conditions[x].op, valueA, conditions[x].value)
//                 // 叶子节点的结果
//                 resultMap.set(conditionsKey, {
//                   op: conditions[x].op,
//                   result: calculateResult,
//                   fatherKey
//                 })
//                 continue;
//               }
//             }
//           }
//           recursion(conditions, conditionsKey)
//           let originConditionResult = resultMap.get(0)
//           while (originConditionResult.result === undefined || originConditionResult.resultList.length < originConditionResult.resultLength) {
//             // 计算出所有的树节点的结果后，遍历每个节点，根据子节点找到父节点，把子节点的结果放入父节点的结果数组里
//             resultMap.forEach((item, key) => {
//               if (item.result !== undefined && item.fatherKey !== undefined) {
//                 let fItem = resultMap.get(item.fatherKey)
//                 if (fItem.resultList.length < fItem.resultLength && fItem.resultKey.indexOf(key) === -1) {
//                   fItem.resultKey.push(key)
//                   fItem.resultList.push(item.result)
//                   resultMap.set(item.fatherKey, fItem)
//                 }
//               }
//             })
//             // 再遍历map，根据节点的op类型和结果数组，算出最终的结果，一层一层的往上抛，最后算出最外层的结果
//             resultMap.forEach(item => {
//               if (item.resultList && item.resultList.length === item.resultLength) {
//                 if (item.op === 'and') {
//                   item.result = item.resultList.indexOf(false) === -1
//                 }
//                 if (item.op === 'or') {
//                   item.result = item.resultList.indexOf(true) !== -1
//                 }
//               }
//             })
//           }

//           const finalResult = resultMap.get(0).result
//           return finalResult
//         } else {
//           return false
//         }
//       } else {
//         const { scope, key, key2 } = this.scope(rule.key)
//         let valueA = scope ? (scope === 'header' ? (key2 ? JSON.parse(header[key2])[key] : header[key]) : (scope === 'user' ? user[key] : params[key])) : event
//         return this.calculate(rule.op, valueA, rule.value)
//       }
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return false
//     }
//   }
//   publish(event: any, planKey: string, ruleKey: string) {
//     try {
//       const { user } = this.configManager.get()
//       const matchInfo = {
//         biz_id: this.stateArray.get(planKey).pattern.biz_id,    // tea 上创建的计划 id
//         plan_id: planKey,    // 计划名称
//         uid: user.user_id || '',
//         did: user.device_id || '',
//         uuid: user.user_unique_id,
//         events: event,    // 事件明细
//         extra: {
//           url: window.location.href,
//         }
//       }
//       this.collect.emit('__cep_match', matchInfo);
//       (window.opener || window.parent).postMessage({
//         type: '__cep_match',
//         data: matchInfo
//       }, '*');

//       this.event('__cep_trigger_sdk', planKey)
//       this.addQuotas(planKey)
//       // reset
//       this.reset(planKey, ruleKey)
//       console.log(this.stateArray)
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return
//     }
//   }
//   addQuotas(planKey: string) {
//     const quotas = this.stateArray.get(planKey).quotas
//     if (quotas && quotas.length) {
//       const localQuotos = this.storage.getItem(this.quotasKey) || {}
//       if (!localQuotos[planKey]) {
//         localQuotos[planKey] = {
//           value: 0,
//           lastTime: Date.now()
//         }
//       }
//       const value = localQuotos[planKey].value
//       if (value === 0) {
//         localQuotos[planKey].lastTime = Date.now()
//       }
//       localQuotos[planKey].value = value + 1
//       this.storage.setItem(this.quotasKey, localQuotos)
//     }
//   }
//   checkQuotas(planKey: string) {
//     const quotas = this.stateArray.get(planKey).quotas
//     const localQuotos = this.storage.getItem(this.quotasKey) || {}
//     if (!localQuotos[planKey]) {
//       localQuotos[planKey] = {
//         value: 0,
//         lastTime: Date.now()
//       }
//     }
//     if (quotas && quotas.length) {
//       const granularity = quotas[0].granularity
//       const value = quotas[0].value
//       if (quotas[0].type !== 'per_keyid') return true
//       const limit = this.release(granularity)
//       if (limit === 'all') {
//         if (localQuotos[planKey].value >= value) {
//           // 达到上限了
//           console.log(`plan:${planKey} limit`)
//           return false
//         }
//         return true
//       } else {
//         // N时间内
//         const rangers = Date.now() - localQuotos[planKey].lastTime
//         if (rangers >= limit) {
//           localQuotos[planKey].value = 0
//           localQuotos[planKey].lastTime = Date.now()
//           this.storage.setItem(this.quotasKey, localQuotos)
//           return true
//         } else {
//           // 上次触发和本次触发时间在范围内
//           if (localQuotos[planKey].value >= value) {
//             // 达到上限了
//             console.log(`plan:${planKey} limit`)
//             return false
//           }
//           return true
//         }
//       }
//     }
//     return true
//   }
//   clearQuotas() {
//     this.storage.setItem(this.quotasKey, null)
//   }
//   release(granularity: string) {
//     if (granularity === 'all') return granularity
//     const _duration = dayjs.duration(granularity)
//     if (_duration) {
//       return _duration.asMilliseconds()
//     } else {
//       return 'all'
//     }
//   }
//   reset(planKey: string, ruleKey: string) {
//     const planStatus = this.stateArray.get(planKey)
//     const singleRule = planStatus.singleRule
//     if (singleRule) {
//       const ruleStatus = planStatus.patternState.get(ruleKey)
//       ruleStatus.isMatch = false
//       ruleStatus.event = null
//       ruleStatus.matchTime = 0
//       ruleStatus.initTime = Date.now()
//       planStatus.patternState.set(ruleKey, ruleStatus)
//     } else {
//       planStatus.patternState.forEach(item => {
//         item.isMatch = false
//         item.event = null
//         item.matchTime = 0
//         item.initTime = Date.now()
//       })
//     }
//     this.stateArray.set(planKey, planStatus)
//     const waitStatus = this.waitArray.get(planKey)
//     if (waitStatus && waitStatus.get(ruleKey)) {
//       clearTimeout(waitStatus.get(ruleKey).waitFn)
//       waitStatus.delete(ruleKey)
//     }
//   }
//   event(eventName: string, planKey?: string) {
//     try {
//       const event = {
//         event: eventName,
//         params: JSON.stringify({
//           plan_id: planKey || '',
//           biz_id: this.stateArray.get(planKey).pattern.biz_id
//         }),
//         local_time_ms: Date.now(),
//       }
//       let { header, user } = this.configManager.get()
//       header['custom'] = JSON.stringify(header['custom'])
//       const data = {
//         events: [event],
//         user,
//         header,
//       }
//       this.collect.requestManager.useRequest({
//         url: this.reportUrl,
//         data: [data]
//       })
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return
//     }
//   }
//   scope(key: string) {
//     let scope = ''
//     let key2 = ''
//     if (key.indexOf('.') > -1) {
//       var tmp = key.split('.')
//       scope = tmp[0]
//       key = tmp[1]
//       if (scope === 'header' && key === 'custom') {
//         key = tmp[2]
//         key2 = tmp[1]
//       }
//     }
//     return { scope, key, key2 }
//   }
//   calculate(op: string, paramA: any, paramB?: any) {
//     try {
//       if (op === '=') {
//         return paramA === paramB
//       } else if (['=', '<', '>', '>=', '<=', '!='].indexOf(op) !== -1) {
//         switch (op) {
//           case '=':
//             return paramA === paramB;
//           case '<':
//             return paramA < paramB;
//           case '>':
//             return paramA > paramB;
//           case '>=':
//             return paramA >= paramB;
//           case '<=':
//             return paramA <= paramB;
//           case '!=':
//             return paramA !== paramB;
//         }
//       } else if (['in', 'contains'].indexOf(op) !== -1) {
//         return paramB.indexOf(paramA) !== -1
//       } else if (['is not null', 'is null'].indexOf(op) !== -1) {
//         if (op === 'is null') {
//           return paramA === null || paramA === undefined
//         } else {
//           return paramA !== null && paramA !== undefined
//         }
//       } else if (['startswith', 'endswith'].indexOf(op) !== -1) {
//         if (op === 'startswith') {
//           return paramA.indexOf(paramB) === 0
//         } else {
//           const start = paramA.length - paramB.length
//           const str = paramA.substr(start, paramB.length)
//           return str === paramB
//         }
//       }
//     } catch (e) {
//       this.collect.emit(DebuggerMesssge.DEBUGGER_MESSAGE, { type: DebuggerMesssge.DEBUGGER_MESSAGE_SDK, info: '发生了异常', level: 'error', time: Date.now(), data: e.message });
//       console.log(e.message)
//       return false
//     }
//   }
// }


// /**@@SCRIPT
// try {
//   const exportCep = (collect: any, config: any) => {
//     const cep = new CepRule()
//     cep.apply(collect, config)
//   }
//   const pluginObject = getLogPluginSpace()
//   if (pluginObject) {
//     pluginObject.LogCep = exportCep
//   }
// } catch (e) {
//   console.log(e)
// }
// @@SCRIPT*/
