// @ies/starling_client@3.6.6/src/index.ts
import Starling from './base';
import { Options } from '@ies/starling_client_core/lib/options';
import { createTracker, channel_type, channel } from './tracker';

type WithOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
export type WebOptions = WithOptional<Options, 'store' | 'request' | 'tracker'> & {
  /**
   * 指定 TEA 上报域名，不同地区对应域名可查看
   * https://bytedance.feishu.cn/wiki/wikcnrILBFaGcjbYQkfAFPljqsg#9CAlE8
   */
  TEAChannel?: channel;
  TEAChannelType?: channel_type;
  TEAChannelDomain?: string;
};

class StarlingWeb extends Starling {
  constructor(options: WebOptions) {
    const { test, TEAChannelDomain = '', mode, TEAChannel, TEAChannelType } = options;

    options.tracker =
      TEAChannelDomain || TEAChannelType || TEAChannel
        ? createTracker({
            isTest: test || mode === 'test',
            channel_domain: TEAChannelDomain,
            channel_type: TEAChannelType,
            channel: TEAChannel,
          })
        : () => {};

    super(options as Options);
  }
}

export default StarlingWeb;
